From 26ba9dab53487c0a76f701cb913b9f73644d7bf5 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 15 Feb 2026 22:30:12 -0800
Subject: [PATCH] feat: Add `.domains` file with custom domain

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.domains b/.domains
new file mode 100644
index 0000000000000000000000000000000000000000..ddec9b960eaed1d83886598926d3545e0512a103
--- /dev/null
+++ b/.domains
@@ -0,0 +1 @@
+upcyclists.org
\ No newline at end of file

