From c83af85234fb185c7b9f6f7202511f8618ba7e0e Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 22 Apr 2026 21:52:13 -0700
Subject: [PATCH] build: Use latest conventions for DOAP and Metainfo files

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/Multiplex.doap b/multiplex.doap
rename from Multiplex.doap
rename to multiplex.doap
index 2b5964b8905b167c6240fa98b3c98900ea528e42..ce6dfe8b4bc1ce9cc678d1659dde17683f7f275a 100644
--- a/Multiplex.doap
+++ b/multiplex.doap
@@ -7,11 +7,8 @@   xmlns="http://usefulinc.com/ns/doap#">
 
   <name xml:lang="en">Multiplex</name>
   <shortdesc xml:lang="en">Watch torrents with your friends</shortdesc>
-  <description xml:lang="en">Multiplex is an app to watch torrents together, providing an experience
-    similar to Apple's SharePlay and Amazon's Prime Video Watch Party.</description>
 
   <homepage rdf:resource="https://github.com/pojntfx/multiplex/" />
-  <download-page rdf:resource="https://flathub.org/apps/com.pojtinger.felicitas.Multiplex" />
   <bug-database rdf:resource="https://github.com/pojntfx/multiplex/issues/" />
 
   <programming-language>Go</programming-language>
@@ -21,6 +18,7 @@
   <maintainer>
     <foaf:Person>
       <foaf:name>Felicitas Pojtinger</foaf:name>
+      <foaf:mbox rdf:resource="mailto:felicitas@pojtinger.com" />
       <foaf:account>
         <foaf:OnlineAccount>
           <foaf:accountServiceHomepage rdf:resource="https://github.com" />
diff --git a/assets/resources/metainfo.xml.in b/assets/resources/metainfo.xml.in
index 65d9228d2f3356dea511510086b1f66e421adaaa..85474e36f73f43bcf32207eec350f747eec6c33f 100644
--- a/assets/resources/metainfo.xml.in
+++ b/assets/resources/metainfo.xml.in
@@ -1,11 +1,24 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <component type="desktop-application">
     <id>com.pojtinger.felicitas.Multiplex</id>
+    <metadata_license>CC0-1.0</metadata_license>
+    <project_license>AGPL-3.0-or-later</project_license>
+
     <name>Multiplex</name>
     <summary>Watch torrents with your friends</summary>
-
-    <metadata_license>CC0-1.0</metadata_license>
-    <project_license>AGPL-3.0-or-later</project_license>
+    <description>
+        <p>Multiplex is an app to watch torrents together, providing an experience similar to
+            Apple&#39;s SharePlay and Amazon&#39;s Prime Video Watch Party.</p>
+        <p>It enables you to:</p>
+        <ul>
+            <li>Stream any file directly using a wide range of video and audio formats with the mpv
+                video player</li>
+            <li>Host online watch parties while preserving your privacy by synchronizing video
+                playback with friends without a central server</li>
+            <li>Bypass internet restrictions by optionally separating the backend and user interface
+                into two separate components</li>
+        </ul>
+    </description>
 
     <developer id="com.pojtinger.felicitas">
         <name translate="no">Felicitas Pojtinger</name>
@@ -18,8 +31,8 @@
     <url type="bugtracker">https://github.com/pojntfx/multiplex/issues</url>
     <url type="homepage">https://github.com/pojntfx/multiplex/</url>
     <url type="donation">https://github.com/sponsors/pojntfx</url>
-    <url type="help">https://matrix.to/#/#multiplex:matrix.org?via=matrix.org</url>
-    <url type="contact">https://felicitas.pojtinger.com/imprint</url>
+    <url type="help">https://matrix.to/#/#multiplex:matrix.org</url>
+    <url type="contact">https://matrix.to/#/#multiplex:matrix.org</url>
     <url type="faq">https://github.com/pojntfx/multiplex?tab=readme-ov-file#tutorial</url>
     <url type="vcs-browser">https://github.com/pojntfx/multiplex/</url>
     <url type="translate">https://hosted.weblate.org/projects/multiplex/</url>
@@ -46,20 +59,6 @@     <branding>
         <color type="primary" scheme_preference="light">#6fb0ff</color>
         <color type="primary" scheme_preference="dark">#0087ff</color>
     </branding>
-
-    <description>
-        <p>Multiplex is an app to watch torrents together, providing an experience similar to
-            Apple&#39;s SharePlay and Amazon&#39;s Prime Video Watch Party.</p>
-        <p>It enables you to:</p>
-        <ul>
-            <li>Stream any file directly using a wide range of video and audio formats with the mpv
-                video player</li>
-            <li>Host online watch parties while preserving your privacy by synchronizing video
-                playback with friends without a central server</li>
-            <li>Bypass internet restrictions by optionally separating the backend and user interface
-                into two separate components</li>
-        </ul>
-    </description>
 
     <screenshots>
         <screenshot type="default">

