From 79aa99a9ed27264ebf65da08f45614c0c3e99dc2 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 22 Apr 2026 22:08:39 -0700
Subject: [PATCH] chore: Prepare new release

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/resources/index.go b/assets/resources/index.go
index fddeae38e960e3074604c129d05b141f73741ed8..d979a52c415a3e4b173244cbeda1495934090e1c 100644
--- a/assets/resources/index.go
+++ b/assets/resources/index.go
@@ -12,7 +12,7 @@ //go:generate sh -c "msgfmt --xml -L metainfo --template ../../assets/resources/metainfo.xml.in -d ../../po -o - -f | sed 's|/LC[-_]MESSAGES/default||g' > ../../assets/resources/metainfo.xml"
 
 const (
 	AppID      = "com.pojtinger.felicitas.Multiplex"
-	AppVersion = "0.1.12"
+	AppVersion = "0.1.13"
 )
 
 //go:generate sh -c "blueprint-compiler batch-compile . . *.blp && glib-compile-resources *.gresource.xml"
diff --git a/assets/resources/metainfo.xml.in b/assets/resources/metainfo.xml.in
index 85474e36f73f43bcf32207eec350f747eec6c33f..c366dc4f5a2e391912d7b2fbe28fc2f3bebbddd5 100644
--- a/assets/resources/metainfo.xml.in
+++ b/assets/resources/metainfo.xml.in
@@ -114,6 +114,13 @@         </screenshot>
     </screenshots>
 
     <releases>
+        <release version="0.1.13" date='2026-04-22T00:00:00Z00:00'>
+            <url type="details">https://github.com/pojntfx/multiplex/releases/tag/v0.1.13</url>
+            <description translate="no">
+                <p>Update to latest GNOME platform, add a keyboard shortcuts dialog, add new
+                    translations, and improve project metadata</p>
+            </description>
+        </release>
         <release version="0.1.12" date='2026-01-21T00:00:00Z00:00'>
             <url type="details">https://github.com/pojntfx/multiplex/releases/tag/v0.1.12</url>
             <description translate="no">

