From 45706694aaf8e419e264024e4949c9cb13589802 Mon Sep 17 00:00:00 2001
From: Julian Sparber <julian@sparber.net>
Date: Tue, 14 Apr 2026 16:50:29 +0200
Subject: [PATCH] Add p2panda_topic_id_new_from_data()

This was also forgotten
---
diff --git a/libp2panda/src/topic.rs b/libp2panda/src/topic.rs
index 62f87881c74fa60f98b8e7aeea377300f811ab1d..491781ae74d70ecb47c857ba6110e2fbb5828e66 100644
--- a/libp2panda/src/topic.rs
+++ b/libp2panda/src/topic.rs
@@ -218,6 +218,13 @@     topic::TopicId::new().into_glib_ptr()
 }
 
 #[unsafe(no_mangle)]
+pub unsafe extern "C" fn p2panda_topic_id_new_from_data(
+    data: *const [u8; 32],
+) -> *mut topic::TopicId {
+    unsafe { topic::TopicId::from_data(*data).into_glib_ptr() }
+}
+
+#[unsafe(no_mangle)]
 pub unsafe extern "C" fn p2panda_topic_id_get_data(topic_id: *mut topic::TopicId) -> *const [u8; 32] {
     unsafe {
         let topic_id = topic::TopicId::from_glib_none(topic_id);

