From 826360f12c6874e81c0783741eb7e51c9260b7d7 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 17 Feb 2026 22:56:39 -0800
Subject: [PATCH] Fix `SetStatus` portal method call by using correct name and dropping `parentWindow` argument

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/background/status.go b/background/status.go
index 9b4314a7a09550223dd2d7b27dcae361d451f1bb..8f6844fc40901e1005a9d3308c8dc37a59860576 100644
--- a/background/status.go
+++ b/background/status.go
@@ -6,14 +6,16 @@ 	"github.com/rymdport/portal/internal/apis"
 	"github.com/rymdport/portal/internal/convert"
 )
 
+const statusCallName = interfaceName + ".SetStatus"
+
 // StatusOptions are the options passed into SetStatus.
 type StatusOptions struct {
 	Message string // A string that will be used as the status message of the application. This should be a single line that can be presented to the user in a list, not a full sentence or paragraph. Must be shorter than 96 characters.
 }
 
 // SetStatus sets the status of the application running in background.
-func SetStatus(parentWindow string, options StatusOptions) error {
+func SetStatus(options StatusOptions) error {
 	data := map[string]dbus.Variant{"message": convert.FromString(options.Message)}
 
-	return apis.CallWithoutResult(requestCallName, parentWindow, data)
+	return apis.CallWithoutResult(statusCallName, data)
 }

