From 8eac2b942d1a6e7e129cfd737ed1e73dc2eb8185 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 20 Mar 2026 23:49:54 -0700
Subject: [PATCH] Use exact description from the spec for the USB portal package

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/usb/usb.go b/usb/usb.go
index eb6a0c1831a1141a3236b2279427e543f095a6cc..1700d93d5ca82d248dc72359a66504b21dd7d482 100644
--- a/usb/usb.go
+++ b/usb/usb.go
@@ -1,4 +1,4 @@
-// Package usb provides access to USB devices through the portal.
+// Package usb lets sandboxed applications monitor and request access to connected USB devices.
 // Upstream API documentation can be found at https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Usb.html.
 package usb
 

