From cc5cca2834af8ad1f365941353243f61d161a977 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 25 Aug 2025 23:25:36 -0700
Subject: [PATCH] Gen: Re-run script

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/v4/adw/adw-action-row.go b/v4/adw/adw-action-row.go
index 6bae8d01e3dc97bad8bed80733dbb720677a5ad2..87b9b48e52d8da671ddf61df8a59e0f91a396b5a 100644
--- a/v4/adw/adw-action-row.go
+++ b/v4/adw/adw-action-row.go
@@ -18,6 +18,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Activate uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/adw/adw-alert-dialog.go b/v4/adw/adw-alert-dialog.go
index ccb239beb8b43f3e5f4b59e3841600ccf7c3557e..768bcf0a2c84b65d1a99f73f82cf77f3f13e9ca2 100644
--- a/v4/adw/adw-alert-dialog.go
+++ b/v4/adw/adw-alert-dialog.go
@@ -19,6 +19,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Response uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/adw/adw-dialog.go b/v4/adw/adw-dialog.go
index b7fe74b427966ec2683375901d8ab9c63e5fc2bc..fe954145ebad070fd57c9f5439195e40031182cc 100644
--- a/v4/adw/adw-dialog.go
+++ b/v4/adw/adw-dialog.go
@@ -18,6 +18,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	CloseAttempt uintptr
+
+	Closed uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/adw/adw-message-dialog.go b/v4/adw/adw-message-dialog.go
index 7b82f1662921ac4d3c855b462392832ba2266b9e..fdae7f1256f3058585f53b0f5de81de17c857829 100644
--- a/v4/adw/adw-message-dialog.go
+++ b/v4/adw/adw-message-dialog.go
@@ -21,6 +21,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Response uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/adw/adw-navigation-view.go b/v4/adw/adw-navigation-view.go
index 7841dc7aa8aea820b77a6082ae1de0721f7e92d5..e3dc3d9d23451057409ea374696a5529865e3115 100644
--- a/v4/adw/adw-navigation-view.go
+++ b/v4/adw/adw-navigation-view.go
@@ -20,6 +20,14 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Showing uintptr
+
+	Shown uintptr
+
+	Hiding uintptr
+
+	Hidden uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/adw/adw-swipeable.go b/v4/adw/adw-swipeable.go
index 37aecf05094c475c5054a23719ff4ded828107bc..791f4afe2a258ac697acc94a4e36b2a8135dfdd1 100644
--- a/v4/adw/adw-swipeable.go
+++ b/v4/adw/adw-swipeable.go
@@ -17,6 +17,16 @@ 	_ structs.HostLayout
 
 	Parent uintptr
 
+	GetDistance uintptr
+
+	GetSnapPoints uintptr
+
+	GetProgress uintptr
+
+	GetCancelProgress uintptr
+
+	GetSwipeArea uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/gdk/gdkcontentprovider.go b/v4/gdk/gdkcontentprovider.go
index aa7124600279c7bf3ced9c315f2cb6240913757b..703bc625cb81f0d8eebbab1ba303cf3f6b518c72 100644
--- a/v4/gdk/gdkcontentprovider.go
+++ b/v4/gdk/gdkcontentprovider.go
@@ -19,6 +19,22 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	ContentChanged uintptr
+
+	AttachClipboard uintptr
+
+	DetachClipboard uintptr
+
+	RefFormats uintptr
+
+	RefStorableFormats uintptr
+
+	WriteMimeTypeAsync uintptr
+
+	WriteMimeTypeFinish uintptr
+
+	GetValue uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gdk/gdkpaintable.go b/v4/gdk/gdkpaintable.go
index df83affc8733d3609166b080114a55c853507042..9fbd1d3b85e7f32b4438f39d7303c3bcf71d1513 100644
--- a/v4/gdk/gdkpaintable.go
+++ b/v4/gdk/gdkpaintable.go
@@ -23,6 +23,18 @@ type PaintableInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Snapshot uintptr
+
+	GetCurrentImage uintptr
+
+	GetFlags uintptr
+
+	GetIntrinsicWidth uintptr
+
+	GetIntrinsicHeight uintptr
+
+	GetIntrinsicAspectRatio uintptr
 }
 
 func (x *PaintableInterface) GoPointer() uintptr {
diff --git a/v4/gdkpixbuf/gdk-pixbuf-animation.go b/v4/gdkpixbuf/gdk-pixbuf-animation.go
index dc9c753f5c2fc7576cf2edcb6c27c0da1c7db108..d44eb642ce278b3b50bc562058805a8b265bf6be 100644
--- a/v4/gdkpixbuf/gdk-pixbuf-animation.go
+++ b/v4/gdkpixbuf/gdk-pixbuf-animation.go
@@ -20,6 +20,14 @@ type PixbufAnimationClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	IsStaticImage uintptr
+
+	GetStaticImage uintptr
+
+	GetSize uintptr
+
+	GetIter uintptr
 }
 
 func (x *PixbufAnimationClass) GoPointer() uintptr {
@@ -33,6 +41,14 @@ type PixbufAnimationIterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetDelayTime uintptr
+
+	GetPixbuf uintptr
+
+	OnCurrentlyLoadingFrame uintptr
+
+	Advance uintptr
 }
 
 func (x *PixbufAnimationIterClass) GoPointer() uintptr {
diff --git a/v4/gdkpixbuf/gdk-pixbuf-io.go b/v4/gdkpixbuf/gdk-pixbuf-io.go
index ca8c8c4fa4cfda4f89f4ea479d0ef901e404fbb5..0461e7fb528b61647cca8c0bb62ca018798718aa 100644
--- a/v4/gdkpixbuf/gdk-pixbuf-io.go
+++ b/v4/gdkpixbuf/gdk-pixbuf-io.go
@@ -338,6 +338,14 @@
 	SaveToCallback PixbufModuleSaveCallbackFunc
 
 	IsSaveOptionSupported PixbufModuleSaveOptionSupportedFunc
+
+	Reserved1 uintptr
+
+	Reserved2 uintptr
+
+	Reserved3 uintptr
+
+	Reserved4 uintptr
 }
 
 func (x *PixbufModule) GoPointer() uintptr {
diff --git a/v4/gdkpixbuf/gdk-pixbuf-loader.go b/v4/gdkpixbuf/gdk-pixbuf-loader.go
index fc39de3764616ec5fc0fd2d8f5e07064d7babc90..36d3292d2b3ebc68bc7d76f3e9ae8e7c365dced9 100644
--- a/v4/gdkpixbuf/gdk-pixbuf-loader.go
+++ b/v4/gdkpixbuf/gdk-pixbuf-loader.go
@@ -16,6 +16,14 @@ type PixbufLoaderClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	SizePrepared uintptr
+
+	AreaPrepared uintptr
+
+	AreaUpdated uintptr
+
+	Closed uintptr
 }
 
 func (x *PixbufLoaderClass) GoPointer() uintptr {
diff --git a/v4/gio/gaction.go b/v4/gio/gaction.go
index c64d0e8ee4f5914cba108410e49a53988159d537..54b37a84494d83901d22f4a8bf7391c30dad0d9d 100644
--- a/v4/gio/gaction.go
+++ b/v4/gio/gaction.go
@@ -16,6 +16,22 @@ type ActionInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetName uintptr
+
+	GetParameterType uintptr
+
+	GetStateType uintptr
+
+	GetStateHint uintptr
+
+	GetEnabled uintptr
+
+	GetState uintptr
+
+	ChangeState uintptr
+
+	Activate uintptr
 }
 
 func (x *ActionInterface) GoPointer() uintptr {
diff --git a/v4/gio/gactiongroup.go b/v4/gio/gactiongroup.go
index 5756657803b9388bd6247aa7bb8127121c122c09..ac04c7e573c456823e8463d06bf3f33b064a855f 100644
--- a/v4/gio/gactiongroup.go
+++ b/v4/gio/gactiongroup.go
@@ -16,6 +16,34 @@ type ActionGroupInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	HasAction uintptr
+
+	ListActions uintptr
+
+	GetActionEnabled uintptr
+
+	GetActionParameterType uintptr
+
+	GetActionStateType uintptr
+
+	GetActionStateHint uintptr
+
+	GetActionState uintptr
+
+	ChangeActionState uintptr
+
+	ActivateAction uintptr
+
+	ActionAdded uintptr
+
+	ActionRemoved uintptr
+
+	ActionEnabledChanged uintptr
+
+	ActionStateChanged uintptr
+
+	QueryAction uintptr
 }
 
 func (x *ActionGroupInterface) GoPointer() uintptr {
diff --git a/v4/gio/gactionmap.go b/v4/gio/gactionmap.go
index 0766bee0e48fef73b275c1db95736765973801bc..9551c6b9514ade764e51a61e73c7cffd7e316f1a 100644
--- a/v4/gio/gactionmap.go
+++ b/v4/gio/gactionmap.go
@@ -26,9 +26,13 @@ 	_ structs.HostLayout
 
 	Name uintptr
 
+	Activate uintptr
+
 	ParameterType uintptr
 
 	State uintptr
+
+	ChangeState uintptr
 
 	Padding [3]uint
 }
@@ -42,6 +46,12 @@ type ActionMapInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	LookupAction uintptr
+
+	AddAction uintptr
+
+	RemoveAction uintptr
 }
 
 func (x *ActionMapInterface) GoPointer() uintptr {
diff --git a/v4/gio/gappinfo.go b/v4/gio/gappinfo.go
index 21d81909bdee5dcaaeca64418e44a2a9dd0b6f72..7998f9ccc37035bf72623c5672ce2fccd440ee40 100644
--- a/v4/gio/gappinfo.go
+++ b/v4/gio/gappinfo.go
@@ -17,6 +17,56 @@ type AppInfoIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Dup uintptr
+
+	Equal uintptr
+
+	GetId uintptr
+
+	GetName uintptr
+
+	GetDescription uintptr
+
+	GetExecutable uintptr
+
+	GetIcon uintptr
+
+	Launch uintptr
+
+	SupportsUris uintptr
+
+	SupportsFiles uintptr
+
+	LaunchUris uintptr
+
+	ShouldShow uintptr
+
+	SetAsDefaultForType uintptr
+
+	SetAsDefaultForExtension uintptr
+
+	AddSupportsType uintptr
+
+	CanRemoveSupportsType uintptr
+
+	RemoveSupportsType uintptr
+
+	CanDelete uintptr
+
+	DoDelete uintptr
+
+	GetCommandline uintptr
+
+	GetDisplayName uintptr
+
+	SetAsLastUsedForType uintptr
+
+	GetSupportedTypes uintptr
+
+	LaunchUrisAsync uintptr
+
+	LaunchUrisFinish uintptr
 }
 
 func (x *AppInfoIface) GoPointer() uintptr {
@@ -27,6 +77,22 @@ type AppLaunchContextClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetDisplay uintptr
+
+	GetStartupNotifyId uintptr
+
+	LaunchFailed uintptr
+
+	Launched uintptr
+
+	LaunchStarted uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
 }
 
 func (x *AppLaunchContextClass) GoPointer() uintptr {
diff --git a/v4/gio/gapplication.go b/v4/gio/gapplication.go
index cc3e1e8ebecda915afdee7b02026d8205b3d210c..9df82e172a0ae742d1f9ee5d8031b792f3deb20a 100644
--- a/v4/gio/gapplication.go
+++ b/v4/gio/gapplication.go
@@ -18,6 +18,36 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Startup uintptr
+
+	Activate uintptr
+
+	Open uintptr
+
+	CommandLine uintptr
+
+	LocalCommandLine uintptr
+
+	BeforeEmit uintptr
+
+	AfterEmit uintptr
+
+	AddPlatformData uintptr
+
+	QuitMainloop uintptr
+
+	RunMainloop uintptr
+
+	Shutdown uintptr
+
+	DbusRegister uintptr
+
+	DbusUnregister uintptr
+
+	HandleLocalOptions uintptr
+
+	NameLost uintptr
+
 	Padding [7]uintptr
 }
 
diff --git a/v4/gio/gapplicationcommandline.go b/v4/gio/gapplicationcommandline.go
index deb22baad3256c5030af21840090cb7ac76260ff..a50ef6a2aaacacba0fee16760d492d036125d736 100644
--- a/v4/gio/gapplicationcommandline.go
+++ b/v4/gio/gapplicationcommandline.go
@@ -19,6 +19,12 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	PrintLiteral uintptr
+
+	PrinterrLiteral uintptr
+
+	GetStdin uintptr
+
 	Padding [11]uintptr
 }
 
diff --git a/v4/gio/gasyncinitable.go b/v4/gio/gasyncinitable.go
index facde86d5c6f1be165f21e799f01aabaa3735474..b6edb83774daba786ce14f471fcbf47c23ef4f87 100644
--- a/v4/gio/gasyncinitable.go
+++ b/v4/gio/gasyncinitable.go
@@ -18,6 +18,10 @@ type AsyncInitableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	InitAsync uintptr
+
+	InitFinish uintptr
 }
 
 func (x *AsyncInitableIface) GoPointer() uintptr {
diff --git a/v4/gio/gasyncresult.go b/v4/gio/gasyncresult.go
index 593ead7978bf2ce220be8ffe10660e1ed8e42bcc..f6e3a3f1a7fef6e7df4b638ac4ddeb97665f44a1 100644
--- a/v4/gio/gasyncresult.go
+++ b/v4/gio/gasyncresult.go
@@ -17,6 +17,12 @@ type AsyncResultIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetUserData uintptr
+
+	GetSourceObject uintptr
+
+	IsTagged uintptr
 }
 
 func (x *AsyncResultIface) GoPointer() uintptr {
diff --git a/v4/gio/gbufferedinputstream.go b/v4/gio/gbufferedinputstream.go
index 1876a237473b7e34d040959b746da5f4a2ccb17e..ffc32ca954a0c2b6206b5d24ca5af972d1936642 100644
--- a/v4/gio/gbufferedinputstream.go
+++ b/v4/gio/gbufferedinputstream.go
@@ -15,6 +15,22 @@ type BufferedInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Fill uintptr
+
+	FillAsync uintptr
+
+	FillFinish uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *BufferedInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gbufferedoutputstream.go b/v4/gio/gbufferedoutputstream.go
index 602955a8ea256123283ffb015e2db505eae2bcae..09bf711fb6785e8fb712717b3dd7368692a1473a 100644
--- a/v4/gio/gbufferedoutputstream.go
+++ b/v4/gio/gbufferedoutputstream.go
@@ -15,6 +15,10 @@ type BufferedOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
 }
 
 func (x *BufferedOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gcancellable.go b/v4/gio/gcancellable.go
index df81103b7a5fa1a2cd5dcde3f4e71ef056b67200..5714be00c7301caa13707d8af1c9121f81877314 100644
--- a/v4/gio/gcancellable.go
+++ b/v4/gio/gcancellable.go
@@ -16,6 +16,18 @@ type CancellableClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Cancelled uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *CancellableClass) GoPointer() uintptr {
diff --git a/v4/gio/gconverter.go b/v4/gio/gconverter.go
index 6cbf17e04af7827bfc0d449b3e32c326355031a8..4ae8fdea74633e91d71240efe7d11b003875f98d 100644
--- a/v4/gio/gconverter.go
+++ b/v4/gio/gconverter.go
@@ -18,6 +18,10 @@ type ConverterIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Convert uintptr
+
+	Reset uintptr
 }
 
 func (x *ConverterIface) GoPointer() uintptr {
diff --git a/v4/gio/gconverterinputstream.go b/v4/gio/gconverterinputstream.go
index b5b22aceff092547a5ad8bf57cf455ea24dd913c..a08dfd4298c73013b96c272e95ae5981e1397e16 100644
--- a/v4/gio/gconverterinputstream.go
+++ b/v4/gio/gconverterinputstream.go
@@ -16,6 +16,16 @@ type ConverterInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *ConverterInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gconverteroutputstream.go b/v4/gio/gconverteroutputstream.go
index adafed5d3997faa3175829d090bf698fe804b64f..0e8b6a66ae333b89212680107f8d3abc9605fc14 100644
--- a/v4/gio/gconverteroutputstream.go
+++ b/v4/gio/gconverteroutputstream.go
@@ -16,6 +16,16 @@ type ConverterOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *ConverterOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gdatagrambased.go b/v4/gio/gdatagrambased.go
index dd23ca070a96699770ebd5e0fd7214bc4aec5f30..d1c7fae2916c3edf389402676eacaf389efbe2a7 100644
--- a/v4/gio/gdatagrambased.go
+++ b/v4/gio/gdatagrambased.go
@@ -20,6 +20,16 @@ type DatagramBasedInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	ReceiveMessages uintptr
+
+	SendMessages uintptr
+
+	CreateSource uintptr
+
+	ConditionCheck uintptr
+
+	ConditionWait uintptr
 }
 
 func (x *DatagramBasedInterface) GoPointer() uintptr {
diff --git a/v4/gio/gdatainputstream.go b/v4/gio/gdatainputstream.go
index abacd173ada647e416b5be113ac24721eb240a3e..34d56715b0f3085ea56b84e4139d3e318ee1c7f9 100644
--- a/v4/gio/gdatainputstream.go
+++ b/v4/gio/gdatainputstream.go
@@ -15,6 +15,16 @@ type DataInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *DataInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gdataoutputstream.go b/v4/gio/gdataoutputstream.go
index ef1f3c1bac6f452030b8f82361ac2e872ba229fc..4c6a52dad358e18058fbac2449981d6d69eff3e7 100644
--- a/v4/gio/gdataoutputstream.go
+++ b/v4/gio/gdataoutputstream.go
@@ -15,6 +15,16 @@ type DataOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *DataOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gdbusinterface.go b/v4/gio/gdbusinterface.go
index bad163681cbc5adc05032dd92d4931523e1f7c6e..275eabed954f76bea038778ea49325cb9634aa22 100644
--- a/v4/gio/gdbusinterface.go
+++ b/v4/gio/gdbusinterface.go
@@ -16,6 +16,14 @@ type DBusInterfaceIface struct {
 	_ structs.HostLayout
 
 	ParentIface uintptr
+
+	GetInfo uintptr
+
+	GetObject uintptr
+
+	SetObject uintptr
+
+	DupObject uintptr
 }
 
 func (x *DBusInterfaceIface) GoPointer() uintptr {
diff --git a/v4/gio/gdbusinterfaceskeleton.go b/v4/gio/gdbusinterfaceskeleton.go
index 3e6791427e5f1cd94057bc6970227f90ff0e0d31..aa073de8283370bc715ca0a546639474e674cdbf 100644
--- a/v4/gio/gdbusinterfaceskeleton.go
+++ b/v4/gio/gdbusinterfaceskeleton.go
@@ -18,7 +18,17 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GetInfo uintptr
+
+	GetVtable uintptr
+
+	GetProperties uintptr
+
+	Flush uintptr
+
 	VfuncPadding [8]uintptr
+
+	GAuthorizeMethod uintptr
 
 	SignalPadding [8]uintptr
 }
diff --git a/v4/gio/gdbusobject.go b/v4/gio/gdbusobject.go
index 4fd10f8dd76ea91656661e67674ad101099a22c2..468afd351946bb4f8f7e76376a6a3cabe6ca8e39 100644
--- a/v4/gio/gdbusobject.go
+++ b/v4/gio/gdbusobject.go
@@ -16,6 +16,16 @@ type DBusObjectIface struct {
 	_ structs.HostLayout
 
 	ParentIface uintptr
+
+	GetObjectPath uintptr
+
+	GetInterfaces uintptr
+
+	GetInterface uintptr
+
+	InterfaceAdded uintptr
+
+	InterfaceRemoved uintptr
 }
 
 func (x *DBusObjectIface) GoPointer() uintptr {
diff --git a/v4/gio/gdbusobjectmanager.go b/v4/gio/gdbusobjectmanager.go
index 5e9939d462cf1ba390ccc44255d613a7f742e7f8..bde34d806c1043553a6106022653857c1537cba3 100644
--- a/v4/gio/gdbusobjectmanager.go
+++ b/v4/gio/gdbusobjectmanager.go
@@ -16,6 +16,22 @@ type DBusObjectManagerIface struct {
 	_ structs.HostLayout
 
 	ParentIface uintptr
+
+	GetObjectPath uintptr
+
+	GetObjects uintptr
+
+	GetObject uintptr
+
+	GetInterface uintptr
+
+	ObjectAdded uintptr
+
+	ObjectRemoved uintptr
+
+	InterfaceAdded uintptr
+
+	InterfaceRemoved uintptr
 }
 
 func (x *DBusObjectManagerIface) GoPointer() uintptr {
diff --git a/v4/gio/gdbusobjectmanagerclient.go b/v4/gio/gdbusobjectmanagerclient.go
index e6a0745bbefb0f5ce313eac9aff4d859f8dad317..64dd008dc3a7b6f985e41ccdc68daf9c41bbb23d 100644
--- a/v4/gio/gdbusobjectmanagerclient.go
+++ b/v4/gio/gdbusobjectmanagerclient.go
@@ -18,6 +18,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	InterfaceProxySignal uintptr
+
+	InterfaceProxyPropertiesChanged uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gio/gdbusobjectskeleton.go b/v4/gio/gdbusobjectskeleton.go
index 83702fcf5e2947e2bdb8a113fcfc0ed13b6e8cee..7d0b35093e87ebc65f11132bb47fceec4fa1ac8b 100644
--- a/v4/gio/gdbusobjectskeleton.go
+++ b/v4/gio/gdbusobjectskeleton.go
@@ -18,6 +18,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	AuthorizeMethod uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gio/gdbusproxy.go b/v4/gio/gdbusproxy.go
index bae37b659943173fde7aef9863130076c7a7dca6..3577371c31a976d9ec8cf24c7ad8a41616715663 100644
--- a/v4/gio/gdbusproxy.go
+++ b/v4/gio/gdbusproxy.go
@@ -18,6 +18,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GPropertiesChanged uintptr
+
+	GSignal uintptr
+
 	Padding [32]uintptr
 }
 
diff --git a/v4/gio/gdebugcontrollerdbus.go b/v4/gio/gdebugcontrollerdbus.go
index f440d699a20eb53358a09352c3645ca55be78e60..776cb5d018d302d21fc6442b6f80612db63e5533 100644
--- a/v4/gio/gdebugcontrollerdbus.go
+++ b/v4/gio/gdebugcontrollerdbus.go
@@ -18,6 +18,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Authorize uintptr
+
 	Padding [12]uintptr
 }
 
diff --git a/v4/gio/gdesktopappinfo.go b/v4/gio/gdesktopappinfo.go
index ea19299666cc0205ef2bcc22267d3f6d0ec3a04d..278f23aa97cab21e351e40a9780bfef2ff9aa8a7 100644
--- a/v4/gio/gdesktopappinfo.go
+++ b/v4/gio/gdesktopappinfo.go
@@ -33,6 +33,8 @@ type DesktopAppInfoLookupIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetDefaultForUriScheme uintptr
 }
 
 func (x *DesktopAppInfoLookupIface) GoPointer() uintptr {
diff --git a/v4/gio/gdrive.go b/v4/gio/gdrive.go
index 270d0db8a67bb4b06c0439a2ec2e2033206274e9..1b1a20dcf17c96587e9b94b81d237c7360a582d2 100644
--- a/v4/gio/gdrive.go
+++ b/v4/gio/gdrive.go
@@ -16,6 +16,70 @@ type DriveIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Changed uintptr
+
+	Disconnected uintptr
+
+	EjectButton uintptr
+
+	GetName uintptr
+
+	GetIcon uintptr
+
+	HasVolumes uintptr
+
+	GetVolumes uintptr
+
+	IsMediaRemovable uintptr
+
+	HasMedia uintptr
+
+	IsMediaCheckAutomatic uintptr
+
+	CanEject uintptr
+
+	CanPollForMedia uintptr
+
+	Eject uintptr
+
+	EjectFinish uintptr
+
+	PollForMedia uintptr
+
+	PollForMediaFinish uintptr
+
+	GetIdentifier uintptr
+
+	EnumerateIdentifiers uintptr
+
+	GetStartStopType uintptr
+
+	CanStart uintptr
+
+	CanStartDegraded uintptr
+
+	Start uintptr
+
+	StartFinish uintptr
+
+	CanStop uintptr
+
+	Stop uintptr
+
+	StopFinish uintptr
+
+	StopButton uintptr
+
+	EjectWithOperation uintptr
+
+	EjectWithOperationFinish uintptr
+
+	GetSortKey uintptr
+
+	GetSymbolicIcon uintptr
+
+	IsRemovable uintptr
 }
 
 func (x *DriveIface) GoPointer() uintptr {
diff --git a/v4/gio/gdtlsconnection.go b/v4/gio/gdtlsconnection.go
index 803f668a65f129d683c8f90abd45133216c0a038..d90e171edc113a2e5fe99d7a3f2498b612f11781 100644
--- a/v4/gio/gdtlsconnection.go
+++ b/v4/gio/gdtlsconnection.go
@@ -17,6 +17,26 @@ type DtlsConnectionInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	AcceptCertificate uintptr
+
+	Handshake uintptr
+
+	HandshakeAsync uintptr
+
+	HandshakeFinish uintptr
+
+	Shutdown uintptr
+
+	ShutdownAsync uintptr
+
+	ShutdownFinish uintptr
+
+	SetAdvertisedProtocols uintptr
+
+	GetNegotiatedProtocol uintptr
+
+	GetBindingData uintptr
 }
 
 func (x *DtlsConnectionInterface) GoPointer() uintptr {
diff --git a/v4/gio/gfile.go b/v4/gio/gfile.go
index 2be45c32ea93843d21501f29b78874628c644bfe..804e76ca6e4a705bd891f44d507d17c60213aee3 100644
--- a/v4/gio/gfile.go
+++ b/v4/gio/gfile.go
@@ -18,7 +18,211 @@ 	_ structs.HostLayout
 
 	GIface uintptr
 
+	Dup uintptr
+
+	Hash uintptr
+
+	Equal uintptr
+
+	IsNative uintptr
+
+	HasUriScheme uintptr
+
+	GetUriScheme uintptr
+
+	GetBasename uintptr
+
+	GetPath uintptr
+
+	GetUri uintptr
+
+	GetParseName uintptr
+
+	GetParent uintptr
+
+	PrefixMatches uintptr
+
+	GetRelativePath uintptr
+
+	ResolveRelativePath uintptr
+
+	GetChildForDisplayName uintptr
+
+	EnumerateChildren uintptr
+
+	EnumerateChildrenAsync uintptr
+
+	EnumerateChildrenFinish uintptr
+
+	QueryInfo uintptr
+
+	QueryInfoAsync uintptr
+
+	QueryInfoFinish uintptr
+
+	QueryFilesystemInfo uintptr
+
+	QueryFilesystemInfoAsync uintptr
+
+	QueryFilesystemInfoFinish uintptr
+
+	FindEnclosingMount uintptr
+
+	FindEnclosingMountAsync uintptr
+
+	FindEnclosingMountFinish uintptr
+
+	SetDisplayName uintptr
+
+	SetDisplayNameAsync uintptr
+
+	SetDisplayNameFinish uintptr
+
+	QuerySettableAttributes uintptr
+
+	QuerySettableAttributesAsync uintptr
+
+	QuerySettableAttributesFinish uintptr
+
+	QueryWritableNamespaces uintptr
+
+	QueryWritableNamespacesAsync uintptr
+
+	QueryWritableNamespacesFinish uintptr
+
+	SetAttribute uintptr
+
+	SetAttributesFromInfo uintptr
+
+	SetAttributesAsync uintptr
+
+	SetAttributesFinish uintptr
+
+	ReadFn uintptr
+
+	ReadAsync uintptr
+
+	ReadFinish uintptr
+
+	AppendTo uintptr
+
+	AppendToAsync uintptr
+
+	AppendToFinish uintptr
+
+	Create uintptr
+
+	CreateAsync uintptr
+
+	CreateFinish uintptr
+
+	Replace uintptr
+
+	ReplaceAsync uintptr
+
+	ReplaceFinish uintptr
+
+	DeleteFile uintptr
+
+	DeleteFileAsync uintptr
+
+	DeleteFileFinish uintptr
+
+	Trash uintptr
+
+	TrashAsync uintptr
+
+	TrashFinish uintptr
+
+	MakeDirectory uintptr
+
+	MakeDirectoryAsync uintptr
+
+	MakeDirectoryFinish uintptr
+
+	MakeSymbolicLink uintptr
+
+	MakeSymbolicLinkAsync uintptr
+
+	MakeSymbolicLinkFinish uintptr
+
+	Copy uintptr
+
+	CopyAsync uintptr
+
+	CopyFinish uintptr
+
+	Move uintptr
+
+	MoveAsync uintptr
+
+	MoveFinish uintptr
+
+	MountMountable uintptr
+
+	MountMountableFinish uintptr
+
+	UnmountMountable uintptr
+
+	UnmountMountableFinish uintptr
+
+	EjectMountable uintptr
+
+	EjectMountableFinish uintptr
+
+	MountEnclosingVolume uintptr
+
+	MountEnclosingVolumeFinish uintptr
+
+	MonitorDir uintptr
+
+	MonitorFile uintptr
+
+	OpenReadwrite uintptr
+
+	OpenReadwriteAsync uintptr
+
+	OpenReadwriteFinish uintptr
+
+	CreateReadwrite uintptr
+
+	CreateReadwriteAsync uintptr
+
+	CreateReadwriteFinish uintptr
+
+	ReplaceReadwrite uintptr
+
+	ReplaceReadwriteAsync uintptr
+
+	ReplaceReadwriteFinish uintptr
+
+	StartMountable uintptr
+
+	StartMountableFinish uintptr
+
+	StopMountable uintptr
+
+	StopMountableFinish uintptr
+
 	SupportsThreadContexts bool
+
+	UnmountMountableWithOperation uintptr
+
+	UnmountMountableWithOperationFinish uintptr
+
+	EjectMountableWithOperation uintptr
+
+	EjectMountableWithOperationFinish uintptr
+
+	PollMountable uintptr
+
+	PollMountableFinish uintptr
+
+	MeasureDiskUsage uintptr
+
+	MeasureDiskUsageAsync uintptr
+
+	MeasureDiskUsageFinish uintptr
 }
 
 func (x *FileIface) GoPointer() uintptr {
diff --git a/v4/gio/gfiledescriptorbased.go b/v4/gio/gfiledescriptorbased.go
index 33d33478cc796cb8cdb9abe449c481698ade7026..40af54ec72882c42b6cc3443d3a0a0d794dbf8fa 100644
--- a/v4/gio/gfiledescriptorbased.go
+++ b/v4/gio/gfiledescriptorbased.go
@@ -15,6 +15,8 @@ type FileDescriptorBasedIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetFd uintptr
 }
 
 func (x *FileDescriptorBasedIface) GoPointer() uintptr {
diff --git a/v4/gio/gfileenumerator.go b/v4/gio/gfileenumerator.go
index 031106c4791889ee1b5c025cf0023810ae05613a..eeaedefea4a3044a9ec5f8dbe7676b6ade60ee01 100644
--- a/v4/gio/gfileenumerator.go
+++ b/v4/gio/gfileenumerator.go
@@ -16,6 +16,32 @@ type FileEnumeratorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	NextFile uintptr
+
+	CloseFn uintptr
+
+	NextFilesAsync uintptr
+
+	NextFilesFinish uintptr
+
+	CloseAsync uintptr
+
+	CloseFinish uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
 }
 
 func (x *FileEnumeratorClass) GoPointer() uintptr {
diff --git a/v4/gio/gfileinputstream.go b/v4/gio/gfileinputstream.go
index 5aac48e48a1292545b03aa7ac9fce1c170f24531..6240fffc46b3cdeed5ef31660f739f52c4a105cf 100644
--- a/v4/gio/gfileinputstream.go
+++ b/v4/gio/gfileinputstream.go
@@ -15,6 +15,28 @@ type FileInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Tell uintptr
+
+	CanSeek uintptr
+
+	Seek uintptr
+
+	QueryInfo uintptr
+
+	QueryInfoAsync uintptr
+
+	QueryInfoFinish uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *FileInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gfileiostream.go b/v4/gio/gfileiostream.go
index e902f4d2daf6ff11d3ebad8db71eb92efaf8821b..266ea95cf101d4113dfad04c3868bbe122a90fff 100644
--- a/v4/gio/gfileiostream.go
+++ b/v4/gio/gfileiostream.go
@@ -15,6 +15,34 @@ type FileIOStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Tell uintptr
+
+	CanSeek uintptr
+
+	Seek uintptr
+
+	CanTruncate uintptr
+
+	TruncateFn uintptr
+
+	QueryInfo uintptr
+
+	QueryInfoAsync uintptr
+
+	QueryInfoFinish uintptr
+
+	GetEtag uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *FileIOStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gfilemonitor.go b/v4/gio/gfilemonitor.go
index ea68a7d87466bd800343b3a7aa9dd65dd48f735b..6bd189416176266616081b5f8fa99bbd3421f62a 100644
--- a/v4/gio/gfilemonitor.go
+++ b/v4/gio/gfilemonitor.go
@@ -16,6 +16,20 @@ type FileMonitorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Changed uintptr
+
+	Cancel uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *FileMonitorClass) GoPointer() uintptr {
diff --git a/v4/gio/gfilenamecompleter.go b/v4/gio/gfilenamecompleter.go
index 19cd1733e7914d79f31b8c6d14521149f3b0394c..aeb38b6385c726ea3a97aba1253957684e918dd5 100644
--- a/v4/gio/gfilenamecompleter.go
+++ b/v4/gio/gfilenamecompleter.go
@@ -16,6 +16,14 @@ type FilenameCompleterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GotCompletionData uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
 }
 
 func (x *FilenameCompleterClass) GoPointer() uintptr {
diff --git a/v4/gio/gfileoutputstream.go b/v4/gio/gfileoutputstream.go
index 9e57ec5eb6313df8f9b50e6c4d5a2f9556f30781..2eae7fe6c30625ccf31c6b3c4601d6f7e25d6d50 100644
--- a/v4/gio/gfileoutputstream.go
+++ b/v4/gio/gfileoutputstream.go
@@ -15,6 +15,34 @@ type FileOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Tell uintptr
+
+	CanSeek uintptr
+
+	Seek uintptr
+
+	CanTruncate uintptr
+
+	TruncateFn uintptr
+
+	QueryInfo uintptr
+
+	QueryInfoAsync uintptr
+
+	QueryInfoFinish uintptr
+
+	GetEtag uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *FileOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gfilterinputstream.go b/v4/gio/gfilterinputstream.go
index 91460d7e896702d07a3e1fd9cf80e3ac6ab19412..41e6f572ddd30aabefdfa7c53b6c5a137a5ae522 100644
--- a/v4/gio/gfilterinputstream.go
+++ b/v4/gio/gfilterinputstream.go
@@ -15,6 +15,12 @@ type FilterInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
 }
 
 func (x *FilterInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gfilteroutputstream.go b/v4/gio/gfilteroutputstream.go
index ef80ee9976243ca12c487b8837537533fa92851f..6fc924f81de4932596b46db54c6681e543e1b99c 100644
--- a/v4/gio/gfilteroutputstream.go
+++ b/v4/gio/gfilteroutputstream.go
@@ -15,6 +15,12 @@ type FilterOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
 }
 
 func (x *FilterOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gicon.go b/v4/gio/gicon.go
index 76b2a42065058bc177c75bb9ef6f1ff89ca163f2..7040547bdfc9b1b005465b85a0e71aa1d79f5e9a 100644
--- a/v4/gio/gicon.go
+++ b/v4/gio/gicon.go
@@ -18,6 +18,16 @@ type IconIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Hash uintptr
+
+	Equal uintptr
+
+	ToTokens uintptr
+
+	FromTokens uintptr
+
+	Serialize uintptr
 }
 
 func (x *IconIface) GoPointer() uintptr {
diff --git a/v4/gio/ginetaddress.go b/v4/gio/ginetaddress.go
index de4ab45e70b086d18e9d8e48d5ba5ddd01dd1e9a..f9a0c743574a125cb5c74f2cba959b1e543b515c 100644
--- a/v4/gio/ginetaddress.go
+++ b/v4/gio/ginetaddress.go
@@ -15,6 +15,10 @@ type InetAddressClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	ToString uintptr
+
+	ToBytes uintptr
 }
 
 func (x *InetAddressClass) GoPointer() uintptr {
diff --git a/v4/gio/ginitable.go b/v4/gio/ginitable.go
index 51c86e533b2195b8a5c59f15f46de3b986406fd0..0136c1aeda32a9ee8ec24e0e9c6b96ccc27a4776 100644
--- a/v4/gio/ginitable.go
+++ b/v4/gio/ginitable.go
@@ -18,6 +18,8 @@ type InitableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Init uintptr
 }
 
 func (x *InitableIface) GoPointer() uintptr {
diff --git a/v4/gio/ginputstream.go b/v4/gio/ginputstream.go
index 7a757bf604b1e3bda829bb711627a28f1fab3749..5d5a01a795075fa65c5b17b5a0662fcf6e525c32 100644
--- a/v4/gio/ginputstream.go
+++ b/v4/gio/ginputstream.go
@@ -16,6 +16,34 @@ type InputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	ReadFn uintptr
+
+	Skip uintptr
+
+	CloseFn uintptr
+
+	ReadAsync uintptr
+
+	ReadFinish uintptr
+
+	SkipAsync uintptr
+
+	SkipFinish uintptr
+
+	CloseAsync uintptr
+
+	CloseFinish uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *InputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/giostream.go b/v4/gio/giostream.go
index 93d56ac9fb01dc40ab2a4066c1ced3abf8d191fd..6ebe6c8a50873ee396e92ad63bf4fd259f49bd17 100644
--- a/v4/gio/giostream.go
+++ b/v4/gio/giostream.go
@@ -16,6 +16,36 @@ type IOStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetInputStream uintptr
+
+	GetOutputStream uintptr
+
+	CloseFn uintptr
+
+	CloseAsync uintptr
+
+	CloseFinish uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
+
+	GReserved8 uintptr
+
+	GReserved9 uintptr
+
+	GReserved10 uintptr
 }
 
 func (x *IOStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/glistmodel.go b/v4/gio/glistmodel.go
index 3f73c2bfd0cb1f8a81d0570ce1bce0367b911de8..51e9a055c0f94455320f546c8589ea0fba7b790d 100644
--- a/v4/gio/glistmodel.go
+++ b/v4/gio/glistmodel.go
@@ -16,6 +16,12 @@ type ListModelInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetItemType uintptr
+
+	GetNItems uintptr
+
+	GetItem uintptr
 }
 
 func (x *ListModelInterface) GoPointer() uintptr {
diff --git a/v4/gio/gloadableicon.go b/v4/gio/gloadableicon.go
index 65bd77e209cd79b552b6c7dd47d0c123dcb4ae7d..4139af2fbd781aad39e9c4680aff79a66c625ccd 100644
--- a/v4/gio/gloadableicon.go
+++ b/v4/gio/gloadableicon.go
@@ -16,6 +16,12 @@ type LoadableIconIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Load uintptr
+
+	LoadAsync uintptr
+
+	LoadFinish uintptr
 }
 
 func (x *LoadableIconIface) GoPointer() uintptr {
diff --git a/v4/gio/gmemoryinputstream.go b/v4/gio/gmemoryinputstream.go
index ec0e8c484728006f1e04e7dbe3edb196d783a234..8a96e7f06c29d86f8f2f4d514d07a66af6d2bfab 100644
--- a/v4/gio/gmemoryinputstream.go
+++ b/v4/gio/gmemoryinputstream.go
@@ -15,6 +15,16 @@ type MemoryInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *MemoryInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gmemorymonitor.go b/v4/gio/gmemorymonitor.go
index 34e0a4af91ffa6d6116c90396e3da54ca916da91..da9c71d97615c1a69979344b930c311fc11a3a89 100644
--- a/v4/gio/gmemorymonitor.go
+++ b/v4/gio/gmemorymonitor.go
@@ -15,6 +15,8 @@ type MemoryMonitorInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	LowMemoryWarning uintptr
 }
 
 func (x *MemoryMonitorInterface) GoPointer() uintptr {
diff --git a/v4/gio/gmemoryoutputstream.go b/v4/gio/gmemoryoutputstream.go
index d3893972cc06bd179f1f430c9ca6f544598d77bc..2be067a2765c126f7a36fa399341867ecd8a4bca 100644
--- a/v4/gio/gmemoryoutputstream.go
+++ b/v4/gio/gmemoryoutputstream.go
@@ -21,6 +21,16 @@ type MemoryOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *MemoryOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gmenumodel.go b/v4/gio/gmenumodel.go
index 3c48e0e01cbaa2f9c0f7ece60553fdd6cd2f805a..21050d6e8eac1425cfe336673cc799a7ae62ee0e 100644
--- a/v4/gio/gmenumodel.go
+++ b/v4/gio/gmenumodel.go
@@ -16,6 +16,8 @@ type MenuAttributeIterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetNext uintptr
 }
 
 func (x *MenuAttributeIterClass) GoPointer() uintptr {
@@ -34,6 +36,8 @@ type MenuLinkIterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetNext uintptr
 }
 
 func (x *MenuLinkIterClass) GoPointer() uintptr {
@@ -52,6 +56,22 @@ type MenuModelClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	IsMutable uintptr
+
+	GetNItems uintptr
+
+	GetItemAttributes uintptr
+
+	IterateItemAttributes uintptr
+
+	GetItemAttributeValue uintptr
+
+	GetItemLinks uintptr
+
+	IterateItemLinks uintptr
+
+	GetItemLink uintptr
 }
 
 func (x *MenuModelClass) GoPointer() uintptr {
diff --git a/v4/gio/gmount.go b/v4/gio/gmount.go
index fde30bce599263e7f7434fd6720cfb50ebb2d85d..ecc6c754b5e35434ebab050379595f361909e170 100644
--- a/v4/gio/gmount.go
+++ b/v4/gio/gmount.go
@@ -16,6 +16,60 @@ type MountIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Changed uintptr
+
+	Unmounted uintptr
+
+	GetRoot uintptr
+
+	GetName uintptr
+
+	GetIcon uintptr
+
+	GetUuid uintptr
+
+	GetVolume uintptr
+
+	GetDrive uintptr
+
+	CanUnmount uintptr
+
+	CanEject uintptr
+
+	Unmount uintptr
+
+	UnmountFinish uintptr
+
+	Eject uintptr
+
+	EjectFinish uintptr
+
+	Remount uintptr
+
+	RemountFinish uintptr
+
+	GuessContentType uintptr
+
+	GuessContentTypeFinish uintptr
+
+	GuessContentTypeSync uintptr
+
+	PreUnmount uintptr
+
+	UnmountWithOperation uintptr
+
+	UnmountWithOperationFinish uintptr
+
+	EjectWithOperation uintptr
+
+	EjectWithOperationFinish uintptr
+
+	GetDefaultLocation uintptr
+
+	GetSortKey uintptr
+
+	GetSymbolicIcon uintptr
 }
 
 func (x *MountIface) GoPointer() uintptr {
diff --git a/v4/gio/gmountoperation.go b/v4/gio/gmountoperation.go
index fd4e953e264318161d49209713806884614b706d..ae2eb6b510ca21efc0e6914e430e473fd44f80a9 100644
--- a/v4/gio/gmountoperation.go
+++ b/v4/gio/gmountoperation.go
@@ -16,6 +16,36 @@ type MountOperationClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	AskPassword uintptr
+
+	AskQuestion uintptr
+
+	Reply uintptr
+
+	Aborted uintptr
+
+	ShowProcesses uintptr
+
+	ShowUnmountProgress uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
+
+	GReserved8 uintptr
+
+	GReserved9 uintptr
 }
 
 func (x *MountOperationClass) GoPointer() uintptr {
diff --git a/v4/gio/gnativevolumemonitor.go b/v4/gio/gnativevolumemonitor.go
index 3a1f42e4d580202f99781968d484e8c6aa5590f9..65e7e57153441b9bb52fc84f88e98799ea84ee0f 100644
--- a/v4/gio/gnativevolumemonitor.go
+++ b/v4/gio/gnativevolumemonitor.go
@@ -12,6 +12,8 @@ type NativeVolumeMonitorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetMountForMountPath uintptr
 }
 
 func (x *NativeVolumeMonitorClass) GoPointer() uintptr {
diff --git a/v4/gio/gnetworkmonitor.go b/v4/gio/gnetworkmonitor.go
index 8903a6eee9175eecba7b7222a41e76fdb3c8d7a2..a118a17192e9aaa838242ea217c9b930b8c34b6a 100644
--- a/v4/gio/gnetworkmonitor.go
+++ b/v4/gio/gnetworkmonitor.go
@@ -17,6 +17,14 @@ type NetworkMonitorInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	NetworkChanged uintptr
+
+	CanReach uintptr
+
+	CanReachAsync uintptr
+
+	CanReachFinish uintptr
 }
 
 func (x *NetworkMonitorInterface) GoPointer() uintptr {
diff --git a/v4/gio/goutputstream.go b/v4/gio/goutputstream.go
index 7ddf0f2f6a12deada4fa2cabfda725ce163e5e5e..b05f6bb8e90003982d78af1844debb0ec581691a 100644
--- a/v4/gio/goutputstream.go
+++ b/v4/gio/goutputstream.go
@@ -16,6 +16,46 @@ type OutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	WriteFn uintptr
+
+	Splice uintptr
+
+	Flush uintptr
+
+	CloseFn uintptr
+
+	WriteAsync uintptr
+
+	WriteFinish uintptr
+
+	SpliceAsync uintptr
+
+	SpliceFinish uintptr
+
+	FlushAsync uintptr
+
+	FlushFinish uintptr
+
+	CloseAsync uintptr
+
+	CloseFinish uintptr
+
+	WritevFn uintptr
+
+	WritevAsync uintptr
+
+	WritevFinish uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
+
+	GReserved8 uintptr
 }
 
 func (x *OutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gpermission.go b/v4/gio/gpermission.go
index 1cba2dae07c348eb95a96e9e9ef490ae722aa603..d2019b724834676d35d2b35dd3aed2096385e071 100644
--- a/v4/gio/gpermission.go
+++ b/v4/gio/gpermission.go
@@ -17,6 +17,18 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Acquire uintptr
+
+	AcquireAsync uintptr
+
+	AcquireFinish uintptr
+
+	Release uintptr
+
+	ReleaseAsync uintptr
+
+	ReleaseFinish uintptr
+
 	Reserved [16]uintptr
 }
 
diff --git a/v4/gio/gpollableinputstream.go b/v4/gio/gpollableinputstream.go
index 6be92533d9ec802aec1dc969ef046494f38c7513..2e2c6122272d1c929d9ed911556a13070c3386f1 100644
--- a/v4/gio/gpollableinputstream.go
+++ b/v4/gio/gpollableinputstream.go
@@ -25,6 +25,14 @@ type PollableInputStreamInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	CanPoll uintptr
+
+	IsReadable uintptr
+
+	CreateSource uintptr
+
+	ReadNonblocking uintptr
 }
 
 func (x *PollableInputStreamInterface) GoPointer() uintptr {
diff --git a/v4/gio/gpollableoutputstream.go b/v4/gio/gpollableoutputstream.go
index b7198752c3454c7ab15361bd979393e9b3275433..4dfe0b5c7dce9044c568a3fa32e5fe0e16403a02 100644
--- a/v4/gio/gpollableoutputstream.go
+++ b/v4/gio/gpollableoutputstream.go
@@ -31,6 +31,16 @@ type PollableOutputStreamInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	CanPoll uintptr
+
+	IsWritable uintptr
+
+	CreateSource uintptr
+
+	WriteNonblocking uintptr
+
+	WritevNonblocking uintptr
 }
 
 func (x *PollableOutputStreamInterface) GoPointer() uintptr {
diff --git a/v4/gio/gproxy.go b/v4/gio/gproxy.go
index cd80f9c1dea96408dfb1de9b7af8275a82c212a5..6700c554b2e2211fcebaa5ac3bf9a81796f0bc5c 100644
--- a/v4/gio/gproxy.go
+++ b/v4/gio/gproxy.go
@@ -16,6 +16,14 @@ type ProxyInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Connect uintptr
+
+	ConnectAsync uintptr
+
+	ConnectFinish uintptr
+
+	SupportsHostname uintptr
 }
 
 func (x *ProxyInterface) GoPointer() uintptr {
diff --git a/v4/gio/gproxyaddressenumerator.go b/v4/gio/gproxyaddressenumerator.go
index 91cb1648deded0d0e137b9e6bb6fd8ba6d8d11c6..04ec56ff31cdebb02ce69cdb8f2686889e0ef6a9 100644
--- a/v4/gio/gproxyaddressenumerator.go
+++ b/v4/gio/gproxyaddressenumerator.go
@@ -13,6 +13,20 @@ type ProxyAddressEnumeratorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
 }
 
 func (x *ProxyAddressEnumeratorClass) GoPointer() uintptr {
diff --git a/v4/gio/gproxyresolver.go b/v4/gio/gproxyresolver.go
index 1ead1a3c6563ddb93094d228bea09bcd533f7393..5aac8361b0297005ad963a7660062ba0aab93984 100644
--- a/v4/gio/gproxyresolver.go
+++ b/v4/gio/gproxyresolver.go
@@ -17,6 +17,14 @@ type ProxyResolverInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	IsSupported uintptr
+
+	Lookup uintptr
+
+	LookupAsync uintptr
+
+	LookupFinish uintptr
 }
 
 func (x *ProxyResolverInterface) GoPointer() uintptr {
diff --git a/v4/gio/gremoteactiongroup.go b/v4/gio/gremoteactiongroup.go
index 89c43ab5ef1819d78b559df26db38f07b429e0e1..0f6cf41d697e0d1f33f3777ab615464c054e8a5a 100644
--- a/v4/gio/gremoteactiongroup.go
+++ b/v4/gio/gremoteactiongroup.go
@@ -16,6 +16,10 @@ type RemoteActionGroupInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	ActivateActionFull uintptr
+
+	ChangeActionStateFull uintptr
 }
 
 func (x *RemoteActionGroupInterface) GoPointer() uintptr {
diff --git a/v4/gio/gresolver.go b/v4/gio/gresolver.go
index b9437dc84bb14e0f8c22307c65032abfcf58b6cc..24793c6f62166e46bb15fa7aaa8436d2f7c058af 100644
--- a/v4/gio/gresolver.go
+++ b/v4/gio/gresolver.go
@@ -16,6 +16,38 @@ type ResolverClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Reload uintptr
+
+	LookupByName uintptr
+
+	LookupByNameAsync uintptr
+
+	LookupByNameFinish uintptr
+
+	LookupByAddress uintptr
+
+	LookupByAddressAsync uintptr
+
+	LookupByAddressFinish uintptr
+
+	LookupService uintptr
+
+	LookupServiceAsync uintptr
+
+	LookupServiceFinish uintptr
+
+	LookupRecords uintptr
+
+	LookupRecordsAsync uintptr
+
+	LookupRecordsFinish uintptr
+
+	LookupByNameWithFlagsAsync uintptr
+
+	LookupByNameWithFlagsFinish uintptr
+
+	LookupByNameWithFlags uintptr
 }
 
 func (x *ResolverClass) GoPointer() uintptr {
diff --git a/v4/gio/gseekable.go b/v4/gio/gseekable.go
index ae1ca9b0c6016371b7eef911ce007cdef0952c2e..0c94711dddf8b6a9e037af7cf9b52a2b39f23968 100644
--- a/v4/gio/gseekable.go
+++ b/v4/gio/gseekable.go
@@ -16,6 +16,16 @@ type SeekableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Tell uintptr
+
+	CanSeek uintptr
+
+	Seek uintptr
+
+	CanTruncate uintptr
+
+	TruncateFn uintptr
 }
 
 func (x *SeekableIface) GoPointer() uintptr {
diff --git a/v4/gio/gsettings.go b/v4/gio/gsettings.go
index f39009acd72987b433e50d3ba4c792bf5efc50fc..9cb4d9a7f6b3e87d13dbbe4b2cb3305e12c0eb44 100644
--- a/v4/gio/gsettings.go
+++ b/v4/gio/gsettings.go
@@ -38,6 +38,14 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	WritableChanged uintptr
+
+	Changed uintptr
+
+	WritableChangeEvent uintptr
+
+	ChangeEvent uintptr
+
 	Padding [20]uintptr
 }
 
diff --git a/v4/gio/gsettingsbackend.go b/v4/gio/gsettingsbackend.go
index ffe181882a21d76c400768cdb9c05c91e6efe4ee..0052bcf2b96eb41be6cc7cc94e0cf349df888e11 100644
--- a/v4/gio/gsettingsbackend.go
+++ b/v4/gio/gsettingsbackend.go
@@ -18,6 +18,26 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Read uintptr
+
+	GetWritable uintptr
+
+	Write uintptr
+
+	WriteTree uintptr
+
+	Reset uintptr
+
+	Subscribe uintptr
+
+	Unsubscribe uintptr
+
+	Sync uintptr
+
+	GetPermission uintptr
+
+	ReadUserValue uintptr
+
 	Padding [23]uintptr
 }
 
diff --git a/v4/gio/gsimpleproxyresolver.go b/v4/gio/gsimpleproxyresolver.go
index 5c1217daa0c4aaebd08a6dcf53eb40a7d70c6319..4036e8a7348e2ed4996b49a47fde4ac62c92d2fa 100644
--- a/v4/gio/gsimpleproxyresolver.go
+++ b/v4/gio/gsimpleproxyresolver.go
@@ -16,6 +16,16 @@ type SimpleProxyResolverClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *SimpleProxyResolverClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocket.go b/v4/gio/gsocket.go
index fd6cd055c43f5c754ec139799d5ec4a1b96ccd96..d22921541d6cf77883a95d0874a44d95abad8870 100644
--- a/v4/gio/gsocket.go
+++ b/v4/gio/gsocket.go
@@ -16,6 +16,26 @@ type SocketClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
+
+	GReserved7 uintptr
+
+	GReserved8 uintptr
+
+	GReserved9 uintptr
+
+	GReserved10 uintptr
 }
 
 func (x *SocketClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketaddress.go b/v4/gio/gsocketaddress.go
index 3f5ef41fc0dece394f7b0ed0befb626ea1baa511..3271588ea5b99599d8b03b108fc2eef55eaec00f 100644
--- a/v4/gio/gsocketaddress.go
+++ b/v4/gio/gsocketaddress.go
@@ -16,6 +16,12 @@ type SocketAddressClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetFamily uintptr
+
+	GetNativeSize uintptr
+
+	ToNative uintptr
 }
 
 func (x *SocketAddressClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketaddressenumerator.go b/v4/gio/gsocketaddressenumerator.go
index ad6a38226cc9a1ab4393ac94d3569acd17a03cb6..27272f8a67b07dde414bc6f33073bbfb9f85ecee 100644
--- a/v4/gio/gsocketaddressenumerator.go
+++ b/v4/gio/gsocketaddressenumerator.go
@@ -17,6 +17,12 @@ type SocketAddressEnumeratorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Next uintptr
+
+	NextAsync uintptr
+
+	NextFinish uintptr
 }
 
 func (x *SocketAddressEnumeratorClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketclient.go b/v4/gio/gsocketclient.go
index 3be0abb97be4e51202a4c14912e0670313ac3790..abfe876c27e95a0fcb030070b63c974d416e4741 100644
--- a/v4/gio/gsocketclient.go
+++ b/v4/gio/gsocketclient.go
@@ -16,6 +16,16 @@ type SocketClientClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Event uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
 }
 
 func (x *SocketClientClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketconnectable.go b/v4/gio/gsocketconnectable.go
index bbd459f0b3ba8882fb56352b860fbe9cbc31e79e..fa8a0984cfe19922e0c3ae669861020b72e7dccd 100644
--- a/v4/gio/gsocketconnectable.go
+++ b/v4/gio/gsocketconnectable.go
@@ -16,6 +16,12 @@ type SocketConnectableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Enumerate uintptr
+
+	ProxyEnumerate uintptr
+
+	ToString uintptr
 }
 
 func (x *SocketConnectableIface) GoPointer() uintptr {
diff --git a/v4/gio/gsocketconnection.go b/v4/gio/gsocketconnection.go
index 665c2017176beb08a20864d10272cd9b788c165f..b081c1c898b90515f48d4d07bbaae61c595474b7 100644
--- a/v4/gio/gsocketconnection.go
+++ b/v4/gio/gsocketconnection.go
@@ -16,6 +16,18 @@ type SocketConnectionClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
 }
 
 func (x *SocketConnectionClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketcontrolmessage.go b/v4/gio/gsocketcontrolmessage.go
index 271a78dbe5cc1e957070a44a7cef3eb6d7aab766..967011f028d65b331441ba264e4e03610aa61c0f 100644
--- a/v4/gio/gsocketcontrolmessage.go
+++ b/v4/gio/gsocketcontrolmessage.go
@@ -16,6 +16,26 @@ type SocketControlMessageClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetSize uintptr
+
+	GetLevel uintptr
+
+	GetType uintptr
+
+	Serialize uintptr
+
+	Deserialize uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *SocketControlMessageClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketlistener.go b/v4/gio/gsocketlistener.go
index 376d289b0590456cbf2581498de3a3fb5f1fc7a1..4beac74cc52dd103add79b91e730a58b648a42a7 100644
--- a/v4/gio/gsocketlistener.go
+++ b/v4/gio/gsocketlistener.go
@@ -17,6 +17,20 @@ type SocketListenerClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Changed uintptr
+
+	Event uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
 }
 
 func (x *SocketListenerClass) GoPointer() uintptr {
diff --git a/v4/gio/gsocketservice.go b/v4/gio/gsocketservice.go
index 02f2793edb2ef2609c8a5d6a41beac11e7b40c1c..92697b5559e30b31394bed3b1983204824490828 100644
--- a/v4/gio/gsocketservice.go
+++ b/v4/gio/gsocketservice.go
@@ -17,6 +17,20 @@ type SocketServiceClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Incoming uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
 }
 
 func (x *SocketServiceClass) GoPointer() uintptr {
diff --git a/v4/gio/gthreadedsocketservice.go b/v4/gio/gthreadedsocketservice.go
index aea1647f74eb0eaf80e8da0f1e70badb6c7a4e96..36e8f4f29cf127666dede34639a6aebcbb228ac1 100644
--- a/v4/gio/gthreadedsocketservice.go
+++ b/v4/gio/gthreadedsocketservice.go
@@ -16,6 +16,18 @@ type ThreadedSocketServiceClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Run uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *ThreadedSocketServiceClass) GoPointer() uintptr {
diff --git a/v4/gio/gtlsbackend.go b/v4/gio/gtlsbackend.go
index 1fc3e7fead9e7b4764e19c488115dad6250198de..a31cf7aa7add02ebb32be6b27fc0cc3afd0a9106 100644
--- a/v4/gio/gtlsbackend.go
+++ b/v4/gio/gtlsbackend.go
@@ -16,6 +16,24 @@ type TlsBackendInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	SupportsTls uintptr
+
+	GetCertificateType uintptr
+
+	GetClientConnectionType uintptr
+
+	GetServerConnectionType uintptr
+
+	GetFileDatabaseType uintptr
+
+	GetDefaultDatabase uintptr
+
+	SupportsDtls uintptr
+
+	GetDtlsClientConnectionType uintptr
+
+	GetDtlsServerConnectionType uintptr
 }
 
 func (x *TlsBackendInterface) GoPointer() uintptr {
diff --git a/v4/gio/gtlscertificate.go b/v4/gio/gtlscertificate.go
index dfecd8fcfda58edcfe80323cbdb132a68d1b48b6..e529f20e83ee2a4f17a1f1a451801159599e3c2c 100644
--- a/v4/gio/gtlscertificate.go
+++ b/v4/gio/gtlscertificate.go
@@ -17,6 +17,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Verify uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gio/gtlsclientconnection.go b/v4/gio/gtlsclientconnection.go
index 2ef6651116a543b2b28b2611c6301baa97534284..b8f9491b6c67a9267779ced9610bc5556b9f67fe 100644
--- a/v4/gio/gtlsclientconnection.go
+++ b/v4/gio/gtlsclientconnection.go
@@ -17,6 +17,8 @@ type TlsClientConnectionInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	CopySessionState uintptr
 }
 
 func (x *TlsClientConnectionInterface) GoPointer() uintptr {
diff --git a/v4/gio/gtlsconnection.go b/v4/gio/gtlsconnection.go
index c3ad7527259e731101d4c638807e58fc7b911a9d..087dde441f454b2acd80f8c499d007b0a0b6bf81 100644
--- a/v4/gio/gtlsconnection.go
+++ b/v4/gio/gtlsconnection.go
@@ -18,6 +18,18 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	AcceptCertificate uintptr
+
+	Handshake uintptr
+
+	HandshakeAsync uintptr
+
+	HandshakeFinish uintptr
+
+	GetBindingData uintptr
+
+	GetNegotiatedProtocol uintptr
+
 	Padding [6]uintptr
 }
 
diff --git a/v4/gio/gtlsdatabase.go b/v4/gio/gtlsdatabase.go
index a88da1bca462f1d0f51500e6ce4e4496ad7dfc4d..0a4961f8906f177db76b6dfc74d83b950e3e26f3 100644
--- a/v4/gio/gtlsdatabase.go
+++ b/v4/gio/gtlsdatabase.go
@@ -20,6 +20,32 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	VerifyChain uintptr
+
+	VerifyChainAsync uintptr
+
+	VerifyChainFinish uintptr
+
+	CreateCertificateHandle uintptr
+
+	LookupCertificateForHandle uintptr
+
+	LookupCertificateForHandleAsync uintptr
+
+	LookupCertificateForHandleFinish uintptr
+
+	LookupCertificateIssuer uintptr
+
+	LookupCertificateIssuerAsync uintptr
+
+	LookupCertificateIssuerFinish uintptr
+
+	LookupCertificatesIssuedBy uintptr
+
+	LookupCertificatesIssuedByAsync uintptr
+
+	LookupCertificatesIssuedByFinish uintptr
+
 	Padding [16]uintptr
 }
 
diff --git a/v4/gio/gtlsinteraction.go b/v4/gio/gtlsinteraction.go
index 0bf9c583dd2d3f112cc638e621b4ee96935de889..56bbfd64e763d4869d8d6412d4757f326526b0b2 100644
--- a/v4/gio/gtlsinteraction.go
+++ b/v4/gio/gtlsinteraction.go
@@ -31,6 +31,18 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	AskPassword uintptr
+
+	AskPasswordAsync uintptr
+
+	AskPasswordFinish uintptr
+
+	RequestCertificate uintptr
+
+	RequestCertificateAsync uintptr
+
+	RequestCertificateFinish uintptr
+
 	Padding [21]uintptr
 }
 
diff --git a/v4/gio/gtlspassword.go b/v4/gio/gtlspassword.go
index 05f353653629e33b4043fc3704f85b550a77824d..34ee9cc0d35cb11dd244c526fdae647b3de20a59 100644
--- a/v4/gio/gtlspassword.go
+++ b/v4/gio/gtlspassword.go
@@ -18,6 +18,12 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GetValue uintptr
+
+	SetValue uintptr
+
+	GetDefaultWarning uintptr
+
 	Padding [4]uintptr
 }
 
diff --git a/v4/gio/gunixcredentialsmessage.go b/v4/gio/gunixcredentialsmessage.go
index cd9ed78aeb03dbb017df7b6617acd037aa38aabb..590a707d6346251c724a8008cf3d8906f535a497 100644
--- a/v4/gio/gunixcredentialsmessage.go
+++ b/v4/gio/gunixcredentialsmessage.go
@@ -16,6 +16,10 @@ type UnixCredentialsMessageClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
 }
 
 func (x *UnixCredentialsMessageClass) GoPointer() uintptr {
diff --git a/v4/gio/gunixfdlist.go b/v4/gio/gunixfdlist.go
index e85730f827f760ea92e5bdfd865627c45c7e4b93..1276f5d3b7854e2d4a46ce2fa5e2992b84869180 100644
--- a/v4/gio/gunixfdlist.go
+++ b/v4/gio/gunixfdlist.go
@@ -16,6 +16,16 @@ type UnixFDListClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *UnixFDListClass) GoPointer() uintptr {
diff --git a/v4/gio/gunixfdmessage.go b/v4/gio/gunixfdmessage.go
index e477e3fa49d94222afb708ac870206320b549179..1ee96cd810eff75ff799a61e35efc26e017714a5 100644
--- a/v4/gio/gunixfdmessage.go
+++ b/v4/gio/gunixfdmessage.go
@@ -16,6 +16,10 @@ type UnixFDMessageClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
 }
 
 func (x *UnixFDMessageClass) GoPointer() uintptr {
diff --git a/v4/gio/gunixinputstream.go b/v4/gio/gunixinputstream.go
index 09808e86b1622c5bbdda6993ee96424f753c3526..491bf092f165e34cd7c814ae77fe54e0c7be2a81 100644
--- a/v4/gio/gunixinputstream.go
+++ b/v4/gio/gunixinputstream.go
@@ -15,6 +15,16 @@ type UnixInputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *UnixInputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gunixoutputstream.go b/v4/gio/gunixoutputstream.go
index 25e8ac22476598f54337fd854f69a87ca9af9c60..3adcfd34b75cac59770c3117924357bd0413131d 100644
--- a/v4/gio/gunixoutputstream.go
+++ b/v4/gio/gunixoutputstream.go
@@ -15,6 +15,16 @@ type UnixOutputStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
 }
 
 func (x *UnixOutputStreamClass) GoPointer() uintptr {
diff --git a/v4/gio/gvfs.go b/v4/gio/gvfs.go
index 680c87702984f73222c1bc6bf14d65667958d32f..62085761ee50ffd31812c18417932008062703e1 100644
--- a/v4/gio/gvfs.go
+++ b/v4/gio/gvfs.go
@@ -24,6 +24,40 @@ type VfsClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	IsActive uintptr
+
+	GetFileForPath uintptr
+
+	GetFileForUri uintptr
+
+	GetSupportedUriSchemes uintptr
+
+	ParseName uintptr
+
+	LocalFileAddInfo uintptr
+
+	AddWritableNamespaces uintptr
+
+	LocalFileSetAttributes uintptr
+
+	LocalFileRemoved uintptr
+
+	LocalFileMoved uintptr
+
+	DeserializeIcon uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
 }
 
 func (x *VfsClass) GoPointer() uintptr {
diff --git a/v4/gio/gvolume.go b/v4/gio/gvolume.go
index 01e9199830b07ca7230a44abd394c33610796335..10910841ad517c51976a342e33202885e3a052e1 100644
--- a/v4/gio/gvolume.go
+++ b/v4/gio/gvolume.go
@@ -16,6 +16,48 @@ type VolumeIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Changed uintptr
+
+	Removed uintptr
+
+	GetName uintptr
+
+	GetIcon uintptr
+
+	GetUuid uintptr
+
+	GetDrive uintptr
+
+	GetMount uintptr
+
+	CanMount uintptr
+
+	CanEject uintptr
+
+	MountFn uintptr
+
+	MountFinish uintptr
+
+	Eject uintptr
+
+	EjectFinish uintptr
+
+	GetIdentifier uintptr
+
+	EnumerateIdentifiers uintptr
+
+	ShouldAutomount uintptr
+
+	GetActivationRoot uintptr
+
+	EjectWithOperation uintptr
+
+	EjectWithOperationFinish uintptr
+
+	GetSortKey uintptr
+
+	GetSymbolicIcon uintptr
 }
 
 func (x *VolumeIface) GoPointer() uintptr {
diff --git a/v4/gio/gvolumemonitor.go b/v4/gio/gvolumemonitor.go
index 6f36bcf9e7685f0b09e869f91c6c7d0f615a43bd..8cb472157695def66c69c743949b54389e13bd95 100644
--- a/v4/gio/gvolumemonitor.go
+++ b/v4/gio/gvolumemonitor.go
@@ -16,6 +16,56 @@ type VolumeMonitorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	VolumeAdded uintptr
+
+	VolumeRemoved uintptr
+
+	VolumeChanged uintptr
+
+	MountAdded uintptr
+
+	MountRemoved uintptr
+
+	MountPreUnmount uintptr
+
+	MountChanged uintptr
+
+	DriveConnected uintptr
+
+	DriveDisconnected uintptr
+
+	DriveChanged uintptr
+
+	IsSupported uintptr
+
+	GetConnectedDrives uintptr
+
+	GetVolumes uintptr
+
+	GetMounts uintptr
+
+	GetVolumeForUuid uintptr
+
+	GetMountForUuid uintptr
+
+	AdoptOrphanMount uintptr
+
+	DriveEjectButton uintptr
+
+	DriveStopButton uintptr
+
+	GReserved1 uintptr
+
+	GReserved2 uintptr
+
+	GReserved3 uintptr
+
+	GReserved4 uintptr
+
+	GReserved5 uintptr
+
+	GReserved6 uintptr
 }
 
 func (x *VolumeMonitorClass) GoPointer() uintptr {
diff --git a/v4/glib/giochannel.go b/v4/glib/giochannel.go
index c0c1dfa442114742eb994c42590721142ca154d2..cd1d34bf4432535b05a1e592ca616f940484b1f6 100644
--- a/v4/glib/giochannel.go
+++ b/v4/glib/giochannel.go
@@ -559,6 +559,22 @@ // A table of functions used to handle different types of #GIOChannel
 // in a generic way.
 type IOFuncs struct {
 	_ structs.HostLayout
+
+	IoRead uintptr
+
+	IoWrite uintptr
+
+	IoSeek uintptr
+
+	IoClose uintptr
+
+	IoCreateWatch uintptr
+
+	IoFree uintptr
+
+	IoSetFlags uintptr
+
+	IoGetFlags uintptr
 }
 
 func (x *IOFuncs) GoPointer() uintptr {
diff --git a/v4/glib/gmain.go b/v4/glib/gmain.go
index f03a282b00ebccabd39518052c46de035969b8fc..4a41ea885529a75865793272a93cea17d5e37017 100644
--- a/v4/glib/gmain.go
+++ b/v4/glib/gmain.go
@@ -1180,6 +1180,12 @@ // The `GSourceCallbackFuncs` struct contains
 // functions for managing callback objects.
 type SourceCallbackFuncs struct {
 	_ structs.HostLayout
+
+	Ref uintptr
+
+	Unref uintptr
+
+	Get uintptr
 }
 
 func (x *SourceCallbackFuncs) GoPointer() uintptr {
@@ -1208,6 +1214,14 @@ // check function, it tests the results of the poll() call to see if the
 // required condition has been met, and returns %TRUE if so.
 type SourceFuncs struct {
 	_ structs.HostLayout
+
+	Prepare uintptr
+
+	Check uintptr
+
+	Dispatch uintptr
+
+	Finalize uintptr
 
 	ClosureCallback SourceFunc
 
diff --git a/v4/glib/gmarkup.go b/v4/glib/gmarkup.go
index df68d865810e01dff2834261f6fe5d0a6e263dd2..6f348ee95f2ebd7b214cbb3f945f7eed7a813e23 100644
--- a/v4/glib/gmarkup.go
+++ b/v4/glib/gmarkup.go
@@ -336,6 +336,16 @@ // from a callback, g_markup_parse_context_parse() will report that error
 // back to its caller.
 type MarkupParser struct {
 	_ structs.HostLayout
+
+	StartElement uintptr
+
+	EndElement uintptr
+
+	Text uintptr
+
+	Passthrough uintptr
+
+	Error uintptr
 }
 
 func (x *MarkupParser) GoPointer() uintptr {
diff --git a/v4/glib/gmem.go b/v4/glib/gmem.go
index d8151d74edad703d4afc835b79b7e8ff0df6db6b..91c313783b058124560901261f485d4789572c35 100644
--- a/v4/glib/gmem.go
+++ b/v4/glib/gmem.go
@@ -16,6 +16,18 @@ //
 // This functions related to this has been deprecated in 2.46, and no longer work.
 type MemVTable struct {
 	_ structs.HostLayout
+
+	Malloc uintptr
+
+	Realloc uintptr
+
+	Free uintptr
+
+	Calloc uintptr
+
+	TryMalloc uintptr
+
+	TryRealloc uintptr
 }
 
 func (x *MemVTable) GoPointer() uintptr {
diff --git a/v4/gobject/gclosure.go b/v4/gobject/gclosure.go
index dd10d4217c84f2a51cbac7ccd50f79b045c18e42..ddfef33e363959c0454f315cb000a3acd967bfc5 100644
--- a/v4/gobject/gclosure.go
+++ b/v4/gobject/gclosure.go
@@ -113,6 +113,8 @@ 	InMarshal uint
 
 	IsInvalid uint
 
+	Marshal uintptr
+
 	Data uintptr
 
 	Notifiers *ClosureNotifyData
diff --git a/v4/gobject/gobject.go b/v4/gobject/gobject.go
index 7ec0148c4e7df87877a8bf1eb18f671200ca1f77..4c4b2596f0b96353fed8c3bf76bcd47766183a50 100644
--- a/v4/gobject/gobject.go
+++ b/v4/gobject/gobject.go
@@ -47,6 +47,22 @@ 	GTypeClass uintptr
 
 	ConstructProperties *glib.SList
 
+	Constructor uintptr
+
+	SetProperty uintptr
+
+	GetProperty uintptr
+
+	Dispose uintptr
+
+	Finalize uintptr
+
+	DispatchPropertiesChanged uintptr
+
+	Notify uintptr
+
+	Constructed uintptr
+
 	Flags uint
 
 	Pdummy [6]uintptr
@@ -91,6 +107,22 @@
 	GTypeClass uintptr
 
 	ConstructProperties *glib.SList
+
+	Constructor uintptr
+
+	SetProperty uintptr
+
+	GetProperty uintptr
+
+	Dispose uintptr
+
+	Finalize uintptr
+
+	DispatchPropertiesChanged uintptr
+
+	Notify uintptr
+
+	Constructed uintptr
 
 	Flags uint
 
diff --git a/v4/gobject/gparam.go b/v4/gobject/gparam.go
index 2ba3f30a7fe7c358e2bd09b6e86f7904c54ef525..ded78b032ad26f175a603675acc863d338220f9e 100644
--- a/v4/gobject/gparam.go
+++ b/v4/gobject/gparam.go
@@ -21,6 +21,14 @@ 	GTypeClass uintptr
 
 	ValueType types.GType
 
+	Finalize uintptr
+
+	ValueSetDefault uintptr
+
+	ValueValidate uintptr
+
+	ValuesCmp uintptr
+
 	Dummy [4]uintptr
 }
 
@@ -111,7 +119,17 @@ 	InstanceSize uint16
 
 	NPreallocs uint16
 
+	InstanceInit uintptr
+
 	ValueType types.GType
+
+	Finalize uintptr
+
+	ValueSetDefault uintptr
+
+	ValueValidate uintptr
+
+	ValuesCmp uintptr
 }
 
 func (x *ParamSpecTypeInfo) GoPointer() uintptr {
diff --git a/v4/gobject/gtype.go b/v4/gobject/gtype.go
index 0c0a2e8334e65416cf354787091052f9fdc42372..4bac3026c9dccec8dbf963f112945a059c590b30 100644
--- a/v4/gobject/gtype.go
+++ b/v4/gobject/gtype.go
@@ -488,9 +488,21 @@ // implementation, to serve as a container for values of a type.
 type TypeValueTable struct {
 	_ structs.HostLayout
 
+	ValueInit uintptr
+
+	ValueFree uintptr
+
+	ValueCopy uintptr
+
+	ValuePeekPointer uintptr
+
 	CollectFormat uintptr
 
+	CollectValue uintptr
+
 	LcopyFormat uintptr
+
+	LcopyValue uintptr
 }
 
 func (x *TypeValueTable) GoPointer() uintptr {
diff --git a/v4/gobject/gtypemodule.go b/v4/gobject/gtypemodule.go
index f3c024b516ba698eda116a517906c5367b32f5c5..fa9f4a5ed311d04c79f7ece5495ebe5549299a14 100644
--- a/v4/gobject/gtypemodule.go
+++ b/v4/gobject/gtypemodule.go
@@ -16,6 +16,18 @@ type TypeModuleClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Load uintptr
+
+	Unload uintptr
+
+	Reserved1 uintptr
+
+	Reserved2 uintptr
+
+	Reserved3 uintptr
+
+	Reserved4 uintptr
 }
 
 func (x *TypeModuleClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkactionable.go b/v4/gtk/gtkactionable.go
index a611916c146770cb699e6ddf61eb7532a6cd2686..306e2fe46c2e8fca079c7179bd7399b2a047764d 100644
--- a/v4/gtk/gtkactionable.go
+++ b/v4/gtk/gtkactionable.go
@@ -16,6 +16,14 @@ type ActionableInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetActionName uintptr
+
+	SetActionName uintptr
+
+	GetActionTargetValue uintptr
+
+	SetActionTargetValue uintptr
 }
 
 func (x *ActionableInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkadjustment.go b/v4/gtk/gtkadjustment.go
index d4973ecdc0ca43dd8e4fee0d6fff26f2dc79f6d3..30b692caf5b627bbb3b5409a97a5fe37affa0d6d 100644
--- a/v4/gtk/gtkadjustment.go
+++ b/v4/gtk/gtkadjustment.go
@@ -16,6 +16,18 @@ type AdjustmentClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Changed uintptr
+
+	ValueChanged uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *AdjustmentClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkapplication.go b/v4/gtk/gtkapplication.go
index 7985d9957be047ea6313b7a2fb68b460bab247ef..bcab43197772d929c3c46f060bca0832af88f99b 100644
--- a/v4/gtk/gtkapplication.go
+++ b/v4/gtk/gtkapplication.go
@@ -18,6 +18,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	WindowAdded uintptr
+
+	WindowRemoved uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkbuildable.go b/v4/gtk/gtkbuildable.go
index 15ee9748aec352b6f12514ac213cfb116feaf381..77a046a45fc19feb183f6dcf44bde18e39bc14c0 100644
--- a/v4/gtk/gtkbuildable.go
+++ b/v4/gtk/gtkbuildable.go
@@ -17,6 +17,26 @@ type BuildableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	SetId uintptr
+
+	GetId uintptr
+
+	AddChild uintptr
+
+	SetBuildableProperty uintptr
+
+	ConstructChild uintptr
+
+	CustomTagStart uintptr
+
+	CustomTagEnd uintptr
+
+	CustomFinished uintptr
+
+	ParserFinished uintptr
+
+	GetInternalChild uintptr
 }
 
 func (x *BuildableIface) GoPointer() uintptr {
@@ -137,6 +157,14 @@
 // A sub-parser for `GtkBuildable` implementations.
 type BuildableParser struct {
 	_ structs.HostLayout
+
+	StartElement uintptr
+
+	EndElement uintptr
+
+	Text uintptr
+
+	Error uintptr
 
 	Padding [4]uintptr
 }
diff --git a/v4/gtk/gtkbuilderscope.go b/v4/gtk/gtkbuilderscope.go
index b104e5b430bcb5534b38ec554af80331d17f6b63..a1a11c65c1f0608f31f8e7de279ca4acf2ceba9b 100644
--- a/v4/gtk/gtkbuilderscope.go
+++ b/v4/gtk/gtkbuilderscope.go
@@ -29,6 +29,12 @@ type BuilderScopeInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	GetTypeFromName uintptr
+
+	GetTypeFromFunction uintptr
+
+	CreateClosure uintptr
 }
 
 func (x *BuilderScopeInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkbutton.go b/v4/gtk/gtkbutton.go
index cdbf1f692a8fda7c5933289ff334bfb4bf5dfac2..6d2b48b999667024ef4f3263d6b480fdbe95969f 100644
--- a/v4/gtk/gtkbutton.go
+++ b/v4/gtk/gtkbutton.go
@@ -17,6 +17,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Clicked uintptr
+
+	Activate uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkcellarea.go b/v4/gtk/gtkcellarea.go
index 6b6c3c0779bc1494b73a685658ee103135de016a..b40ba481717c0ffa3d032823c885f5b55647f259 100644
--- a/v4/gtk/gtkcellarea.go
+++ b/v4/gtk/gtkcellarea.go
@@ -27,6 +27,44 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Add uintptr
+
+	Remove uintptr
+
+	Foreach uintptr
+
+	ForeachAlloc uintptr
+
+	Event uintptr
+
+	Snapshot uintptr
+
+	ApplyAttributes uintptr
+
+	CreateContext uintptr
+
+	CopyContext uintptr
+
+	GetRequestMode uintptr
+
+	GetPreferredWidth uintptr
+
+	GetPreferredHeightForWidth uintptr
+
+	GetPreferredHeight uintptr
+
+	GetPreferredWidthForHeight uintptr
+
+	SetCellProperty uintptr
+
+	GetCellProperty uintptr
+
+	Focus uintptr
+
+	IsActivatable uintptr
+
+	Activate uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkcellareacontext.go b/v4/gtk/gtkcellareacontext.go
index 9d470ec42a1861d0b090794eb9fda124faf6e053..f906184b2d25915f7e562465fd49146500479d2e 100644
--- a/v4/gtk/gtkcellareacontext.go
+++ b/v4/gtk/gtkcellareacontext.go
@@ -16,6 +16,14 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Allocate uintptr
+
+	Reset uintptr
+
+	GetPreferredHeightForWidth uintptr
+
+	GetPreferredWidthForHeight uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkcelleditable.go b/v4/gtk/gtkcelleditable.go
index 1efd36fea57a675489d640acfd45c1c4c7a1ede9..e181d20608b7e6ec89dd3102ee2736814d8bf398 100644
--- a/v4/gtk/gtkcelleditable.go
+++ b/v4/gtk/gtkcelleditable.go
@@ -15,6 +15,12 @@ type CellEditableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	EditingDone uintptr
+
+	RemoveWidget uintptr
+
+	StartEditing uintptr
 }
 
 func (x *CellEditableIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkcelllayout.go b/v4/gtk/gtkcelllayout.go
index ac21bab7dec5e1c6e57b32a45c8ec20abae23815..3dbd38a0a3b14b37513cfbe78d51f510890179e1 100644
--- a/v4/gtk/gtkcelllayout.go
+++ b/v4/gtk/gtkcelllayout.go
@@ -20,6 +20,24 @@ type CellLayoutIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	PackStart uintptr
+
+	PackEnd uintptr
+
+	Clear uintptr
+
+	AddAttribute uintptr
+
+	SetCellDataFunc uintptr
+
+	ClearAttributes uintptr
+
+	Reorder uintptr
+
+	GetCells uintptr
+
+	GetArea uintptr
 }
 
 func (x *CellLayoutIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkcellrenderer.go b/v4/gtk/gtkcellrenderer.go
index 2db6f293cbc3d31b420b760bc45b32de87c09f8a..6be6211aedf520ca315739379aa49ede41913476 100644
--- a/v4/gtk/gtkcellrenderer.go
+++ b/v4/gtk/gtkcellrenderer.go
@@ -18,6 +18,28 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GetRequestMode uintptr
+
+	GetPreferredWidth uintptr
+
+	GetPreferredHeightForWidth uintptr
+
+	GetPreferredHeight uintptr
+
+	GetPreferredWidthForHeight uintptr
+
+	GetAlignedArea uintptr
+
+	Snapshot uintptr
+
+	Activate uintptr
+
+	StartEditing uintptr
+
+	EditingCanceled uintptr
+
+	EditingStarted uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkcellrenderertext.go b/v4/gtk/gtkcellrenderertext.go
index 882eacb28b8f6d9e83940766df68209433897321..7f0874d33e6d637b22b67c3df4fb6b8d1544e635 100644
--- a/v4/gtk/gtkcellrenderertext.go
+++ b/v4/gtk/gtkcellrenderertext.go
@@ -17,6 +17,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Edited uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkcheckbutton.go b/v4/gtk/gtkcheckbutton.go
index 913fb80ebc98718d26e7c5fb35812a9363223abf..ebbe7256916c5a4d60ceb48fe173d75fd84224f6 100644
--- a/v4/gtk/gtkcheckbutton.go
+++ b/v4/gtk/gtkcheckbutton.go
@@ -17,6 +17,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Toggled uintptr
+
+	Activate uintptr
+
 	Padding [7]uintptr
 }
 
diff --git a/v4/gtk/gtkcolorchooser.go b/v4/gtk/gtkcolorchooser.go
index dedbbeb2e9f1a56efbc2009e56bdf30a9f5f3fdb..7dad73ebcf78919cac2bc83c407684194633d14e 100644
--- a/v4/gtk/gtkcolorchooser.go
+++ b/v4/gtk/gtkcolorchooser.go
@@ -16,6 +16,14 @@ 	_ structs.HostLayout
 
 	BaseInterface uintptr
 
+	GetRgba uintptr
+
+	SetRgba uintptr
+
+	AddPalette uintptr
+
+	ColorActivated uintptr
+
 	Padding [12]uintptr
 }
 
diff --git a/v4/gtk/gtkcombobox.go b/v4/gtk/gtkcombobox.go
index c2c4e916e8bd67b744f1beefb46f0df593f602f8..3b5d6404e84bab7bde4a6ff76f1ce14eebf63863 100644
--- a/v4/gtk/gtkcombobox.go
+++ b/v4/gtk/gtkcombobox.go
@@ -18,6 +18,12 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Changed uintptr
+
+	FormatEntryText uintptr
+
+	Activate uintptr
+
 	Padding [7]uintptr
 }
 
diff --git a/v4/gtk/gtkdialog.go b/v4/gtk/gtkdialog.go
index 12488f0f6f69ffe368b1592422c9669fd12b4c7d..db6375b77d2f14b9cf67b3fcebefc567dea041ca 100644
--- a/v4/gtk/gtkdialog.go
+++ b/v4/gtk/gtkdialog.go
@@ -19,6 +19,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Response uintptr
+
+	Close uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkdrawingarea.go b/v4/gtk/gtkdrawingarea.go
index d12e3109798603eb536e01bcaf378d6804e5b9fc..4763861639c07e7313c4ed53a05d1b66db3094d7 100644
--- a/v4/gtk/gtkdrawingarea.go
+++ b/v4/gtk/gtkdrawingarea.go
@@ -24,6 +24,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Resize uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkeditable.go b/v4/gtk/gtkeditable.go
index 2637a020f3cfd32f9db84023d26beb91cf95705e..3d33d86011f42f666e124730ca65753d3db7962a 100644
--- a/v4/gtk/gtkeditable.go
+++ b/v4/gtk/gtkeditable.go
@@ -15,6 +15,24 @@ type EditableInterface struct {
 	_ structs.HostLayout
 
 	BaseIface uintptr
+
+	InsertText uintptr
+
+	DeleteText uintptr
+
+	Changed uintptr
+
+	GetText uintptr
+
+	DoInsertText uintptr
+
+	DoDeleteText uintptr
+
+	GetSelectionBounds uintptr
+
+	SetSelectionBounds uintptr
+
+	GetDelegate uintptr
 }
 
 func (x *EditableInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkentry.go b/v4/gtk/gtkentry.go
index 70abfcf0762112c45bbeb811d895396e630469b4..53ab2af9310e4b41453ad3d75502732b7e41ceaf 100644
--- a/v4/gtk/gtkentry.go
+++ b/v4/gtk/gtkentry.go
@@ -25,6 +25,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Activate uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkentrybuffer.go b/v4/gtk/gtkentrybuffer.go
index 22664fd83aab815ff96bc3b9930262f42523ba1d..68b332b45bdd8e5bd3c418ac646096622889723d 100644
--- a/v4/gtk/gtkentrybuffer.go
+++ b/v4/gtk/gtkentrybuffer.go
@@ -16,6 +16,34 @@ type EntryBufferClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	InsertedText uintptr
+
+	DeletedText uintptr
+
+	GetText uintptr
+
+	GetLength uintptr
+
+	InsertText uintptr
+
+	DeleteText uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
+
+	GtkReserved6 uintptr
+
+	GtkReserved7 uintptr
+
+	GtkReserved8 uintptr
 }
 
 func (x *EntryBufferClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkfilter.go b/v4/gtk/gtkfilter.go
index 9c08384aec665017d7fd0bf8792d9fcccd102224..046550dd95764e4d76b89ec9efff4b3344387fe4 100644
--- a/v4/gtk/gtkfilter.go
+++ b/v4/gtk/gtkfilter.go
@@ -16,6 +16,26 @@ type FilterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Match uintptr
+
+	GetStrictness uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
+
+	GtkReserved6 uintptr
+
+	GtkReserved7 uintptr
+
+	GtkReserved8 uintptr
 }
 
 func (x *FilterClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkflowbox.go b/v4/gtk/gtkflowbox.go
index 0c36740fe8818e348119a9a20549676f68fc02f1..f8d08101e22a9862e513a5e77df0395c2679fe41 100644
--- a/v4/gtk/gtkflowbox.go
+++ b/v4/gtk/gtkflowbox.go
@@ -38,6 +38,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Activate uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkfontchooser.go b/v4/gtk/gtkfontchooser.go
index 043f7faa47b42fc3afa5526ac7c7d0bae132acc8..ffea928b8b0b60d8f86314d3e71dac1d3e75b171 100644
--- a/v4/gtk/gtkfontchooser.go
+++ b/v4/gtk/gtkfontchooser.go
@@ -24,6 +24,20 @@ 	_ structs.HostLayout
 
 	BaseIface uintptr
 
+	GetFontFamily uintptr
+
+	GetFontFace uintptr
+
+	GetFontSize uintptr
+
+	SetFilterFunc uintptr
+
+	FontActivated uintptr
+
+	SetFontMap uintptr
+
+	GetFontMap uintptr
+
 	Padding [10]uintptr
 }
 
diff --git a/v4/gtk/gtkframe.go b/v4/gtk/gtkframe.go
index dbca285df518ff31fc4a21dd3220f283d25fb5dc..981140d7f35008363eb12f38fcfaf009582048a4 100644
--- a/v4/gtk/gtkframe.go
+++ b/v4/gtk/gtkframe.go
@@ -16,6 +16,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	ComputeChildAllocation uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkglarea.go b/v4/gtk/gtkglarea.go
index ccc2778fc5b7a89eb43beaf1b91526d9eb60a1b5..174652771ff98b0ee0c05cfa52da34bd3ee5cfc6 100644
--- a/v4/gtk/gtkglarea.go
+++ b/v4/gtk/gtkglarea.go
@@ -19,6 +19,12 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Render uintptr
+
+	Resize uintptr
+
+	CreateContext uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkimcontext.go b/v4/gtk/gtkimcontext.go
index 699eb32d1ddfdf52f7b6503c64e65631432dabdc..c34b7e88815884c438615b0464485358628d0720 100644
--- a/v4/gtk/gtkimcontext.go
+++ b/v4/gtk/gtkimcontext.go
@@ -18,6 +18,52 @@ type IMContextClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	PreeditStart uintptr
+
+	PreeditEnd uintptr
+
+	PreeditChanged uintptr
+
+	Commit uintptr
+
+	RetrieveSurrounding uintptr
+
+	DeleteSurrounding uintptr
+
+	SetClientWidget uintptr
+
+	GetPreeditString uintptr
+
+	FilterKeypress uintptr
+
+	FocusIn uintptr
+
+	FocusOut uintptr
+
+	Reset uintptr
+
+	SetCursorLocation uintptr
+
+	SetUsePreedit uintptr
+
+	SetSurrounding uintptr
+
+	GetSurrounding uintptr
+
+	SetSurroundingWithSelection uintptr
+
+	GetSurroundingWithSelection uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
 }
 
 func (x *IMContextClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkimmulticontext.go b/v4/gtk/gtkimmulticontext.go
index 8a4b6ce13ddff7646def9028988a491feaedc685..d714b274c31190ec7a2cd5e3053e91c018a58265 100644
--- a/v4/gtk/gtkimmulticontext.go
+++ b/v4/gtk/gtkimmulticontext.go
@@ -14,6 +14,14 @@ type IMMulticontextClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *IMMulticontextClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtklayoutmanager.go b/v4/gtk/gtklayoutmanager.go
index 2816716c9c9ca14c5ff3be00c31c722c961ea7e2..f3a0c88ffab1266f8afd150c10452ff7a6795997 100644
--- a/v4/gtk/gtklayoutmanager.go
+++ b/v4/gtk/gtklayoutmanager.go
@@ -19,7 +19,19 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GetRequestMode uintptr
+
+	Measure uintptr
+
+	Allocate uintptr
+
 	LayoutChildType types.GType
+
+	CreateLayoutChild uintptr
+
+	Root uintptr
+
+	Unroot uintptr
 
 	Padding [16]uintptr
 }
diff --git a/v4/gtk/gtklistbox.go b/v4/gtk/gtklistbox.go
index 2213d056b2784cc4c1ff33015f763ba0d8c82b3b..0691dbd2aa8ead417f7fb41691de92ce396d1fd2 100644
--- a/v4/gtk/gtklistbox.go
+++ b/v4/gtk/gtklistbox.go
@@ -44,6 +44,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Activate uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkmediafile.go b/v4/gtk/gtkmediafile.go
index a4d68df8be91bf9b7003ae25b5181424cadee364..8088a12d2fd69facad93cc9b4e65675259c403ba 100644
--- a/v4/gtk/gtkmediafile.go
+++ b/v4/gtk/gtkmediafile.go
@@ -17,6 +17,18 @@ type MediaFileClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Open uintptr
+
+	Close uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *MediaFileClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkmediastream.go b/v4/gtk/gtkmediastream.go
index 1ff5efcc45fb62282048464297c6eb433172ead5..3c05ae927a14af66efc76b75278c9e48d657017f 100644
--- a/v4/gtk/gtkmediastream.go
+++ b/v4/gtk/gtkmediastream.go
@@ -17,6 +17,34 @@ type MediaStreamClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Play uintptr
+
+	Pause uintptr
+
+	Seek uintptr
+
+	UpdateAudio uintptr
+
+	Realize uintptr
+
+	Unrealize uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
+
+	GtkReserved6 uintptr
+
+	GtkReserved7 uintptr
+
+	GtkReserved8 uintptr
 }
 
 func (x *MediaStreamClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkmountoperation.go b/v4/gtk/gtkmountoperation.go
index 7d7f93fd1ce24da528ae5be1e64025768c78c99f..ed365171b59cc77563032a3c9a9d9abe77bed6ee 100644
--- a/v4/gtk/gtkmountoperation.go
+++ b/v4/gtk/gtkmountoperation.go
@@ -17,6 +17,14 @@ type MountOperationClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *MountOperationClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtknativedialog.go b/v4/gtk/gtknativedialog.go
index 40de261a9a1afc15b2ea30f9c19b05e44418f4e9..cf991a0467464d81cfe76c129a1433fd61e7e867 100644
--- a/v4/gtk/gtknativedialog.go
+++ b/v4/gtk/gtknativedialog.go
@@ -17,6 +17,20 @@ type NativeDialogClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Response uintptr
+
+	Show uintptr
+
+	Hide uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *NativeDialogClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkpopover.go b/v4/gtk/gtkpopover.go
index a54cd664aa5347508b481c9301080f89b7f7839a..634e07981fdc39494a336a7fa6c0cef5dcb7d8de 100644
--- a/v4/gtk/gtkpopover.go
+++ b/v4/gtk/gtkpopover.go
@@ -19,6 +19,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Closed uintptr
+
+	ActivateDefault uintptr
+
 	Reserved [8]uintptr
 }
 
diff --git a/v4/gtk/gtkprintoperation.go b/v4/gtk/gtkprintoperation.go
index 2d55562f28619cb779d3ea8514aaeb4c15e91841..ec7cb040e150c5e4d98a30a07aac2a496f8f868f 100644
--- a/v4/gtk/gtkprintoperation.go
+++ b/v4/gtk/gtkprintoperation.go
@@ -24,6 +24,28 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Done uintptr
+
+	BeginPrint uintptr
+
+	Paginate uintptr
+
+	RequestPageSetup uintptr
+
+	DrawPage uintptr
+
+	EndPrint uintptr
+
+	StatusChanged uintptr
+
+	CreateCustomWidget uintptr
+
+	CustomWidgetApply uintptr
+
+	Preview uintptr
+
+	UpdateCustomWidget uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkprintoperationpreview.go b/v4/gtk/gtkprintoperationpreview.go
index c32115b62019a61dc19cb25eb66e1b19f0009861..89801d3df9fb1b2a7e94fd664208ea609fe0a61b 100644
--- a/v4/gtk/gtkprintoperationpreview.go
+++ b/v4/gtk/gtkprintoperationpreview.go
@@ -14,6 +14,32 @@ type PrintOperationPreviewIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	Ready uintptr
+
+	GotPageSize uintptr
+
+	RenderPage uintptr
+
+	IsSelected uintptr
+
+	EndPreview uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
+
+	GtkReserved6 uintptr
+
+	GtkReserved7 uintptr
+
+	GtkReserved8 uintptr
 }
 
 func (x *PrintOperationPreviewIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkrange.go b/v4/gtk/gtkrange.go
index e87af6d7550866eb53838fa6134872e2bbe7d5ca..8ac3c5f351c9ff4dd11bce39b4be7a092914a64c 100644
--- a/v4/gtk/gtkrange.go
+++ b/v4/gtk/gtkrange.go
@@ -18,6 +18,16 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	ValueChanged uintptr
+
+	AdjustBounds uintptr
+
+	MoveSlider uintptr
+
+	GetRangeBorder uintptr
+
+	ChangeValue uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkrecentmanager.go b/v4/gtk/gtkrecentmanager.go
index e64da13a236be61f37dfd9e336fd4911393227fc..27d995e78ce9767dbb45804e7ea2c9d1ab8de018 100644
--- a/v4/gtk/gtkrecentmanager.go
+++ b/v4/gtk/gtkrecentmanager.go
@@ -330,6 +330,16 @@ type RecentManagerClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Changed uintptr
+
+	GtkRecent1 uintptr
+
+	GtkRecent2 uintptr
+
+	GtkRecent3 uintptr
+
+	GtkRecent4 uintptr
 }
 
 func (x *RecentManagerClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkscale.go b/v4/gtk/gtkscale.go
index 9745171ff3d4aba2d2480780ea0fe53c383238b1..48ceb8fad9ddbb89aa9ece8459ba42354b67ccb7 100644
--- a/v4/gtk/gtkscale.go
+++ b/v4/gtk/gtkscale.go
@@ -20,6 +20,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	GetLayoutOffsets uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkscalebutton.go b/v4/gtk/gtkscalebutton.go
index 9f4a3de451a3355f601e3bd49dfe105f81f3b15e..e6fe768cc5306fe464f75dd0a3450c30c936b11e 100644
--- a/v4/gtk/gtkscalebutton.go
+++ b/v4/gtk/gtkscalebutton.go
@@ -17,6 +17,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	ValueChanged uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkscrollable.go b/v4/gtk/gtkscrollable.go
index 3bd7093c5308e9134faf5377570d37c99b11dc34..b1181009d2d9e425fed308112992fc7746973ae2 100644
--- a/v4/gtk/gtkscrollable.go
+++ b/v4/gtk/gtkscrollable.go
@@ -15,6 +15,8 @@ type ScrollableInterface struct {
 	_ structs.HostLayout
 
 	BaseIface uintptr
+
+	GetBorder uintptr
 }
 
 func (x *ScrollableInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkselectionmodel.go b/v4/gtk/gtkselectionmodel.go
index 4f5d223c51b50f2b583a4f716c84aa3c42ef516c..9a8c3f11f8aab9bbb808c28362024adc0fa3af3d 100644
--- a/v4/gtk/gtkselectionmodel.go
+++ b/v4/gtk/gtkselectionmodel.go
@@ -26,6 +26,24 @@ type SelectionModelInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	IsSelected uintptr
+
+	GetSelectionInRange uintptr
+
+	SelectItem uintptr
+
+	UnselectItem uintptr
+
+	SelectRange uintptr
+
+	UnselectRange uintptr
+
+	SelectAll uintptr
+
+	UnselectAll uintptr
+
+	SetSelection uintptr
 }
 
 func (x *SelectionModelInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtkshortcutmanager.go b/v4/gtk/gtkshortcutmanager.go
index 9695dd8fed58075fc31f0c2635221bb7ee30458a..045c678bf9b9513b5d55c1638458b5f6efb2eb72 100644
--- a/v4/gtk/gtkshortcutmanager.go
+++ b/v4/gtk/gtkshortcutmanager.go
@@ -17,6 +17,10 @@ type ShortcutManagerInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	AddController uintptr
+
+	RemoveController uintptr
 }
 
 func (x *ShortcutManagerInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtksorter.go b/v4/gtk/gtksorter.go
index 9062b8a4f6aba0358d781871f96e1e725c90b1b3..02506f495a3606373b5b672759ae579796c1f359 100644
--- a/v4/gtk/gtksorter.go
+++ b/v4/gtk/gtksorter.go
@@ -17,6 +17,26 @@ type SorterClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Compare uintptr
+
+	GetOrder uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
+
+	GtkReserved5 uintptr
+
+	GtkReserved6 uintptr
+
+	GtkReserved7 uintptr
+
+	GtkReserved8 uintptr
 }
 
 func (x *SorterClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtkstylecontext.go b/v4/gtk/gtkstylecontext.go
index 448de30a3b1c914461112ea739ee0e8f19e527b3..6da087629379b403ffd356ca3237c7aac03be716 100644
--- a/v4/gtk/gtkstylecontext.go
+++ b/v4/gtk/gtkstylecontext.go
@@ -16,6 +16,16 @@ type StyleContextClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Changed uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *StyleContextClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtksymbolicpaintable.go b/v4/gtk/gtksymbolicpaintable.go
index 350d3a07b9f4d9170fcd4df526e006199455d9f2..0c51b45588530eac4d5ba22885973e9908e47f8a 100644
--- a/v4/gtk/gtksymbolicpaintable.go
+++ b/v4/gtk/gtksymbolicpaintable.go
@@ -17,6 +17,8 @@ type SymbolicPaintableInterface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	SnapshotSymbolic uintptr
 }
 
 func (x *SymbolicPaintableInterface) GoPointer() uintptr {
diff --git a/v4/gtk/gtktextbuffer.go b/v4/gtk/gtktextbuffer.go
index 698ae8176a4ee575df0bf9c3fd5ed9c771dc2e52..00125880235360d20a5a8a6d60acb8a2b37cd020 100644
--- a/v4/gtk/gtktextbuffer.go
+++ b/v4/gtk/gtktextbuffer.go
@@ -18,6 +18,44 @@ type TextBufferClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	InsertText uintptr
+
+	InsertPaintable uintptr
+
+	InsertChildAnchor uintptr
+
+	DeleteRange uintptr
+
+	Changed uintptr
+
+	ModifiedChanged uintptr
+
+	MarkSet uintptr
+
+	MarkDeleted uintptr
+
+	ApplyTag uintptr
+
+	RemoveTag uintptr
+
+	BeginUserAction uintptr
+
+	EndUserAction uintptr
+
+	PasteDone uintptr
+
+	Undo uintptr
+
+	Redo uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *TextBufferClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtktextchild.go b/v4/gtk/gtktextchild.go
index 61baeefe680005e9876b9c020b5fcc486b7fe273..78311afaa565967390792abae940bfc3ae9740bb 100644
--- a/v4/gtk/gtktextchild.go
+++ b/v4/gtk/gtktextchild.go
@@ -15,6 +15,14 @@ type TextChildAnchorClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *TextChildAnchorClass) GoPointer() uintptr {
diff --git a/v4/gtk/gtktextview.go b/v4/gtk/gtktextview.go
index cf50f56538974f0d5c8b131314ff5412909a493d..e43f250ca513fdcc5706a2d72555eb1530a2679e 100644
--- a/v4/gtk/gtktextview.go
+++ b/v4/gtk/gtktextview.go
@@ -20,6 +20,32 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	MoveCursor uintptr
+
+	SetAnchor uintptr
+
+	InsertAtCursor uintptr
+
+	DeleteFromCursor uintptr
+
+	Backspace uintptr
+
+	CutClipboard uintptr
+
+	CopyClipboard uintptr
+
+	PasteClipboard uintptr
+
+	ToggleOverwrite uintptr
+
+	CreateBuffer uintptr
+
+	SnapshotLayer uintptr
+
+	ExtendSelection uintptr
+
+	InsertEmoji uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtktogglebutton.go b/v4/gtk/gtktogglebutton.go
index a4e05f43836436728df6d24d0f32a33eef116a8c..0c1a0b26088db124af041c070721599944ff430e 100644
--- a/v4/gtk/gtktogglebutton.go
+++ b/v4/gtk/gtktogglebutton.go
@@ -17,6 +17,8 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Toggled uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtktreednd.go b/v4/gtk/gtktreednd.go
index 020e6d0a358c359104c2ad5fc17d1deb4ae06eb5..1476955baeda852c17d8c8706cf171da403d97a6 100644
--- a/v4/gtk/gtktreednd.go
+++ b/v4/gtk/gtktreednd.go
@@ -16,6 +16,10 @@ type TreeDragDestIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	DragDataReceived uintptr
+
+	RowDropPossible uintptr
 }
 
 func (x *TreeDragDestIface) GoPointer() uintptr {
@@ -26,6 +30,12 @@ type TreeDragSourceIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	RowDraggable uintptr
+
+	DragDataGet uintptr
+
+	DragDataDelete uintptr
 }
 
 func (x *TreeDragSourceIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtktreemodel.go b/v4/gtk/gtktreemodel.go
index 008316d5170fde08bbc2a1b798f9386c526be79d..631f62a54a2d51215c3fedeb1207070daf5d8bab 100644
--- a/v4/gtk/gtktreemodel.go
+++ b/v4/gtk/gtktreemodel.go
@@ -72,6 +72,46 @@ type TreeModelIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	RowChanged uintptr
+
+	RowInserted uintptr
+
+	RowHasChildToggled uintptr
+
+	RowDeleted uintptr
+
+	RowsReordered uintptr
+
+	GetFlags uintptr
+
+	GetNColumns uintptr
+
+	GetColumnType uintptr
+
+	GetIter uintptr
+
+	GetPath uintptr
+
+	GetValue uintptr
+
+	IterNext uintptr
+
+	IterPrevious uintptr
+
+	IterChildren uintptr
+
+	IterHasChild uintptr
+
+	IterNChildren uintptr
+
+	IterNthChild uintptr
+
+	IterParent uintptr
+
+	RefNode uintptr
+
+	UnrefNode uintptr
 }
 
 func (x *TreeModelIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtktreemodelfilter.go b/v4/gtk/gtktreemodelfilter.go
index 3917465bd7a971bcf46d2ddeeee87c1ec8deff70..ae11e96b8ef285ef3918b69004dd63513cebb4a9 100644
--- a/v4/gtk/gtktreemodelfilter.go
+++ b/v4/gtk/gtktreemodelfilter.go
@@ -29,6 +29,10 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Visible uintptr
+
+	Modify uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtktreesortable.go b/v4/gtk/gtktreesortable.go
index 13a480ca7200bf348a6815a4ab762651bfef66d3..319d1b96142bad2b8c9f96e3ef311e9d7297ad9d 100644
--- a/v4/gtk/gtktreesortable.go
+++ b/v4/gtk/gtktreesortable.go
@@ -29,6 +29,18 @@ type TreeSortableIface struct {
 	_ structs.HostLayout
 
 	GIface uintptr
+
+	SortColumnChanged uintptr
+
+	GetSortColumnId uintptr
+
+	SetSortColumnId uintptr
+
+	SetSortFunc uintptr
+
+	SetDefaultSortFunc uintptr
+
+	HasDefaultSortFunc uintptr
 }
 
 func (x *TreeSortableIface) GoPointer() uintptr {
diff --git a/v4/gtk/gtktreeview.go b/v4/gtk/gtktreeview.go
index 917a17678e5bfe17af858d435172d089f2f402db..ac9a770d99217d820459b1b079b1106f658874d8 100644
--- a/v4/gtk/gtktreeview.go
+++ b/v4/gtk/gtktreeview.go
@@ -43,6 +43,36 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	RowActivated uintptr
+
+	TestExpandRow uintptr
+
+	TestCollapseRow uintptr
+
+	RowExpanded uintptr
+
+	RowCollapsed uintptr
+
+	ColumnsChanged uintptr
+
+	CursorChanged uintptr
+
+	MoveCursor uintptr
+
+	SelectAll uintptr
+
+	UnselectAll uintptr
+
+	SelectCursorRow uintptr
+
+	ToggleCursorRow uintptr
+
+	ExpandCollapseCursorRow uintptr
+
+	SelectCursorParent uintptr
+
+	StartInteractiveSearch uintptr
+
 	Reserved [16]uintptr
 }
 
diff --git a/v4/gtk/gtkwidget.go b/v4/gtk/gtkwidget.go
index 83316f719108fb2d4361de9141083bdeeba9ea27..85f5dfb668a1d4b2c1a9289fd69ab964eab9f7e5 100644
--- a/v4/gtk/gtkwidget.go
+++ b/v4/gtk/gtkwidget.go
@@ -82,6 +82,56 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	Show uintptr
+
+	Hide uintptr
+
+	Map uintptr
+
+	Unmap uintptr
+
+	Realize uintptr
+
+	Unrealize uintptr
+
+	Root uintptr
+
+	Unroot uintptr
+
+	SizeAllocate uintptr
+
+	StateFlagsChanged uintptr
+
+	DirectionChanged uintptr
+
+	GetRequestMode uintptr
+
+	Measure uintptr
+
+	MnemonicActivate uintptr
+
+	GrabFocus uintptr
+
+	Focus uintptr
+
+	SetFocusChild uintptr
+
+	MoveFocus uintptr
+
+	KeynavFailed uintptr
+
+	QueryTooltip uintptr
+
+	ComputeExpand uintptr
+
+	CssChanged uintptr
+
+	SystemSettingChanged uintptr
+
+	Snapshot uintptr
+
+	Contains uintptr
+
 	Priv *WidgetClassPrivate
 
 	Padding [8]uintptr
diff --git a/v4/gtk/gtkwindow.go b/v4/gtk/gtkwindow.go
index 34fa947d4909b3368ef04274cdf9edff6485798e..4ef5655ee912c19b763e2968e023e10edaa0b194 100644
--- a/v4/gtk/gtkwindow.go
+++ b/v4/gtk/gtkwindow.go
@@ -20,6 +20,16 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	ActivateFocus uintptr
+
+	ActivateDefault uintptr
+
+	KeysChanged uintptr
+
+	EnableDebugging uintptr
+
+	CloseRequest uintptr
+
 	Padding [8]uintptr
 }
 
diff --git a/v4/gtk/gtkwindowgroup.go b/v4/gtk/gtkwindowgroup.go
index 999125fe2c56e9068a68547ec5b67776743878a5..a829ee9c81fe42840d17b06199f1fe6684241c33 100644
--- a/v4/gtk/gtkwindowgroup.go
+++ b/v4/gtk/gtkwindowgroup.go
@@ -16,6 +16,14 @@ type WindowGroupClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GtkReserved1 uintptr
+
+	GtkReserved2 uintptr
+
+	GtkReserved3 uintptr
+
+	GtkReserved4 uintptr
 }
 
 func (x *WindowGroupClass) GoPointer() uintptr {
diff --git a/v4/pango/pango-attributes.go b/v4/pango/pango-attributes.go
index 95a9edabd1746618e2ac036492547f84e54e5538..c2631992744ac8822618478a67eda968901d81d1 100644
--- a/v4/pango/pango-attributes.go
+++ b/v4/pango/pango-attributes.go
@@ -26,6 +26,12 @@ type AttrClass struct {
 	_ structs.HostLayout
 
 	Type AttrType
+
+	Copy uintptr
+
+	Destroy uintptr
+
+	Equal uintptr
 }
 
 func (x *AttrClass) GoPointer() uintptr {
diff --git a/v4/pango/pango-font.go b/v4/pango/pango-font.go
index fdb82cd146f6038cc603d5473bd84ba05fbc949e..90c00381fdae15b841c8588e41240c06ced54d0b 100644
--- a/v4/pango/pango-font.go
+++ b/v4/pango/pango-font.go
@@ -16,6 +16,22 @@ type FontClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	Describe uintptr
+
+	GetCoverage uintptr
+
+	GetGlyphExtents uintptr
+
+	GetMetrics uintptr
+
+	GetFontMap uintptr
+
+	DescribeAbsolute uintptr
+
+	GetFeatures uintptr
+
+	CreateHbFont uintptr
 }
 
 func (x *FontClass) GoPointer() uintptr {
@@ -480,6 +496,20 @@ type FontFaceClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetFaceName uintptr
+
+	Describe uintptr
+
+	ListSizes uintptr
+
+	IsSynthesized uintptr
+
+	GetFamily uintptr
+
+	PangoReserved3 uintptr
+
+	PangoReserved4 uintptr
 }
 
 func (x *FontFaceClass) GoPointer() uintptr {
@@ -490,6 +520,18 @@ type FontFamilyClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	ListFaces uintptr
+
+	GetName uintptr
+
+	IsMonospace uintptr
+
+	IsVariable uintptr
+
+	GetFace uintptr
+
+	PangoReserved2 uintptr
 }
 
 func (x *FontFamilyClass) GoPointer() uintptr {
diff --git a/v4/pango/pango-fontmap.go b/v4/pango/pango-fontmap.go
index a371d79728227f739e0c19a607403cfb7f1243cf..37613d3ae4704e82c0136c76a3835b4b9f126154 100644
--- a/v4/pango/pango-fontmap.go
+++ b/v4/pango/pango-fontmap.go
@@ -18,7 +18,21 @@ 	_ structs.HostLayout
 
 	ParentClass uintptr
 
+	LoadFont uintptr
+
+	ListFamilies uintptr
+
+	LoadFontset uintptr
+
 	ShapeEngineType uintptr
+
+	GetSerial uintptr
+
+	Changed uintptr
+
+	GetFamily uintptr
+
+	GetFace uintptr
 }
 
 func (x *FontMapClass) GoPointer() uintptr {
diff --git a/v4/pango/pango-fontset.go b/v4/pango/pango-fontset.go
index 6338e131b35187410c62dbff5c97492e23c05b43..9a3201179b861d1ca236b12987b1fa636c32fe9a 100644
--- a/v4/pango/pango-fontset.go
+++ b/v4/pango/pango-fontset.go
@@ -22,6 +22,22 @@ type FontsetClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	GetFont uintptr
+
+	GetMetrics uintptr
+
+	GetLanguage uintptr
+
+	Foreach uintptr
+
+	PangoReserved1 uintptr
+
+	PangoReserved2 uintptr
+
+	PangoReserved3 uintptr
+
+	PangoReserved4 uintptr
 }
 
 func (x *FontsetClass) GoPointer() uintptr {
diff --git a/v4/pango/pango-renderer.go b/v4/pango/pango-renderer.go
index f5c7eadc16f98d03cdad0e6fe9e49376282666f0..0438ed5d1f333e1ae2976c2c16290124f08b31fd 100644
--- a/v4/pango/pango-renderer.go
+++ b/v4/pango/pango-renderer.go
@@ -31,6 +31,34 @@ type RendererClass struct {
 	_ structs.HostLayout
 
 	ParentClass uintptr
+
+	DrawGlyphs uintptr
+
+	DrawRectangle uintptr
+
+	DrawErrorUnderline uintptr
+
+	DrawShape uintptr
+
+	DrawTrapezoid uintptr
+
+	DrawGlyph uintptr
+
+	PartChanged uintptr
+
+	Begin uintptr
+
+	End uintptr
+
+	PrepareRun uintptr
+
+	DrawGlyphItem uintptr
+
+	PangoReserved2 uintptr
+
+	PangoReserved3 uintptr
+
+	PangoReserved4 uintptr
 }
 
 func (x *RendererClass) GoPointer() uintptr {

