From 815cc7544abf379a229b794780ebf1d1b11b1294 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 25 Aug 2025 23:46:02 -0700
Subject: [PATCH] GIR: Export pass and util functions to allow generating from 3rd party repos

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/gen.go b/gen.go
index 921109acc601f9c8c4b9db7e0ca12e5021e2c31d..17fff64ade2eb721048b40402cbfdd11092ae72f 100644
--- a/gen.go
+++ b/gen.go
@@ -6,8 +6,8 @@ 	"path/filepath"
 	"strings"
 	"text/template"
 
-	"github.com/jwijenbergh/puregotk/internal/gir/pass"
-	"github.com/jwijenbergh/puregotk/internal/gir/util"
+	"github.com/jwijenbergh/puregotk/pkg/gir/pass"
+	"github.com/jwijenbergh/puregotk/pkg/gir/util"
 )
 
 //go:generate go run gen.go
diff --git a/pkg/gir/pass/pass.go b/pkg/gir/pass/pass.go
new file mode 100644
index 0000000000000000000000000000000000000000..730e7576bb172daf3d4415508531b554e53c0697
--- /dev/null
+++ b/pkg/gir/pass/pass.go
@@ -0,0 +1,11 @@
+package pass
+
+import "github.com/jwijenbergh/puregotk/internal/gir/pass"
+
+type (
+	Pass = pass.Pass
+)
+
+var (
+	New = pass.New
+)
diff --git a/pkg/gir/util/util.go b/pkg/gir/util/util.go
new file mode 100644
index 0000000000000000000000000000000000000000..4f5c539929b966cfe58b5f045aa6e736126877bd
--- /dev/null
+++ b/pkg/gir/util/util.go
@@ -0,0 +1,8 @@
+package util
+
+import "github.com/jwijenbergh/puregotk/internal/gir/util"
+
+var (
+	ConvertArgs      = util.ConvertArgs
+	ConvertArgsComma = util.ConvertArgsComma
+)

