From 08dfbe517df1cd7d90d94c6e6e18be3185fd6cc0 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 26 Aug 2025 00:03:02 -0700
Subject: [PATCH] GIR: Export core functions to allow generating from 3rd party repos

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/pkg/core/core.go b/pkg/core/core.go
new file mode 100644
index 0000000000000000000000000000000000000000..f6ad08fc09a5509ff5526537d88714ae943f84c0
--- /dev/null
+++ b/pkg/core/core.go
@@ -0,0 +1,9 @@
+package core
+
+import "github.com/jwijenbergh/puregotk/internal/core"
+
+var (
+	PuregoSafeRegister = core.PuregoSafeRegister
+	GetPath           = core.GetPath
+	GoString          = core.GoString
+)
\ No newline at end of file
diff --git a/templates/glib b/templates/glib
index c37715351160f0b06ad79986484bbbf2f1f473b4..86a98bf5c25f2d247cc2e8732b2b49e5904b2ef8 100644
--- a/templates/glib
+++ b/templates/glib
@@ -5,7 +5,7 @@ 	"fmt"
 	"sync"
 
 	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/puregotk/internal/core"
+	"github.com/jwijenbergh/puregotk/pkg/core"
 )
 
 var callbacks = struct {
diff --git a/templates/go b/templates/go
index 3bc6fa583d766515fd22da01f41ee4fa46e6228c..438ec83a052d5920de70ee5f9807295de7b2e54b 100644
--- a/templates/go
+++ b/templates/go
@@ -25,7 +25,7 @@ {{- if $NeedsPurego}}
 	"github.com/jwijenbergh/purego"
 {{- end}}
 {{- if .NeedsInit}}
-	"github.com/jwijenbergh/puregotk/internal/core"
+	"github.com/jwijenbergh/puregotk/pkg/core"
 {{- end}}
 )
 {{end}}

