From 31a0b0558339801558952c9541ac047d18c89f29 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 08 Oct 2025 21:40:13 -0700
Subject: [PATCH] Gen: Use OverrideX and GetX as callback override functions

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/templates/go b/templates/go
index bd883ae9bcf271021a66fa823cb0e9597b0e5fb2..68d4676eb94747fa08038f12ce7f6d772949697b 100644
--- a/templates/go
+++ b/templates/go
@@ -84,9 +84,9 @@ }
 {{end}}
 
 {{range .CallbackAccessors -}}
-{{if .Doc}}// SetCallback{{.Name}} sets the callback function.
-{{.Doc}}{{else}}// SetCallback{{.Name}} sets the callback function.{{end}}
-func (x *{{$outer.Name}}) SetCallback{{.Name}}(cb {{.CallbackType}}) {
+{{if .Doc}}// Override{{.Name}} sets the callback function.
+{{.Doc}}{{else}}// Override{{.Name}} sets the callback function.{{end}}
+func (x *{{$outer.Name}}) Override{{.Name}}(cb {{.CallbackType}}) {
      if cb == nil {
           x.x{{.Name}} = 0
      } else {
@@ -100,9 +100,9 @@           })
      }
 }
 
-{{if .Doc}}// GetCallback{{.Name}} gets the callback function.
-{{.Doc}}{{else}}// GetCallback{{.Name}} gets the callback function.{{end}}
-func (x *{{$outer.Name}}) GetCallback{{.Name}}() {{.CallbackType}} {
+{{if .Doc}}// Get{{.Name}} gets the callback function.
+{{.Doc}}{{else}}// Get{{.Name}} gets the callback function.{{end}}
+func (x *{{$outer.Name}}) Get{{.Name}}() {{.CallbackType}} {
      if x.x{{.Name}} == 0 {
           return nil
      }

