From 26498607114ebb1989ebdd0893b9c077f58fd70a Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 17 Oct 2025 20:59:19 -0700
Subject: [PATCH] GIR: Remove `extract-specs.sh` tool in favor of the one in root

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/gir/spec/extract-specs.sh b/internal/gir/spec/extract-specs.sh
deleted file mode 100755
index c4f2ac7405bf6690707c327d3780026f660644cf..0000000000000000000000000000000000000000
--- a/internal/gir/spec/extract-specs.sh
+++ /dev/null
@@ -1,3 +0,0 @@
-#!/bin/bash
-
-for f in *.gir; do flatpak run --filesystem="${PWD}" --command=bash org.gnome.Sdk -c "cp /usr/share/gir-1.0/${f} ${PWD}/${f}"; done

