From 5bdb85b39f65f226bf8011475f411f6a3db946d5 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 31 Oct 2025 19:46:57 -0700
Subject: [PATCH] GIR: Add simple pointer-based setter for `TypePointerVal` in generator

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/gir/types/types.go b/internal/gir/types/types.go
index 883baefda07cc8081728a1808b17fac8eb99330b..d86568943b6c05efe65ba18d2c13c4bf034d71c0 100644
--- a/internal/gir/types/types.go
+++ b/internal/gir/types/types.go
@@ -769,6 +769,8 @@ 	case "float64":
 		return "TypeDoubleVal", "SetDouble", "GetDouble"
 	case "string":
 		return "TypeStringVal", "SetString", "GetString"
+	case "uintptr":
+		return "TypePointerVal", "SetPointer", "GetPointer"
 	case "[]string":
 		return "BoxedStrv", "SetBoxed", "GetBoxed"
 	default:

