From 8fc2e0d2b1cfd596fa6784202020827964b2c642 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 12 Dec 2025 19:23:12 -0800
Subject: [PATCH] GIR: Simplify primitive check function

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/gir/types/template.go b/internal/gir/types/template.go
index 4f940b4611b736cd81d1c2e628f9501d38282b66..bc4f2ee1e1c7477c2db24e9f5ed729fb96abd531 100644
--- a/internal/gir/types/template.go
+++ b/internal/gir/types/template.go
@@ -142,12 +142,14 @@ 	f.Pure.Call = append(f.Pure.Call, c)
 	f.Pure.Full = append(f.Pure.Full, n+" "+t)
 }
 
-// isGoPrimitive checks if a type name (without [] or *) is a Go primitive type
+// isGoPrimitive checks if a scalar or vector type name is a Go primitive type
 func isGoPrimitive(typeName string) bool {
-	baseType := strings.TrimLeft(typeName, "[]*")       // Strip [] and * prefixes to get the base type
-	if idx := strings.Index(baseType, "]"); idx != -1 { // Handle cases like "[2]byte" by removing the array size
-		baseType = baseType[idx+1:]
-	}
+	baseType := strings.TrimLeftFunc(
+		typeName,
+		func(r rune) bool {
+			return r == '*' || r == '[' || r == ']' || (r >= '0' && r <= '9') // Strip pointer, slice and array prefixes
+		},
+	)
 
 	for _, goType := range convList {
 		if goType == baseType {
@@ -172,7 +174,7 @@
 	stars := strings.Count(goType, "*")
 	goType = util.NormalizeNamespace(ns, goType, true)
 
-	if kind != OtherType && kind != UnknownType && !isGoPrimitive(goType) { // // Only add namespace for non-primitive types
+	if kind != OtherType && kind != UnknownType && !isGoPrimitive(goType) { // Only add namespace for non-primitive types
 		goType = util.AddNamespace(goType, ins)
 	}
 	if stars > 0 {

