From 01a0f049fbe96a3558cb8267b590cceeb83ca03b Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 08 Dec 2025 22:54:39 -0800
Subject: [PATCH] Examples: Remove dead links leftover from import into repo
---
diff --git a/examples/mylib-gtk-meson/src/docs.toml.in b/examples/mylib-gtk-meson/src/docs.toml.in
index d01a1cc852a2cd35d7ea7973eb178d9a835ca75b..56152c0bd70268e0f5cc766bb781f61e306c3db1 100644
--- a/examples/mylib-gtk-meson/src/docs.toml.in
+++ b/examples/mylib-gtk-meson/src/docs.toml.in
@@ -1,10 +1,10 @@
 [library]
     version = @PACKAGE_VERSION@
-    authors = "puregotk"
+    authors = "Felicitas Pojtinger"
     license = "AGPL-3.0"
-    website_url = "https://github.com/nicholasrobinson/puregotk"
-    browse_url = "https://github.com/nicholasrobinson/puregotk"
-    repository_url = "https://github.com/nicholasrobinson/puregotk.git"
+    website_url = "https://github.com/jwijenbergh/puregotk"
+    browse_url = "https://github.com/jwijenbergh/puregotk"
+    repository_url = "https://github.com/jwijenbergh/puregotk.git"
     description = "Example GTK library built with Go"
     devhelp = true
     search_index = true
@@ -25,4 +25,4 @@     show_index_summary = true
     show_class_hierarchy = true
 
 [source-location]
-    base_url = "https://github.com/nicholasrobinson/puregotk/blob/main/examples/mylib-gtk-meson"
+    base_url = "https://github.com/jwijenbergh/puregotk/blob/main/examples/mylib-gtk-meson"

