From a789f5f992e64e00636a8351bde73468c29fd927 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 08 Dec 2025 22:34:30 -0800
Subject: [PATCH] Examples: Rename `mylib-gtk-go-meson` to `mylib-gtk-meson-go`
---
diff --git a/examples/myapp-gnome-gomod/go.mod b/examples/myapp-gnome-gomod/go.mod
index 8f49f1a70087e97b3768f55501ca227d178319b7..003c3029baa55a221b91516420d0342ffc12f2f9 100644
--- a/examples/myapp-gnome-gomod/go.mod
+++ b/examples/myapp-gnome-gomod/go.mod
@@ -6,7 +6,7 @@ tool github.com/dennwc/flatpak-go-mod
 
 require (
 	github.com/jwijenbergh/puregotk v0.0.0-00010101000000-000000000000
-	github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000
+	github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000
 	github.com/pojntfx/go-gettext v0.2.0
 )
 
@@ -19,5 +19,5 @@ )
 
 replace (
 	github.com/jwijenbergh/puregotk v0.0.0-00010101000000-000000000000 => ../..
-	github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000 => ../mylib-gtk-go-meson
+	github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000 => ../mylib-gtk-meson-go
 )
diff --git a/examples/myapp-gnome-gomod/io.github.puregotk.MyAppGnomeGomod.json b/examples/myapp-gnome-gomod/io.github.puregotk.MyAppGnomeGomod.json
index a19adeb49c77cc7f33f741d0356d62bd1cf0f0e6..b48086f26bc84e309d6d451bd0341fe313cd8d8a 100644
--- a/examples/myapp-gnome-gomod/io.github.puregotk.MyAppGnomeGomod.json
+++ b/examples/myapp-gnome-gomod/io.github.puregotk.MyAppGnomeGomod.json
@@ -46,8 +46,8 @@           "dest": "vendor/github.com/jwijenbergh/puregotk"
         },
         {
           "type": "dir",
-          "path": "../mylib-gtk-go-meson",
-          "dest": "vendor/github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson"
+          "path": "../mylib-gtk-meson-go",
+          "dest": "vendor/github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go"
         },
         "go.mod.json"
       ]
diff --git a/examples/myapp-gnome-gomod/modules.txt b/examples/myapp-gnome-gomod/modules.txt
index 39e422f3ec7eb0f6927e4eda97e1d85d0a2bc2d3..eb22d10167fd062e82959213801c2782eb8f3cd7 100644
--- a/examples/myapp-gnome-gomod/modules.txt
+++ b/examples/myapp-gnome-gomod/modules.txt
@@ -36,9 +36,9 @@ github.com/jwijenbergh/puregotk/v4/graphene
 github.com/jwijenbergh/puregotk/v4/gsk
 github.com/jwijenbergh/puregotk/v4/gtk
 github.com/jwijenbergh/puregotk/v4/pango
-# github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000 => ../mylib-gtk-go-meson
+# github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000 => ../mylib-gtk-meson-go
 ## explicit; go 1.25.0
-github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson/mylibgtkmeson
+github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go/mylibgtkmeson
 # github.com/pojntfx/go-gettext v0.2.0
 ## explicit; go 1.25.0
 github.com/pojntfx/go-gettext/pkg/i18n
diff --git a/examples/myapp-gnome-gomod/src/window.go b/examples/myapp-gnome-gomod/src/window.go
index 39503c7192c805ed61960706854b216d40212e5b..486aea5df73f66b2d725fd38ce8ae012921ccd4c 100644
--- a/examples/myapp-gnome-gomod/src/window.go
+++ b/examples/myapp-gnome-gomod/src/window.go
@@ -6,7 +6,7 @@ 	"runtime"
 	"time"
 	"unsafe"
 
-	"github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson/mylibgtkmeson"
+	"github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go/mylibgtkmeson"
 	"github.com/jwijenbergh/puregotk/v4/adw"
 	"github.com/jwijenbergh/puregotk/v4/glib"
 	"github.com/jwijenbergh/puregotk/v4/gobject"
diff --git a/examples/myapp-gnome-meson/go.mod b/examples/myapp-gnome-meson/go.mod
index 5d89505b2f8e2fe8fcb06a49790a291b89cb86b6..abc8d438dbf4272d23c3a76e840d29b71df5e484 100644
--- a/examples/myapp-gnome-meson/go.mod
+++ b/examples/myapp-gnome-meson/go.mod
@@ -6,7 +6,7 @@ tool github.com/dennwc/flatpak-go-mod
 
 require (
 	github.com/jwijenbergh/puregotk v0.0.0-00010101000000-000000000000
-	github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000
+	github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000
 	github.com/pojntfx/go-gettext v0.2.0
 )
 
@@ -19,5 +19,5 @@ )
 
 replace (
 	github.com/jwijenbergh/puregotk v0.0.0-00010101000000-000000000000 => ../..
-	github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000 => ../mylib-gtk-go-meson
+	github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000 => ../mylib-gtk-meson-go
 )
diff --git a/examples/myapp-gnome-meson/io.github.puregotk.MyAppGnomeMeson.json b/examples/myapp-gnome-meson/io.github.puregotk.MyAppGnomeMeson.json
index fa5686bf9f0f76ce5812c0a6e0c41b6b3b6d16e8..0d0d22d6d0ae65e8307522d449e3b5859d10f6c9 100644
--- a/examples/myapp-gnome-meson/io.github.puregotk.MyAppGnomeMeson.json
+++ b/examples/myapp-gnome-meson/io.github.puregotk.MyAppGnomeMeson.json
@@ -41,8 +41,8 @@           "dest": "vendor/github.com/jwijenbergh/puregotk"
         },
         {
           "type": "dir",
-          "path": "../mylib-gtk-go-meson",
-          "dest": "vendor/github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson"
+          "path": "../mylib-gtk-meson-go",
+          "dest": "vendor/github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go"
         },
         "go.mod.json"
       ]
diff --git a/examples/myapp-gnome-meson/modules.txt b/examples/myapp-gnome-meson/modules.txt
index 39e422f3ec7eb0f6927e4eda97e1d85d0a2bc2d3..eb22d10167fd062e82959213801c2782eb8f3cd7 100644
--- a/examples/myapp-gnome-meson/modules.txt
+++ b/examples/myapp-gnome-meson/modules.txt
@@ -36,9 +36,9 @@ github.com/jwijenbergh/puregotk/v4/graphene
 github.com/jwijenbergh/puregotk/v4/gsk
 github.com/jwijenbergh/puregotk/v4/gtk
 github.com/jwijenbergh/puregotk/v4/pango
-# github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson v0.0.0-00010101000000-000000000000 => ../mylib-gtk-go-meson
+# github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go v0.0.0-00010101000000-000000000000 => ../mylib-gtk-meson-go
 ## explicit; go 1.25.0
-github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson/mylibgtkmeson
+github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go/mylibgtkmeson
 # github.com/pojntfx/go-gettext v0.2.0
 ## explicit; go 1.25.0
 github.com/pojntfx/go-gettext/pkg/i18n
diff --git a/examples/myapp-gnome-meson/src/window.go b/examples/myapp-gnome-meson/src/window.go
index cdd6b05d01929905fa26522b67e1f05eb3f9305b..c7a5e177091680010e1e1ff3558364fb1ea74bb9 100644
--- a/examples/myapp-gnome-meson/src/window.go
+++ b/examples/myapp-gnome-meson/src/window.go
@@ -6,7 +6,7 @@ 	"runtime"
 	"time"
 	"unsafe"
 
-	"github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson/mylibgtkmeson"
+	"github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go/mylibgtkmeson"
 	"github.com/jwijenbergh/puregotk/v4/adw"
 	"github.com/jwijenbergh/puregotk/v4/glib"
 	"github.com/jwijenbergh/puregotk/v4/gobject"
diff --git a/examples/mylib-gtk-go-meson/gen.go b/examples/mylib-gtk-meson-go/gen.go
rename from examples/mylib-gtk-go-meson/gen.go
rename to examples/mylib-gtk-meson-go/gen.go
diff --git a/examples/mylib-gtk-go-meson/gen.sh b/examples/mylib-gtk-meson-go/gen.sh
rename from examples/mylib-gtk-go-meson/gen.sh
rename to examples/mylib-gtk-meson-go/gen.sh
diff --git a/examples/mylib-gtk-go-meson/go.mod b/examples/mylib-gtk-meson-go/go.mod
rename from examples/mylib-gtk-go-meson/go.mod
rename to examples/mylib-gtk-meson-go/go.mod
index a7272c46ba2641b1ca2d636aaefe40a24ea15b6c..200f97bcc7ed53167c9ffea19f635ac30f1f93be 100644
--- a/examples/mylib-gtk-go-meson/go.mod
+++ b/examples/mylib-gtk-meson-go/go.mod
@@ -1,4 +1,4 @@
-module github.com/jwijenbergh/puregotk/examples/mylib-gtk-go-meson
+module github.com/jwijenbergh/puregotk/examples/mylib-gtk-meson-go
 
 go 1.25.0
 
diff --git a/examples/mylib-gtk-go-meson/go.sum b/examples/mylib-gtk-meson-go/go.sum
rename from examples/mylib-gtk-go-meson/go.sum
rename to examples/mylib-gtk-meson-go/go.sum
diff --git a/examples/mylib-gtk-go-meson/internal/gir/spec/MyLibGtkMeson-0.1.gir b/examples/mylib-gtk-meson-go/internal/gir/spec/MyLibGtkMeson-0.1.gir
rename from examples/mylib-gtk-go-meson/internal/gir/spec/MyLibGtkMeson-0.1.gir
rename to examples/mylib-gtk-meson-go/internal/gir/spec/MyLibGtkMeson-0.1.gir
diff --git a/examples/mylib-gtk-go-meson/mylibgtkmeson/window.go b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
rename from examples/mylib-gtk-go-meson/mylibgtkmeson/window.go
rename to examples/mylib-gtk-meson-go/mylibgtkmeson/window.go

