From a4946c4f07703c70d234a8064787e99de51ce456 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 11 Dec 2025 23:31:09 -0800
Subject: [PATCH] Gen: Re-run script
---
diff --git a/v4/webkit/WebKitAuthenticationRequest.go b/v4/webkit/WebKitAuthenticationRequest.go
index 75f19b0c564b8dd69ccd3e62dbc6f9e49520d8d6..df87b446cbebb132ba1ad595a5c5aa0a2a37ec47 100644
--- a/v4/webkit/WebKitAuthenticationRequest.go
+++ b/v4/webkit/WebKitAuthenticationRequest.go
@@ -333,4 +333,8 @@ 	core.PuregoSafeRegister(&xAuthenticationRequestIsRetry, libs, "webkit_authentication_request_is_retry")
 	core.PuregoSafeRegister(&xAuthenticationRequestSetCanSaveCredentials, libs, "webkit_authentication_request_set_can_save_credentials")
 	core.PuregoSafeRegister(&xAuthenticationRequestSetProposedCredential, libs, "webkit_authentication_request_set_proposed_credential")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	AuthenticationRequestGLibType()
 }
diff --git a/v4/webkit/WebKitAutomationSession.go b/v4/webkit/WebKitAutomationSession.go
index dd592ecae829169a9525d12e87e19b4ac94b7836..d7f255ac14ee7aa328ae62d925e00820c229f239 100644
--- a/v4/webkit/WebKitAutomationSession.go
+++ b/v4/webkit/WebKitAutomationSession.go
@@ -200,4 +200,8 @@ 	core.PuregoSafeRegister(&xAutomationSessionGetApplicationInfo, libs, "webkit_automation_session_get_application_info")
 	core.PuregoSafeRegister(&xAutomationSessionGetId, libs, "webkit_automation_session_get_id")
 	core.PuregoSafeRegister(&xAutomationSessionSetApplicationInfo, libs, "webkit_automation_session_set_application_info")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	AutomationSessionGLibType()
 }
diff --git a/v4/webkit/WebKitBackForwardList.go b/v4/webkit/WebKitBackForwardList.go
index 0cc9e2f6c08630146cd157a70f37fbb0e97f4c5d..559c0c50bf43342d28c6672fe099a3d4cb2dba7e 100644
--- a/v4/webkit/WebKitBackForwardList.go
+++ b/v4/webkit/WebKitBackForwardList.go
@@ -225,4 +225,8 @@ 	core.PuregoSafeRegister(&xBackForwardListGetForwardListWithLimit, libs, "webkit_back_forward_list_get_forward_list_with_limit")
 	core.PuregoSafeRegister(&xBackForwardListGetLength, libs, "webkit_back_forward_list_get_length")
 	core.PuregoSafeRegister(&xBackForwardListGetNthItem, libs, "webkit_back_forward_list_get_nth_item")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	BackForwardListGLibType()
 }
diff --git a/v4/webkit/WebKitBackForwardListItem.go b/v4/webkit/WebKitBackForwardListItem.go
index 9ed254b89b80b1343813dea1c2c2932132d8caad..3fdf0911dd65d391ef541410c722ebe73adc4a4a 100644
--- a/v4/webkit/WebKitBackForwardListItem.go
+++ b/v4/webkit/WebKitBackForwardListItem.go
@@ -103,4 +103,8 @@ 	core.PuregoSafeRegister(&xBackForwardListItemGetOriginalUri, libs, "webkit_back_forward_list_item_get_original_uri")
 	core.PuregoSafeRegister(&xBackForwardListItemGetTitle, libs, "webkit_back_forward_list_item_get_title")
 	core.PuregoSafeRegister(&xBackForwardListItemGetUri, libs, "webkit_back_forward_list_item_get_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	BackForwardListItemGLibType()
 }
diff --git a/v4/webkit/WebKitColorChooserRequest.go b/v4/webkit/WebKitColorChooserRequest.go
index 066f0954fa9eff0c255eadfd506809851a41104f..617f951fd391cad86abae93789f13384c5410ed9 100644
--- a/v4/webkit/WebKitColorChooserRequest.go
+++ b/v4/webkit/WebKitColorChooserRequest.go
@@ -181,4 +181,8 @@ 	core.PuregoSafeRegister(&xColorChooserRequestGetElementRectangle, libs, "webkit_color_chooser_request_get_element_rectangle")
 	core.PuregoSafeRegister(&xColorChooserRequestGetRgba, libs, "webkit_color_chooser_request_get_rgba")
 	core.PuregoSafeRegister(&xColorChooserRequestSetRgba, libs, "webkit_color_chooser_request_set_rgba")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	ColorChooserRequestGLibType()
 }
diff --git a/v4/webkit/WebKitContextMenu.go b/v4/webkit/WebKitContextMenu.go
index acea4f67814b70d99797b515e791d03b0dc3533b..bacaa7d1907b6ac6cc83e42a1dd33ae22879fbf0 100644
--- a/v4/webkit/WebKitContextMenu.go
+++ b/v4/webkit/WebKitContextMenu.go
@@ -330,4 +330,8 @@ 	core.PuregoSafeRegister(&xContextMenuRemove, libs, "webkit_context_menu_remove")
 	core.PuregoSafeRegister(&xContextMenuRemoveAll, libs, "webkit_context_menu_remove_all")
 	core.PuregoSafeRegister(&xContextMenuSetUserData, libs, "webkit_context_menu_set_user_data")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	ContextMenuGLibType()
 }
diff --git a/v4/webkit/WebKitContextMenuItem.go b/v4/webkit/WebKitContextMenuItem.go
index 3ff2718f8796e161c351230bd9637ae9b981c16e..4a5e4952ea0a4f87ca45f7eb766d067a0836243a 100644
--- a/v4/webkit/WebKitContextMenuItem.go
+++ b/v4/webkit/WebKitContextMenuItem.go
@@ -252,4 +252,8 @@ 	core.PuregoSafeRegister(&xContextMenuItemGetSubmenu, libs, "webkit_context_menu_item_get_submenu")
 	core.PuregoSafeRegister(&xContextMenuItemIsSeparator, libs, "webkit_context_menu_item_is_separator")
 	core.PuregoSafeRegister(&xContextMenuItemSetSubmenu, libs, "webkit_context_menu_item_set_submenu")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	ContextMenuItemGLibType()
 }
diff --git a/v4/webkit/WebKitCookieManager.go b/v4/webkit/WebKitCookieManager.go
index 206a1322c1c601b54cb1b70d905ae397e6fc8ae4..4e336fa8649f3f6b7776f8f5df79d584bc5edc63 100644
--- a/v4/webkit/WebKitCookieManager.go
+++ b/v4/webkit/WebKitCookieManager.go
@@ -351,4 +351,8 @@ 	core.PuregoSafeRegister(&xCookieManagerReplaceCookiesFinish, libs, "webkit_cookie_manager_replace_cookies_finish")
 	core.PuregoSafeRegister(&xCookieManagerSetAcceptPolicy, libs, "webkit_cookie_manager_set_accept_policy")
 	core.PuregoSafeRegister(&xCookieManagerSetPersistentStorage, libs, "webkit_cookie_manager_set_persistent_storage")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	CookieManagerGLibType()
 }
diff --git a/v4/webkit/WebKitDownload.go b/v4/webkit/WebKitDownload.go
index 5cebd8ff08693f71bd781cf973d8bdd24f37563d..795314dea9120fae4587b7bbd735ab0e1be503f9 100644
--- a/v4/webkit/WebKitDownload.go
+++ b/v4/webkit/WebKitDownload.go
@@ -412,4 +412,8 @@ 	core.PuregoSafeRegister(&xDownloadGetWebView, libs, "webkit_download_get_web_view")
 	core.PuregoSafeRegister(&xDownloadSetAllowOverwrite, libs, "webkit_download_set_allow_overwrite")
 	core.PuregoSafeRegister(&xDownloadSetDestination, libs, "webkit_download_set_destination")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	DownloadGLibType()
 }
diff --git a/v4/webkit/WebKitEditorState.go b/v4/webkit/WebKitEditorState.go
index 07d4cb2541d8b6ffd5a12d4b094fcd840bd674a7..52795c62e2fc3a11a819c2f4db7935db9397f57d 100644
--- a/v4/webkit/WebKitEditorState.go
+++ b/v4/webkit/WebKitEditorState.go
@@ -188,4 +188,8 @@ 	core.PuregoSafeRegister(&xEditorStateIsPasteAvailable, libs, "webkit_editor_state_is_paste_available")
 	core.PuregoSafeRegister(&xEditorStateIsRedoAvailable, libs, "webkit_editor_state_is_redo_available")
 	core.PuregoSafeRegister(&xEditorStateIsUndoAvailable, libs, "webkit_editor_state_is_undo_available")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	EditorStateGLibType()
 }
diff --git a/v4/webkit/WebKitFaviconDatabase.go b/v4/webkit/WebKitFaviconDatabase.go
index f9c57727d87e6cba425417e8c6b73e0b6922078d..b2924e80d86292bd39e9aa9c9a3299e8cfdce2e3 100644
--- a/v4/webkit/WebKitFaviconDatabase.go
+++ b/v4/webkit/WebKitFaviconDatabase.go
@@ -194,4 +194,8 @@ 	core.PuregoSafeRegister(&xFaviconDatabaseGetFavicon, libs, "webkit_favicon_database_get_favicon")
 	core.PuregoSafeRegister(&xFaviconDatabaseGetFaviconFinish, libs, "webkit_favicon_database_get_favicon_finish")
 	core.PuregoSafeRegister(&xFaviconDatabaseGetFaviconUri, libs, "webkit_favicon_database_get_favicon_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	FaviconDatabaseGLibType()
 }
diff --git a/v4/webkit/WebKitFileChooserRequest.go b/v4/webkit/WebKitFileChooserRequest.go
index f229a39a72edc9f86a6a102bee97a6c8dcdaeb49..dfbeac583e4f656fc0110c5679ed6d94322212b8 100644
--- a/v4/webkit/WebKitFileChooserRequest.go
+++ b/v4/webkit/WebKitFileChooserRequest.go
@@ -218,4 +218,8 @@ 	core.PuregoSafeRegister(&xFileChooserRequestGetSelectMultiple, libs, "webkit_file_chooser_request_get_select_multiple")
 	core.PuregoSafeRegister(&xFileChooserRequestGetSelectedFiles, libs, "webkit_file_chooser_request_get_selected_files")
 	core.PuregoSafeRegister(&xFileChooserRequestSelectFiles, libs, "webkit_file_chooser_request_select_files")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	FileChooserRequestGLibType()
 }
diff --git a/v4/webkit/WebKitFindController.go b/v4/webkit/WebKitFindController.go
index 282abb0ca96f656bb967304610444936734add71..c24a6ec8a77999ebd4ad778ec7b37a3bedd54cf0 100644
--- a/v4/webkit/WebKitFindController.go
+++ b/v4/webkit/WebKitFindController.go
@@ -345,4 +345,8 @@ 	core.PuregoSafeRegister(&xFindControllerSearchFinish, libs, "webkit_find_controller_search_finish")
 	core.PuregoSafeRegister(&xFindControllerSearchNext, libs, "webkit_find_controller_search_next")
 	core.PuregoSafeRegister(&xFindControllerSearchPrevious, libs, "webkit_find_controller_search_previous")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	FindControllerGLibType()
 }
diff --git a/v4/webkit/WebKitFormSubmissionRequest.go b/v4/webkit/WebKitFormSubmissionRequest.go
index df9f94fc168f23bd4a6a64b3fde0497d5bcf5326..b4ae23eae3f961fca301d5c4a575eae2f48b1caa 100644
--- a/v4/webkit/WebKitFormSubmissionRequest.go
+++ b/v4/webkit/WebKitFormSubmissionRequest.go
@@ -98,4 +98,8 @@
 	core.PuregoSafeRegister(&xFormSubmissionRequestListTextFields, libs, "webkit_form_submission_request_list_text_fields")
 	core.PuregoSafeRegister(&xFormSubmissionRequestSubmit, libs, "webkit_form_submission_request_submit")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	FormSubmissionRequestGLibType()
 }
diff --git a/v4/webkit/WebKitGeolocationManager.go b/v4/webkit/WebKitGeolocationManager.go
index a589dc6039baa81743b083bcf4559df3f6ab44bc..64f68a22e37a5f6a5d9fb1be6eea7e50ff856ce3 100644
--- a/v4/webkit/WebKitGeolocationManager.go
+++ b/v4/webkit/WebKitGeolocationManager.go
@@ -267,4 +267,8 @@ 	core.PuregoSafeRegister(&xGeolocationManagerFailed, libs, "webkit_geolocation_manager_failed")
 	core.PuregoSafeRegister(&xGeolocationManagerGetEnableHighAccuracy, libs, "webkit_geolocation_manager_get_enable_high_accuracy")
 	core.PuregoSafeRegister(&xGeolocationManagerUpdatePosition, libs, "webkit_geolocation_manager_update_position")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	GeolocationManagerGLibType()
 }
diff --git a/v4/webkit/WebKitHitTestResult.go b/v4/webkit/WebKitHitTestResult.go
index ae0e40c6cc324e3db6a435e254510ce121dbc5b9..da638ebe53ae758613d3b834594bf6b9e5e68ad3 100644
--- a/v4/webkit/WebKitHitTestResult.go
+++ b/v4/webkit/WebKitHitTestResult.go
@@ -352,4 +352,8 @@ 	core.PuregoSafeRegister(&xHitTestResultGetLinkTitle, libs, "webkit_hit_test_result_get_link_title")
 	core.PuregoSafeRegister(&xHitTestResultGetLinkUri, libs, "webkit_hit_test_result_get_link_uri")
 	core.PuregoSafeRegister(&xHitTestResultGetMediaUri, libs, "webkit_hit_test_result_get_media_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	HitTestResultGLibType()
 }
diff --git a/v4/webkit/WebKitInputMethodContext.go b/v4/webkit/WebKitInputMethodContext.go
index f2d6cb0b9a7ef05e5d1e37918ef2263104484b37..fc206ce6b2359d2e83b458ff92467da069569bb5 100644
--- a/v4/webkit/WebKitInputMethodContext.go
+++ b/v4/webkit/WebKitInputMethodContext.go
@@ -1235,4 +1235,8 @@ 	core.PuregoSafeRegister(&xInputMethodContextSetEnablePreedit, libs, "webkit_input_method_context_set_enable_preedit")
 	core.PuregoSafeRegister(&xInputMethodContextSetInputHints, libs, "webkit_input_method_context_set_input_hints")
 	core.PuregoSafeRegister(&xInputMethodContextSetInputPurpose, libs, "webkit_input_method_context_set_input_purpose")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	InputMethodContextGLibType()
 }
diff --git a/v4/webkit/WebKitMediaKeySystemPermissionRequest.go b/v4/webkit/WebKitMediaKeySystemPermissionRequest.go
index 3840cfed0ce3284859a712966010da54fcf147af..7d77f79641b2e2fc14df5a9c3342745fa61c4a9e 100644
--- a/v4/webkit/WebKitMediaKeySystemPermissionRequest.go
+++ b/v4/webkit/WebKitMediaKeySystemPermissionRequest.go
@@ -98,4 +98,8 @@ 	core.PuregoSafeRegister(&xMediaKeySystemPermissionGetName, libs, "webkit_media_key_system_permission_get_name")
 
 	core.PuregoSafeRegister(&xMediaKeySystemPermissionRequestGLibType, libs, "webkit_media_key_system_permission_request_get_type")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	MediaKeySystemPermissionRequestGLibType()
 }
diff --git a/v4/webkit/WebKitNavigationPolicyDecision.go b/v4/webkit/WebKitNavigationPolicyDecision.go
index 8d1b6b700dee182f60e6f5ed739c1d23f1de2206..cbfe101367356d93adef7bdd98149ac45d7e8dd2 100644
--- a/v4/webkit/WebKitNavigationPolicyDecision.go
+++ b/v4/webkit/WebKitNavigationPolicyDecision.go
@@ -86,4 +86,8 @@ 	core.PuregoSafeRegister(&xNavigationPolicyDecisionGLibType, libs, "webkit_navigation_policy_decision_get_type")
 
 	core.PuregoSafeRegister(&xNavigationPolicyDecisionGetNavigationAction, libs, "webkit_navigation_policy_decision_get_navigation_action")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	NavigationPolicyDecisionGLibType()
 }
diff --git a/v4/webkit/WebKitNetworkSession.go b/v4/webkit/WebKitNetworkSession.go
index 248df216151e1551ac9adff3506006a227b1865d..7bd406a7d8ea1cb06363191c05f3d6fcaaee6ade 100644
--- a/v4/webkit/WebKitNetworkSession.go
+++ b/v4/webkit/WebKitNetworkSession.go
@@ -419,4 +419,8 @@
 	core.PuregoSafeRegister(&xNetworkSessionGetDefault, libs, "webkit_network_session_get_default")
 	core.PuregoSafeRegister(&xNetworkSessionSetMemoryPressureSettings, libs, "webkit_network_session_set_memory_pressure_settings")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	NetworkSessionGLibType()
 }
diff --git a/v4/webkit/WebKitNotification.go b/v4/webkit/WebKitNotification.go
index d2d34a8360e856c2f9eba540d068ef279675f629..6bd09010b168cd00cb57bbeebdd5ab51f5f518f0 100644
--- a/v4/webkit/WebKitNotification.go
+++ b/v4/webkit/WebKitNotification.go
@@ -203,4 +203,8 @@ 	core.PuregoSafeRegister(&xNotificationGetId, libs, "webkit_notification_get_id")
 	core.PuregoSafeRegister(&xNotificationGetTag, libs, "webkit_notification_get_tag")
 	core.PuregoSafeRegister(&xNotificationGetTitle, libs, "webkit_notification_get_title")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	NotificationGLibType()
 }
diff --git a/v4/webkit/WebKitOptionMenu.go b/v4/webkit/WebKitOptionMenu.go
index 3980d322672f1944f9136085ac183e4b21cfba9c..498441c59fee4e478f3f54dd405588597fc244a6 100644
--- a/v4/webkit/WebKitOptionMenu.go
+++ b/v4/webkit/WebKitOptionMenu.go
@@ -178,4 +178,8 @@ 	core.PuregoSafeRegister(&xOptionMenuGetItem, libs, "webkit_option_menu_get_item")
 	core.PuregoSafeRegister(&xOptionMenuGetNItems, libs, "webkit_option_menu_get_n_items")
 	core.PuregoSafeRegister(&xOptionMenuSelectItem, libs, "webkit_option_menu_select_item")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	OptionMenuGLibType()
 }
diff --git a/v4/webkit/WebKitPermissionRequest.go b/v4/webkit/WebKitPermissionRequest.go
index 96e2aae7d15f9b678a3bf0cae956794e656bfa76..82b092cc1bef960af4ca6343a5a662a2fea3365c 100644
--- a/v4/webkit/WebKitPermissionRequest.go
+++ b/v4/webkit/WebKitPermissionRequest.go
@@ -140,4 +140,8 @@
 	core.PuregoSafeRegister(&XWebkitPermissionRequestAllow, libs, "webkit_permission_request_allow")
 	core.PuregoSafeRegister(&XWebkitPermissionRequestDeny, libs, "webkit_permission_request_deny")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	PermissionRequestGLibType()
 }
diff --git a/v4/webkit/WebKitPolicyDecision.go b/v4/webkit/WebKitPolicyDecision.go
index f022daaa3764d72e7d22203e2d56bbf713a9919a..101faba365ed162d3a3672a69a0db2ad7a293213 100644
--- a/v4/webkit/WebKitPolicyDecision.go
+++ b/v4/webkit/WebKitPolicyDecision.go
@@ -333,4 +333,8 @@ 	core.PuregoSafeRegister(&xPolicyDecisionIgnore, libs, "webkit_policy_decision_ignore")
 	core.PuregoSafeRegister(&xPolicyDecisionUse, libs, "webkit_policy_decision_use")
 	core.PuregoSafeRegister(&xPolicyDecisionUseWithPolicies, libs, "webkit_policy_decision_use_with_policies")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	PolicyDecisionGLibType()
 }
diff --git a/v4/webkit/WebKitPrintOperation.go b/v4/webkit/WebKitPrintOperation.go
index 4bbb5afa2762e427d1497ee0cb69a6011102929f..885a0a8255f7996c513ce3b39dc338a9969a0116 100644
--- a/v4/webkit/WebKitPrintOperation.go
+++ b/v4/webkit/WebKitPrintOperation.go
@@ -269,4 +269,8 @@ 	core.PuregoSafeRegister(&xPrintOperationRunDialog, libs, "webkit_print_operation_run_dialog")
 	core.PuregoSafeRegister(&xPrintOperationSetPageSetup, libs, "webkit_print_operation_set_page_setup")
 	core.PuregoSafeRegister(&xPrintOperationSetPrintSettings, libs, "webkit_print_operation_set_print_settings")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	PrintOperationGLibType()
 }
diff --git a/v4/webkit/WebKitResponsePolicyDecision.go b/v4/webkit/WebKitResponsePolicyDecision.go
index d7b3f56e1b7a050952eece2e5981794530b97ed9..f21c5c1e1c7868a5447b5c315f15f04328f57412 100644
--- a/v4/webkit/WebKitResponsePolicyDecision.go
+++ b/v4/webkit/WebKitResponsePolicyDecision.go
@@ -135,4 +135,8 @@ 	core.PuregoSafeRegister(&xResponsePolicyDecisionGetResponse, libs, "webkit_response_policy_decision_get_response")
 	core.PuregoSafeRegister(&xResponsePolicyDecisionIsMainFrameMainResource, libs, "webkit_response_policy_decision_is_main_frame_main_resource")
 	core.PuregoSafeRegister(&xResponsePolicyDecisionIsMimeTypeSupported, libs, "webkit_response_policy_decision_is_mime_type_supported")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	ResponsePolicyDecisionGLibType()
 }
diff --git a/v4/webkit/WebKitSecurityManager.go b/v4/webkit/WebKitSecurityManager.go
index 44d0b5a186d0e2ad3e7cd15203d6d8e127c02251..89ffa07009961307d8444bd32371cbecffc0a917 100644
--- a/v4/webkit/WebKitSecurityManager.go
+++ b/v4/webkit/WebKitSecurityManager.go
@@ -221,4 +221,8 @@ 	core.PuregoSafeRegister(&xSecurityManagerUriSchemeIsLocal, libs, "webkit_security_manager_uri_scheme_is_local")
 	core.PuregoSafeRegister(&xSecurityManagerUriSchemeIsNoAccess, libs, "webkit_security_manager_uri_scheme_is_no_access")
 	core.PuregoSafeRegister(&xSecurityManagerUriSchemeIsSecure, libs, "webkit_security_manager_uri_scheme_is_secure")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	SecurityManagerGLibType()
 }
diff --git a/v4/webkit/WebKitSettings.go b/v4/webkit/WebKitSettings.go
index 657c2e64b1117a52801fae77739ee191c1adba85..c095a1af828d669bd433701d213a0657b8a5f135 100644
--- a/v4/webkit/WebKitSettings.go
+++ b/v4/webkit/WebKitSettings.go
@@ -2620,4 +2620,8 @@ 	core.PuregoSafeRegister(&xSettingsGetAllFeatures, libs, "webkit_settings_get_all_features")
 	core.PuregoSafeRegister(&xSettingsGetDevelopmentFeatures, libs, "webkit_settings_get_development_features")
 	core.PuregoSafeRegister(&xSettingsGetExperimentalFeatures, libs, "webkit_settings_get_experimental_features")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	SettingsGLibType()
 }
diff --git a/v4/webkit/WebKitURIRequest.go b/v4/webkit/WebKitURIRequest.go
index 2acb5b4e260e086e910ac58290b3dae403e7b43e..bd61d970c085d9c73c5347d8e82026a9bbedb6f6 100644
--- a/v4/webkit/WebKitURIRequest.go
+++ b/v4/webkit/WebKitURIRequest.go
@@ -144,4 +144,8 @@ 	core.PuregoSafeRegister(&xURIRequestGetHttpMethod, libs, "webkit_uri_request_get_http_method")
 	core.PuregoSafeRegister(&xURIRequestGetUri, libs, "webkit_uri_request_get_uri")
 	core.PuregoSafeRegister(&xURIRequestSetUri, libs, "webkit_uri_request_set_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	URIRequestGLibType()
 }
diff --git a/v4/webkit/WebKitURIResponse.go b/v4/webkit/WebKitURIResponse.go
index 2016eacd69320972626e9754f596707d99ded354..b8f16bb3c9accd7e4fd910b281f4b402c69d8e2c 100644
--- a/v4/webkit/WebKitURIResponse.go
+++ b/v4/webkit/WebKitURIResponse.go
@@ -188,4 +188,8 @@ 	core.PuregoSafeRegister(&xURIResponseGetStatusCode, libs, "webkit_uri_response_get_status_code")
 	core.PuregoSafeRegister(&xURIResponseGetSuggestedFilename, libs, "webkit_uri_response_get_suggested_filename")
 	core.PuregoSafeRegister(&xURIResponseGetUri, libs, "webkit_uri_response_get_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	URIResponseGLibType()
 }
diff --git a/v4/webkit/WebKitURISchemeRequest.go b/v4/webkit/WebKitURISchemeRequest.go
index d2e2ce84b6457ede1b3da31d4276808655f7c33e..2eb0a441c7da27f5186341544b7db3fd92cb6ac8 100644
--- a/v4/webkit/WebKitURISchemeRequest.go
+++ b/v4/webkit/WebKitURISchemeRequest.go
@@ -191,4 +191,8 @@ 	core.PuregoSafeRegister(&xURISchemeRequestGetScheme, libs, "webkit_uri_scheme_request_get_scheme")
 	core.PuregoSafeRegister(&xURISchemeRequestGetUri, libs, "webkit_uri_scheme_request_get_uri")
 	core.PuregoSafeRegister(&xURISchemeRequestGetWebView, libs, "webkit_uri_scheme_request_get_web_view")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	URISchemeRequestGLibType()
 }
diff --git a/v4/webkit/WebKitURISchemeResponse.go b/v4/webkit/WebKitURISchemeResponse.go
index f933a1ac0830b04cbf55a7682af98a7fca062947..b3fa24499546b702095b5072ed105a2acdc6fc2a 100644
--- a/v4/webkit/WebKitURISchemeResponse.go
+++ b/v4/webkit/WebKitURISchemeResponse.go
@@ -139,4 +139,8 @@ 	core.PuregoSafeRegister(&xURISchemeResponseSetContentType, libs, "webkit_uri_scheme_response_set_content_type")
 	core.PuregoSafeRegister(&xURISchemeResponseSetHttpHeaders, libs, "webkit_uri_scheme_response_set_http_headers")
 	core.PuregoSafeRegister(&xURISchemeResponseSetStatus, libs, "webkit_uri_scheme_response_set_status")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	URISchemeResponseGLibType()
 }
diff --git a/v4/webkit/WebKitUserContentFilterStore.go b/v4/webkit/WebKitUserContentFilterStore.go
index d00e82dd358535c9c19e4cebc6c447ad7c88aaeb..957c8396e46fc40201de128d9b4b0cf7fb110219 100644
--- a/v4/webkit/WebKitUserContentFilterStore.go
+++ b/v4/webkit/WebKitUserContentFilterStore.go
@@ -287,4 +287,8 @@ 	core.PuregoSafeRegister(&xUserContentFilterStoreSaveFinish, libs, "webkit_user_content_filter_store_save_finish")
 	core.PuregoSafeRegister(&xUserContentFilterStoreSaveFromFile, libs, "webkit_user_content_filter_store_save_from_file")
 	core.PuregoSafeRegister(&xUserContentFilterStoreSaveFromFileFinish, libs, "webkit_user_content_filter_store_save_from_file_finish")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	UserContentFilterStoreGLibType()
 }
diff --git a/v4/webkit/WebKitUserContentManager.go b/v4/webkit/WebKitUserContentManager.go
index b8cb3f9164e55a5d2037092c6faa42580927a017..ea0cb5286416444b44f46ad427e2581f5b84fc22 100644
--- a/v4/webkit/WebKitUserContentManager.go
+++ b/v4/webkit/WebKitUserContentManager.go
@@ -418,4 +418,8 @@ 	core.PuregoSafeRegister(&xUserContentManagerRemoveScript, libs, "webkit_user_content_manager_remove_script")
 	core.PuregoSafeRegister(&xUserContentManagerRemoveStyleSheet, libs, "webkit_user_content_manager_remove_style_sheet")
 	core.PuregoSafeRegister(&xUserContentManagerUnregisterScriptMessageHandler, libs, "webkit_user_content_manager_unregister_script_message_handler")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	UserContentManagerGLibType()
 }
diff --git a/v4/webkit/WebKitUserMediaPermissionRequest.go b/v4/webkit/WebKitUserMediaPermissionRequest.go
index 185a1a8dc43805b571a1cf49fa43cf27dd8c212e..f20017f5a3bd2280cda7782a15914a35dde207b8 100644
--- a/v4/webkit/WebKitUserMediaPermissionRequest.go
+++ b/v4/webkit/WebKitUserMediaPermissionRequest.go
@@ -131,4 +131,8 @@ 	core.PuregoSafeRegister(&xUserMediaPermissionIsForVideoDevice, libs, "webkit_user_media_permission_is_for_video_device")
 
 	core.PuregoSafeRegister(&xUserMediaPermissionRequestGLibType, libs, "webkit_user_media_permission_request_get_type")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	UserMediaPermissionRequestGLibType()
 }
diff --git a/v4/webkit/WebKitUserMessage.go b/v4/webkit/WebKitUserMessage.go
index 42227bcfc07ad986bae0aae6dd6db80792cc9fb2..548f87a420c8821475337555b6f676f5546c8f82 100644
--- a/v4/webkit/WebKitUserMessage.go
+++ b/v4/webkit/WebKitUserMessage.go
@@ -228,4 +228,8 @@ 	core.PuregoSafeRegister(&xUserMessageGetName, libs, "webkit_user_message_get_name")
 	core.PuregoSafeRegister(&xUserMessageGetParameters, libs, "webkit_user_message_get_parameters")
 	core.PuregoSafeRegister(&xUserMessageSendReply, libs, "webkit_user_message_send_reply")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	UserMessageGLibType()
 }
diff --git a/v4/webkit/WebKitWebContext.go b/v4/webkit/WebKitWebContext.go
index 205a45c3d72d1dad01b408b77dbe1aaf62c1f130..b0e8884b768383c51ae03c11ada5709b365258eb 100644
--- a/v4/webkit/WebKitWebContext.go
+++ b/v4/webkit/WebKitWebContext.go
@@ -647,4 +647,8 @@ 	core.PuregoSafeRegister(&xWebContextSetWebProcessExtensionsInitializationUserData, libs, "webkit_web_context_set_web_process_extensions_initialization_user_data")
 
 	core.PuregoSafeRegister(&xWebContextGetDefault, libs, "webkit_web_context_get_default")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebContextGLibType()
 }
diff --git a/v4/webkit/WebKitWebInspector.go b/v4/webkit/WebKitWebInspector.go
index ad6eea008ac4a2bf3c00921a08d6b832556e7554..4c65e4773706d8d06aff1a3fe9594f13d3d2c0b8 100644
--- a/v4/webkit/WebKitWebInspector.go
+++ b/v4/webkit/WebKitWebInspector.go
@@ -366,4 +366,8 @@ 	core.PuregoSafeRegister(&xWebInspectorGetWebView, libs, "webkit_web_inspector_get_web_view")
 	core.PuregoSafeRegister(&xWebInspectorIsAttached, libs, "webkit_web_inspector_is_attached")
 	core.PuregoSafeRegister(&xWebInspectorShow, libs, "webkit_web_inspector_show")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebInspectorGLibType()
 }
diff --git a/v4/webkit/WebKitWebResource.go b/v4/webkit/WebKitWebResource.go
index 234269a259331a49ad52e8f7e9face572abbbb5d..a37b7f73a2ebe62928883d2c4227168fc1a6d944 100644
--- a/v4/webkit/WebKitWebResource.go
+++ b/v4/webkit/WebKitWebResource.go
@@ -267,4 +267,8 @@ 	core.PuregoSafeRegister(&xWebResourceGetDataFinish, libs, "webkit_web_resource_get_data_finish")
 	core.PuregoSafeRegister(&xWebResourceGetResponse, libs, "webkit_web_resource_get_response")
 	core.PuregoSafeRegister(&xWebResourceGetUri, libs, "webkit_web_resource_get_uri")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebResourceGLibType()
 }
diff --git a/v4/webkit/WebKitWebView.go b/v4/webkit/WebKitWebView.go
index 6d2713511e6c2f864b439275e18d93d29e470021..7b230d68ca4766a54073a6cec6ad344d1c1cb949 100644
--- a/v4/webkit/WebKitWebView.go
+++ b/v4/webkit/WebKitWebView.go
@@ -4605,4 +4605,8 @@ 	core.PuregoSafeRegister(&xWebViewStopLoading, libs, "webkit_web_view_stop_loading")
 	core.PuregoSafeRegister(&xWebViewTerminateWebProcess, libs, "webkit_web_view_terminate_web_process")
 	core.PuregoSafeRegister(&xWebViewTryClose, libs, "webkit_web_view_try_close")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebViewGLibType()
 }
diff --git a/v4/webkit/WebKitWebsiteDataAccessPermissionRequest.go b/v4/webkit/WebKitWebsiteDataAccessPermissionRequest.go
index 645dfe0501dab41e55ea32cd33aa938b31ac7ccf..98e974012dbe2d5c0eaf4999d65188e7a82b55c3 100644
--- a/v4/webkit/WebKitWebsiteDataAccessPermissionRequest.go
+++ b/v4/webkit/WebKitWebsiteDataAccessPermissionRequest.go
@@ -104,4 +104,8 @@
 	core.PuregoSafeRegister(&xWebsiteDataAccessPermissionRequestGetCurrentDomain, libs, "webkit_website_data_access_permission_request_get_current_domain")
 	core.PuregoSafeRegister(&xWebsiteDataAccessPermissionRequestGetRequestingDomain, libs, "webkit_website_data_access_permission_request_get_requesting_domain")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebsiteDataAccessPermissionRequestGLibType()
 }
diff --git a/v4/webkit/WebKitWebsiteDataManager.go b/v4/webkit/WebKitWebsiteDataManager.go
index 53108dfa362161a24812aa151f285a1e3de557fe..2fa458c53b52147ba222f4bd7467db4b5be72321 100644
--- a/v4/webkit/WebKitWebsiteDataManager.go
+++ b/v4/webkit/WebKitWebsiteDataManager.go
@@ -524,4 +524,8 @@ 	core.PuregoSafeRegister(&xWebsiteDataManagerRemove, libs, "webkit_website_data_manager_remove")
 	core.PuregoSafeRegister(&xWebsiteDataManagerRemoveFinish, libs, "webkit_website_data_manager_remove_finish")
 	core.PuregoSafeRegister(&xWebsiteDataManagerSetFaviconsEnabled, libs, "webkit_website_data_manager_set_favicons_enabled")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebsiteDataManagerGLibType()
 }
diff --git a/v4/webkit/WebKitWebsitePolicies.go b/v4/webkit/WebKitWebsitePolicies.go
index e7ff279654aa2d198c3dd200a4aa8b8fbecc08f1..8076fd8dfdcec6ef6a0b5adca5b6b49ddf0993ed 100644
--- a/v4/webkit/WebKitWebsitePolicies.go
+++ b/v4/webkit/WebKitWebsitePolicies.go
@@ -137,4 +137,8 @@ 	core.PuregoSafeRegister(&xNewWebsitePoliciesWithPolicies, libs, "webkit_website_policies_new_with_policies")
 
 	core.PuregoSafeRegister(&xWebsitePoliciesGetAutoplayPolicy, libs, "webkit_website_policies_get_autoplay_policy")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WebsitePoliciesGLibType()
 }
diff --git a/v4/webkit/WebKitWindowProperties.go b/v4/webkit/WebKitWindowProperties.go
index ec59e85d3d7d507b70dbbd23c444f4efdba9de3d..ffcb542a68de38e354b86aa38bfddcd3fd1f87dd 100644
--- a/v4/webkit/WebKitWindowProperties.go
+++ b/v4/webkit/WebKitWindowProperties.go
@@ -336,4 +336,8 @@ 	core.PuregoSafeRegister(&xWindowPropertiesGetScrollbarsVisible, libs, "webkit_window_properties_get_scrollbars_visible")
 	core.PuregoSafeRegister(&xWindowPropertiesGetStatusbarVisible, libs, "webkit_window_properties_get_statusbar_visible")
 	core.PuregoSafeRegister(&xWindowPropertiesGetToolbarVisible, libs, "webkit_window_properties_get_toolbar_visible")
 
+	// Manually register types since they aren't being automatically registered when
+	// the library is loaded
+	// See https://bugs.webkit.org/show_bug.cgi?id=175937
+	WindowPropertiesGLibType()
 }

