From df692eb08726f66ab8b7d19b011f7a2938d1e0fa Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 28 Jan 2026 23:47:01 -0800
Subject: [PATCH] GIR: Fix type conversion list to use LLP64 integer lengths correctly for `gint` and `guint`

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/gir/types/types.go b/internal/gir/types/types.go
index bc9d4447f15d5b624a4351392675d6ba7ee3f54d..252b02668fe85cd96e4001be3e9914340ae18732 100644
--- a/internal/gir/types/types.go
+++ b/internal/gir/types/types.go
@@ -12,22 +12,38 @@ 	"github.com/jwijenbergh/puregotk/internal/gir/util"
 )
 
 // convList maps the given GIR primitive type to a Go builtin type.
-// See https://github.com/diamondburned/gotk4/blob/fd960d20b525a07580938d10a214336bafb47d12/gir/girgen/types/types.go#LL483C1-L512C2
+// Based on https://github.com/diamondburned/gotk4/blob/fd960d20b525a07580938d10a214336bafb47d12/gir/girgen/types/types.go#LL483C1-L512C2
+// Adapted to match https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models; we only support Linux on AMD64/ARM64, so this is based on LLP64
+// | Data model | short | int | long int | long long | Pointer, size_t | Sample operating systems |
+// |------------|-------|-----|----------|-----------|-----------------|--------------------------|
+// | ILP32 | 16 | 32 | 32 | 64 | 32 | x32 and arm64ilp32 ABIs on Linux systems; MIPS N32 ABI |
+// | LLP64 | 16 | 32 | 32 | 64 | 64 | Microsoft Windows (x86-64, IA-64, and ARM64) using Visual C++; and MinGW |
+// | LP64 | 16 | 32 | 64 | 64 | 64 | Most Unix and Unix-like systems, e.g., Solaris, Linux, BSD, macOS. Windows when using Cygwin; z/OS |
+// | ILP64 | 16 | 64 | 64 | 64 | 64 | HAL Computer Systems port of Solaris to the SPARC64 |
+// | SILP64 | 64 | 64 | 64 | 64 | 64 | Classic UNICOS (versus UNICOS/mp, etc.) |
 var convList = map[string]string{
 	"none":     "",
 	"gboolean": "bool",
 	"gfloat":   "float32",
 	"gdouble":  "float64",
-	"gint":     "int",
-	"gssize":   "int",
-	"gint8":    "int8",
-	"gint16":   "int16",
-	"gshort":   "int16",
-	"gint32":   "int32",
-	"glong":    "int32",
-	"int32":    "int32",
-	"gint64":   "int64",
-	"guint":    "uint",
+
+	// Not just `int` since the Go int is 64 bit on 64 bit systems and 32 bit on 32 bit systems,
+	// whereas the LLP64 int is always 32-bit
+	"gint": "int32",
+
+	"gssize": "int",
+	"gint8":  "int8",
+	"gint16": "int16",
+	"gshort": "int16",
+	"gint32": "int32",
+	"glong":  "int32",
+	"int32":  "int32",
+	"gint64": "int64",
+
+	// Not just `uint` since the Go uint is 64 bit on 64 bit systems and 32 bit on 32 bit systems,
+	// whereas the LLP64 int is always 32-bit
+	"guint": "uint32",
+
 	"gsize":    "uint",
 	"guchar":   "byte",
 	"gchar":    "byte",

