From b4c7ccb6d00da76f84263961132202e68d1a8ffc Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 22 Apr 2026 22:09:59 -0700
Subject: [PATCH] chore: Prepare new release

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/resources/index.go b/assets/resources/index.go
index 494da8a0314cbd77f236d12a4a2a96050512b4fc..e60dddce88864dbc457e11f1aa73f9b9d2f4b12a 100644
--- a/assets/resources/index.go
+++ b/assets/resources/index.go
@@ -12,7 +12,7 @@ //go:generate sh -c "msgfmt --xml -L metainfo --template ../../assets/resources/metainfo.xml.in -d ../../po -o - -f | sed 's|/LC[-_]MESSAGES/default||g' > ../../assets/resources/metainfo.xml"
 
 const (
 	AppID      = "com.pojtinger.felicitas.Sessions"
-	AppVersion = "0.1.16"
+	AppVersion = "0.1.17"
 )
 
 //go:generate sh -c "blueprint-compiler batch-compile . . *.blp && glib-compile-resources *.gresource.xml"
diff --git a/assets/resources/metainfo.xml.in b/assets/resources/metainfo.xml.in
index 666a704799c67c3ce6ac6ca1abc95ea708dfb32e..c220c62d62faa5e8cf5ec6b20cc4e7bec8a980c6 100644
--- a/assets/resources/metainfo.xml.in
+++ b/assets/resources/metainfo.xml.in
@@ -72,6 +72,12 @@         </screenshot>
     </screenshots>
 
     <releases>
+        <release version="0.1.17" date='2026-04-22T00:00:00Z00:00'>
+            <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.17</url>
+            <description translate="no">
+                <p>Add Bengali translations and bump dependencies</p>
+            </description>
+        </release>
         <release version="0.1.16" date='2026-04-08T00:00:00Z00:00'>
             <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.16</url>
             <description translate="no">

