From 54ad63440170d113a58f54384b652defed8d3470 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 07 May 2026 19:02:06 -0700
Subject: [PATCH] chore: Prepare new release

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/resources/index.go b/assets/resources/index.go
index e60dddce88864dbc457e11f1aa73f9b9d2f4b12a..1b440150dc2656e88ba6f416830dddfc9038e9a3 100644
--- a/assets/resources/index.go
+++ b/assets/resources/index.go
@@ -12,7 +12,7 @@ //go:generate sh -c "msgfmt --xml -L metainfo --template ../../assets/resources/metainfo.xml.in -d ../../po -o - -f | sed 's|/LC[-_]MESSAGES/default||g' > ../../assets/resources/metainfo.xml"
 
 const (
 	AppID      = "com.pojtinger.felicitas.Sessions"
-	AppVersion = "0.1.17"
+	AppVersion = "0.1.18"
 )
 
 //go:generate sh -c "blueprint-compiler batch-compile . . *.blp && glib-compile-resources *.gresource.xml"
diff --git a/assets/resources/metainfo.xml.in b/assets/resources/metainfo.xml.in
index c220c62d62faa5e8cf5ec6b20cc4e7bec8a980c6..037f2850023895cec75bac83213c4bb0cdbe021b 100644
--- a/assets/resources/metainfo.xml.in
+++ b/assets/resources/metainfo.xml.in
@@ -72,6 +72,12 @@         </screenshot>
     </screenshots>
 
     <releases>
+        <release version="0.1.18" date='2026-05-07T00:00:00Z00:00'>
+            <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.18</url>
+            <description translate="no">
+                <p>Add alarming state with flashing display and update translations</p>
+            </description>
+        </release>
         <release version="0.1.17" date='2026-04-22T00:00:00Z00:00'>
             <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.17</url>
             <description translate="no">

