From 8b59dfdb585a34c97788f46329f11a7be20b29ea Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 05 May 2026 13:58:03 -0700
Subject: [PATCH] feat: Allow transitioning with plus/minus timer state back into counting down state from counting down state

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/sessions-fsm/main.go b/cmd/sessions-fsm/main.go
index 4731d4721a3d71475fb25ede77f4558713daecb9..0354f8984d7ae31075107f56043e8c13f32407f3 100644
--- a/cmd/sessions-fsm/main.go
+++ b/cmd/sessions-fsm/main.go
@@ -53,7 +53,6 @@ 	s.machine.
 		Configure(statePaused).
 		PermitReentry(triggerPlusTimer, s.mustBeBelowMaxRemainingTime).
 		OnEntryFrom(triggerPlusTimer, s.increaseRemainingTime)
-
 	s.machine.
 		Configure(statePaused).
 		PermitReentry(triggerMinusTimer, s.mustBeAboveMinRemainingTime).
@@ -67,6 +66,15 @@ 		Configure(stateDragging).
 		Permit(triggerStopDragging, stateCountingDown, s.validRemainingTime)
 
 	s.machine.Configure(stateCountingDown).Permit(triggerStartDragging, stateDragging)
+
+	s.machine.
+		Configure(stateCountingDown).
+		PermitReentry(triggerPlusTimer, s.mustBeBelowMaxRemainingTime).
+		OnEntryFrom(triggerPlusTimer, s.increaseRemainingTime)
+	s.machine.
+		Configure(stateCountingDown).
+		PermitReentry(triggerMinusTimer, s.mustBeAboveMinRemainingTime).
+		OnEntryFrom(triggerMinusTimer, s.decreaseRemainingTime)
 
 	return s
 }
@@ -162,5 +170,17 @@ 	}
 
 	if err := s.StopDragging(ctx, initialRemainingTime+remainingTimerAdjustmentInterval*2); err != nil {
 		panic(err)
+	}
+
+	for range 50 {
+		if err := s.PlusTimer(ctx); err != nil {
+			panic(err)
+		}
+	}
+
+	for range 30 {
+		if err := s.MinusTimer(ctx); err != nil {
+			panic(err)
+		}
 	}
 }
diff --git a/cmd/sessions-fsm/main_test.go b/cmd/sessions-fsm/main_test.go
index affba3878e3dbde3170dcd135492f6e20e899cd4..f3e64b33b750f448b5074e86470eb049d2d6dbea 100644
--- a/cmd/sessions-fsm/main_test.go
+++ b/cmd/sessions-fsm/main_test.go
@@ -31,21 +31,28 @@ 			expectErrAt: 1,
 		},
 	}
 	for _, tt := range plusTimerTests {
-		t.Run(
-			fmt.Sprintf("initial %v plusTimes %v", tt.initial, tt.plusTimes),
-			func(t *testing.T) {
-				s := newStateMachine(tt.initial)
+		for _, fromCountingDown := range []bool{true, false} {
+			t.Run(
+				fmt.Sprintf("initial %v plusTimes %v fromCountingDown %v", tt.initial, tt.plusTimes, fromCountingDown),
+				func(t *testing.T) {
+					s := newStateMachine(tt.initial)
+
+					if fromCountingDown {
+						require.NoError(t, s.StartDragging(t.Context()))
+						require.NoError(t, s.StopDragging(t.Context(), initialRemainingTime))
+					}
 
-				for i := range tt.plusTimes {
-					err := s.PlusTimer(t.Context())
-					if i == tt.expectErrAt {
-						require.Error(t, err)
-					} else {
-						require.NoError(t, err)
+					for i := range tt.plusTimes {
+						err := s.PlusTimer(t.Context())
+						if i == tt.expectErrAt {
+							require.Error(t, err)
+						} else {
+							require.NoError(t, err)
+						}
 					}
-				}
-			},
-		)
+				},
+			)
+		}
 	}
 }
 
@@ -72,21 +79,28 @@ 			expectErrAt: 1,
 		},
 	}
 	for _, tt := range minusTimerTests {
-		t.Run(
-			fmt.Sprintf("initial %v plusTimes %v", tt.initial, tt.minusTimes),
-			func(t *testing.T) {
-				s := newStateMachine(tt.initial)
+		for _, fromCountingDown := range []bool{true, false} {
+			t.Run(
+				fmt.Sprintf("initial %v plusTimes %v fromCountingDown %v", tt.initial, tt.minusTimes, fromCountingDown),
+				func(t *testing.T) {
+					s := newStateMachine(tt.initial)
+
+					if fromCountingDown {
+						require.NoError(t, s.StartDragging(t.Context()))
+						require.NoError(t, s.StopDragging(t.Context(), initialRemainingTime))
+					}
 
-				for i := range tt.minusTimes {
-					err := s.MinusTimer(t.Context())
-					if i == tt.expectErrAt {
-						require.Error(t, err)
-					} else {
-						require.NoError(t, err)
+					for i := range tt.minusTimes {
+						err := s.MinusTimer(t.Context())
+						if i == tt.expectErrAt {
+							require.Error(t, err)
+						} else {
+							require.NoError(t, err)
+						}
 					}
-				}
-			},
-		)
+				},
+			)
+		}
 	}
 }
 

