From a708bd482207e6a41f83774bf3d348381fa65901 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 14 May 2026 19:05:40 -0700
Subject: [PATCH] feat: Stop timer when dragging from counting down state so that we don't continue counting down

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/sessions-fsm/main.go b/cmd/sessions-fsm/main.go
index 29577aba9491c4173778001deb831e7aee877d13..bd7bd680a3a08d2292a24996cb013a4552508d00 100644
--- a/cmd/sessions-fsm/main.go
+++ b/cmd/sessions-fsm/main.go
@@ -67,6 +67,7 @@ 	hooks *hooks
 
 	machine         *stateless.StateMachine
 	ticker          *time.Ticker
+	tickerCtx       context.Context
 	cancelTickerCtx context.CancelFunc
 }
 
@@ -111,8 +112,8 @@
 	// TODO: When incrementing/decrementing from the counting down state, we need to round the current remaining
 	// time down to the nearest lower multiple of remainingTimerAdjustmentInterval, increment from there, set that as the
 	// new initial time (& maybe call the onInitialRemainingTimeChange hook? Might not be necessary, we will update the UI with the onCurrentRemainingTimeTick),
-	// stop the _internal_ timer (without calling the hooks/changing state), manually call onCurrentRemainingTimeTick to reflect the new position, and
-	// restart the _internal_ timer (without calling the hooks/changing state)
+	// stop the _internal_ timer (without calling the hooks/changing state), manually call onCurrentRemainingTimeTick to reflect the new position. We do not
+	// need to manually re-start the internal timer, that should automatically happen when entering stateCountingDown already
 
 	// From stopped state, we can start dragging
 	s.machine.Configure(stateStopped).Permit(triggerStartDragging, stateDragging)
@@ -127,8 +128,12 @@
 	// When we enter the counting down state by stopping to drag, we set the initial remaining time
 	s.machine.Configure(stateCountingDown).OnEntryFrom(triggerStopDragging, s.setInitialRemainingTime)
 
-	// From counting down state, we can start dragging as well
-	s.machine.Configure(stateCountingDown).Permit(triggerStartDragging, stateDragging)
+	// From counting down state, we can start dragging as well. When we start dragging while in counting down state,
+	// we stop the timer
+	s.machine.
+		Configure(stateCountingDown).
+		Permit(triggerStartDragging, stateDragging).
+		OnExitWith(triggerStartDragging, s.stopTimerWithoutHooks)
 
 	// From counting down state, we can stop/reset and then restart the timer
 	s.machine.Configure(stateCountingDown).Permit(triggerStopTimer, stateStopped)
@@ -270,13 +275,12 @@ 	}
 
 	s.currentRemainingTime = s.initialRemainingTime
 	s.ticker = time.NewTicker(tickerInterval)
-	tickerCtx, cancelTickerCtx := context.WithCancel(s.ctx)
-	s.cancelTickerCtx = cancelTickerCtx
+	s.tickerCtx, s.cancelTickerCtx = context.WithCancel(s.ctx)
 
 	go func() {
 		for {
 			select {
-			case <-tickerCtx.Done(): // tickerCtx derives from s.ctx so this catches both
+			case <-s.tickerCtx.Done(): // tickerCtx derives from s.ctx so this catches both
 				return
 
 			case <-s.ticker.C:
@@ -313,13 +317,21 @@ 	if err := s.hooks.onBeforeStoppingTimer(ctx); err != nil {
 		return err
 	}
 
-	s.ticker.Stop()
-	s.cancelTickerCtx()
+	if err := s.stopTimerWithoutHooks(ctx, args...); err != nil {
+		return err
+	}
 
 	s.log.InfoContext(ctx, "Calling onAfterStoppingTimer hook")
 	if err := s.hooks.onAfterStoppingTimer(ctx); err != nil {
 		return err
 	}
+
+	return nil
+}
+
+func (s *stateMachine) stopTimerWithoutHooks(ctx context.Context, args ...any) error {
+	s.ticker.Stop()
+	s.cancelTickerCtx()
 
 	return nil
 }
diff --git a/cmd/sessions-fsm/main_test.go b/cmd/sessions-fsm/main_test.go
index 9ff3c5e3b673eb2e4ef7971a19d936a6158ab7e6..62609aa6ac78952627b6eaf3027dc13c1305ffc1 100644
--- a/cmd/sessions-fsm/main_test.go
+++ b/cmd/sessions-fsm/main_test.go
@@ -2,6 +2,7 @@ package main
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"testing"
 	"testing/synctest"
@@ -167,9 +168,10 @@ }
 
 func TestStartDragging(t *testing.T) {
 	var startDraggingTests = []struct {
-		name      string
-		prepare   func(*stateMachine) error
-		expectErr bool
+		name         string
+		prepare      func(*stateMachine) error
+		expectErr    bool
+		postRunCheck func(*stateMachine) error
 	}{
 		{
 			name: "can transition from initial state to dragging",
@@ -207,6 +209,24 @@ 				return sm.timerFinished(t.Context())
 			},
 			expectErr: true,
 		},
+		{
+			name: "transitioning from counting down state to dragging cancels running timer",
+			prepare: func(sm *stateMachine) error {
+				if err := sm.StartDragging(t.Context()); err != nil {
+					return err
+				}
+
+				return sm.StopDragging(t.Context(), initialRemainingTime)
+			},
+			expectErr: false,
+			postRunCheck: func(sm *stateMachine) error {
+				if sm.tickerCtx.Err() == nil {
+					return errors.New("timer is still running")
+				}
+
+				return nil
+			},
+		},
 	}
 	for _, tt := range startDraggingTests {
 		t.Run(
@@ -238,6 +258,10 @@ 				if tt.expectErr {
 					require.Error(t, err)
 				} else {
 					require.NoError(t, err)
+				}
+
+				if hook := tt.postRunCheck; hook != nil {
+					require.NoError(t, hook(s))
 				}
 			},
 		)

