From a90722847965c9e0b9e5dab735e632d03c5aa30a Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 05 May 2026 13:44:12 -0700
Subject: [PATCH] feat: Allow transitioning from counting down state to dragging state

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/sessions-fsm/main.go b/cmd/sessions-fsm/main.go
index 8cac8d221f2acfa182cf86c91f9a1f2536a8951d..4731d4721a3d71475fb25ede77f4558713daecb9 100644
--- a/cmd/sessions-fsm/main.go
+++ b/cmd/sessions-fsm/main.go
@@ -66,6 +66,8 @@ 	s.machine.
 		Configure(stateDragging).
 		Permit(triggerStopDragging, stateCountingDown, s.validRemainingTime)
 
+	s.machine.Configure(stateCountingDown).Permit(triggerStartDragging, stateDragging)
+
 	return s
 }
 
@@ -151,6 +153,14 @@ 		panic(err)
 	}
 
 	if err := s.StopDragging(ctx, initialRemainingTime+remainingTimerAdjustmentInterval*3); err != nil {
+		panic(err)
+	}
+
+	if err := s.StartDragging(ctx); err != nil {
+		panic(err)
+	}
+
+	if err := s.StopDragging(ctx, initialRemainingTime+remainingTimerAdjustmentInterval*2); err != nil {
 		panic(err)
 	}
 }
diff --git a/cmd/sessions-fsm/main_test.go b/cmd/sessions-fsm/main_test.go
index 5807ac999d3e3edb4ba678a9779d63a5c2bc3e05..affba3878e3dbde3170dcd135492f6e20e899cd4 100644
--- a/cmd/sessions-fsm/main_test.go
+++ b/cmd/sessions-fsm/main_test.go
@@ -110,6 +110,17 @@ 				return sm.StartDragging(t.Context())
 			},
 			expectErr: true,
 		},
+		{
+			name: "can transition from counting down state to dragging",
+			prepare: func(sm *stateMachine) error {
+				if err := sm.StartDragging(t.Context()); err != nil {
+					return err
+				}
+
+				return sm.StopDragging(t.Context(), initialRemainingTime)
+			},
+			expectErr: false,
+		},
 	}
 	for _, tt := range startDraggingTests {
 		t.Run(

