From f17720b65e84292a0b9162be3a6aecae469c56fd Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 14 May 2026 01:58:46 -0700
Subject: [PATCH] fix: Don't comment out onStopAlarm

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/sessions-fsm/main.go b/cmd/sessions-fsm/main.go
index 4f658ca801634feccdcad4d5c37223ab7a6ba419..29577aba9491c4173778001deb831e7aee877d13 100644
--- a/cmd/sessions-fsm/main.go
+++ b/cmd/sessions-fsm/main.go
@@ -55,7 +55,7 @@ 	onAfterStoppingTimer  func(ctx context.Context) error
 
 	onStartAlarm func(ctx context.Context) error
 
-	// onStopAlarm func(ctx context.Context) error
+	onStopAlarm func(ctx context.Context) error
 }
 
 type stateMachine struct {

