From c23636fe2c885cf48e020bab19454444ac6a47f7 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 26 May 2026 23:37:46 -0700
Subject: [PATCH] chore: Prepare new release

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/resources/index.go b/assets/resources/index.go
index 1b440150dc2656e88ba6f416830dddfc9038e9a3..101f963490a5fb40acb4ef4b2496fc463f6a5b00 100644
--- a/assets/resources/index.go
+++ b/assets/resources/index.go
@@ -12,7 +12,7 @@ //go:generate sh -c "msgfmt --xml -L metainfo --template ../../assets/resources/metainfo.xml.in -d ../../po -o - -f | sed 's|/LC[-_]MESSAGES/default||g' > ../../assets/resources/metainfo.xml"
 
 const (
 	AppID      = "com.pojtinger.felicitas.Sessions"
-	AppVersion = "0.1.18"
+	AppVersion = "0.1.19"
 )
 
 //go:generate sh -c "blueprint-compiler batch-compile . . *.blp && glib-compile-resources *.gresource.xml"
diff --git a/assets/resources/metainfo.xml.in b/assets/resources/metainfo.xml.in
index 037f2850023895cec75bac83213c4bb0cdbe021b..2e13038da7a61dba4d126a08396729972a606d8f 100644
--- a/assets/resources/metainfo.xml.in
+++ b/assets/resources/metainfo.xml.in
@@ -72,6 +72,12 @@         </screenshot>
     </screenshots>
 
     <releases>
+        <release version="0.1.19" date='2026-05-26T00:00:00Z00:00'>
+            <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.19</url>
+            <description translate="no">
+                <p>Improve internals to better handle unexpected states and add new icon</p>
+            </description>
+        </release>
         <release version="0.1.18" date='2026-05-07T00:00:00Z00:00'>
             <url type="details">https://github.com/pojntfx/sessions/releases/tag/v0.1.18</url>
             <description translate="no">

