From e1e719ac1668b47310cbe3f96717b49519593308 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 22 May 2026 20:39:31 -0700
Subject: [PATCH] feat: Add support for adjusting initial remaining time with plus and minus buttons via state machine

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/components/main_window.go b/internal/components/main_window.go
index 948adf7e5d6565736d28b3f8b44ad392f013d999..cda0c2527463282e6161c49033d121dbc02d380c 100644
--- a/internal/components/main_window.go
+++ b/internal/components/main_window.go
@@ -83,16 +83,28 @@
 	onDialDragEnd := func() {}
 	dial.ConnectDragEnd(&onDialDragEnd)
 
+	lastInitialRemainingTime := time.Second * time.Duration(window.settings.GetInt64(resources.SchemaLastPositionKey))
+	window.dialWidget.SetRemainingTime(int(lastInitialRemainingTime.Seconds()))
+
 	window.s = state.NewStateMachine(
 		window.ctx,
-		time.Second*time.Duration(window.settings.GetInt64(resources.SchemaLastPositionKey)),
+		lastInitialRemainingTime,
 		window.log,
 		&state.Hooks{
 			OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
 			OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
 
-			OnInitialRemainingTimeChange: func(ctx context.Context, initialRemainingTime time.Duration) error { return nil },
-			OnCurrentRemainingTimeTick:   func(ctx context.Context, currentRemainingTime time.Duration) error { return nil },
+			OnInitialRemainingTimeChange: func(ctx context.Context, initialRemainingTime time.Duration) error {
+				fn := glib.SourceFunc(func(u uintptr) bool {
+					window.dialWidget.SetRemainingTime(int(initialRemainingTime.Seconds()))
+
+					return false
+				})
+				glib.IdleAdd(&fn, 0)
+
+				return nil
+			},
+			OnCurrentRemainingTimeTick: func(ctx context.Context, currentRemainingTime time.Duration) error { return nil },
 
 			OnBeforeStoppingTimer: func(ctx context.Context) error { return nil },
 			OnAfterStoppingTimer:  func(ctx context.Context) error { return nil },
@@ -112,12 +124,24 @@ 	toggleTimerAction.ConnectActivate(&onToggleTimer)
 	window.AddAction(toggleTimerAction)
 
 	addTimeAction := gio.NewSimpleAction("addTime", nil)
-	onAddTime := func(gio.SimpleAction, uintptr) {}
+	onAddTime := func(gio.SimpleAction, uintptr) {
+		if err := window.s.PlusTimer(window.ctx); err != nil {
+			window.log.Error("Could not add time to timer", "err", err)
+
+			return
+		}
+	}
 	addTimeAction.ConnectActivate(&onAddTime)
 	window.AddAction(addTimeAction)
 
 	removeTimeAction := gio.NewSimpleAction("removeTime", nil)
-	onRemoveTime := func(gio.SimpleAction, uintptr) {}
+	onRemoveTime := func(gio.SimpleAction, uintptr) {
+		if err := window.s.MinusTimer(window.ctx); err != nil {
+			window.log.Error("Could not remove time from timer", "err", err)
+
+			return
+		}
+	}
 	removeTimeAction.ConnectActivate(&onRemoveTime)
 	window.AddAction(removeTimeAction)
 

