From 83d94c812d37bedbc925dc85d69ec14b6a9e1474 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 20 May 2026 22:57:48 -0700
Subject: [PATCH] feat: Start including using state machine in main app, switch to using `time.Duration` internally

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/components/application.go b/internal/components/application.go
index e5143ffe6edf2ccd0733b6e6f098c213cdc213f3..96bbf79b0fca64f7171381419cee1c9d473d4797 100644
--- a/internal/components/application.go
+++ b/internal/components/application.go
@@ -1,6 +1,7 @@
 package components
 
 import (
+	"context"
 	"log/slog"
 	"runtime"
 	"unsafe"
@@ -12,6 +13,7 @@ 	"codeberg.org/puregotk/puregotk/v4/gobject"
 	"codeberg.org/puregotk/puregotk/v4/gtk"
 	. "github.com/pojntfx/go-gettext/pkg/i18n"
 	"github.com/pojntfx/sessions/assets/resources"
+	"github.com/pojntfx/sessions/pkg/state"
 )
 
 var (
@@ -25,9 +27,12 @@ 	window      *MainWindow
 	aboutDialog *adw.AboutDialog
 	settings    *gio.Settings
 	log         *slog.Logger
+
+	ctx context.Context
+	s   *state.StateMachine
 }
 
-func NewApplication(settings *gio.Settings, log *slog.Logger, FirstPropertyNameVar string, varArgs ...interface{}) Application {
+func NewApplication(ctx context.Context, settings *gio.Settings, log *slog.Logger, FirstPropertyNameVar string, varArgs ...interface{}) Application {
 	obj := gobject.NewObject(gTypeApplication, FirstPropertyNameVar, varArgs...)
 
 	var v Application
@@ -36,6 +41,8 @@
 	app := (*Application)(unsafe.Pointer(obj.GetData(dataKeyGoInstance)))
 	app.settings = settings
 	app.log = log
+
+	app.ctx = ctx
 
 	return v
 }
@@ -85,9 +92,10 @@
 			var app gtk.Application
 			a.Cast(&app)
 
-			obj := NewMainWindow(&sessionsApp.Application, sessionsApp.log, "application", app)
+			obj := NewMainWindow(sessionsApp.ctx, &sessionsApp.Application, sessionsApp.log, "application", app)
 
 			sessionsApp.window = (*MainWindow)(unsafe.Pointer(obj.GetData(dataKeyGoInstance)))
+			sessionsApp.window.ctx = sessionsApp.ctx
 			sessionsApp.window.settings = sessionsApp.settings
 			sessionsApp.window.LoadLastPosition()
 
diff --git a/internal/components/dial.go b/internal/components/dial.go
index 0fb37d03687543aedbbbf122db0dc73ad60593b0..59b7eeae987adc1ff7b0c1f2c1b70fd10466fdd9 100644
--- a/internal/components/dial.go
+++ b/internal/components/dial.go
@@ -151,7 +151,7 @@ 				strokeBorder(innerBorderPath)
 			}
 
 			if dialW.totalSec > 0 {
-				progress := float64(dialW.totalSec) / maxDialValue
+				progress := float64(dialW.totalSec) / maxDialValue.Seconds()
 				end := -math.Pi/2 + 2*math.Pi*progress
 				var angle float64
 				var lineColor gdk.RGBA
diff --git a/internal/components/main_window.go b/internal/components/main_window.go
index 5e95d4c05203faaa7582a952d2441613a77f9fa5..a296a76203f734ff4b852abb7a62b7875fae2bef 100644
--- a/internal/components/main_window.go
+++ b/internal/components/main_window.go
@@ -1,6 +1,7 @@
 package components
 
 import (
+	"context"
 	"fmt"
 	"log/slog"
 	"math"
@@ -15,6 +16,7 @@ 	"codeberg.org/puregotk/puregotk/v4/gobject"
 	"codeberg.org/puregotk/puregotk/v4/gtk"
 	. "github.com/pojntfx/go-gettext/pkg/i18n"
 	"github.com/pojntfx/sessions/assets/resources"
+	"github.com/pojntfx/sessions/pkg/state"
 	"github.com/rymdport/portal/background"
 )
 
@@ -23,8 +25,8 @@ 	gTypeMainWindow gobject.Type
 )
 
 const (
-	minDialValue = 30
-	maxDialValue = 3600
+	minDialValue = time.Second * 30
+	maxDialValue = state.MaxRemainingTime
 
 	notificationIdVar = "session-finished"
 )
@@ -50,9 +52,12 @@ 	timer                 uint32
 	dragging              bool
 	paused                bool
 	held                  bool
+
+	ctx context.Context
+	s   *state.StateMachine
 }
 
-func NewMainWindow(app *adw.Application, log *slog.Logger, FirstPropertyNameVar string, varArgs ...interface{}) MainWindow {
+func NewMainWindow(ctx context.Context, app *adw.Application, log *slog.Logger, FirstPropertyNameVar string, varArgs ...interface{}) MainWindow {
 	obj := gobject.NewObject(gTypeMainWindow, FirstPropertyNameVar, varArgs...)
 
 	var v MainWindow
@@ -69,14 +74,38 @@ 	window.dialArea.Append(&dial.Widget)
 	window.dialWidget = &dial
 	window.setupDialGestures()
 
+	window.ctx = ctx
+
 	return v
 }
 
 func (w *MainWindow) LoadLastPosition() {
 	lastPosition := w.settings.GetInt64(resources.SchemaLastPositionKey)
-	if lastPosition >= minDialValue && lastPosition <= maxDialValue {
+	if lastPosition >= int64(minDialValue.Seconds()) && lastPosition <= int64(maxDialValue.Seconds()) {
 		w.totalSec = int(lastPosition)
 	}
+
+	w.s = state.NewStateMachine(
+		w.ctx,
+		time.Second*time.Duration(w.totalSec),
+		w.log,
+		&state.Hooks{
+			OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
+			OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
+
+			OnInitialRemainingTimeChange: func(ctx context.Context, initialRemainingTime time.Duration) error { return nil },
+			OnCurrentRemainingTimeTick:   func(ctx context.Context, currentRemainingTime time.Duration) error { return nil },
+
+			OnBeforeStoppingTimer: func(ctx context.Context) error { return nil },
+			OnAfterStoppingTimer:  func(ctx context.Context) error { return nil },
+
+			OnStartAlarm: func(ctx context.Context) error { return nil },
+
+			OnStopAlarm: func(ctx context.Context) error { return nil },
+
+			OnPermittedTriggersChange: func(ctx context.Context, permittedTriggers []state.Trigger) error { return nil },
+		},
+	)
 }
 
 func (w *MainWindow) SaveLastPosition() {
@@ -192,16 +221,16 @@ 		w.actionButton.SetLabel(L("_Stop"))
 		w.actionButton.RemoveCssClass("suggested-action")
 		w.actionButton.AddCssClass("destructive-action")
 
-		w.plusButton.SetSensitive(w.totalSec < maxDialValue)
-		w.minusButton.SetSensitive(w.totalSec > minDialValue)
+		w.plusButton.SetSensitive(w.totalSec < int(maxDialValue.Seconds()))
+		w.minusButton.SetSensitive(w.totalSec > int(minDialValue.Seconds()))
 	} else {
 		w.actionButton.SetIconName("media-playback-start-symbolic")
 		w.actionButton.SetLabel(L("_Start Timer"))
 		w.actionButton.RemoveCssClass("destructive-action")
 		w.actionButton.AddCssClass("suggested-action")
 
-		w.plusButton.SetSensitive(w.totalSec < maxDialValue)
-		w.minusButton.SetSensitive(w.totalSec > minDialValue)
+		w.plusButton.SetSensitive(w.totalSec < int(maxDialValue.Seconds()))
+		w.minusButton.SetSensitive(w.totalSec > int(minDialValue.Seconds()))
 	}
 }
 
@@ -324,7 +353,7 @@ 	if intervals == 0 {
 		intervals = 120
 	}
 
-	w.totalSec = intervals * minDialValue
+	w.totalSec = intervals * int(minDialValue.Seconds())
 
 	if w.paused {
 		w.remain = time.Duration(w.totalSec) * time.Second
@@ -399,8 +428,8 @@ 	if w.alarming {
 		return
 	}
 
-	if w.totalSec < maxDialValue {
-		w.totalSec += minDialValue
+	if w.totalSec < int(maxDialValue.Seconds()) {
+		w.totalSec += int(minDialValue.Seconds())
 		if w.running {
 			w.remain = time.Duration(w.totalSec) * time.Second
 		}
@@ -416,8 +445,8 @@ 	if w.alarming {
 		return
 	}
 
-	if w.totalSec > minDialValue {
-		w.totalSec -= minDialValue
+	if w.totalSec > int(minDialValue.Seconds()) {
+		w.totalSec -= int(minDialValue.Seconds())
 		if w.running {
 			w.remain = time.Duration(w.totalSec) * time.Second
 		}
diff --git a/main.go b/main.go
index 7a2ad6eaac7198b069c22b078a3e931c611aeb97..a224cfa6777da9dd852927a03a99ec45b1811791 100644
--- a/main.go
+++ b/main.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"context"
 	"errors"
 	"log/slog"
 	"os"
@@ -36,6 +37,9 @@ 	gio.ResourcesRegister(resource)
 }
 
 func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	var settings gio.Settings
 	if SchemaDir == "" {
 		settings = *gio.NewSettings(resources.AppID)
@@ -54,6 +58,7 @@ 		settings = *gio.NewSettingsFull(schema, nil, schema.GetPath())
 	}
 
 	app := components.NewApplication(
+		ctx,
 		&settings,
 		slog.Default(),
 		"application_id", resources.AppID,
diff --git a/pkg/state/constants.go b/pkg/state/constants.go
index 7f4098c20c43081e18b169fd6186899b725a99e3..a2c8caeb0070cda0786ec13beb4013a65d63efcb 100644
--- a/pkg/state/constants.go
+++ b/pkg/state/constants.go
@@ -37,6 +37,6 @@ 	DefaultInitialRemainingTime      = time.Minute * 5
 	RemainingTimerAdjustmentInterval = time.Second * 30
 
 	minRemainingTime = time.Duration(0)
-	maxRemainingTime = time.Hour
+	MaxRemainingTime = time.Hour
 	tickerInterval   = time.Second
 )
diff --git a/pkg/state/machine.go b/pkg/state/machine.go
index 02bd519915192326fa088aa316b1d072824ece96..f782f665766210e4df3d68b9431ab0fe9d727ba5 100644
--- a/pkg/state/machine.go
+++ b/pkg/state/machine.go
@@ -177,7 +177,7 @@ }
 
 func (s *StateMachine) mustBeBelowMaxInitialRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := s.initialRemainingTime + RemainingTimerAdjustmentInterval
-	if newInitialRemainingTime > maxRemainingTime {
+	if newInitialRemainingTime > MaxRemainingTime {
 		return false
 	}
 
@@ -219,7 +219,7 @@ 	}
 
 	newInitialRemainingTime := args[0].(time.Duration)
 	if newInitialRemainingTime < minRemainingTime ||
-		newInitialRemainingTime > maxRemainingTime ||
+		newInitialRemainingTime > MaxRemainingTime ||
 		newInitialRemainingTime%RemainingTimerAdjustmentInterval != 0 {
 		return false
 	}
@@ -269,7 +269,7 @@ }
 
 func (s *StateMachine) mustBeBelowMaxCurrentRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := getInitialRemainingTimeFromCurrentRemainingTime(s.currentRemainingTime, 1)
-	if newInitialRemainingTime > maxRemainingTime {
+	if newInitialRemainingTime > MaxRemainingTime {
 		return false
 	}
 
diff --git a/pkg/state/machine_test.go b/pkg/state/machine_test.go
index 69236cb47a075eaf40803203242f20ba1073ba6a..69a38244f1c9bbe6ae658e43758875f7a2f7b64e 100644
--- a/pkg/state/machine_test.go
+++ b/pkg/state/machine_test.go
@@ -28,12 +28,12 @@ 			plusTimes:   1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     maxRemainingTime - RemainingTimerAdjustmentInterval,
+			initial:     MaxRemainingTime - RemainingTimerAdjustmentInterval,
 			plusTimes:   1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     maxRemainingTime - RemainingTimerAdjustmentInterval,
+			initial:     MaxRemainingTime - RemainingTimerAdjustmentInterval,
 			plusTimes:   2,
 			expectErrAt: 1,
 		},
@@ -125,7 +125,7 @@ 		minusTimes,
 		expectErrAt int
 	}{
 		{
-			initial:     maxRemainingTime,
+			initial:     MaxRemainingTime,
 			minusTimes:  1,
 			expectErrAt: -1,
 		},
@@ -353,7 +353,7 @@ 			expectErr: true,
 		},
 		{
 			name:          "can not transition from dragging state to counting down state with initial remaining time above maximum remaining time",
-			remainingTime: maxRemainingTime + RemainingTimerAdjustmentInterval,
+			remainingTime: MaxRemainingTime + RemainingTimerAdjustmentInterval,
 			prepare: func(sm *StateMachine) error {
 				return sm.StartDragging(t.Context())
 			},

