From 684550340423b369526eefcffb95f1e9b440447e Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 20 May 2026 23:45:30 -0700
Subject: [PATCH] fix: Always start from min initial remaining time and make sure we can't set initial remaining time below it

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/sessions-fsm/main.go b/cmd/sessions-fsm/main.go
index 7acdb652d654ef652616906910d9048191bf6319..6604e60c51ae5318377366427216ecdf6c2cc2ec 100644
--- a/cmd/sessions-fsm/main.go
+++ b/cmd/sessions-fsm/main.go
@@ -15,7 +15,7 @@ 	defer cancel()
 
 	s := state.NewStateMachine(
 		ctx,
-		state.DefaultInitialRemainingTime,
+		state.MinInitialRemainingTime,
 		slog.Default(),
 		&state.Hooks{
 			OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
diff --git a/internal/components/main_window.go b/internal/components/main_window.go
index c26d337fcc620ad726e3868ddfffa98299f6595d..821f9c9b8193e160f9b584206edf9f559c997b87 100644
--- a/internal/components/main_window.go
+++ b/internal/components/main_window.go
@@ -25,8 +25,8 @@ 	gTypeMainWindow gobject.Type
 )
 
 const (
-	minDialValue = time.Second * 30
-	maxDialValue = state.MaxRemainingTime
+	minDialValue = state.MinInitialRemainingTime
+	maxDialValue = state.MaxInitialRemainingTime
 
 	notificationIdVar = "session-finished"
 )
diff --git a/pkg/state/constants.go b/pkg/state/constants.go
index a2c8caeb0070cda0786ec13beb4013a65d63efcb..8fbc789c109901e6178353a9015a8a668672bbb0 100644
--- a/pkg/state/constants.go
+++ b/pkg/state/constants.go
@@ -36,7 +36,8 @@ const (
 	DefaultInitialRemainingTime      = time.Minute * 5
 	RemainingTimerAdjustmentInterval = time.Second * 30
 
-	minRemainingTime = time.Duration(0)
-	MaxRemainingTime = time.Hour
-	tickerInterval   = time.Second
+	MinInitialRemainingTime = RemainingTimerAdjustmentInterval
+	MaxInitialRemainingTime = time.Hour
+
+	tickerInterval = time.Second
 )
diff --git a/pkg/state/machine.go b/pkg/state/machine.go
index f782f665766210e4df3d68b9431ab0fe9d727ba5..cebaa18a03ece1e79793f5fc7e1993a6d91baee8 100644
--- a/pkg/state/machine.go
+++ b/pkg/state/machine.go
@@ -177,7 +177,7 @@ }
 
 func (s *StateMachine) mustBeBelowMaxInitialRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := s.initialRemainingTime + RemainingTimerAdjustmentInterval
-	if newInitialRemainingTime > MaxRemainingTime {
+	if newInitialRemainingTime > MaxInitialRemainingTime {
 		return false
 	}
 
@@ -200,7 +200,7 @@ }
 
 func (s *StateMachine) mustBeAboveMinInitialRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := s.initialRemainingTime - RemainingTimerAdjustmentInterval
-	if newInitialRemainingTime < minRemainingTime {
+	if newInitialRemainingTime < MinInitialRemainingTime {
 		return false
 	}
 
@@ -218,8 +218,8 @@ 		return false
 	}
 
 	newInitialRemainingTime := args[0].(time.Duration)
-	if newInitialRemainingTime < minRemainingTime ||
-		newInitialRemainingTime > MaxRemainingTime ||
+	if newInitialRemainingTime < MinInitialRemainingTime ||
+		newInitialRemainingTime > MaxInitialRemainingTime ||
 		newInitialRemainingTime%RemainingTimerAdjustmentInterval != 0 {
 		return false
 	}
@@ -269,7 +269,7 @@ }
 
 func (s *StateMachine) mustBeBelowMaxCurrentRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := getInitialRemainingTimeFromCurrentRemainingTime(s.currentRemainingTime, 1)
-	if newInitialRemainingTime > MaxRemainingTime {
+	if newInitialRemainingTime > MaxInitialRemainingTime {
 		return false
 	}
 
@@ -292,7 +292,7 @@ }
 
 func (s *StateMachine) mustBeAboveMinCurrentRemainingTime(ctx context.Context, args ...any) bool {
 	newInitialRemainingTime := getInitialRemainingTimeFromCurrentRemainingTime(s.currentRemainingTime, -1)
-	if newInitialRemainingTime < minRemainingTime {
+	if newInitialRemainingTime < MinInitialRemainingTime {
 		return false
 	}
 
diff --git a/pkg/state/machine_test.go b/pkg/state/machine_test.go
index 69a38244f1c9bbe6ae658e43758875f7a2f7b64e..0e56fdc2ebde5ac78546bccb43620b6e2481b16d 100644
--- a/pkg/state/machine_test.go
+++ b/pkg/state/machine_test.go
@@ -23,17 +23,17 @@ 		plusTimes,
 		expectErrAt int
 	}{
 		{
-			initial:     time.Duration(0),
+			initial:     MinInitialRemainingTime,
 			plusTimes:   1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     MaxRemainingTime - RemainingTimerAdjustmentInterval,
+			initial:     MaxInitialRemainingTime - RemainingTimerAdjustmentInterval,
 			plusTimes:   1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     MaxRemainingTime - RemainingTimerAdjustmentInterval,
+			initial:     MaxInitialRemainingTime - RemainingTimerAdjustmentInterval,
 			plusTimes:   2,
 			expectErrAt: 1,
 		},
@@ -47,7 +47,7 @@ 					var (
 						onBeforeStartingTimerCalled = 0
 						onAfterStartingTimerCalled  = 0
 
-						internalInitialRemainingTime = time.Duration(0)
+						internalInitialRemainingTime = tt.initial
 					)
 					s := newTestingStateMachine(
 						t,
@@ -125,20 +125,25 @@ 		minusTimes,
 		expectErrAt int
 	}{
 		{
-			initial:     MaxRemainingTime,
+			initial:     MaxInitialRemainingTime,
 			minusTimes:  1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     minRemainingTime + RemainingTimerAdjustmentInterval,
+			initial:     MinInitialRemainingTime + RemainingTimerAdjustmentInterval,
 			minusTimes:  1,
 			expectErrAt: -1,
 		},
 		{
-			initial:     minRemainingTime + RemainingTimerAdjustmentInterval,
+			initial:     MinInitialRemainingTime + RemainingTimerAdjustmentInterval,
 			minusTimes:  2,
 			expectErrAt: 1,
 		},
+		{
+			initial:     MinInitialRemainingTime,
+			minusTimes:  1,
+			expectErrAt: 0,
+		},
 	}
 	for _, tt := range minusTimerTests {
 		for _, fromCountingDown := range []bool{true, false} {
@@ -149,7 +154,7 @@ 					var (
 						onBeforeStartingTimerCalled = 0
 						onAfterStartingTimerCalled  = 0
 
-						internalInitialRemainingTime = time.Duration(0)
+						internalInitialRemainingTime = tt.initial
 					)
 					s := newTestingStateMachine(
 						t,
@@ -208,9 +213,17 @@ 						}
 					}
 
 					if fromCountingDown {
-						// The timer should still be running if we increased during the counting down phase
-						require.Equal(t, 2, onBeforeStartingTimerCalled)
-						require.Equal(t, 2, onAfterStartingTimerCalled)
+						if tt.expectErrAt == 0 {
+							// If we tried to decrease during the counting down phase, but failed on the first decrease,
+							// then we only start the timer once (when we initially started dragging), we don't restart
+							// the timer when fail to decrease
+							require.Equal(t, 1, onBeforeStartingTimerCalled)
+							require.Equal(t, 1, onAfterStartingTimerCalled)
+						} else {
+							// The timer should still be running if we decreased during the counting down phase
+							require.Equal(t, 2, onBeforeStartingTimerCalled)
+							require.Equal(t, 2, onAfterStartingTimerCalled)
+						}
 					}
 
 					require.Equal(t, expectedInitialRemainingTime, internalInitialRemainingTime)
@@ -288,7 +301,7 @@ 			tt.name,
 			func(t *testing.T) {
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
 						OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
@@ -345,7 +358,7 @@ 			onAfterStartingTimerCalled:  1,
 		},
 		{
 			name:          "can not transition from dragging state to counting down state with initial remaining time below minimum remaining time",
-			remainingTime: minRemainingTime - RemainingTimerAdjustmentInterval,
+			remainingTime: MinInitialRemainingTime - RemainingTimerAdjustmentInterval,
 			prepare: func(sm *StateMachine) error {
 				return sm.StartDragging(t.Context())
 			},
@@ -353,7 +366,7 @@ 			expectErr: true,
 		},
 		{
 			name:          "can not transition from dragging state to counting down state with initial remaining time above maximum remaining time",
-			remainingTime: MaxRemainingTime + RemainingTimerAdjustmentInterval,
+			remainingTime: MaxInitialRemainingTime + RemainingTimerAdjustmentInterval,
 			prepare: func(sm *StateMachine) error {
 				return sm.StartDragging(t.Context())
 			},
@@ -384,11 +397,11 @@ 				var (
 					onBeforeStartingTimerCalled = 0
 					onAfterStartingTimerCalled  = 0
 
-					internalInitialRemainingTime = time.Duration(0)
+					internalInitialRemainingTime = MinInitialRemainingTime
 				)
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error {
 							onBeforeStartingTimerCalled++
@@ -498,7 +511,7 @@ 					onAfterStoppingTimerCalled  = 0
 				)
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
 						OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
@@ -576,7 +589,7 @@ 					onAfterStartingTimerCalled  = 0
 				)
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error {
 							onBeforeStartingTimerCalled++
@@ -649,7 +662,7 @@ 			func(t *testing.T) {
 				onStartAlarmCalled := 0
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
 						OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
@@ -729,7 +742,7 @@ 					onAfterStoppingTimerCalled  = 0
 				)
 				s := newTestingStateMachine(
 					t,
-					0,
+					MinInitialRemainingTime,
 					&Hooks{
 						OnBeforeStartingTimer: func(ctx context.Context) error { return nil },
 						OnAfterStartingTimer:  func(ctx context.Context) error { return nil },
@@ -806,7 +819,7 @@ 				require.NoError(t, s.PlusTimer(t.Context()))
 
 				require.NoError(t, s.StartTimer(t.Context()))
 
-				time.Sleep(RemainingTimerAdjustmentInterval * 3) // We run this *3, not *2, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
+				time.Sleep(RemainingTimerAdjustmentInterval * 4) // We run this *4, not *3, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
 
 				require.NoError(t, s.StopAlarming(t.Context()))
 			},
@@ -814,7 +827,7 @@
 			onBeforeStartingTimerCalled: 1,
 			onAfterStartingTimerCalled:  1,
 
-			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 2,
+			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 3,
 
 			onBeforeStoppingTimerCalled: 1,
 			onAfterStoppingTimerCalled:  1,
@@ -841,7 +854,7 @@ 				require.NoError(t, s.PlusTimer(t.Context()))
 
 				require.NoError(t, s.StartTimer(t.Context()))
 
-				time.Sleep(RemainingTimerAdjustmentInterval * 5) // We run this *5, not *4, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
+				time.Sleep(RemainingTimerAdjustmentInterval * 6) // We run this *6, not *5, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
 
 				require.NoError(t, s.StopAlarming(t.Context()))
 			},
@@ -849,7 +862,7 @@
 			onBeforeStartingTimerCalled: 1,
 			onAfterStartingTimerCalled:  1,
 
-			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 4,
+			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 5,
 
 			onBeforeStoppingTimerCalled: 1,
 			onAfterStoppingTimerCalled:  1,
@@ -876,13 +889,13 @@ 				require.NoError(t, s.PlusTimer(t.Context()))
 
 				require.NoError(t, s.StartTimer(t.Context()))
 
-				time.Sleep(RemainingTimerAdjustmentInterval * 5) // We run this *5, not *4, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
+				time.Sleep(RemainingTimerAdjustmentInterval * 6) // We run this *6, not *5, to make sure that `onCurrentRemainingTimeTick` no longer fires after stopping
 			},
 
 			onBeforeStartingTimerCalled: 1,
 			onAfterStartingTimerCalled:  1,
 
-			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 4,
+			internalInitialRemainingTime: RemainingTimerAdjustmentInterval * 5,
 
 			onBeforeStoppingTimerCalled: 1,
 			onAfterStoppingTimerCalled:  1,
@@ -912,7 +925,7 @@ 					var (
 						onBeforeStartingTimerCalled = 0
 						onAfterStartingTimerCalled  = 0
 
-						internalInitialRemainingTime = time.Duration(0)
+						internalInitialRemainingTime = MinInitialRemainingTime
 
 						onBeforeStoppingTimerCalled = 0
 						onAfterStoppingTimerCalled  = 0
@@ -925,7 +938,7 @@ 						internalPermittedTriggers = []Trigger{}
 					)
 					s := newTestingStateMachine(
 						t,
-						0,
+						MinInitialRemainingTime,
 						&Hooks{
 							OnBeforeStartingTimer: func(ctx context.Context) error {
 								onBeforeStartingTimerCalled++

