From 4a604f0f973513035d3a7f451a57cd35720480a4 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 25 May 2026 23:24:28 -0700
Subject: [PATCH] feat: Support storing last saved position in GSettings

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/components/main_window.go b/internal/components/main_window.go
index 760c756329342a73152dda07c9990cda9dcfad58..8a181bc08724b86cff10127a60688f2e8e369a96 100644
--- a/internal/components/main_window.go
+++ b/internal/components/main_window.go
@@ -151,6 +151,7 @@ 				lastInitialRemainingTime = initialRemainingTime
 
 				fn := glib.SourceFunc(func(u uintptr) bool {
 					window.dialWidget.SetRemainingTime(int(lastInitialRemainingTime.Seconds()))
+					window.settings.SetInt64(resources.SchemaLastPositionKey, int64(lastInitialRemainingTime.Seconds()))
 
 					return false
 				})

