From 7094fac6c47676a9138a8a5cf0e7b95d26f1ec7d Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 03 Apr 2026 23:48:19 -0700
Subject: [PATCH] feat: Add GitHub and web icons to icon component

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/js/icons.js b/assets/js/icons.js
index 65062fc3bd623bada1ba4eb4f82b76459520447e..b647fe7efeb931b0c10d65022dd9d0aba77bc2ca 100644
--- a/assets/js/icons.js
+++ b/assets/js/icons.js
@@ -26,6 +26,8 @@   "icon-mastodon": "fa-brands fa-mastodon",
   "icon-bluesky": "fa-brands fa-bluesky",
   "icon-linkedin": "fa-brands fa-linkedin",
   "icon-paypal": "fa-brands fa-paypal",
+  "icon-github": "fa-brands fa-github",
+  "icon-globe": "fa-solid fa-globe",
   "icon-external": "fa-solid fa-arrow-up-right-from-square",
   "icon-email": "fa-solid fa-envelope",
 };

