From 6690306bee4e3c8f86d26ab30e86e7a5052bc5ed Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 01 Apr 2026 14:55:41 -0700
Subject: [PATCH] refactor: Switch to using custom domain

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/hugo.yaml b/hugo.yaml
index 1da7344d576d581429bf5aa0eab6309f027d89ea..6ee17546555ad4c69b47a3cb546b63230a2105b2 100644
--- a/hugo.yaml
+++ b/hugo.yaml
@@ -1,4 +1,4 @@
-baseURL: https://vanlug.github.io/
+baseURL: https://vanlug.ca/
 locale: en-CA
 languageCode: en-CA
 title: "VanLUG: Vancouver Linux Users Group"
diff --git a/main.go b/main.go
index 92b60549b90aa4e16d516db8e3e7f04b28d0a3ed..4a2249fb88b77496d775e0f84a17e2772da70de2 100644
--- a/main.go
+++ b/main.go
@@ -34,7 +34,7 @@ 	}
 
 	siteURL := os.Getenv("SITE_URL")
 	if siteURL == "" {
-		siteURL = "https://vanlug.github.io/"
+		siteURL = "https://vanlug.ca/"
 	}
 
 	mastodonServer := os.Getenv("MASTODON_SERVER")
diff --git a/static/CNAME b/static/CNAME
new file mode 100644
index 0000000000000000000000000000000000000000..db13d870043d07c3bf796f80c8683d0934bdb8c6
--- /dev/null
+++ b/static/CNAME
@@ -0,0 +1 @@
+vanlug.ca

