From 7a16865d70877c1a8ac4864039400fa41dc54569 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 01 Apr 2026 00:25:41 -0700
Subject: [PATCH] fix: Show favourites and reposts next to each other in Mastodon sidebar

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/js/mastodon-feed.js b/assets/js/mastodon-feed.js
index 7b30770ce6d05ee2781f703b35899fb06f365e51..267f5a857d8c2a72c61d7a8fdff3d30c9ecddcab 100644
--- a/assets/js/mastodon-feed.js
+++ b/assets/js/mastodon-feed.js
@@ -8,7 +8,7 @@     <div data-slot="media"></div>
     <div class="pf-v6-l-flex pf-m-gap-sm pf-m-align-items-center pf-v6-u-font-size-xs pf-v6-u-text-color-subtle">
       <img data-slot="avatar" style="width: 1rem; height: 1rem; border-radius: 50%;" />
       <span data-slot="timestamp"></span>
-      <span data-slot="stats"></span>
+      <span class="pf-v6-l-flex pf-m-gap-sm pf-m-align-items-center" data-slot="stats"></span>
     </div>
   </a>`;
 

