From 3322c75f38b1b5751b22f05971bebfc456f65624 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 31 Mar 2026 22:51:34 -0700
Subject: [PATCH] fix: Use Canadian date formats for `toLocaleDateString`

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/static/main.js b/static/main.js
index 64786c1e8e0e5259a5a7c3624ae1364f623a84b6..95d13219c2240254a570be3c54ec5158ffeaa435 100644
--- a/static/main.js
+++ b/static/main.js
@@ -97,12 +97,12 @@     const start = new Date(evt.start_at);
     const days = Math.ceil((start - new Date()) / 86400000);
     return {
       name: evt.name || "VanLUG Meeting",
-      date: start.toLocaleDateString("en-US", {
+      date: start.toLocaleDateString("en-CA", {
         weekday: "short",
         month: "short",
         day: "numeric",
       }),
-      time: start.toLocaleTimeString("en-US", {
+      time: start.toLocaleTimeString("en-CA", {
         hour: "numeric",
         minute: "2-digit",
       }),

