From 5c603d1768a5efa9577d375232333d5d8649e539 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 27 May 2026 00:26:24 -0700
Subject: [PATCH] fix: Don't break dates on small screen sizes for upcoming events list

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/js/luma-upcoming-events.js b/assets/js/luma-upcoming-events.js
index b079fc86242b633a361ba31dcbbece4a05fdef65..5075e727083b39185da085bdae2150b00664ebcc 100644
--- a/assets/js/luma-upcoming-events.js
+++ b/assets/js/luma-upcoming-events.js
@@ -9,7 +9,7 @@       <div class="pf-v6-l-flex pf-m-gap-md pf-m-align-items-center">
         <span class="pf-v6-u-display-inline-block pf-v6-u-display-none-on-md" data-slot="cover-mobile"></span>
         <div>
           <div class="pf-v6-u-font-weight-bold pf-v6-u-display-block pf-v6-u-display-none-on-md" data-slot="name-mobile"></div>
-          <div><span data-slot="date"></span> <span class="pf-v6-u-font-size-xs" data-slot="relative"></span></div>
+          <div style="white-space: nowrap;"><span data-slot="date"></span> <span class="pf-v6-u-font-size-xs" data-slot="relative"></span></div>
           <small class="pf-v6-u-text-color-subtle" data-slot="time"></small>
         </div>
       </div>

