From 4c5c3c22864b13ccd3e7580cf543b9ccce563097 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 27 May 2026 00:18:18 -0700
Subject: [PATCH] fix: If an event is happening today, make it display "today"

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/js/events.js b/assets/js/events.js
index 69b748db257b2f48503f53cfffaac201542768b9..5661ead01c00de976bdc3a2fb74ed8f9e1a3bca4 100644
--- a/assets/js/events.js
+++ b/assets/js/events.js
@@ -5,7 +5,11 @@ };
 
 export const formatEvent = (evt, api) => {
   const start = new Date(evt.start_at);
-  const days = Math.ceil((start - new Date()) / 86400000);
+  const today = new Date();
+  today.setHours(0, 0, 0, 0);
+  const eventDay = new Date(start);
+  eventDay.setHours(0, 0, 0, 0);
+  const days = Math.round((eventDay - today) / 86400000);
   return {
     apiId: evt.api_id,
     name: evt.name || "VanLUG Meeting",

