From d949e9ef70e50624fecfc442d5debd806b23299e Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 23 Jul 2026 18:40:38 -0700
Subject: [PATCH] feat: Write quadlet logs to journald and show how to retrieve them

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/README.md b/README.md
index 30aebd7dd900125e23438457408fe3c9d6a8985c..20b008b0dc98af68ff43f1ea9a8681c40072d3db 100644
--- a/README.md
+++ b/README.md
@@ -20,6 +20,9 @@ $ tofu init
 $ tofu apply
 $ ansible-galaxy collection install cloud.terraform ansible.posix
 $ ansible-playbook site.yaml # You might want to change `binary_path` in `inventory.yaml` first
+# To get logs for quadlets
+$ ssh setup@vanlug.p8.lu
+[setup@vanlug-main ~]$ sudo journalctl _SYSTEMD_USER_UNIT=caddy.service -f
 ```
 
 ## License
diff --git a/caddy.container b/caddy.container
index 86bdc101153a614c87c9f4a0281062542d08644f..70ae25d57b77a5a33c6248c6d040f2391544602f 100644
--- a/caddy.container
+++ b/caddy.container
@@ -7,6 +7,7 @@ PublishPort=8443:443
 PublishPort=8443:443/udp
 Volume=caddy_data:/data
 Volume=/home/caddy/.config/caddy:/etc/caddy:ro,Z
+LogDriver=journald
 
 [Service]
 Restart=always

