From d601b3566234443a6cd1946bc70bb06f537bff95 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 21 Jul 2026 22:41:29 -0700
Subject: [PATCH] feat: Add Ansible provider to OpenTofu and set up basic Ansible playbook boilerplate with OpenTofu inventory provider

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.terraform.lock.hcl b/.terraform.lock.hcl
index 31be13c7a3845568a11e7b1356095c4f945b1cc7..020fc5d941d559572bf138d3356d349cb8c4db49 100644
--- a/.terraform.lock.hcl
+++ b/.terraform.lock.hcl
@@ -1,6 +1,40 @@
 # This file is maintained automatically by "tofu init".
 # Manual edits may be lost in future updates.
 
+provider "registry.opentofu.org/ansible/ansible" {
+  version     = "1.4.0"
+  constraints = "~> 1.4.0"
+  hashes = [
+    "h1:3BLWpuJUFvdjHr5qKDhzCCSfpMkYy/itbXgDWc3wkFA=",
+    "h1:6C8Fo1tBd8GukVQqz9lCzETOb/zW6A2k3wCVlMjAp/E=",
+    "h1:8UUj0/Wh1ouz1LU9sg/m4G9bRxMEW+COPUuUN0dwD0s=",
+    "h1:B5hHbn7cL+yJ4xU7eTaqSsKL5jOtUBO7/vWoDcOTfuc=",
+    "h1:C+BNERRzLHkZh08sWRNJrItCXGWC5vYAqvdBUC9pjH4=",
+    "h1:DHcrxMD63RBSSyqWsSxR/9ODwaP/586gMQc7DsxZLQM=",
+    "h1:F7w05ppMX7ZE3TjiGt9pE8LQoRAYgjArzZAnr5OHD1A=",
+    "h1:O3QR17mD+CLmw9+u34KembOMRNwh1LjkiURdFHMrOkY=",
+    "h1:PLQHXVpF2xZUiqJsHLNhp8B+2ndFeaSbhTvQ2RIFRJg=",
+    "h1:YqtY4N1TQAYGnZT8E7RcqV6vGNvoQIQ1kqlEHn9W+b8=",
+    "h1:bifTrFyAiUmPnis9Yh6Ifnb9AhZi6pExLh+ndSw2KaU=",
+    "h1:nfB30icaMbt+fnJauE+Jt1oe1JWupPV9btoXGQFOVsA=",
+    "h1:tdexOiIceKW/W0M3OfcLqwdCrFR/zwud9qRr/BXI3k4=",
+    "zh:0d45166bf99e50a28a1221d17a13bf8fd461109e15675984768403f7c429fe09",
+    "zh:31cb7f7745f2d5ca88936ba3c86e408f242ed0754f4de02f85dc821cf8b46a3c",
+    "zh:3c9d29e378426a1f9926deb2aa71bcac27a50af46e0948896bfdcde5a9969e76",
+    "zh:3ff407b93aec8a9fae1aeb71214a1619ba06d28972ee514b403884ad88a994c9",
+    "zh:6149d11d87cff5ab3a55a490f95f9562845c910c74ae63f5722e753802d67ff6",
+    "zh:753edbf93b7112d868f55ac10a827ba581200d6aacf5355e4bf98ba07116a427",
+    "zh:90db772188e09afdaae43ff2a3d4c3a745203c8f0d63388b01ba6381ae4e0d07",
+    "zh:9966aa595937c426a4e114db91bb24d5f63184a9e0862195dc0a7bebfdfc7fc9",
+    "zh:a97fba47457b919556def655d5bc33354fedd39933b31e4a236ef5cf99030298",
+    "zh:b20001789127619435f99be52c32ed834ac74f16e1a3dbed7267a12be5254148",
+    "zh:b545ef33e31403854c7b39846f966508ea6e77fe455d5735f801c842210623dd",
+    "zh:b99e7e64786a1b22ac0f5972079f0c2934e7f58a8afd7d021217ef439e0c46ea",
+    "zh:f569b65999264a9416862bca5cd2a6177d94ccb0424f3a4ef424428912b9cb3c",
+    "zh:ff46f8ee2cb0bb9615e64230702645500c96546ccf5338b161c5ddcfd713f81d",
+  ]
+}
+
 provider "registry.opentofu.org/hetznercloud/hcloud" {
   version     = "1.66.1"
   constraints = "~> 1.45"
diff --git a/README.md b/README.md
index 0037fe37a65fa017f2751a0ac560d6eb369aa2a5..a30a9fbc009f239b3158f5a0f87619fe01f03743 100644
--- a/README.md
+++ b/README.md
@@ -15,8 +15,11 @@
 ```shell
 $ git clone https://codeberg.org/pojntfx/vanlug-infrastructure
 $ cd vanlug-infrastructure
-$ source .env.local # See .env for required secrets; ask @pojntfx:beeper.com for access
+$ source .env.local # See `.env` for required secrets; ask `@pojntfx:beeper.com` for access
 $ tofu init
+$ tofu apply
+$ ansible-galaxy collection install cloud.terraform
+$ ansible-playbook site.yaml # You might want to change `binary_path` in `inventory.yaml` first
 ```
 
 ## License
diff --git a/ansible.cfg b/ansible.cfg
new file mode 100644
index 0000000000000000000000000000000000000000..f8382ee25476460dc58b6f8aa65baf019bf56d45
--- /dev/null
+++ b/ansible.cfg
@@ -0,0 +1,2 @@
+[defaults]
+inventory = inventory.yaml
diff --git a/inventory.yaml b/inventory.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..bf0e0551da6ae44928f447e21f9239999dc1c293
--- /dev/null
+++ b/inventory.yaml
@@ -0,0 +1,2 @@
+plugin: cloud.terraform.terraform_provider
+binary_path: /home/linuxbrew/.linuxbrew/bin/tofu
diff --git a/main.tofu b/main.tofu
index 801fd5b8abb7d2ea4e56d09fe720ae1312182e99..29fa8d13464dbc0f078382f1feac0e8730decb00 100644
--- a/main.tofu
+++ b/main.tofu
@@ -75,3 +75,18 @@   #     user_data # We can't update cloud-init user data without destroying the server and recreating it
   #   ]
   # }
 }
+
+resource "ansible_group" "group" {
+  name = "vanlug"
+}
+
+resource "ansible_host" "vanlug_main" {
+  name   = "vanlug_main"
+  groups = ["vanlug"]
+
+  variables = {
+    ansible_host = hcloud_server.vanlug_main.ipv6_address
+    ansible_user = var.ssh_setup_user
+  }
+}
+
diff --git a/providers.tofu b/providers.tofu
index 17f7abbffae081494cbe02f42e19cffa1d423faa..41fb5455fb300fb8351d978aea0ceb8da84aadbc 100644
--- a/providers.tofu
+++ b/providers.tofu
@@ -16,6 +16,11 @@     hcloud = {
       source  = "hetznercloud/hcloud"
       version = "~> 1.45"
     }
+
+    ansible = {
+      version = "~> 1.4.0"
+      source  = "ansible/ansible"
+    }
   }
 }
 
diff --git a/site.yaml b/site.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..2351866769dd29c991d30921dab61013edc7837e
--- /dev/null
+++ b/site.yaml
@@ -0,0 +1,11 @@
+- name: Configure all hosts
+  hosts: all
+  gather_facts: false
+  tasks:
+    - name: Get hostname
+      ansible.builtin.command: hostname
+      changed_when: false
+      register: hostname
+    - name: Show hostname
+      ansible.builtin.debug:
+        msg: "{{ hostname.stdout }}"

