From 43aa82d28f15b856d40fa622275f58bd68d9593a Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 22 Jul 2026 21:51:40 -0700
Subject: [PATCH] feat: Enable pipelining for Ansible

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/ansible.cfg b/ansible.cfg
index f8382ee25476460dc58b6f8aa65baf019bf56d45..dc6a777dc71fc5788e4cb74188bba6a8f204e22f 100644
--- a/ansible.cfg
+++ b/ansible.cfg
@@ -1,2 +1,5 @@
 [defaults]
 inventory = inventory.yaml
+
+[ssh_connection]
+pipelining = true

