From ce090a58330441bc30e264ee09dcb94fd92325e8 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 21 Jul 2026 22:42:36 -0700
Subject: [PATCH] chore: Enable protection for Hetzner server in Terraform

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/main.tofu b/main.tofu
index 29fa8d13464dbc0f078382f1feac0e8730decb00..0e82b4c00be8f108069cc8c1f629e2e299cc550c 100644
--- a/main.tofu
+++ b/main.tofu
@@ -58,22 +58,22 @@
   backups = true
 
   # # Data is stored on the server's volume directly; we should not destroy it
-  # delete_protection  = true
-  # rebuild_protection = true
+  delete_protection  = true
+  rebuild_protection = true
 
   user_data = templatefile("cloud-init.yaml", {
     ssh_setup_user = var.ssh_setup_user
     ssh_public_key = var.ssh_public_key
   })
 
-  # lifecycle {
-  #   prevent_destroy = true # Data is stored on the server's volume directly; we should not destroy it
+  lifecycle {
+    prevent_destroy = true # Data is stored on the server's volume directly; we should not destroy it
 
-  #   ignore_changes = [
-  #     ssh_keys, # We can't update SSH keys without destroying the server and recreating it
-  #     user_data # We can't update cloud-init user data without destroying the server and recreating it
-  #   ]
-  # }
+    ignore_changes = [
+      ssh_keys, # We can't update SSH keys without destroying the server and recreating it
+      user_data # We can't update cloud-init user data without destroying the server and recreating it
+    ]
+  }
 }
 
 resource "ansible_group" "group" {

