From 43b8f50c9dda61b95102fae9cc165105655395e0 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 29 Jul 2026 17:25:02 -0700
Subject: [PATCH] feat: Add Docker-in-Docker support via rootless Podman

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/forgejo-runner.container b/forgejo-runner.container
index a634853c4b92a3baac335c8932bc31e241750bdd..63fe28ed0cf91ea12b33af27085b05bd7e3654d7 100644
--- a/forgejo-runner.container
+++ b/forgejo-runner.container
@@ -1,16 +1,27 @@
+[Unit]
+## For rootless Podman-in-Podman
+Requires=podman.socket
+After=podman.socket
+
 [Container]
 ContainerName=forgejo-runner
 Exec=/bin/forgejo-runner --config config.yml daemon
 Image=code.forgejo.org/forgejo/runner:12
 Label=io.containers.autoupdate=registry
 Volume=forgejo_runner_data:/data
+# See https://docs.podman.io/en/latest/markdown/podman-system-service.1.html#run-the-command-in-a-systemd-service
+Volume=%t/podman/podman.sock:/var/run/docker.sock
+# See https://docs.docker.com/reference/cli/docker/#host
+Environment=DOCKER_HOST=unix:///var/run/docker.sock
+# Process stays running as the `setup` user, but GID 0 since inside this rootless container, GID 0 maps to the unprivileged forgejo-runner group on the host, which owns podman.sock
+GroupAdd=root
 LogDriver=journald
 NoNewPrivileges=true
 ReadOnly=true
 
 [Service]
 # See https://code.forgejo.org/forgejo/runner/src/branch/main/examples/docker-compose/compose-forgejo-and-runner.yml#L69-L81
-ExecStartPre=/usr/bin/podman run --rm -v forgejo_runner_data:/data -w /data code.forgejo.org/forgejo/runner:12 bash -c 'forgejo-runner generate-config > config.yml && sed -i -e "s|  connections:|  connections:\\n    default:\\n      url: https://git.vanlug.p8.lu\\n      uuid: {{ forgejo_runner_uuid | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}\\n      token: {{ forgejo_runner_shared_secret | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}|" config.yml'
+ExecStartPre=/usr/bin/podman run --rm -v forgejo_runner_data:/data -w /data code.forgejo.org/forgejo/runner:12 bash -c 'forgejo-runner generate-config > config.yml && sed -i -e "s|  connections:|  connections:\\n    default:\\n      url: https://git.vanlug.p8.lu\\n      uuid: {{ forgejo_runner_uuid | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}\\n      token: {{ forgejo_runner_shared_secret | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}|" config.yml && sed -i -e "s|  labels: \[\]|  labels:\\n    - ubuntu-latest:docker://node:lts|" config.yml'
 
 [Install]
 WantedBy=default.target
diff --git a/forgejo-runner.yaml b/forgejo-runner.yaml
index e88259c9f8f0f1572132a8a52b19e22dfdc4eabe..85867ba6fc888413d21c24b9ad5479d1dbda848f 100644
--- a/forgejo-runner.yaml
+++ b/forgejo-runner.yaml
@@ -35,6 +35,18 @@         forgejo_runner_uuid: '{{ lookup("ansible.builtin.env", "FORGEJO_RUNNER_UUID") }}'
       no_log: true # The file includes the shared secret and UUID
       # This will only run on changes to the file; see https://docs.ansible.com/projects/ansible/latest/playbook_guide/playbooks_handlers.html#id1
       notify: Restart Forgejo runner user service
+    # For rootless Podman-in-Podman
+    - name: Enable Podman API user service
+      become: true
+      become_user: forgejo-runner
+      ansible.builtin.systemd_service:
+        name: podman.socket
+        state: started
+        enabled: true
+        daemon_reload: true
+        scope: user
+      environment:
+        XDG_RUNTIME_DIR: "/run/user/{{ forgejo_runner_user.uid }}"
     - name: Enable Forgejo runner user service
       become: true
       become_user: forgejo-runner

