From 40f5fc7ad25c13c4aee2603cf0bef48f2a2a178c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 29 Jul 2026 22:56:23 -0700
Subject: [PATCH] feat: Add IPv6 support for Forgejo runner

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/forgejo-runner.container b/forgejo-runner.container
index 63fe28ed0cf91ea12b33af27085b05bd7e3654d7..9807c3d9c0a16fa869364c871e6cbbe7e7f876ec 100644
--- a/forgejo-runner.container
+++ b/forgejo-runner.container
@@ -21,7 +21,7 @@ ReadOnly=true
 
 [Service]
 # See https://code.forgejo.org/forgejo/runner/src/branch/main/examples/docker-compose/compose-forgejo-and-runner.yml#L69-L81
-ExecStartPre=/usr/bin/podman run --rm -v forgejo_runner_data:/data -w /data code.forgejo.org/forgejo/runner:12 bash -c 'forgejo-runner generate-config > config.yml && sed -i -e "s|  connections:|  connections:\\n    default:\\n      url: https://git.vanlug.p8.lu\\n      uuid: {{ forgejo_runner_uuid | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}\\n      token: {{ forgejo_runner_shared_secret | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}|" config.yml && sed -i -e "s|  labels: \[\]|  labels:\\n    - ubuntu-latest:docker://node:lts|" config.yml'
+ExecStartPre=/usr/bin/podman run --rm -v forgejo_runner_data:/data -w /data code.forgejo.org/forgejo/runner:12 bash -c 'forgejo-runner generate-config > config.yml && sed -i -e "s|  connections:|  connections:\\n    default:\\n      url: https://git.vanlug.p8.lu\\n      uuid: {{ forgejo_runner_uuid | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}\\n      token: {{ forgejo_runner_shared_secret | replace("\\", "\\\\") | replace("|", "\\|") | replace("&", "\\&") }}|" config.yml && sed -i -e "s|  labels: \[\]|  labels:\\n    - ubuntu-latest:docker://node:lts|" config.yml && sed -i -e "s|  enable_ipv6: false|  enable_ipv6: true|" config.yml'
 
 [Install]
 WantedBy=default.target

