From 7a3add5a6a7801bba33bda7efe6d848571b643af Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 29 Jul 2026 22:56:02 -0700
Subject: [PATCH] feat: Add wildcard domain support for git-pages

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/git-pages.container b/git-pages.container
index d8122aef21aa238af55f0d58c16925e9549bb9de..06b39035cd3579bc6df5def4ed582dc903b8731f 100644
--- a/git-pages.container
+++ b/git-pages.container
@@ -6,6 +6,11 @@ # We use systemd-socket-proxyd on 3003/3005, so we listen on 3004/3006 instead
 PublishPort=3004:3000
 PublishPort=3006:3001
 Volume=git_pages_data:/app/data
+# See https://git-pages.org/running-a-server/#wildcard-domains
+# We're not supposed to re-use the same second-level domain (vanlug.p8.lu)
+# for user-generated pages, but since we only allow board members to public
+# sites, this is fine for us
+Environment="PAGES_WILDCARD=[{\"Domain\":\"pages.vanlug.p8.lu\",\"CloneURL\":\"http://host.containers.internal:3000/<user>/<project>.git\",\"IndexRepo\":\"<user>.pages.vanlug.p8.lu\",\"IndexRepoBranch\":\"main\",\"Authorization\":\"forgejo\"}]"
 LogDriver=journald
 NoNewPrivileges=true
 ReadOnly=true

