From 0af86591c3485f65c9b1d26474e3d71538eab1a4 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 25 Jul 2026 16:29:54 -0700
Subject: [PATCH] fix: Publish port `2222` for Forgejo

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/forgejo.container b/forgejo.container
index fdba09f64fb7e1ddad6230724540a6246b5783a3..e3867ea143681ce7ef968c93258c595b254a3f0a 100644
--- a/forgejo.container
+++ b/forgejo.container
@@ -3,7 +3,7 @@ ContainerName=forgejo
 Image=codeberg.org/forgejo/forgejo:16-rootless
 Label=io.containers.autoupdate=registry
 PublishPort=3000:3000
-PublishPort=2222:22
+PublishPort=2222:2222
 Volume=forgejo_data:/var/lib/gitea
 Environment=FORGEJO__security__INSTALL_LOCK=true
 Environment=FORGEJO__security__GLOBAL_TWO_FACTOR_REQUIREMENT=all

