From a1a897a630aedd07c78c0d6c5f451cd5c0dafbc5 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 04 Aug 2026 22:46:12 -0700
Subject: [PATCH] feat: Add external link markers to external links in headerbar and footer

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/custom/templates/custom/extra_links.tmpl b/custom/templates/custom/extra_links.tmpl
index 4c5205d4a35401830cd5a731c1c4def1f6fa6c26..67b862fd8b51b4cfa11526c5ea3478c65dd4da7f 100644
--- a/custom/templates/custom/extra_links.tmpl
+++ b/custom/templates/custom/extra_links.tmpl
@@ -1 +1 @@
-<a class="item" target="_blank" href="https://vanlug.ca/">{{svg "octicon-link-external"}} vanlug.ca</a>
+<a class="item" target="_blank" href="https://vanlug.ca/"><span class="tw-mr-1">vanlug.ca</span>{{svg "octicon-link-external"}}</a>
diff --git a/custom/templates/custom/extra_links_footer.tmpl b/custom/templates/custom/extra_links_footer.tmpl
index 9b42f994207e81876a521486e84044d32d56c095..27cfe3e8101b8ccb80fe978de1274c68b487d260 100644
--- a/custom/templates/custom/extra_links_footer.tmpl
+++ b/custom/templates/custom/extra_links_footer.tmpl
@@ -1,2 +1,2 @@
-<a class="item" target="_blank" href="https://vanlug.ca/terms/">Terms</a>
-<a class="item" target="_blank" href="https://vanlug.ca/privacy/">Privacy</a>
+<a class="item" target="_blank" href="https://vanlug.ca/terms/"><span class="tw-mr-1">Terms</span>{{svg "octicon-link-external"}}</a>
+<a class="item" target="_blank" href="https://vanlug.ca/privacy/"><span class="tw-mr-1">Privacy</span>{{svg "octicon-link-external"}}</a>

