From 31be9e0e031371ff83ef6dda1e479e5255891188 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 15 Aug 2026 10:50:51 -0700
Subject: [PATCH] feat: Add terms, privacy and CoC to sign in/up pages

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/custom/templates/custom/extra_links_footer.tmpl b/custom/templates/custom/extra_links_footer.tmpl
index 27cfe3e8101b8ccb80fe978de1274c68b487d260..51f12c2f83edff7f775e4c2e72238672c07c00d4 100644
--- a/custom/templates/custom/extra_links_footer.tmpl
+++ b/custom/templates/custom/extra_links_footer.tmpl
@@ -1,2 +1,3 @@
+<a class="item" target="_blank" href="https://vanlug.ca/conduct/"><span class="tw-mr-1">Code of Conduct</span>{{svg "octicon-link-external"}}</a>
 <a class="item" target="_blank" href="https://vanlug.ca/terms/"><span class="tw-mr-1">Terms</span>{{svg "octicon-link-external"}}</a>
 <a class="item" target="_blank" href="https://vanlug.ca/privacy/"><span class="tw-mr-1">Privacy</span>{{svg "octicon-link-external"}}</a>
diff --git a/custom/templates/custom/signin_mid.tmpl b/custom/templates/custom/signin_mid.tmpl
new file mode 100644
index 0000000000000000000000000000000000000000..7ba16aa00a654a1549cd1f2a821bc749c3ddddfe
--- /dev/null
+++ b/custom/templates/custom/signin_mid.tmpl
@@ -0,0 +1,10 @@
+<div class="ui warning message tw-mb-4">
+    <p>
+        By signing in or creating an account, you agree to our
+        <a target="_blank" href="https://vanlug.ca/terms/">terms {{svg "octicon-link-external"}}</a>
+        and
+        <a target="_blank" href="https://vanlug.ca/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
+        and to abide by our
+        <a target="_blank" href="https://vanlug.ca/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
+    </p>
+</div>
diff --git a/custom/templates/custom/signup_top.tmpl b/custom/templates/custom/signup_top.tmpl
new file mode 100644
index 0000000000000000000000000000000000000000..a9bbbf2e17464847a1fbb112d2a968fae8e9f808
--- /dev/null
+++ b/custom/templates/custom/signup_top.tmpl
@@ -0,0 +1,10 @@
+<div class="ui warning message tw-flex tw-mt-0 tw-mb-4 tw-text-start">
+	<p>
+		By signing in or creating an account, you agree to our
+		<a target="_blank" href="https://vanlug.ca/terms/">terms {{svg "octicon-link-external"}}</a>
+		and
+		<a target="_blank" href="https://vanlug.ca/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
+		and to abide by our
+		<a target="_blank" href="https://vanlug.ca/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
+	</p>
+</div>
\ No newline at end of file

