From c389f10b3fa6ecbdd14c84527e17f16f782a55e3 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 15 Aug 2026 11:35:57 -0700
Subject: [PATCH] fix: Use new runner OCI image

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/forgejo-runner-generate-config.sh b/forgejo-runner-generate-config.sh
index fd9b1bcfbe21c7d22855613dd92325b50e23cde4..4a4d640a6c3cc9320fc15340054c76f71840af18 100644
--- a/forgejo-runner-generate-config.sh
+++ b/forgejo-runner-generate-config.sh
@@ -5,7 +5,7 @@
 forgejo-runner generate-config >config.yml
 
 sed -i \
-	-e "s|  labels: \[\]|  labels:\n    - ubuntu-latest:docker://git.vanlug.p8.lu/pojntfx/forgejo-runner-docker-image:latest|" \
+	-e "s|  labels: \[\]|  labels:\n    - ubuntu-latest:docker://git.vanlug.p8.lu/pojntfx/forgejo-runner-meta:latest|" \
 	-e "s|  enable_ipv6: false|  enable_ipv6: true|" \
 	-e "s|  options:.*|  options: --security-opt label=disable --cap-add=all --volume /etc/resolv.conf:/etc/resolv.conf:ro --runtime=runsc|" \
 	-e "s|  valid_volumes: \[\]|  valid_volumes:\n    - /etc/resolv.conf|" \

