From be24f51a27ebff6ba71cca629b70c427dac4bf27 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 18 Aug 2026 22:24:19 -0700
Subject: [PATCH] fix: Allow Caddy to reload it's own configuration

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/caddy.cil b/caddy.cil
index 5ba0408af264f238ce4d6366ef4b623a36f56bb6..caf35ebe60712d12462c0d856eb6a68584afebc0 100644
--- a/caddy.cil
+++ b/caddy.cil
@@ -49,6 +49,8 @@     ; Dialing permission for upstreams
     (allow process forgejo_port_t ( tcp_socket ( name_connect )))
     (allow process git_pages_port_t ( tcp_socket ( name_connect )))
     (allow process anubis_port_t ( tcp_socket ( name_connect )))
+    ; Admin API to apply changes from Caddyfle
+    (allow process unreserved_port_t ( tcp_socket ( name_connect )))
 
     ; Custom rules end here
 )
\ No newline at end of file

