From 6b2ab81b1b483c8d55ef9b2cddde9205a7861b2c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 21 Aug 2026 19:38:13 -0700
Subject: [PATCH] chore: Use the `.p8.lu` domain consistently

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/config/botPolicies.yaml b/config/botPolicies.yaml
index b5219118c91536ef01b86fb4cc649a2f8128ac25..fba399d4f67d2f28c05233a3a9b7f96b95bf94a5 100644
--- a/config/botPolicies.yaml
+++ b/config/botPolicies.yaml
@@ -8,6 +8,6 @@   footer: Please see the imprint for our code of conduct, terms and privacy policy.
   page:
     title: Code of Conduct, Terms and Privacy Policy
     body: >-
-      <a target="_blank" href="https://vanlug.ca/conduct/">Code of Conduct</a>
-      <a target="_blank" href="https://vanlug.ca/terms/">Terms</a>
-      <a target="_blank" href="https://vanlug.ca/privacy/">Privacy</a>
+      <a target="_blank" href="https://vanlug.p8.lu/conduct/">Code of Conduct</a>
+      <a target="_blank" href="https://vanlug.p8.lu/terms/">Terms</a>
+      <a target="_blank" href="https://vanlug.p8.lu/privacy/">Privacy</a>
diff --git a/config/custom/templates/custom/extra_links.tmpl b/config/custom/templates/custom/extra_links.tmpl
index 67b862fd8b51b4cfa11526c5ea3478c65dd4da7f..fcb2ee7861b666305fa2f50c4aae59031b1d39b6 100644
--- a/config/custom/templates/custom/extra_links.tmpl
+++ b/config/custom/templates/custom/extra_links.tmpl
@@ -1 +1 @@
-<a class="item" target="_blank" href="https://vanlug.ca/"><span class="tw-mr-1">vanlug.ca</span>{{svg "octicon-link-external"}}</a>
+<a class="item" target="_blank" href="https://vanlug.p8.lu/"><span class="tw-mr-1">vanlug.p8.lu</span>{{svg "octicon-link-external"}}</a>
diff --git a/config/custom/templates/custom/extra_links_footer.tmpl b/config/custom/templates/custom/extra_links_footer.tmpl
index 51f12c2f83edff7f775e4c2e72238672c07c00d4..f25eb7ebc0d5e93ad8c5c822885f5475a7ddbd53 100644
--- a/config/custom/templates/custom/extra_links_footer.tmpl
+++ b/config/custom/templates/custom/extra_links_footer.tmpl
@@ -1,3 +1,3 @@
-<a class="item" target="_blank" href="https://vanlug.ca/conduct/"><span class="tw-mr-1">Code of Conduct</span>{{svg "octicon-link-external"}}</a>
-<a class="item" target="_blank" href="https://vanlug.ca/terms/"><span class="tw-mr-1">Terms</span>{{svg "octicon-link-external"}}</a>
-<a class="item" target="_blank" href="https://vanlug.ca/privacy/"><span class="tw-mr-1">Privacy</span>{{svg "octicon-link-external"}}</a>
+<a class="item" target="_blank" href="https://vanlug.p8.lu/conduct/"><span class="tw-mr-1">Code of Conduct</span>{{svg "octicon-link-external"}}</a>
+<a class="item" target="_blank" href="https://vanlug.p8.lu/terms/"><span class="tw-mr-1">Terms</span>{{svg "octicon-link-external"}}</a>
+<a class="item" target="_blank" href="https://vanlug.p8.lu/privacy/"><span class="tw-mr-1">Privacy</span>{{svg "octicon-link-external"}}</a>
diff --git a/config/custom/templates/custom/signin_mid.tmpl b/config/custom/templates/custom/signin_mid.tmpl
index 7ba16aa00a654a1549cd1f2a821bc749c3ddddfe..a4cd664db02da9c60e0a144355e0750b55b20577 100644
--- a/config/custom/templates/custom/signin_mid.tmpl
+++ b/config/custom/templates/custom/signin_mid.tmpl
@@ -1,10 +1,10 @@
 <div class="ui warning message tw-mb-4">
     <p>
         By signing in or creating an account, you agree to our
-        <a target="_blank" href="https://vanlug.ca/terms/">terms {{svg "octicon-link-external"}}</a>
+        <a target="_blank" href="https://vanlug.p8.lu/terms/">terms {{svg "octicon-link-external"}}</a>
         and
-        <a target="_blank" href="https://vanlug.ca/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
+        <a target="_blank" href="https://vanlug.p8.lu/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
         and to abide by our
-        <a target="_blank" href="https://vanlug.ca/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
+        <a target="_blank" href="https://vanlug.p8.lu/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
     </p>
 </div>
diff --git a/config/custom/templates/custom/signup_top.tmpl b/config/custom/templates/custom/signup_top.tmpl
index a9bbbf2e17464847a1fbb112d2a968fae8e9f808..1f27d084a03a4f7358d898838196fb91e2d0c2de 100644
--- a/config/custom/templates/custom/signup_top.tmpl
+++ b/config/custom/templates/custom/signup_top.tmpl
@@ -1,10 +1,10 @@
 <div class="ui warning message tw-flex tw-mt-0 tw-mb-4 tw-text-start">
 	<p>
 		By signing in or creating an account, you agree to our
-		<a target="_blank" href="https://vanlug.ca/terms/">terms {{svg "octicon-link-external"}}</a>
+		<a target="_blank" href="https://vanlug.p8.lu/terms/">terms {{svg "octicon-link-external"}}</a>
 		and
-		<a target="_blank" href="https://vanlug.ca/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
+		<a target="_blank" href="https://vanlug.p8.lu/privacy/">privacy policy {{svg "octicon-link-external"}}</a>,
 		and to abide by our
-		<a target="_blank" href="https://vanlug.ca/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
+		<a target="_blank" href="https://vanlug.p8.lu/conduct/">code of conduct {{svg "octicon-link-external"}}</a>.
 	</p>
 </div>
\ No newline at end of file

