From 1bd01ee2830c5e7a9023bd08c38c9dee40d1bf0d Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 26 May 2025 23:39:08 -0700
Subject: [PATCH] fix: Also use new asset import syntax for page background

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/src/App.tsx b/src/App.tsx
index 1a998d57c60d46467102b30f561f27ed28b082a3..5485d4cb5f9a294dbc97471477a4b12c1abf2bbc 100644
--- a/src/App.tsx
+++ b/src/App.tsx
@@ -1,10 +1,8 @@
 import { createBrowserRouter, RouterProvider } from "react-router-dom";
-// @ts-ignore
 import "semantic-ui-css/semantic.min.css";
 import { Container } from "semantic-ui-react";
 import constants from "./constants.json";
 import { Home } from "./Home";
-// @ts-ignore
 import { PageLayout } from "./PageLayout";
 import { Person } from "./Person";
 
diff --git a/src/PageLayout.tsx b/src/PageLayout.tsx
index 49e2fa5f7fc35d37f03e80ccd52dee7f7f439760..2a4198e3e52e01e67ea2d8b62496d196cac25607 100644
--- a/src/PageLayout.tsx
+++ b/src/PageLayout.tsx
@@ -1,9 +1,7 @@
-import { Button, ButtonGroup, Segment } from "semantic-ui-react";
-import { Link } from "react-router-dom";
-import styled from "@emotion/styled";
 import { css, Global } from "@emotion/react";
-// @ts-ignore
-import Bg from "../bg.webp";
+import styled from "@emotion/styled";
+import { Link } from "react-router-dom";
+import { Button, ButtonGroup, Segment } from "semantic-ui-react";
 
 const Wrapper = styled.div`
   min-height: 100%;
@@ -77,7 +75,10 @@             height: 100%;
           }
         }
         body {
-          background-image: url(${Bg}) !important;
+          background-image: url(${new URL(
+            "../bg.webp",
+            import.meta.url
+          ).toString()}) !important;
           background-position: 50% !important;
           background-attachment: fixed !important;
           background-repeat: no-repeat !important;

