From cfc88d5b49cd74a1a2f2d58b20c08fa473bb4778 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 21 Mar 2026 23:37:27 -0700
Subject: [PATCH] chore: Bump copyright

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/README.md b/README.md
index b470a6e5fecc886e30fce02fcc22b219aa86c025..319726db9b07db404bbab08a100f17b0ba18bdb8 100644
--- a/README.md
+++ b/README.md
@@ -51,6 +51,6 @@ ```
 
 ## License
 
-Family Site (c) 2025 Felicitas Pojtinger
+Family Site (c) 2026 Felicitas Pojtinger
 
 SPDX-License-Identifier: AGPL-3.0
diff --git a/src/PageLayout.tsx b/src/PageLayout.tsx
index 2a4198e3e52e01e67ea2d8b62496d196cac25607..e1530642cdd1eb8ead04f35f634edea29360d1b3 100644
--- a/src/PageLayout.tsx
+++ b/src/PageLayout.tsx
@@ -77,7 +77,7 @@         }
         body {
           background-image: url(${new URL(
             "../bg.webp",
-            import.meta.url
+            import.meta.url,
           ).toString()}) !important;
           background-position: 50% !important;
           background-attachment: fixed !important;
@@ -126,7 +126,7 @@       <Button
         color="green"
         as="a"
         href="https://github.com/pojntfx/pojtinger.com#license"
-        content="© 2025 Felicitas Pojtinger and contributors (AGPL-3.0)"
+        content="© 2026 Felicitas Pojtinger and contributors (AGPL-3.0)"
         target="_blank"
       />
 

