From a787905eb6a1cc192261c9e6943a0ecf06b05ac8 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 04 Jun 2026 22:17:17 -0700
Subject: [PATCH] feat: Add link to NetteART for Nette

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/src/Person.tsx b/src/Person.tsx
index 7ae2ca443b654473faffff50315a7bd925ce4f60..5038ba3996c06ce69eabc455ea1ee9deb42d7cdb 100644
--- a/src/Person.tsx
+++ b/src/Person.tsx
@@ -47,6 +47,19 @@           icon="mail"
           content="Mail"
           target="_blank"
         />
+        {props.person.websiteLink && (
+          <>
+            <div className="ui divider"></div>
+            <HeaderButton
+              as="a"
+              href={props.person.websiteLink}
+              icon="external"
+              secondary
+              content={props.person.websiteTitle || ""}
+              target="_blank"
+            />
+          </>
+        )}
       </Grid.Column>
     </Grid.Row>
   </HeaderWrapper>
diff --git a/src/constants.json b/src/constants.json
index f868521bdf5f7d8a7ba07516b456e53475c4ff87..5baa429fe7f86458bc5976e26430e605a7ef124e 100644
--- a/src/constants.json
+++ b/src/constants.json
@@ -7,7 +7,9 @@       "username": "@an.nette.pojtinger",
       "followLink": "https://www.instagram.com/an.nette.pojtinger/",
       "emailLink": "mailto:nette@pojtinger.com",
       "path": "/nette",
-      "avatarLink": "https://cloud.pojtinger.com/apps/files_sharing/publicpreview/CzS9ikQ8pbSdpAw?file=/Nette.jpg&fileId=2090309&x=3456&y=2160&a=true&etag=afb0f18888fc9b901912d84ebc72fb38"
+      "avatarLink": "https://cloud.pojtinger.com/apps/files_sharing/publicpreview/CzS9ikQ8pbSdpAw?file=/Nette.jpg&fileId=2090309&x=3456&y=2160&a=true&etag=afb0f18888fc9b901912d84ebc72fb38",
+      "websiteLink": "https://netteart.com/",
+      "websiteTitle": "NetteART"
     },
     {
       "name": "Michael",
diff --git a/src/constants.ts b/src/constants.ts
index 5d6d318a5172583dc3291de8df0bb9ef69aa28f8..fbb51ff22a1c2b322eb6a3464e08d24115e79f58 100644
--- a/src/constants.ts
+++ b/src/constants.ts
@@ -6,4 +6,6 @@   avatarLink: string;
   username: string;
   followLink: string;
   emailLink: string;
+  websiteLink?: string;
+  websiteTitle?: string;
 }

