From 7a772b9be819a8acdf9a87005e79ac9a79309549 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 16 Apr 2026 20:50:29 -0700
Subject: [PATCH] Merge branch gnome-build-meta:master into rework-docs---
diff --git a/elements/core-deps/evolution-data-server.bst b/elements/core-deps/evolution-data-server.bst
index 08250e846f1c2cddd69210fe7332708e0cf747ed..adef716f35afa92c8339ba51221802331954e212 100644
--- a/elements/core-deps/evolution-data-server.bst
+++ b/elements/core-deps/evolution-data-server.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:evolution-data-server.git
   track: master
-  ref: 3.60.0-14-g97a3e486ec58b881768fcf81fceb5bd1260fd4fc
+  ref: 3.60.0-16-ge611b762a89c58ce7fd9d4540665ed64c0240e6b
 
 build-depends:
 - sdk/gobject-introspection.bst
diff --git a/elements/core-deps/libgsf.bst b/elements/core-deps/libgsf.bst
index 6c0bb13a2eac6ac3c468d23ed22ad26e44e73791..45a416bd2822e018607851e5475674427b024633 100644
--- a/elements/core-deps/libgsf.bst
+++ b/elements/core-deps/libgsf.bst
@@ -3,7 +3,7 @@ sources:
 - kind: git_repo
   url: gnome:libgsf.git
   track: master
-  ref: LIBGSF_1_14_56-3-gd048fe91d63e0ef4541cccaf068a346bdacfb980
+  ref: LIBGSF_1_14_56-5-g4eadd61fbfc94a195fb53fb3010eaa17e0dc1777
 build-depends:
 - sdk/gobject-introspection.bst
 - sdk/gtk-doc.bst
diff --git a/elements/core-deps/localsearch.bst b/elements/core-deps/localsearch.bst
index 4d56621713ca616070cbe70fc6242acbf84a97c4..5be8cc9718c3bfeb3185f92ac04c25cb5233245b 100644
--- a/elements/core-deps/localsearch.bst
+++ b/elements/core-deps/localsearch.bst
@@ -3,7 +3,7 @@ sources:
 - kind: git_repo
   url: gnome:localsearch.git
   track: main
-  ref: 3.11.0-4-g511244d9a593a5e3c2f3478efc5b014a971470cd
+  ref: 3.11.1-0-g052bb7d3cd2e0ad308f98664dddee4902b8ec10d
 build-depends:
 - core-deps/python-dbusmock.bst
 - freedesktop-sdk.bst:components/asciidoc.bst
diff --git a/elements/core-deps/totem-pl-parser.bst b/elements/core-deps/totem-pl-parser.bst
index 822c511dc00e56da36438ff6532b6280344d40c3..0606f1947ffdd22e7183c73522e3d42d865f8150 100644
--- a/elements/core-deps/totem-pl-parser.bst
+++ b/elements/core-deps/totem-pl-parser.bst
@@ -3,7 +3,7 @@ sources:
 - kind: git_repo
   url: gnome:totem-pl-parser.git
   track: master
-  ref: 3.26.7-0-g050ad13280cd97b574d935045da987cc31016651
+  ref: 3.26.7-1-gb59867f870c819d3ea787a4fcba3fadaa366f90a
 build-depends:
 - sdk/gobject-introspection.bst
 - freedesktop-sdk.bst:public-stacks/buildsystem-meson.bst
diff --git a/elements/core-deps/vte.bst b/elements/core-deps/vte.bst
index 6f87b93cffd97bafd24db4c76160112126eb7ca1..645e42199b64a92abffc3b25b435a1509c70e2f6 100644
--- a/elements/core-deps/vte.bst
+++ b/elements/core-deps/vte.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:vte.git
   track: master
-  ref: 0.81.0-80-g77fe661760d2afac1e4d5a4b7eba477cbed6b430
+  ref: 0.81.0-81-gfc2760558f98f8a5808154264c3f4b74d7a3e19c
 
 build-depends:
 - sdk/gi-docgen.bst
diff --git a/elements/core/calls.bst b/elements/core/calls.bst
index 455bfd7a684573a77949a651c870257592d2cafb..c8497a3f66b1e50ff2f8992bbcaacd08080fe167 100644
--- a/elements/core/calls.bst
+++ b/elements/core/calls.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:calls.git
   track: main
-  ref: v50.0-3-gdfb1c6c57675a4132af2b881cdf34e244529010d
+  ref: v50.0-4-g56d401fba24f7795bcde14dc94d3cb6e04d4cb35
 # Needs to be updated manually until we have a wrap-git plugin
 - kind: git_repo
   url: gnome_gitlab:World/Phosh/libcall-ui.git
diff --git a/elements/core/epiphany.bst b/elements/core/epiphany.bst
index 4061bb2e84cb5e2a6ddc3e0831f58fb66f950658..9a3345339357b1ef75104f4d343313465d7aeaf6 100644
--- a/elements/core/epiphany.bst
+++ b/elements/core/epiphany.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:epiphany.git
   track: main
-  ref: 50.3-13-g621905b15df21177380a598c7f8479f6d91ad437
+  ref: 50.3-16-g5e03dd45f03d45eef4ca61ad5f33f6b448150954
 
 build-depends:
 - sdk/blueprint-compiler.bst
diff --git a/elements/core/gdm.bst b/elements/core/gdm.bst
index 54c9062ffc61b86649ed645e3519237b0a84e7fe..7c99663ecce23f6a8b45b30976ed8a7a995a5970 100644
--- a/elements/core/gdm.bst
+++ b/elements/core/gdm.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gdm.git
   track: main
-  ref: 50.0-8-g64533cdb09005e826ef6a10731b90175152f5ad5
+  ref: 50.0-10-g9caa2488fbb16068ad117b9cc3aaaec20c5a45b6
 - kind: local
   path: files/gdm
   directory: gdm-config
diff --git a/elements/core/gnome-boxes.bst b/elements/core/gnome-boxes.bst
index 68e9e1d3ff36c1a1f34acea8467ea67246e7d871..1548442a601594ba66a59a1fbdc13d509bf17917 100644
--- a/elements/core/gnome-boxes.bst
+++ b/elements/core/gnome-boxes.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-boxes.git
   track: main
-  ref: 50.0-11-g5b0727f2228ff61d8d8c8dad33afa13544207d9c
+  ref: 50.0-14-gbdb517b9acda31df4537758370626d8638731edc
 - kind: git_module
   url: gnome_gitlab:felipeborges/libovf-glib.git
   path: subprojects/libovf-glib
diff --git a/elements/core/gnome-calculator.bst b/elements/core/gnome-calculator.bst
index 1b51b53aa3eefd0ff6f40493e1b957397a3f8a13..301dfa0e76fe13304ffc85ad79e9ef1ecb524f40 100644
--- a/elements/core/gnome-calculator.bst
+++ b/elements/core/gnome-calculator.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-calculator.git
   track: main
-  ref: 50.0-10-gb8c42074248fe9f11d1f39f96de598207072eec7
+  ref: 50.0-14-g53bb400c6947c33b19c2028cb06b456551f792a8
 
 build-depends:
 - sdk/blueprint-compiler.bst
diff --git a/elements/core/gnome-clocks.bst b/elements/core/gnome-clocks.bst
index ccbb0e1815ad4674efe1fe8a9cc0dc003f848081..45b0b4c393d135c85742b113cca5a4cf0909c162 100644
--- a/elements/core/gnome-clocks.bst
+++ b/elements/core/gnome-clocks.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-clocks.git
   track: master
-  ref: 50.0-5-gd270b9336346ed1f035c22b5c5d1532162264d8c
+  ref: 50.0-6-g5ba6dea5ade064076e8c4d4d65e481e8652c4d21
 
 build-depends:
 - core-deps/vorbis-tools.bst
diff --git a/elements/core/gnome-control-center.bst b/elements/core/gnome-control-center.bst
index c82cce37ea3e4a86710d189dd0789a07c0e9fc5f..40dc6b678c0bc55e18a032a5c1b4ab66d8a4cb34 100644
--- a/elements/core/gnome-control-center.bst
+++ b/elements/core/gnome-control-center.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-control-center.git
   track: main
-  ref: 50.0-75-geea3ace61442e9a0db9990e7c375188443d922e3
+  ref: 50.0-109-gd92819a174cea471412360f4d17809a2ddf904a1
 - kind: git_module
   url: gnome:libgnome-volume-control.git
   path: subprojects/gvc
diff --git a/elements/core/gnome-maps.bst b/elements/core/gnome-maps.bst
index 472c0736a86d0cea3a7375b50c77e551a9bd7a2a..60f9d84042f5c4f6d74a5f6c788d9bf1635d15cb 100644
--- a/elements/core/gnome-maps.bst
+++ b/elements/core/gnome-maps.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-maps.git
   track: main
-  ref: 50.rc-37-gaf88ace6542b10bbbb05aab773ad2fb147175e59
+  ref: 50.rc-45-gd9bc3b7765c0779eb6fad8872502c9fe0860b804
 
 build-depends:
 - freedesktop-sdk.bst:components/desktop-file-utils.bst
diff --git a/elements/core/gnome-settings-daemon.bst b/elements/core/gnome-settings-daemon.bst
index 49ed24a5ff610bfafacde645367460961e535298..2a0655d57ea19c7eedd2e856490f324f17ac5d29 100644
--- a/elements/core/gnome-settings-daemon.bst
+++ b/elements/core/gnome-settings-daemon.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-settings-daemon.git
   track: main
-  ref: 50.0-0-g0037c5aab6ea7141640c1919d2fae7a604c61551
+  ref: 50.1-0-gec681847221cf44e658363b9d8137b4cbae7b321
 - kind: git_module
   url: gnome:libgnome-volume-control.git
   path: subprojects/gvc
diff --git a/elements/core/gnome-shell.bst b/elements/core/gnome-shell.bst
index 23a97413a550253716dde9f56296464862a5336b..0b3c967c700d61228cfb03c8dd281979c3a15bbe 100644
--- a/elements/core/gnome-shell.bst
+++ b/elements/core/gnome-shell.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-shell.git
   track: main
-  ref: 50.0-33-gfbd2b6b82305891b0f0f397adff11c109baad3d7
+  ref: 50.1-15-g8b930bdb9c950e118a2c53f1b576f73ac579cadd
 # FIXME: Keep this in sync with the repo until we can use wraps as sources
 # https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/subprojects/gvc.wrap?ref_type=heads
 #
diff --git a/elements/core/gnome-software.bst b/elements/core/gnome-software.bst
index 3dc1911c262f3d63d7504737ebde96578b28ba2b..61052057d3d492fdbd73bef5f8ad594cdc5cb8fb 100644
--- a/elements/core/gnome-software.bst
+++ b/elements/core/gnome-software.bst
@@ -9,6 +9,8 @@ - kind: git_module
   url: gnome_gitlab:mwleeds/gnome-pwa-list.git
   path: subprojects/gnome-pwa-list
   ref: 48ac9f7ed6a639eab7648f3e3b19105694eaf306
+- kind: patch
+  path: patches/gnome-software/describe.patch
 
 build-depends:
 - sdk/gtk-doc.bst
diff --git a/elements/core/gnome-system-monitor.bst b/elements/core/gnome-system-monitor.bst
index 35e2e9aaf47aa8ac8fbfe35095f82c81f122fef5..670d270c1a53dedeb6d5bfcad0453e67374808b2 100644
--- a/elements/core/gnome-system-monitor.bst
+++ b/elements/core/gnome-system-monitor.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gnome-system-monitor.git
   track: main
-  ref: 50.0-6-g24f6b576b7f5039a05fb991eabc585bace434a87
+  ref: 50.0-9-gce333e7c04f6b9575568fc3af48a7eca362e7b71
 
 build-depends:
 - core-deps/catch2.bst
diff --git a/elements/core/mutter.bst b/elements/core/mutter.bst
index cb2bf50d2fbb9575fd3b556eeb9d028223624d23..3483e535140dfc1a215ecf4ec3e9ce28453b33c6 100644
--- a/elements/core/mutter.bst
+++ b/elements/core/mutter.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:mutter.git
   track: main
-  ref: 50.0-37-gaa5b4892aabcca49047f32865dfb65198dab2ec3
+  ref: 50.1-0-g8e9bb806da4a6a4fc87aa4a481b228dba6735cbd
 # FIXME: Keep this in sync with the repo until we can use wraps as sources
 # https://gitlab.gnome.org/GNOME/mutter/-/blob/main/subprojects/gvdb.wrap?ref_type=heads
 - kind: git_repo
diff --git a/elements/core/nautilus.bst b/elements/core/nautilus.bst
index 9bc0ef5429944e4f6c5d934f0e26f4f67f33e2d2..fa4b457ec02c20e8c8e7a1d90d2d4226e22061af 100644
--- a/elements/core/nautilus.bst
+++ b/elements/core/nautilus.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:nautilus.git
   track: main
-  ref: 50.0-78-gae22997924f34d4a97f2eefc892c4ea26a810b79
+  ref: 50.0-85-g696ec02f29a4d32c6fb7579556bab49ded8be101
 
 build-depends:
 - sdk/blueprint-compiler.bst
diff --git a/elements/core/orca.bst b/elements/core/orca.bst
index 7ff83b6b3c389867cab9544eb0cf14aff697373f..69899ce10486f122088171b71cba5c99a3783e61 100644
--- a/elements/core/orca.bst
+++ b/elements/core/orca.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:orca.git
   track: main
-  ref: 50.0-189-g1c2e0fd67f20b35f360e18a0686651b0ba0d9d59
+  ref: 50.0-192-g75cd32032d642cb746da476d4c4f5b9aa555c017
 
 build-depends:
 - freedesktop-sdk.bst:components/systemd.bst
diff --git a/elements/core/snapshot.bst b/elements/core/snapshot.bst
index 80835a770be5c41d04936a8ef15c76f4ef0e290e..bc3c80734e977bc27131973928eff7a6164307d2 100644
--- a/elements/core/snapshot.bst
+++ b/elements/core/snapshot.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:snapshot.git
   track: main
-  ref: 50.0-1-ga1b7ef52d25c2a3db311a4df3c597b99cb15d163
+  ref: 50.0-2-g79ae4187c051825ef9e60c9aaebbe35976d9b05b
 - kind: cargo2
   ref:
   - kind: registry
diff --git a/elements/sdk/gcr.bst b/elements/sdk/gcr.bst
index f1131b7f48b94ccec05ed77a53c9d43f803156bb..66ad659c8c83ad86a918c438729fcab8ed72c7e4 100644
--- a/elements/sdk/gcr.bst
+++ b/elements/sdk/gcr.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gcr.git
   track: main
-  ref: 4.4.0.1-21-g33d78d804c4c2131fedcca0512a9e45b384c4ba6
+  ref: 4.4.0.1-22-g91d2a8d7806fe315fb3bd442a7b7af38955ae623
 
 build-depends:
 - sdk/gi-docgen.bst
diff --git a/elements/sdk/gsettings-desktop-schemas.bst b/elements/sdk/gsettings-desktop-schemas.bst
index 5b64c7438f95af8e126b009d670c041bf6568469..edd0489314b50a0cedafc1cb6e7cd832dd854714 100644
--- a/elements/sdk/gsettings-desktop-schemas.bst
+++ b/elements/sdk/gsettings-desktop-schemas.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gsettings-desktop-schemas.git
   track: main
-  ref: 50.0-6-g8c672b3a0b659ddafb361411f820589627653642
+  ref: 50.1-0-gea4bd542cfd931be783667fb84e02414f04e3ba9
 
 build-depends:
 - sdk/gobject-introspection.bst
diff --git a/elements/sdk/gtk-doc.bst b/elements/sdk/gtk-doc.bst
index 9f6378b5d52bdbb8a096d64f9532a324e788ebf9..b38c5d80e0a2633dce8fddd5b00b0825c32f8c5d 100644
--- a/elements/sdk/gtk-doc.bst
+++ b/elements/sdk/gtk-doc.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gtk-doc.git
   track: master
-  ref: 1.36.1-0-g6d87f9b2dadc8977a000d7a66d98ac662fa803b7
+  ref: 1.36.1-1-gcf004748a68e947e5cd995d3791c7b1ae2e0f2bf
 
 build-depends:
 - freedesktop-sdk.bst:components/itstool.bst
diff --git a/elements/sdk/gtk.bst b/elements/sdk/gtk.bst
index 45cc146240c5a404682332ef0f5aacaef70320bb..35f9cab30588d9bb1d3eadf1fb26e2db073b351a 100644
--- a/elements/sdk/gtk.bst
+++ b/elements/sdk/gtk.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:gtk.git
   track: main
-  ref: 4.22.0-713-g1245603862ed482537d68b1983386b32f233695c
+  ref: 4.23.0-15-gb67affadeeefe2c2165748407539165bea626001
 
 build-depends:
 - sdk-deps/sassc.bst
diff --git a/elements/sdk/libadwaita.bst b/elements/sdk/libadwaita.bst
index 841e24f7025dc044d6b102949fdc2d07eb6d51bc..00ca472c600ed7cc9b5ef7b44f88e9792b565bb8 100644
--- a/elements/sdk/libadwaita.bst
+++ b/elements/sdk/libadwaita.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:libadwaita.git
   track: main
-  ref: 1.9.0-33-gd7c93316faf11da78940381f479125d9e797edd3
+  ref: 1.9.0-45-gda740214b6c4af456996e620be59f91f19e216f5
 
 build-depends:
 - sdk-deps/sassc.bst
diff --git a/elements/sdk/libdex.bst b/elements/sdk/libdex.bst
index 2250522258aac0fe924ad007facf8f73f6fe64dd..cd2d08803520d57b97181d3266223648ce44ae60 100644
--- a/elements/sdk/libdex.bst
+++ b/elements/sdk/libdex.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:libdex.git
   track: main
-  ref: 1.1.0-2-g0d1e1052d01763b5608fb9a20940499eb31d45b9
+  ref: 1.1.0-4-gc73e762c0bef00ee4fc7510b74b13c3492f4fc84
 
 build-depends:
 - sdk/gi-docgen.bst
diff --git a/elements/sdk/libsoup.bst b/elements/sdk/libsoup.bst
index 227618e4354e7c33435ce1c92412b7de9c671182..1bb758497a036c065cf09d8906c5a875021ddd89 100644
--- a/elements/sdk/libsoup.bst
+++ b/elements/sdk/libsoup.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:libsoup.git
   track: master
-  ref: 3.6.4-128-g53a609e9c73161335fbd6dde34cf614fb37683e9
+  ref: 3.6.4-136-g1d408d3d7e5770e9fd0db89653bb4c8490d90071
 
 build-depends:
 - sdk/gi-docgen.bst
diff --git a/elements/sdk/tinysparql.bst b/elements/sdk/tinysparql.bst
index 6b431c00a12738857884041548848e29e35b838b..dab381c41d25d23d191479c764e4bcb43f6a37f9 100644
--- a/elements/sdk/tinysparql.bst
+++ b/elements/sdk/tinysparql.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:tinysparql.git
   track: main
-  ref: 3.11.0-15-g7afef0dbb3333180b6133ea3b819ca15c4ab929e
+  ref: 3.11.1-0-ga9677b6c350f2f9e179a265549e312bee31ae5ac
 # Need to be updated manually until we get a meson wrap source
 - kind: remote
   url: snowballstem:libstemmer_c-3.0.1.tar.gz
diff --git a/elements/sdk/yelp.bst b/elements/sdk/yelp.bst
index 3a0222d63b251861d4405cf203446c57716657d8..d8a31944ce80cc8db42f1bc3c11ee7ded4c47fe0 100644
--- a/elements/sdk/yelp.bst
+++ b/elements/sdk/yelp.bst
@@ -4,7 +4,7 @@ sources:
 - kind: git_repo
   url: gnome:yelp.git
   track: master
-  ref: 49.0-23-gb96450ac5156aab597fc9cc34621175f2ef28cfe
+  ref: 49.0-24-gd1fa128fc48219bc2c00b566152ed381f40a92f8
 
 build-depends:
 - freedesktop-sdk.bst:components/appstream.bst
diff --git a/patches/gnome-software/describe.patch b/patches/gnome-software/describe.patch
new file mode 100644
index 0000000000000000000000000000000000000000..daa1c62dabc9a243aedd15b7c9fa4c90771313aa
--- /dev/null
+++ b/patches/gnome-software/describe.patch
@@ -0,0 +1,184 @@
+commit dae27ea08152f3aefeb5fd548f795216c8e17baa
+Author: Valentin David <me@valentindavid.com>
+Date:   Wed Apr 15 20:05:37 2026 +0200
+
+    plugins/systemd-sysupdate: Use describe to find status of update
+    
+    An update could be fully installed, pending, partial or incomplete.
+
+diff --git a/plugins/systemd-sysupdate/gs-plugin-systemd-sysupdate.c b/plugins/systemd-sysupdate/gs-plugin-systemd-sysupdate.c
+index 001c919c4..6e982cf67 100644
+--- a/plugins/systemd-sysupdate/gs-plugin-systemd-sysupdate.c
++++ b/plugins/systemd-sysupdate/gs-plugin-systemd-sysupdate.c
+@@ -11,6 +11,7 @@
+ #include <gio/gio.h>
+ #include <glib.h>
+ #include <glib/gi18n.h>
++#include <json-glib/json-glib.h>
+ #include <malloc.h>
+ #include <xmlb.h>
+ 
+@@ -62,6 +63,10 @@
+ typedef struct {
+ 	GsSystemdSysupdateTarget *proxy;
+ 	gboolean is_valid;
++	gboolean is_installed;
++	gboolean is_pending;
++	gboolean is_partial;
++	gboolean is_incomplete;
+ 	gchar *id; /* (owned) (not nullable) */
+ 	gchar *class; /* (owned) (not nullable) */
+ 	gchar *name; /* (owned) (not nullable) */
+@@ -115,21 +120,21 @@ target_item_get_id (TargetItem *target)
+ }
+ 
+ static gboolean
+-target_item_is_available (TargetItem *target)
++target_item_is_pending (TargetItem *target)
+ {
+-	return target->latest_version != NULL;
++	return target->is_installed && target->is_pending && !target->is_partial  && !target->is_incomplete;
+ }
+ 
+ static gboolean
+ target_item_is_installed (TargetItem *target)
+ {
+-	return target->current_version != NULL;
++	return target->is_installed && !target->is_pending && !target->is_partial  && !target->is_incomplete;
+ }
+ 
+ static gboolean
+ target_item_is_updatable (TargetItem *target)
+ {
+-	return target_item_is_available (target) && target_item_is_installed (target);
++	return target->is_pending || target->is_incomplete || target->is_partial || !target->is_installed;
+ }
+ 
+ static gboolean
+@@ -596,6 +601,11 @@ gs_plugin_systemd_sysupdate_target_refresh_metadata_check_new_cb (GObject      *
+                                                                   GAsyncResult *result,
+                                                                   gpointer      user_data);
+ 
++static void
++gs_plugin_systemd_sysupdate_target_refresh_metadata_describe_cb (GObject      *source_object,
++								 GAsyncResult *result,
++								 gpointer      user_data);
++
+ static gboolean
+ gs_plugin_systemd_sysupdate_target_refresh_metadata_finish (GsPlugin      *plugin,
+                                                             GAsyncResult  *result,
+@@ -833,10 +843,16 @@ update_app_for_target (GsPluginSystemdSysupdate *self,
+ {
+ 	const gchar *app_version = NULL;
+ 	GsAppState app_state = GS_APP_STATE_UNKNOWN;
++	GsSizeType size_type = GS_SIZE_TYPE_UNKNOWN;
+ 
+-	if (target_item_is_updatable (target)) {
++	if (target_item_is_pending (target)) {
++		app_version = target->latest_version;
++		app_state = GS_APP_STATE_UPDATABLE;
++		size_type = GS_SIZE_TYPE_VALID;
++	} else if (target_item_is_updatable (target)) {
+ 		app_version = target->latest_version;
+ 		app_state = GS_APP_STATE_UPDATABLE;
++		size_type = GS_SIZE_TYPE_UNKNOWABLE;
+ 	} else if (target_item_is_installed (target)) {
+ 		if (g_strcmp0 (target->class, "host") == 0) {
+ 			app_version = self->os_version;
+@@ -844,10 +860,12 @@ update_app_for_target (GsPluginSystemdSysupdate *self,
+ 			app_version = target->current_version;
+ 		}
+ 		app_state = GS_APP_STATE_INSTALLED;
++		size_type = GS_SIZE_TYPE_VALID;
+ 	}
+ 
+ 	gs_app_set_version (app, app_version);
+ 	gs_app_set_state (app, app_state);
++	gs_app_set_size_download (app, size_type, 0);
+ }
+ 
+ /* Remove the given job. It is called when the server notified us a job
+@@ -2015,9 +2033,9 @@ gs_plugin_systemd_sysupdate_target_refresh_metadata_check_new_cb (GObject      *
+ 	GsPluginSystemdSysupdateTargetRefreshMetadataData *data = NULL;
+ 	g_autoptr(GTask) task = g_steal_pointer (&user_data);
+ 	g_autoptr(GError) local_error = NULL;
+-	GsPluginSystemdSysupdate *self = g_task_get_source_object (task);
+-	g_autoptr(GsApp) app = NULL;
+ 	g_autofree gchar *latest_version = NULL;
++	const gchar *version = NULL;
++	GCancellable *cancellable = g_task_get_cancellable (task);
+ 
+ 	/* currently, the returned result contains only one string
+ 	 * representing the latest version found in the server. However,
+@@ -2039,6 +2057,71 @@ gs_plugin_systemd_sysupdate_target_refresh_metadata_check_new_cb (GObject      *
+ 		data->target->latest_version = g_steal_pointer (&latest_version);
+ 	}
+ 
++	version = data->target->latest_version != NULL ? data->target->latest_version
++	                                         : data->target->current_version;
++
++	gs_systemd_sysupdate_target_call_describe (data->target->proxy,
++	                                           version,
++	                                           SYSUPDATED_TARGET_DESCRIBE_FLAGS_NONE,
++	                                           G_DBUS_CALL_FLAGS_NONE,
++	                                           SYSUPDATED_TARGET_DESCRIBE_TIMEOUT_MS,
++	                                           cancellable,
++	                                           gs_plugin_systemd_sysupdate_target_refresh_metadata_describe_cb,
++	                                           g_steal_pointer (&task));
++}
++
++static void
++gs_plugin_systemd_sysupdate_target_refresh_metadata_describe_cb (GObject      *source_object,
++								 GAsyncResult *result,
++								 gpointer      user_data)
++{
++	g_autoptr(GTask) task = g_steal_pointer (&user_data);
++	GsPluginSystemdSysupdateTargetRefreshMetadataData *data = g_task_get_task_data (task);
++	g_autoptr(GsApp) app = NULL;
++	g_autoptr(GError) local_error = NULL;
++	GsPluginSystemdSysupdate *self = g_task_get_source_object (task);
++	g_autofree gchar *json = NULL;
++	g_autoptr(JsonParser) json_parser = NULL;
++	JsonNode *json_root;
++	JsonObject *json_item;
++
++	if (!gs_systemd_sysupdate_target_call_describe_finish (GS_SYSTEMD_SYSUPDATE_TARGET (source_object),
++	                                                       &json,
++	                                                       result,
++	                                                       &local_error)) {
++		g_task_return_error (task, g_steal_pointer (&local_error));
++		return;
++	}
++
++	json_parser = json_parser_new_immutable ();
++	if (!json_parser_load_from_data (json_parser, json, -1, &local_error)) {
++		g_task_return_error (task, g_steal_pointer (&local_error));
++		return;
++	}
++
++	json_root = json_parser_get_root (json_parser);
++	if (json_root == NULL) {
++		g_task_return_new_error (task, G_IO_ERROR, GS_PLUGIN_ERROR_FAILED,
++		                         "json from describe call has no root: %s", target_item_get_id (data->target));
++		return;
++	}
++	if (json_node_get_node_type (json_root) != JSON_NODE_OBJECT) {
++		g_task_return_new_error (task, G_IO_ERROR, GS_PLUGIN_ERROR_FAILED,
++		                         "json from describe call is not an object: %s", target_item_get_id (data->target));
++		return;
++	}
++	json_item = json_node_get_object (json_root);
++	if (json_item == NULL) {
++		g_task_return_new_error (task, G_IO_ERROR, GS_PLUGIN_ERROR_FAILED,
++		                         "json from describe call is not an object: %s", target_item_get_id (data->target));
++		return;
++	}
++
++	data->target->is_installed = json_object_get_boolean_member (json_item, "installed");
++	data->target->is_pending = json_object_get_boolean_member (json_item, "pending");
++	data->target->is_partial = json_object_get_boolean_member (json_item, "partial");
++	data->target->is_incomplete = json_object_get_boolean_member (json_item, "incomplete");
++
+ 	/* update app state base on the target's new version */
+ 	app = get_or_create_app_for_target (self, data->target, &local_error);
+ 	if (app == NULL) {

