From 2290d79ffda8752ec67f21033d5c84dbbf18976c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 26 May 2026 22:25:40 -0700
Subject: [PATCH] docs: Symlink secure boot keys to previous location

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/files/boot-keys/README.md b/files/boot-keys/README.md
new file mode 120000
index 0000000000000000000000000000000000000000..b062697d65370ef185f0e209ffb36ac8ff7c577c
--- /dev/null
+++ b/files/boot-keys/README.md
@@ -0,0 +1 @@
+../../docs/keys.md
\ No newline at end of file

