From 7fdb6b3ffea3f007fbd5e651a7383db9318d9e12 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 04 Jun 2026 22:52:20 -0700
Subject: [PATCH] feat: Add `theme` color and equivalents support

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/scss/main.scss b/assets/scss/main.scss
index a99f415d07febda494c88e7bd4c58fe606c1352f..86ed2ebc1eabb6b43e5d0ca6eeb2e19d31506e35 100644
--- a/assets/scss/main.scss
+++ b/assets/scss/main.scss
@@ -24,15 +24,22 @@   --pf-t--global--font--family--body: "Ubuntu", sans-serif;
   --pf-t--global--font--family--heading: "Ubuntu", sans-serif;
 }
 
+html {
+  // On some platforms (e.g. Safari on macOS 26+) this is used instead of `theme-color`
+  // See https://grooovinger.com/notes/2026-02-27-safari-26-header-background
+  background-color: var(--pf-t--global--background--color--secondary--default);
+}
+
 .pf-v6-c-page__main-container--home {
   @extend .pf-v6-u-m-0;
   border-radius: 0;
   min-height: 100%;
   border: 0;
 
+  // Keep this in sync with the <html> background-color
   --bg-tint: color-mix(
     in srgb,
-    var(--pf-t--global--background--color--primary--default) 90%,
+    var(--pf-t--global--background--color--secondary--default) 90%,
     transparent
   );
 
diff --git a/layouts/baseof.html b/layouts/baseof.html
index 7a0e10a881465f6199eb51fb5519cf82e7fcb9d2..e464d10f97956e3ea069535a67ecbc5f993ce052 100644
--- a/layouts/baseof.html
+++ b/layouts/baseof.html
@@ -31,8 +31,9 @@       {{ end }}
     {{ end }}
     <link rel="icon" href="{{ "design/icon.svg" | relURL }}" />
 
-    <meta name="theme-color" content="#000000" media="(prefers-color-scheme: light)" />
-    <meta name="theme-color" content="#000000" media="(prefers-color-scheme: dark)" />
+    <!-- Keep these in sync with the <html> element's background colour -->
+    <meta name="theme-color" content="#f2f2f2" media="(prefers-color-scheme: light)" />
+    <meta name="theme-color" content="#151515" media="(prefers-color-scheme: dark)" />
 
     <meta name="author" content="{{ .Site.Data.site.author }}" />
     <meta name="description" content="{{ .Site.Data.site.description }}" />

