From 6cb6901a9ca7a9a75bc500409bde564677781a96 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 04 Jun 2026 20:20:04 -0700
Subject: [PATCH] feat: Add support for high-contrast mode

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/js/head.js b/assets/js/head.js
index a9b98a695a1b3730bc149ba2eb94ee827cd07fb0..6d06cdbc6ccfe5878b9e8b5f5e2b739b2568d516 100644
--- a/assets/js/head.js
+++ b/assets/js/head.js
@@ -1,10 +1,20 @@
-const applyTheme = () =>
+const applyTheme = () => {
   document.documentElement.classList.toggle(
     "pf-v6-theme-dark",
     matchMedia("(prefers-color-scheme: dark)").matches,
   );
+  document.documentElement.classList.toggle(
+    "pf-v6-theme-high-contrast",
+    matchMedia("(prefers-contrast: more)").matches,
+  );
+};
+
 applyTheme();
 matchMedia("(prefers-color-scheme: dark)").addEventListener(
   "change",
   applyTheme,
 );
+matchMedia("(prefers-contrast: more)").addEventListener(
+  "change",
+  applyTheme,
+);

