From 6530e17a9d2accc229384908febc48228793c210 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 04 Jun 2026 22:24:06 -0700
Subject: [PATCH] feat: Add `netteart.com` as the custom domain

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.domains b/.domains
new file mode 100644
index 0000000000000000000000000000000000000000..dd83c15d90d8d34cfeb12a92b1e7dfe79ac7680b
--- /dev/null
+++ b/.domains
@@ -0,0 +1 @@
+netteart.com
\ No newline at end of file

