From aceecf259ce3df31a6f1254bfead256939f8167f Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 04 Jun 2026 22:04:00 -0700
Subject: [PATCH] feat: Add background picture to home page

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/scss/main.scss b/assets/scss/main.scss
index 55597ed570b6cb5771796644fed20c2ccc860f88..a99f415d07febda494c88e7bd4c58fe606c1352f 100644
--- a/assets/scss/main.scss
+++ b/assets/scss/main.scss
@@ -28,6 +28,19 @@ .pf-v6-c-page__main-container--home {
   @extend .pf-v6-u-m-0;
   border-radius: 0;
   min-height: 100%;
+  border: 0;
+
+  --bg-tint: color-mix(
+    in srgb,
+    var(--pf-t--global--background--color--primary--default) 90%,
+    transparent
+  );
+
+  background-image:
+    linear-gradient(var(--bg-tint), var(--bg-tint)), url(/img/bg.webp);
+  background-position: center;
+  background-repeat: no-repeat;
+  background-size: cover;
 }
 
 .logo--home {
@@ -51,3 +64,7 @@   --pf-v6-c-masthead--c-toolbar--Width: auto;
   --pf-v6-c-masthead__logo--Width: 38px;
   --pf-v6-c-masthead__main--MarginInlineEnd: 1rem;
 }
+
+.footer--home {
+  background: transparent;
+}
diff --git a/layouts/_partials/footer.html b/layouts/_partials/footer.html
index f8004f8e6e21a023496c15200af279a60c910714..de9b4ab318402e6d8daeea5b48308992f88ea663 100644
--- a/layouts/_partials/footer.html
+++ b/layouts/_partials/footer.html
@@ -1,4 +1,4 @@
-<footer class="pf-v6-c-page__main-section pf-m-no-fill">
+<footer class="pf-v6-c-page__main-section pf-m-no-fill {{ if .IsHome }}footer--home{{ end }}">
   <div class="pf-v6-c-page__main-body">
     <div
       class="pf-v6-l-grid pf-m-gutter pf-v6-u-py-xl-on-sm pf-v6-u-py-0-on-md pf-v6-u-align-items-center"
diff --git a/static/img/bg.webp b/static/img/bg.webp
new file mode 100644
index 0000000000000000000000000000000000000000..76f4e0386ef9090a2f7a4ad920c5835a55336813
Binary files /dev/null and b/static/img/bg.webp differ

