From cbb6c02e9ca0ffb121a9435f900516a82e012ab2 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 28 Jun 2026 12:44:10 -0700
Subject: [PATCH] chore: Also add `www` subdomain to `.domains` file

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.domains b/.domains
index dd83c15d90d8d34cfeb12a92b1e7dfe79ac7680b..c5818bd59d41bf5d5acc2bca6668de12028bc50d 100644
--- a/.domains
+++ b/.domains
@@ -1 +1,2 @@
-netteart.com
\ No newline at end of file
+netteart.com
+www.netteart.com
\ No newline at end of file

