From b118ef81e11abbc8d6def59c729357a17def728c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 28 Jun 2026 12:53:31 -0700
Subject: [PATCH] chore: Drop `www` subdomain again and use a redirect instead

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.domains b/.domains
index c5818bd59d41bf5d5acc2bca6668de12028bc50d..dd83c15d90d8d34cfeb12a92b1e7dfe79ac7680b 100644
--- a/.domains
+++ b/.domains
@@ -1,2 +1 @@
-netteart.com
-www.netteart.com
\ No newline at end of file
+netteart.com
\ No newline at end of file

