From 59211b03b8d3550d92e4589d5360df366e425f95 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 28 Jun 2026 16:28:27 -0700
Subject: [PATCH] feat: Start adding support for clamped footer to page

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/scss/main.scss b/assets/scss/main.scss
index 86ed2ebc1eabb6b43e5d0ca6eeb2e19d31506e35..2bc1e662c7d1c7f4b38f71218b38abf2a9d4296f 100644
--- a/assets/scss/main.scss
+++ b/assets/scss/main.scss
@@ -70,8 +70,21 @@   --pf-v6-c-masthead__content--Order: 0;
   --pf-v6-c-masthead--c-toolbar--Width: auto;
   --pf-v6-c-masthead__logo--Width: 38px;
   --pf-v6-c-masthead__main--MarginInlineEnd: 1rem;
+
+  justify-content: center;
 }
 
 .footer--home {
   background: transparent;
 }
+
+.pf-v6-c-page__main-body {
+  max-width: var(--pf-v6-c-page--section--m-limit-width--MaxWidth);
+}
+
+// TODO: This seems to collapse the section down to <992px, unlike `<main>`,
+// which is bounded by `var(--pf-v6-c-page--section--m-limit-width--MaxWidth)`.
+// We should fix this so that both the footer and content are clamped in the same way.
+footer.pf-v6-c-page__main-section {
+  align-items: center;
+}
diff --git a/layouts/baseof.html b/layouts/baseof.html
index 726bbe2dfdd92df1a4a5602c1b63e61695954d6d..35dfb98dca77d42418b17a3fa5740de2567c064e 100644
--- a/layouts/baseof.html
+++ b/layouts/baseof.html
@@ -122,6 +122,7 @@         <div
           class="pf-v6-c-page__main"
           tabindex="-1"
           id="main-content"
+          style="--pf-v6-c-page--section--m-limit-width--MaxWidth: 62rem"
         >
           {{ block "main" . }}{{ end }}
         </div>
diff --git a/layouts/single.html b/layouts/single.html
index d2378190f6fdb76b50a15f70070f69d6e2f2288a..ea9d65a9dc7ed0aee4ed2739c65bd59acaefe470 100644
--- a/layouts/single.html
+++ b/layouts/single.html
@@ -3,7 +3,6 @@ <main class="pf-v6-u-flex-1 pf-v6-u-display-flex pf-v6-u-flex-direction-column">
   <!-- #main-title is in masthead -->
   <section
     class="pf-v6-c-page__main-section pf-m-limit-width pf-m-align-center"
-    style="--pf-v6-c-page--section--m-limit-width--MaxWidth: 62rem"
     aria-labelledby="main-title"
   >
     <div class="pf-v6-c-page__main-body">

