From afcbe54a1f027f959bc33b170a0a809c6b4040cf Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 08 Jul 2026 19:49:08 -0700
Subject: [PATCH] feat: Make headers sticky-top

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/scss/main.scss b/assets/scss/main.scss
index 9950d9cf5676049a82a0fcc3601363758ddb9eea..e7d689069f517f67630ad297910b1479e59ad07a 100644
--- a/assets/scss/main.scss
+++ b/assets/scss/main.scss
@@ -324,6 +324,9 @@   border-radius: 0;
   min-height: 100%;
   border: 0;
 
+  max-height: none;
+  overflow: visible;
+
   // On the home page we want the page itself to scroll so that `#`-based navigation
   // works as expected
   .pf-v6-c-page__main {
@@ -492,6 +495,16 @@   display: flex;
   justify-content: center;
 }
 
+.portfolio-group {
+  position: relative;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+
+  // We need to make the portfolio groups full-width again
+  width: 100%;
+}
+
 .portfolio-section {
   width: 100%;
   align-items: center;
@@ -590,6 +603,9 @@   }
 }
 
 .portfolio-section--striped {
+  --pf-v6-c-page--section--m-sticky-top--BackgroundColor: var(
+    --pf-t--global--background--color--secondary--default
+  );
   background: var(--pf-t--global--background--color--secondary--default);
 }
 
diff --git a/layouts/_partials/section-home.html b/layouts/_partials/section-home.html
index fcddb79cfe26000552a0d88c71bff130df686329..4886e6cc9ec27c97723d13f2d31a3955dfc80954 100644
--- a/layouts/_partials/section-home.html
+++ b/layouts/_partials/section-home.html
@@ -78,8 +78,9 @@ {{ range $i, $group := hugo.Data.portfolioitems }} {{ if not ( eq $i 0 ) }}
 <hr class="pf-v6-c-divider" />
 {{ end }}
 
+<div class="portfolio-group">
 <section
-  class="pf-v6-c-page__main-section pf-v6-c-page__main-section--home pf-m-limit-width pf-m-align-center pf-v6-u-pt-lg pf-v6-u-pb-lg portfolio-section pf-v6-u-px-0 {{ if eq (mod $i 2 ) 0 }}portfolio-section--striped{{ end }}"
+  class="pf-v6-c-page__main-section pf-v6-c-page__main-section--home pf-m-limit-width pf-m-align-center pf-m-sticky-top pf-v6-u-pt-lg pf-v6-u-pb-lg portfolio-section pf-v6-u-px-0 {{ if eq (mod $i 2 ) 0 }}portfolio-section--striped{{ end }}"
   id="{{ $group.id }}"
 >
   <div class="portfolio-tabs-wrapper">
@@ -199,6 +200,7 @@     </section>
     {{ end }}
   </div>
 </section>
+</div>
 {{ end }}
 
 <div class="section--full-width">

