From dbc2e9469801aecd7c791d678e984c4f3199c4fc Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 25 Aug 2026 23:46:31 -0700
Subject: [PATCH] refactor: Rework subtitles and headings with better prose

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/assets/scss/main.scss b/assets/scss/main.scss
index 4b35b73005132dc5d14a782390a7b23f61502eb8..e819ab055b286014f3857671604ce82afdeaabe0 100644
--- a/assets/scss/main.scss
+++ b/assets/scss/main.scss
@@ -498,7 +498,9 @@   justify-content: end;
 }
 
 // The 300 weight isn't included in PatternFly's tokens
-.na-hero__subtitle {
+.na-hero__subtitle,
+.na-portfolio__subtitle,
+.na-outro__subtitle {
   --pf-v6-c-title--m-md--FontWeight: 300;
 }
 
diff --git a/data/portfolioitems.yaml b/data/portfolioitems.yaml
index d5506195655dda5b185402378c89b6e78e58aa1e..4c16e8c3d41717aace630323f0cab060596b2812 100644
--- a/data/portfolioitems.yaml
+++ b/data/portfolioitems.yaml
@@ -2,7 +2,7 @@ - title: Broschüren, Plakate & Flyer
   subtitle: Klassische Printmedien für Tourismus, Events und Kommunen.
   id: drucksachen
   items:
-    - title: Broschüren & Bücher
+    - title: Broschüren & Magazine
       panelID: drucksachen-panel-broschueren
       tabID: drucksachen-tab-broschueren
       items:
diff --git a/layouts/_partials/section-home.html b/layouts/_partials/section-home.html
index a1dc22db672586949e9cb5fffba665a9dcc10a5d..a3dfc412353c0539df405c7c1501b73fb3d43d58 100644
--- a/layouts/_partials/section-home.html
+++ b/layouts/_partials/section-home.html
@@ -97,7 +97,7 @@       {{ $group.title }}
     </h2>
 
     <p
-      class="pf-v6-c-title pf-v6-u-text-color-subtle pf-m-md pf-v6-u-mb-md pf-v6-u-px-lg"
+      class="pf-v6-c-title pf-v6-u-text-color-subtle pf-m-md pf-v6-u-mb-md pf-v6-u-px-lg na-portfolio__subtitle"
     >
       {{ $group.subtitle }}
     </p>
@@ -253,12 +253,12 @@           class="na-outro__logo"
         />
 
         <h2 class="pf-v6-c-title pf-m-4xl pf-m-page-title pf-v6-u-mb-sm">
-          Ihr Projekt mit meiner Gestaltung
+          Sie haben Interesse an meiner Arbeit?
         </h2>
 
-        <p class="pf-v6-c-title pf-m-md pf-v6-u-mb-md">
-          Ob Broschüre, Kampagne oder komplette Markenidentität - ich freue mich
-          auf Ihre Anfrage.
+        <p class="pf-v6-c-title pf-m-md pf-v6-u-mb-md na-outro__subtitle">
+          Dann freue ich mich auf Ihre Nachricht und darauf, auch Ihr Projekt
+          kreativ umzusetzen.
         </p>
 
         <a class="pf-v6-c-button pf-m-primary" href="mailto:info@netteart.com">

