From b92595ad1d34dff91bd7ef0a61ee561b8d4041e1 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 08 Jun 2026 23:42:18 -0700
Subject: [PATCH] feat: Add URL to parsed PRs

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/gatekeeper-list-prs/main.go b/cmd/gatekeeper-list-prs/main.go
index 2b20f2d1c65c3b9d940974d79ff6b4b3f34d9ab2..eee4ea07620f6541569869e7a34b6a59c0d4ab91 100644
--- a/cmd/gatekeeper-list-prs/main.go
+++ b/cmd/gatekeeper-list-prs/main.go
@@ -43,6 +43,7 @@ 	Repo  RepoRepoConfig `yaml:"repo"`
 }
 
 type PullRequest struct {
+	URL       string    `yaml:"url"`
 	Title     string    `yaml:"title"`
 	Labels    []string  `yaml:"labels"` // TODO: Use a well-known list of labels (e.g. "blocked")
 	Checklist Checklist `yaml:"checklist"`
@@ -145,6 +146,7 @@ 			return pullRequests, err
 		}
 
 		pullRequests = append(pullRequests, PullRequest{
+			URL:       pr.GetHTMLURL(),
 			Title:     pr.GetTitle(),
 			Labels:    labels,
 			Checklist: checklist,
@@ -191,6 +193,7 @@ 			return pullRequests, err
 		}
 
 		pullRequests = append(pullRequests, PullRequest{
+			URL:       pr.URL,
 			Title:     pr.Title,
 			Labels:    labels,
 			Checklist: checklist,

