From a3eba169bf53a95bd6cab95bee72cceac597c97c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 16 Jun 2026 21:45:17 -0700
Subject: [PATCH] refactor: Flip checkbox state order

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/gatekeeper-create-response/main.go b/cmd/gatekeeper-create-response/main.go
index 56ace9a650b1ecda025450bf39f7da31546f3557..8699efc7985b970038d8e8e068e0955aef9a8f62 100644
--- a/cmd/gatekeeper-create-response/main.go
+++ b/cmd/gatekeeper-create-response/main.go
@@ -218,9 +218,9 @@ 						return
 					}
 
 					// Cycle through statuses
-					nextStatus := item.GetStatus() + 1
-					if nextStatus > statusIdPass {
-						nextStatus = statusIdPending
+					nextStatus := item.GetStatus() - 1
+					if nextStatus < statusIdPending {
+						nextStatus = statusIdPass
 					}
 
 					item.SetStatus(nextStatus)

