From aebb39951bb2cf87d1f31ecbfd195b2efdfcef00 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 17 Jun 2026 00:36:07 -0700
Subject: [PATCH] feat: Use Adwaita colours for quad-state check button

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/gatekeeper-create-response/main.go b/cmd/gatekeeper-create-response/main.go
index ae2cc9cbbe30b5d7168bf29af560d69bcf668403..6318bfc730796d8df0504efc84fe5b614d7a24af 100644
--- a/cmd/gatekeeper-create-response/main.go
+++ b/cmd/gatekeeper-create-response/main.go
@@ -141,25 +141,33 @@ 	case statusIdPending:
 		check.SetActive(false)
 		check.SetInconsistent(false)
 
-		check.RemoveCssClass("fail")
+		check.RemoveCssClass("pass-check")
+		check.RemoveCssClass("conditional-check")
+		check.RemoveCssClass("fail-check")
 
 	case statusIdFail:
 		check.SetActive(false)
 		check.SetInconsistent(true)
 
-		check.AddCssClass("fail")
+		check.RemoveCssClass("pass-check")
+		check.RemoveCssClass("conditional-check")
+		check.AddCssClass("fail-check")
 
 	case statusIdConditional:
 		check.SetActive(false)
 		check.SetInconsistent(true)
 
-		check.RemoveCssClass("fail")
+		check.RemoveCssClass("pass-check")
+		check.AddCssClass("conditional-check")
+		check.RemoveCssClass("fail-check")
 
 	case statusIdPass:
 		check.SetActive(true)
 		check.SetInconsistent(false)
 
-		check.RemoveCssClass("fail")
+		check.AddCssClass("pass-check")
+		check.RemoveCssClass("conditional-check")
+		check.RemoveCssClass("fail-check")
 	}
 
 	setRadioState(a, status)
@@ -184,9 +192,18 @@
 	flatGuidelines := flattenGuidelines(guidelines, 0)
 
 	csp := gtk.NewCssProvider()
-	csp.LoadFromString(`.fail check {
-  background-color: var(--error-bg-color);
-}`)
+	csp.LoadFromString(`.pass-check check {
+  background-color: var(--success-bg-color);
+}
+
+.conditional-check check {
+  background-color: var(--warning-bg-color);
+}
+
+.fail-check check {
+  background-color: var(--destructive-bg-color);
+}
+`)
 
 	app := adw.NewApplication("com.pojtinger.felicitas.GatekeeperCreateResponse", gio.GApplicationDefaultFlagsValue)
 

