From fec842bd0b96377888dd1afa3e236e48ab1da61e Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 26 Jun 2026 20:28:49 -0700
Subject: [PATCH] feat: Scroll to selected value in list view when switching selection

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/gatekeeper-create-response/main.go b/cmd/gatekeeper-create-response/main.go
index 1d00165c2a50a2b7635bf196fd9b8894754e0c7b..92e55bc7e3fa86c27196b1367f2b5fc04c0fe75c 100644
--- a/cmd/gatekeeper-create-response/main.go
+++ b/cmd/gatekeeper-create-response/main.go
@@ -898,6 +898,10 @@
 			// There are no top-level leaves, so we're safe to always assume
 			// that len(parentHeadings) > 1
 			wt.SetTitle(strings.Join(item.GetParentHeadings(), " › "))
+
+			si := gtk.NewScrollInfo()
+			si.SetEnableVertical(true)
+			lw.ScrollTo(pos, gtk.ListScrollSelectValue, si)
 		}
 		sl.ConnectNotify(&onConnectNotify)
 

