From 586a4c2fb9dcc229eab0e1ec4cd9cd043458c08e Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 14 Apr 2026 02:21:47 -0700
Subject: [PATCH] Expose `p2panda_topic_id_new_from_data` and `p2panda_topic_id_get_data` not just in GIR headers but also Rust

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/libp2panda/src/topic.rs b/libp2panda/src/topic.rs
index e6eb4b0cd8ea17360398f22f70a9cc98d924ac26..a70d98bc3e9523e3f27a85855c4d56d454acf087 100644
--- a/libp2panda/src/topic.rs
+++ b/libp2panda/src/topic.rs
@@ -218,6 +218,23 @@     topic::TopicId::new().into_glib_ptr()
 }
 
 #[unsafe(no_mangle)]
+pub unsafe extern "C" fn p2panda_topic_id_new_from_data(
+    data: *const [u8; 32],
+) -> *mut topic::TopicId {
+    unsafe { topic::TopicId::from_data(*data).into_glib_ptr() }
+}
+
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn p2panda_topic_id_get_data(
+    topic_id: *mut topic::TopicId,
+) -> *const [u8; 32] {
+    unsafe {
+        let topic_id = topic::TopicId::from_glib_none(topic_id);
+        topic_id.into_glib_ptr() as *const [u8; 32]
+    }
+}
+
+#[unsafe(no_mangle)]
 pub unsafe extern "C" fn p2panda_topic_id_copy(
     topic_id: *mut topic::TopicId,
 ) -> *mut topic::TopicId {

