From a6f1356a28a9db7578231b03ff707747346703e6 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 27 May 2026 23:26:25 -0700
Subject: [PATCH] build: Swap PWA build base to Alpine edge so that we can install dart-sass

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/.github/workflows/hydrun.yaml b/.github/workflows/hydrun.yaml
index e9731a75bb96587390e74ce4949b44135fe30ee6..ae11d6c3445ee4b2c2a77ea20be4f01f2eaddfe2 100644
--- a/.github/workflows/hydrun.yaml
+++ b/.github/workflows/hydrun.yaml
@@ -47,7 +47,7 @@             dst: out/*
             runner: ubuntu-latest
           - id: pwa
             src: .
-            os: golang:alpine
+            os: alpine:edge
             flags: -e '-v /tmp/ccache:/root/.cache/go-build'
             cmd: ./Hydrunfile pwa
             dst: out/*
diff --git a/Hydrunfile b/Hydrunfile
index 982411be0e6f0b460c6ae35cc61c75d348db62ef..06674d32bf6dc4b5aa394436e140d0b196b95a70 100755
--- a/Hydrunfile
+++ b/Hydrunfile
@@ -47,8 +47,8 @@ if [ "$1" = "pwa" ]; then
     # Setup repositories
     echo 'https://dl-cdn.alpinelinux.org/alpine/edge/testing' >>/etc/apk/repositories
 
-    # Install Hugo
-    apk add curl make hugo dart-sass npm
+    # Install Hugo and Go
+    apk add curl make hugo dart-sass npm go git
 
     # Generate dependencies
     make depend

