From 6d71a563c2dd7f6b1ef4f0b7466b77310c21b9fa Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 27 May 2026 23:11:48 -0700
Subject: [PATCH] feat: Allow showing reblogs in Mastodon API if requested

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/api/mastodon/index.go b/api/mastodon/index.go
index 6a600e9ffc7321513487918636db49f6e0164e07..20b2ff3d2973751bfc8c5339d823a0871584a988 100644
--- a/api/mastodon/index.go
+++ b/api/mastodon/index.go
@@ -33,6 +33,16 @@ 	ReblogsCount   int `json:"reblogsCount"`
 	FavoritesCount int `json:"favoritesCount"`
 
 	URL string `json:"url"`
+
+	Author    Author  `json:"author"`
+	BoostedBy *Author `json:"boostedBy,omitempty"`
+}
+
+type Author struct {
+	UserDisplayName       string `json:"userDisplayName"`
+	UserName              string `json:"username"`
+	UserURL               string `json:"userUrl"`
+	UserProfilePictureURL string `json:"userProfilePictureURL"`
 }
 
 type Media struct {
@@ -48,6 +58,8 @@ 		w.Write([]byte("missing username query parameter: "))
 
 		panic("missing username query parameter")
 	}
+
+	includeReblogs := r.URL.Query().Has("includeReblogs")
 
 	client := mastodon.NewClient(&mastodon.Config{
 		Server:       server,
@@ -90,20 +102,50 @@ 	if err != nil {
 		panic(err)
 	}
 
+	feedAuthor := Author{
+		UserDisplayName:       account.DisplayName,
+		UserName:              account.Username,
+		UserURL:               account.URL,
+		UserProfilePictureURL: account.AvatarStatic,
+	}
+
 	for _, sourceToot := range sourceToots {
-		// Skip toots are empty, i.e. reblogs
-		if strings.TrimSpace(sourceToot.Content) == "" && len(sourceToot.MediaAttachments) <= 0 {
+		// Hide any non-public toots
+		// See https://docs.joinmastodon.org/entities/Status/#visibility
+		if sourceToot.Visibility != "public" {
 			continue
 		}
 
 		toot := Toot{}
 
+		contentSource := sourceToot
+		toot.Author = feedAuthor
+
+		if sourceToot.Reblog != nil {
+			if !includeReblogs {
+				continue
+			}
+
+			contentSource = sourceToot.Reblog
+			toot.Author = Author{
+				UserDisplayName:       sourceToot.Reblog.Account.DisplayName,
+				UserName:              sourceToot.Reblog.Account.Username,
+				UserURL:               sourceToot.Reblog.Account.URL,
+				UserProfilePictureURL: sourceToot.Reblog.Account.AvatarStatic,
+			}
+			boostedBy := feedAuthor
+			toot.BoostedBy = &boostedBy
+		} else if strings.TrimSpace(sourceToot.Content) == "" && len(sourceToot.MediaAttachments) <= 0 {
+			// Skip empty toots
+			continue
+		}
+
 		toot.Timestamp = sourceToot.CreatedAt.Format(time.RFC3339)
-		toot.Body = sourceToot.Content
+		toot.Body = contentSource.Content
 
 		images := []Media{}
 
-		for _, attachment := range sourceToot.MediaAttachments {
+		for _, attachment := range contentSource.MediaAttachments {
 			images = append(images, Media{
 				URL:     attachment.URL,
 				AltText: attachment.Description,
@@ -113,11 +155,11 @@ 		}
 
 		toot.Media = images
 
-		toot.RepliesCount = int(sourceToot.RepliesCount)
-		toot.ReblogsCount = int(sourceToot.ReblogsCount)
-		toot.FavoritesCount = int(sourceToot.FavouritesCount)
+		toot.RepliesCount = int(contentSource.RepliesCount)
+		toot.ReblogsCount = int(contentSource.ReblogsCount)
+		toot.FavoritesCount = int(contentSource.FavouritesCount)
 
-		toot.URL = sourceToot.URL
+		toot.URL = contentSource.URL
 
 		toots = append(toots, toot)
 	}

