From 703daa22bd956ce859c3c2dba1727fc7327b7a91 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Fri, 12 Sep 2025 14:10:09 +0200
Subject: [PATCH] feat: Use better map tiles and fix build

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/components/Map.tsx b/components/Map.tsx
index 8e9eef3e8198557dc21afbd45031f37adcf72ec4..366510e7525eaaeaa95b247b205710e388ddfd65 100644
--- a/components/Map.tsx
+++ b/components/Map.tsx
@@ -8,7 +8,7 @@       {...props}
       width="100%"
       height="100%"
       mapboxAccessToken="pk.eyJ1IjoicG9qbnRmeCIsImEiOiJjbWZnb25lNjYwMzBtMmxwdDloanpkaTZ5In0.aG9eHAQc1ZSixnwpJKhNfQ"
-      mapStyle="mapbox://styles/mapbox/light-v9"
+      mapStyle="mapbox://styles/mapbox/streets-v9"
       initialViewState={{
         latitude: 48.50468,
         longitude: 8.363728,
diff --git a/pages/index.tsx b/pages/index.tsx
index 46e7dddc78b8cb280791b1abd74f1c00d7c59908..05a5fd1370bad39488ae3ac89781e8ee77cd114c 100644
--- a/pages/index.tsx
+++ b/pages/index.tsx
@@ -158,10 +158,11 @@   grid-template-columns: 2fr 1fr;
   min-height: 30rem;
 `;
 
-const SplitImage = styled(FullscreenImage)`
+const SplitImage = styled(FullscreenImage)<{ $objectPosition?: string }>`
   width: 100%;
   height: 100%;
   object-fit: cover;
+  object-position: ${props => props.$objectPosition || 'center'};
 `;
 
 const SplitText = styled.div`
@@ -472,7 +473,7 @@               </SplitText>
               <SplitImage
                 alt="Behandlungszimmer"
                 src="/room.jpg"
-                style={{ objectPosition: "center 40%" }}
+                $objectPosition="center 40%"
               />
             </RightSplit>
           </Container>

