From 479c59541e4bd33ab536d0dbe57d3d2f6ec854bd Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 17 May 2026 23:26:21 -0700
Subject: [PATCH] chore: Drop Lernstunden section

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/pages/index.tsx b/pages/index.tsx
index 65a9e93e80279e70cc2026335afd685a79d6f650..c1a5fad75f89faf1f4450e5795ebea4a9e51ffd2 100644
--- a/pages/index.tsx
+++ b/pages/index.tsx
@@ -639,8 +639,7 @@               </EventPrimary>
 
               <Event style={{ gridArea: "e12a" }}>
                 <Typography variant="body1">
-                  Re-Zertifizierung <Focus>Boeger®-Therapie Update</Focus> (60
-                  Lernstunden)
+                  Re-Zertifizierung <Focus>Boeger®-Therapie Update</Focus>
                 </Typography>
               </Event>
               <EventRight style={{ gridArea: "e12b" }}>

