From 0d35abd39bba8899f3a0901207dd843979ad899a Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 08 Jul 2026 19:39:03 -0700
Subject: [PATCH] refactor: Swap out Heilpraktikerin with Cibionic on home page

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/pages/index.tsx b/pages/index.tsx
index c1a5fad75f89faf1f4450e5795ebea4a9e51ffd2..b88a90558c42cf7726d28dcb6ab1104f7f992264 100644
--- a/pages/index.tsx
+++ b/pages/index.tsx
@@ -632,8 +632,8 @@                 <Typography variant="body1">Heute</Typography>
               </EventLeftPrimary>
               <EventPrimary style={{ gridArea: "e11c" }}>
                 <Typography variant="body1">
-                  <MobileOnly>Heute: </MobileOnly>In Ausbildung zur{" "}
-                  <Focus>Heilpraktikerin</Focus>
+                  <MobileOnly>Heute: </MobileOnly>Zusätzliche Arbeit mit dem
+                  neuen <Focus>Cibionic 2.0</Focus> - „eine Idee wird Realität“
                 </Typography>
               </EventPrimary>
 

