From 51138b500a98bb4d3c09aa9b6cfed452a6566d8c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Wed, 10 Dec 2025 22:29:37 -0800
Subject: [PATCH] fix: Manually initialize WebView type for now
---
diff --git a/senbara-gnome/cmd/senbara-gnome/main.go b/senbara-gnome/cmd/senbara-gnome/main.go
index 517907325ca87415bb82a5b7ddeac767fc9f5840..b932d90a0ca7cb7f22236dcb3cddf48dd475f16f 100644
--- a/senbara-gnome/cmd/senbara-gnome/main.go
+++ b/senbara-gnome/cmd/senbara-gnome/main.go
@@ -49,6 +49,8 @@ func main() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	_ = webkit.WebViewGLibType() // TODO: WebKit should initialize itself automatically
+
 	level := new(slog.LevelVar)
 	log := slog.New(slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{
 		Level: level,

