From bd243a8057fe5f44a4d7a9c15b989cf4144b1993 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 19 Feb 2026 21:44:04 -0800
Subject: [PATCH] chore: Bump `puregotk` to latest synthetic branch and port to new 32-bit data types

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/go.work.sum b/go.work.sum
index c144c9f214a757e8bbb55b40a2ecfe0fc075a5e2..629ae6332963b317b1ab9fac17e58e0cc0c61bf2 100644
--- a/go.work.sum
+++ b/go.work.sum
@@ -113,8 +113,6 @@ github.com/go-logr/logr v1.2.3/go.mod h1:jdQByPbusPIv2/zmleS9BjJVeZ6kBagPoEUsqbVz/1A=
 github.com/go-ole/go-ole v1.2.6/go.mod h1:pprOEPIfldk/42T2oK7lQ4v4JSDwmV0As9GaiUsvbm0=
 github.com/go-openapi/jsonpointer v0.21.0/go.mod h1:IUyH9l/+uyhIYQ/PXVA41Rexl+kOkAPDdXEYns6fzUY=
 github.com/go-openapi/swag v0.23.0/go.mod h1:esZ8ITTYEsH1V2trKHjAN8Ai7xHb8RV+YSZ577vPjgQ=
-github.com/go-openapi/swag v0.23.1 h1:lpsStH0n2ittzTnbaSloVZLuB5+fvSY/+hnagBjSNZU=
-github.com/go-openapi/swag v0.23.1/go.mod h1:STZs8TbRvEQQKUA+JZNAm3EWlgaOBGpyFDqQnDHMef0=
 github.com/go-openapi/swag/jsonname v0.25.1/go.mod h1:71Tekow6UOLBD3wS7XhdT98g5J5GR13NOTQ9/6Q11Zo=
 github.com/go-playground/assert/v2 v2.2.0/go.mod h1:VDjEfimB/XKnb+ZQfWdccd7VUvScMdVu0Titje2rxJ4=
 github.com/go-playground/locales v0.14.1/go.mod h1:hxrqLVvrK65+Rwrd5Fc6F2O76J/NuW9t0sjnWqG1slY=
@@ -224,7 +222,6 @@ github.com/magiconair/properties v1.8.9 h1:nWcCbLq1N2v/cpNsy5WvQ37Fb+YElfq20WJ/a8RkpQM=
 github.com/magiconair/properties v1.8.9/go.mod h1:Dhd985XPs7jluiymwWYZ0G4Z61jb3vdS329zhj2hYo0=
 github.com/mailgun/raymond/v2 v2.0.48/go.mod h1:lsgvL50kgt1ylcFJYZiULi5fjPBkkhNfj4KA0W54Z18=
 github.com/mailru/easyjson v0.7.7/go.mod h1:xzfreul335JAWq5oZzymOObrkdz5UnU4kGfJJLY9Nlc=
-github.com/mailru/easyjson v0.9.0/go.mod h1:1+xMtQp2MRNVL/V1bOzuP3aP8VNwRW55fQUto+XFtTU=
 github.com/matryer/try v0.0.0-20161228173917-9ac251b645a2/go.mod h1:0KeJpeMD6o+O4hW7qJOT7vyQPKrWmj26uf5wMc/IiIs=
 github.com/mattn/go-colorable v0.1.13/go.mod h1:7S9/ev0klgBDR4GtXTXX8a3vIGJpMovkB8vQcUbaXHg=
 github.com/mattn/go-isatty v0.0.16/go.mod h1:kYGgaQfpe5nmfYZH+SKPsOc2e4SrIfOl2e/yFXSvRLM=
@@ -261,12 +258,14 @@ github.com/pkg/diff v0.0.0-20210226163009-20ebb0f2a09e/go.mod h1:pJLUxLENpZxwdsKMEsNbx1VGcRFpLqf3715MtcvvzbA=
 github.com/pkg/errors v0.9.1/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pkg/sftp v1.13.7/go.mod h1:KMKI0t3T6hfA+lTR/ssZdunHo+uwq7ghoN09/FSu3DY=
 github.com/planetscale/vtprotobuf v0.6.1-0.20240319094008-0393e58bdf10/go.mod h1:t/avpk3KcrXxUnYOhZhMXJlSEyie6gQbtLq5NM3loB8=
-github.com/pojntfx/go-gettext v0.2.0 h1:Nw3Hyy0Fef3pOuvYYPZESmbOT+IdnYjqx9/ywFo46GU=
-github.com/pojntfx/go-gettext v0.2.0/go.mod h1:eGt2r0JKvlPZjiz+GqEDtg92//F30LEV/IB0h42bk40=
+github.com/pojntfx/purego v0.0.0-20260128031338-5429b1e47a6d h1:f4fm8/eZJvaE/tfBomboi8CQByYmbXKxHcP6PBwAM04=
+github.com/pojntfx/purego v0.0.0-20260128031338-5429b1e47a6d/go.mod h1:amE5lVGstyh2GTEIHFEbgJHmoMD0llIuapnzIlZx7Fc=
 github.com/pojntfx/puregotk v0.0.0-20251009044016-417c19274d31 h1:aTLkBlf2AqI1ivMmW4QfB4quHoPqZNulf1HZJyTRyfQ=
 github.com/pojntfx/puregotk v0.0.0-20251009044016-417c19274d31/go.mod h1:9FUAoCJsQgot2Zt6tvy1ZrviyMe71bNhWhwo3z36XQY=
 github.com/pojntfx/puregotk v0.0.0-20251011060225-c87603a2de88/go.mod h1:9FUAoCJsQgot2Zt6tvy1ZrviyMe71bNhWhwo3z36XQY=
 github.com/pojntfx/puregotk v0.0.0-20251011062924-170bec63e7cb/go.mod h1:9FUAoCJsQgot2Zt6tvy1ZrviyMe71bNhWhwo3z36XQY=
+github.com/pojntfx/puregotk v0.0.0-20260220051800-f8a38f1e0894 h1:pPTVC5Q/5DJ46RzXKISgtUbZgHl/xampBJtBaPOd1ZQ=
+github.com/pojntfx/puregotk v0.0.0-20260220051800-f8a38f1e0894/go.mod h1:mZu8qSvCSqtVL+xoiebZuicnJijwrLfGRs6/eTVgFlc=
 github.com/power-devops/perfstat v0.0.0-20210106213030-5aafc221ea8c/go.mod h1:OmDBASR4679mdNQnz2pUhc2G8CO2JrUAVFDRBDP/hJE=
 github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4/go.mod h1:xMI15A0UPsDsEKsMN9yxemIoYk6Tm2C1GtYGdfGttqA=
 github.com/prometheus/procfs v0.0.0-20190425082905-87a4384529e0/go.mod h1:TjEm7ze935MbeOT/UhFTIMYKhuLP4wbCsTZCD3I8kEA=
@@ -297,7 +296,6 @@ github.com/spf13/cobra v1.7.0/go.mod h1:uLxZILRyS/50WlhOIKD7W6V5bgeIt+4sICxh6uRMrb0=
 github.com/spf13/pflag v1.0.5/go.mod h1:McXfInJRrz4CZXVZOBLb0bTZqETkiAhM9Iw0y3An2Bg=
 github.com/stretchr/objx v0.4.0/go.mod h1:YvHI0jy2hoMjB+UWwv71VJQ9isScKT/TqJzVSSt89Yw=
 github.com/stretchr/objx v0.5.0/go.mod h1:Yh+to48EsGEfYuaHDzXPcE3xhTkx73EhmCGUpEOglKo=
-github.com/stretchr/objx v0.5.2/go.mod h1:FRsXN1f5AsAjCGJKqEizvkpNtU+EGNCLh3NxZ/8L+MA=
 github.com/stretchr/testify v1.2.2/go.mod h1:a8OnRcib4nhh0OaRAV+Yts87kKdq0PP7pXfy6kDkUVs=
 github.com/stretchr/testify v1.6.1/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=
 github.com/stretchr/testify v1.7.1/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=
@@ -306,7 +304,6 @@ github.com/stretchr/testify v1.8.1/go.mod h1:w2LPCIKwWwSfY2zedu0+kehJoqGctiVI29o6fzry7u4=
 github.com/stretchr/testify v1.8.2/go.mod h1:w2LPCIKwWwSfY2zedu0+kehJoqGctiVI29o6fzry7u4=
 github.com/stretchr/testify v1.8.4/go.mod h1:sz/lmYIOXD/1dqDmKjjqLyZ2RngseejIcXlSw2iwfAo=
 github.com/stretchr/testify v1.9.0/go.mod h1:r2ic/lqez/lEtzL7wO/rwa5dbSLXVDPFyf8C91i36aY=
-github.com/stretchr/testify v1.10.0/go.mod h1:r2ic/lqez/lEtzL7wO/rwa5dbSLXVDPFyf8C91i36aY=
 github.com/tdewolff/minify/v2 v2.12.9/go.mod h1:qOqdlDfL+7v0/fyymB+OP497nIxJYSvX4MQWA8OoiXU=
 github.com/tdewolff/parse/v2 v2.6.8/go.mod h1:XHDhaU6IBgsryfdnpzUXBlT6leW/l25yrFBTEb4eIyM=
 github.com/tdewolff/test v1.0.9/go.mod h1:6DAvZliBAAnD7rhVgwaM7DE5/d9NMOAJ09SqYqeK4QE=
diff --git a/senbara-gnome/cmd/senbara-gnome/callbacks.go b/senbara-gnome/cmd/senbara-gnome/callbacks.go
index b052c0de7761863908f669641ff4d4b8ff244eac..35591ac5953733bc6edab4c968da1a0cdd01a26b 100644
--- a/senbara-gnome/cmd/senbara-gnome/callbacks.go
+++ b/senbara-gnome/cmd/senbara-gnome/callbacks.go
@@ -11,7 +11,7 @@ 	"github.com/jwijenbergh/puregotk/v4/gtk"
 )
 
 func glibDateTimeFromGo(t time.Time) *glib.DateTime {
-	return glib.NewDateTimeLocal(t.Year(), int(t.Month()), t.Day(), t.Hour(), t.Minute(), float64(t.Second()))
+	return glib.NewDateTimeLocal(int32(t.Year()), int32(t.Month()), int32(t.Day()), int32(t.Hour()), int32(t.Minute()), float64(t.Second()))
 }
 
 func connectButtonClicked(button *gtk.Button, fn func()) {
diff --git a/senbara-gnome/cmd/senbara-gnome/main.go b/senbara-gnome/cmd/senbara-gnome/main.go
index fc2e1b4d332095a9dafdf2b99f24feb354e3a487..1ae58b964da8229dc8638037a6491c6c5cc81811 100644
--- a/senbara-gnome/cmd/senbara-gnome/main.go
+++ b/senbara-gnome/cmd/senbara-gnome/main.go
@@ -89,7 +89,7 @@ 			func() string {
 				return nv.GetVisiblePage().GetTag()
 			},
 			func(tags []string, position int) {
-				nv.ReplaceWithTags(tags, position)
+				nv.ReplaceWithTags(tags, int32(position))
 			},
 			func(ud userData) {
 				u = ud
@@ -4041,7 +4041,7 @@ 					journalEntriesEditPageSaveButton.SetActionTargetValue(glib.NewVariantInt64(*res.JSON200.Id))
 
 					journalEntriesEditPageTitle.SetSubtitle(*res.JSON200.Title)
 
-					journalEntriesEditPageRatingToggleGroup.SetActive(3 - uint(*res.JSON200.Rating)) // The toggle group is zero-indexed, but the rating is one-indexed
+					journalEntriesEditPageRatingToggleGroup.SetActive(uint32(3 - uint(*res.JSON200.Rating))) // The toggle group is zero-indexed, but the rating is one-indexed
 
 					journalEntriesEditPageTitleInput.SetText(*res.JSON200.Title)
 
@@ -4398,7 +4398,7 @@ 		a.AddWindow(&w.Window)
 	}
 	a.ConnectActivate(&onActivate)
 
-	onOpen := func(_ gio.Application, filesPtr uintptr, nFiles int, hint string) {
+	onOpen := func(_ gio.Application, filesPtr uintptr, nFiles int32, hint string) {
 		if w.GoPointer() == 0 {
 			a.Activate()
 		} else {
@@ -4544,11 +4544,11 @@ 		if signedOut && nextURL == resources.PageHome {
 			nextURL = resources.PageIndex
 		}
 
-		nv.ReplaceWithTags([]string{nextURL}, 1)
+		nv.ReplaceWithTags([]string{nextURL}, int32(1))
 	}
 	a.ConnectOpen(&onOpen)
 
-	if code := a.Run(len(os.Args), os.Args); code > 0 {
-		os.Exit(code)
+	if code := a.Run(int32(len(os.Args)), os.Args); code > 0 {
+		os.Exit(int(code))
 	}
 }
diff --git a/senbara-gnome/go.mod b/senbara-gnome/go.mod
index 0657fc703225b2a392471b6df34199b693405192..53ee9325505d2f3da6f9c6678a678b2c39a69b2b 100644
--- a/senbara-gnome/go.mod
+++ b/senbara-gnome/go.mod
@@ -38,4 +38,7 @@ 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
-replace github.com/jwijenbergh/puregotk v0.0.0-20251201161753-28ec1479c381 => ../../puregotk
+replace (
+	github.com/jwijenbergh/purego v0.0.0-20251017112123-b71757b9ba42 => github.com/pojntfx/purego v0.0.0-20260128031338-5429b1e47a6d
+	github.com/jwijenbergh/puregotk v0.0.0-20251201161753-28ec1479c381 => github.com/pojntfx/puregotk v0.0.0-20260220051800-f8a38f1e0894
+)
diff --git a/senbara-gnome/go.sum b/senbara-gnome/go.sum
index 39b32a4bfb28ef91d40e1d3994ab94564087f3f3..89f4499d7ade267801b9d0f52c309672dc0f8a2f 100644
--- a/senbara-gnome/go.sum
+++ b/senbara-gnome/go.sum
@@ -32,10 +32,6 @@ github.com/google/uuid v1.6.0/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
 github.com/josharian/intern v1.0.0 h1:vlS4z54oSdjm0bgjRigI+G1HpF+tI+9rE5LLzOg8HmY=
 github.com/josharian/intern v1.0.0/go.mod h1:5DoeVV0s6jJacbCEi61lwdGj/aVlrQvzHFFd8Hwg//Y=
 github.com/juju/gnuflag v0.0.0-20171113085948-2ce1bb71843d/go.mod h1:2PavIy+JPciBPrBUjwbNvtwB6RQlve+hkpll6QSNmOE=
-github.com/jwijenbergh/purego v0.0.0-20251017112123-b71757b9ba42 h1:kQ0LX4ffJvtONhymro3HzrB8XUa6dIGgrbnkl38Rgl0=
-github.com/jwijenbergh/purego v0.0.0-20251017112123-b71757b9ba42/go.mod h1:amE5lVGstyh2GTEIHFEbgJHmoMD0llIuapnzIlZx7Fc=
-github.com/jwijenbergh/puregotk v0.0.0-20251201161753-28ec1479c381 h1:Cy6in3ph8eGnl0w71S0NnH8ZCj0YhZZHzrwYEgtbtII=
-github.com/jwijenbergh/puregotk v0.0.0-20251201161753-28ec1479c381/go.mod h1:mZu8qSvCSqtVL+xoiebZuicnJijwrLfGRs6/eTVgFlc=
 github.com/kr/pretty v0.3.1 h1:flRD4NNwYAUpkphVc1HcthR4KEIFJ65n8Mw5qdRn3LE=
 github.com/kr/pretty v0.3.1/go.mod h1:hoEshYVHaxMs3cyo3Yncou5ZscifuDolrwPKZanG3xk=
 github.com/kr/text v0.2.0 h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=
@@ -59,6 +55,10 @@ github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2 h1:Jamvg5psRIccs7FGNTlIRMkT8wgtp5eCXdBlqhYGL6U=
 github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
 github.com/pojntfx/go-gettext v0.2.0 h1:Nw3Hyy0Fef3pOuvYYPZESmbOT+IdnYjqx9/ywFo46GU=
 github.com/pojntfx/go-gettext v0.2.0/go.mod h1:eGt2r0JKvlPZjiz+GqEDtg92//F30LEV/IB0h42bk40=
+github.com/pojntfx/purego v0.0.0-20260128031338-5429b1e47a6d h1:f4fm8/eZJvaE/tfBomboi8CQByYmbXKxHcP6PBwAM04=
+github.com/pojntfx/purego v0.0.0-20260128031338-5429b1e47a6d/go.mod h1:amE5lVGstyh2GTEIHFEbgJHmoMD0llIuapnzIlZx7Fc=
+github.com/pojntfx/puregotk v0.0.0-20260220051800-f8a38f1e0894 h1:pPTVC5Q/5DJ46RzXKISgtUbZgHl/xampBJtBaPOd1ZQ=
+github.com/pojntfx/puregotk v0.0.0-20260220051800-f8a38f1e0894/go.mod h1:mZu8qSvCSqtVL+xoiebZuicnJijwrLfGRs6/eTVgFlc=
 github.com/pojntfx/senbara/senbara-common v0.0.0-20250520062435-d85e71a7a89f h1:oK4Sg9qjfW3XKPqwbkIVtj3CaQtl5xy98EnSF7GIUKs=
 github.com/pojntfx/senbara/senbara-common v0.0.0-20250520062435-d85e71a7a89f/go.mod h1:qOP28T7iYf/hUUHckc8QUdeSBv+wx0hrzvtXUuPwMe4=
 github.com/pojntfx/senbara/senbara-rest v0.0.0-20250520062435-d85e71a7a89f h1:WwMvnNYJtch1GdA6E/EjrBs5E6c/IXEqYN4EXf0VT2A=

