From 5098aaad57c14af274cc31bd099b0efbe0b890d7 Mon Sep 17 00:00:00 2001
From: Jeroen Wijenbergh <jeroen.wijenbergh@geant.org>
Date: Tue, 24 Feb 2026 10:49:49 +0100
Subject: [PATCH] Go module: Move to codeberg
---
diff --git a/callback_test.go b/callback_test.go
index 2bc4ba4b32ba69028cfa84d9788ada33aef75ea9..d26187b029727c4a21c7516cbcd0b5f511a750ed 100644
--- a/callback_test.go
+++ b/callback_test.go
@@ -13,7 +13,7 @@ 	"runtime"
 	"testing"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 // TestCallGoFromSharedLib is a test that checks for stack corruption on arm64
diff --git a/cgo.go b/cgo.go
index 1bea1bbecc03841c3b4e6e4f4a4689d314631e5a..e659837a141ef6b549593fbe774b5170878c79ab 100644
--- a/cgo.go
+++ b/cgo.go
@@ -15,5 +15,5 @@ // it is since we don't use `import "C"`!
 import (
 	_ "runtime/cgo"
 
-	_ "github.com/jwijenbergh/purego/internal/cgo"
+	_ "codeberg.org/puregotk/purego/internal/cgo"
 )
diff --git a/dlfcn_android.go b/dlfcn_android.go
index 030d9f5a3a8ef291794295d005a31dec5d43b379..a5ec5d4c148944167873f3ea65a3fff8c6977780 100644
--- a/dlfcn_android.go
+++ b/dlfcn_android.go
@@ -3,7 +3,7 @@ // SPDX-FileCopyrightText: 2024 The Ebitengine Authors
 
 package purego
 
-import "github.com/jwijenbergh/purego/internal/cgo"
+import "codeberg.org/puregotk/purego/internal/cgo"
 
 // Source for constants: https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/include/dlfcn.h
 
diff --git a/dlfcn_test.go b/dlfcn_test.go
index a76693f2009602dfa2376afd79867bc56595ad2e..b6fc6d46497c3d1b12ad2e6baf0bfbcd0523b409 100644
--- a/dlfcn_test.go
+++ b/dlfcn_test.go
@@ -11,7 +11,7 @@ 	"path/filepath"
 	"testing"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 func TestSimpleDlsym(t *testing.T) {
diff --git a/examples/libc/main.go b/examples/libc/main.go
index 535e69bbdab3a0db014e2d9c64a6a2ea01e249fe..a18f13144c855996fe97ff17ff1f358512605924 100644
--- a/examples/libc/main.go
+++ b/examples/libc/main.go
@@ -9,7 +9,7 @@ import (
 	"fmt"
 	"runtime"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 func getSystemLibrary() string {
diff --git a/examples/libc/main_unix.go b/examples/libc/main_unix.go
index 90da093670d60d21497edd725d7fc5d61e81961a..99065f7e8caca2bb1b156eb2e7976d5b5c59be47 100644
--- a/examples/libc/main_unix.go
+++ b/examples/libc/main_unix.go
@@ -5,7 +5,7 @@ //go:build darwin || freebsd || linux || netbsd
 
 package main
 
-import "github.com/jwijenbergh/purego"
+import "codeberg.org/puregotk/purego"
 
 func openLibrary(name string) (uintptr, error) {
 	return purego.Dlopen(name, purego.RTLD_NOW|purego.RTLD_GLOBAL)
diff --git a/examples/objc/main_darwin.go b/examples/objc/main_darwin.go
index 384a3cf685ac04d83b34792458c57f6ef36fc1bc..92a7cd93f503bd083a03a0fa5a2ef21926740434 100644
--- a/examples/objc/main_darwin.go
+++ b/examples/objc/main_darwin.go
@@ -7,7 +7,7 @@ import (
 	"fmt"
 	"reflect"
 
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 var (
diff --git a/examples/protocol-dumper/main_darwin.go b/examples/protocol-dumper/main_darwin.go
index 8a07b14296c2ffd2d925b5154a67c8ddd359cba1..b6071a1172ff83bb8050b8269a8c3d7dd3b06e16 100644
--- a/examples/protocol-dumper/main_darwin.go
+++ b/examples/protocol-dumper/main_darwin.go
@@ -24,7 +24,7 @@ 	"strconv"
 	"strings"
 	"text/template"
 
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 var protocols = []string{
@@ -79,7 +79,7 @@
 import (
 	"log"
 
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 func init() {
diff --git a/examples/window/main_darwin.go b/examples/window/main_darwin.go
index 8e61675888b49ee68639aab1cba4328860098c0c..c35910e25edd157886658cc61abe0891381d8ca9 100644
--- a/examples/window/main_darwin.go
+++ b/examples/window/main_darwin.go
@@ -6,8 +6,8 @@
 import (
 	"runtime"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 const (
diff --git a/examples/window/main_windows.go b/examples/window/main_windows.go
index e74410515af6eedd29461507c1558888d15d13fc..8bc92f4c52324bac1e74bbb75de6f64852fdfef2 100644
--- a/examples/window/main_windows.go
+++ b/examples/window/main_windows.go
@@ -8,7 +8,7 @@ 	"runtime"
 	"syscall"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 const (
diff --git a/func.go b/func.go
index 41b3deebbd222067ace2858db5bf083e4ff9a4e2..1188d6e9d7b4f8808f03a98c4ff25598338affa8 100644
--- a/func.go
+++ b/func.go
@@ -12,8 +12,8 @@ 	"reflect"
 	"runtime"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego/internal/strings"
-	"github.com/jwijenbergh/purego/internal/xreflect"
+	"codeberg.org/puregotk/purego/internal/strings"
+	"codeberg.org/puregotk/purego/internal/xreflect"
 )
 
 const (
diff --git a/func_sysv_test.go b/func_sysv_test.go
index eb6aad11948cf6ff7ab916a908aac2790cdf5332..78a2a0ec2f68f5780a398bd546193c6a49d89f4f 100644
--- a/func_sysv_test.go
+++ b/func_sysv_test.go
@@ -8,7 +8,7 @@
 import (
 	"testing"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 func TestUnrefCallback(t *testing.T) {
diff --git a/func_test.go b/func_test.go
index 562f68a5ff8db5b5b4473bafa4fc3ee55851c073..cbfec952b7838e6b6d90134596c2322eda03067d 100644
--- a/func_test.go
+++ b/func_test.go
@@ -15,8 +15,8 @@ 	"sync"
 	"testing"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/internal/load"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/internal/load"
 )
 
 func getSystemLibrary() (string, error) {
diff --git a/go.mod b/go.mod
index 73f140d3458c37ca2fc270e15cee4c29375d4289..62453ff8639256724b3db9536795c51b2100b492 100644
--- a/go.mod
+++ b/go.mod
@@ -1,3 +1,3 @@
-module github.com/jwijenbergh/purego
+module codeberg.org/puregotk/purego
 
 go 1.18
diff --git a/internal/buildtest/main.go b/internal/buildtest/main.go
index b353f20c5f4ac9ffd03c5d5f9fdcd89e2245f792..705213e2e0651ec1f999fa205d9bd19732ca03bf 100644
--- a/internal/buildtest/main.go
+++ b/internal/buildtest/main.go
@@ -6,7 +6,7 @@
 package main
 
 import (
-	_ "github.com/jwijenbergh/purego"
+	_ "codeberg.org/puregotk/purego"
 )
 
 import "C"
diff --git a/internal/fakecgo/doc.go b/internal/fakecgo/doc.go
index 6fab36d9c5d5d4a6d0646f618ca151d3bf12f255..f77f436403e37f0a34f4ba471b2567d0a34c2633 100644
--- a/internal/fakecgo/doc.go
+++ b/internal/fakecgo/doc.go
@@ -21,7 +21,7 @@ // that fakecgo does not implement at the moment.
 //
 // # Usage
 //
-// Using fakecgo is easy just import _ "github.com/jwijenbergh/purego" and then
+// Using fakecgo is easy just import _ "codeberg.org/puregotk/purego" and then
 // set the environment variable CGO_ENABLED=0.
 // The recommended usage for fakecgo is to prefer using runtime/cgo if possible
 // but if cross-compiling or fast build times are important fakecgo is available.
diff --git a/internal/fakecgo/freebsd.go b/internal/fakecgo/freebsd.go
index 9ed4d99e376d41db3971b4e18ccf5ff7320b41cd..75cae973e8befcd82b1a0037af6180f137f4f166 100644
--- a/internal/fakecgo/freebsd.go
+++ b/internal/fakecgo/freebsd.go
@@ -15,7 +15,7 @@
 // Note: when building with cross-compiling or CGO_ENABLED=0, add
 // the following argument to `go` so that these symbols are defined by
 // making fakecgo the Cgo.
-//   -gcflags="github.com/jwijenbergh/purego/internal/fakecgo=-std"
+//   -gcflags="codeberg.org/puregotk/purego/internal/fakecgo=-std"
 
 //go:linkname _environ environ
 //go:linkname _progname __progname
diff --git a/internal/load/load_unix.go b/internal/load/load_unix.go
index 3b528f4a8178843fb5f384fccc704e7f01b65068..1ed6a108640a65a0a051ca4f315b9f29480248e3 100644
--- a/internal/load/load_unix.go
+++ b/internal/load/load_unix.go
@@ -5,7 +5,7 @@ //go:build darwin || freebsd || linux || netbsd
 
 package load
 
-import "github.com/jwijenbergh/purego"
+import "codeberg.org/puregotk/purego"
 
 func OpenLibrary(name string) (uintptr, error) {
 	return purego.Dlopen(name, purego.RTLD_NOW|purego.RTLD_GLOBAL)
diff --git a/nocgo.go b/nocgo.go
index 81c043cc214b49e4ef4fed8102186f08ed4998d4..30bd699037edacb77e2bcfc2fd38190bf29a4c00 100644
--- a/nocgo.go
+++ b/nocgo.go
@@ -22,4 +22,4 @@ // )
 // importing fakecgo will set these (using //go:linkname) with functions written
 // entirely in Go (except for some assembly trampolines to change GCC ABI to Go ABI).
 // Doing so makes it possible to build applications that call into C without CGO_ENABLED=1.
-import _ "github.com/jwijenbergh/purego/internal/fakecgo"
+import _ "codeberg.org/puregotk/purego/internal/fakecgo"
diff --git a/objc/objc_block_darwin.go b/objc/objc_block_darwin.go
index 11a645d02cbf8dd51a86b4266ad48f13706868de..c68a563c20d304739b24f5034d5c8ee73bfa64f6 100644
--- a/objc/objc_block_darwin.go
+++ b/objc/objc_block_darwin.go
@@ -9,8 +9,8 @@ 	"reflect"
 	"sync"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/internal/xreflect"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/internal/xreflect"
 )
 
 const (
@@ -95,7 +95,7 @@
 // blockCache is a thread safe cache of block layouts.
 //
 // It takes advantage of the block being the first argument of a block call being the block closure,
-// only invoking [github.com/jwijenbergh/purego.NewCallback] when it encounters a new function type (rather than on for every block creation).
+// only invoking [codeberg.org/puregotk/purego.NewCallback] when it encounters a new function type (rather than on for every block creation).
 // This should mitigate block creations putting pressure on the callback limit.
 type blockCache struct {
 	sync.Mutex
diff --git a/objc/objc_block_darwin_test.go b/objc/objc_block_darwin_test.go
index b444b9b99135f26bdfbadb11aef6f93715e778a2..06ca579200483a63b2051722b16fff4a9614d351 100644
--- a/objc/objc_block_darwin_test.go
+++ b/objc/objc_block_darwin_test.go
@@ -7,8 +7,8 @@ import (
 	"fmt"
 	"testing"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 func ExampleNewBlock() {
diff --git a/objc/objc_runtime_darwin.go b/objc/objc_runtime_darwin.go
index 265c475d8f2a0b559edcca1550060358a217cfef..292fcf38f8640fa2820969c86c53ce88da008500 100644
--- a/objc/objc_runtime_darwin.go
+++ b/objc/objc_runtime_darwin.go
@@ -15,9 +15,9 @@ 	"runtime"
 	"unicode"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/internal/strings"
-	"github.com/jwijenbergh/purego/internal/xreflect"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/internal/strings"
+	"codeberg.org/puregotk/purego/internal/xreflect"
 )
 
 // TODO: support try/catch?
diff --git a/objc/objc_runtime_darwin_test.go b/objc/objc_runtime_darwin_test.go
index da927e0c70accd3459fc6875a90603cea6df8fd2..c940123e9b6cb07ba7e1297b019ad1c3b5ad2582 100644
--- a/objc/objc_runtime_darwin_test.go
+++ b/objc/objc_runtime_darwin_test.go
@@ -9,8 +9,8 @@ 	"log"
 	"reflect"
 	"testing"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/objc"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/objc"
 )
 
 func ExampleRegisterClass_helloworld() {
diff --git a/struct_test.go b/struct_test.go
index cc819ea2ca0dc9bf4b18de876510c2477031bebe..43dd9fe0e583e1b01f091e3564e63e451ddd8832 100644
--- a/struct_test.go
+++ b/struct_test.go
@@ -13,7 +13,7 @@ 	"runtime"
 	"testing"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
+	"codeberg.org/puregotk/purego"
 )
 
 func TestRegisterFunc_structArgs(t *testing.T) {
diff --git a/syscall_cgo_linux.go b/syscall_cgo_linux.go
index 626b510f25081f6d91cffb0f871374c07260d833..cb6f13cc3971bbfb2ffa29f0a7e28f5399e00180 100644
--- a/syscall_cgo_linux.go
+++ b/syscall_cgo_linux.go
@@ -6,7 +6,7 @@
 package purego
 
 import (
-	"github.com/jwijenbergh/purego/internal/cgo"
+	"codeberg.org/puregotk/purego/internal/cgo"
 )
 
 var syscall15XABI0 = uintptr(cgo.Syscall15XABI0)
diff --git a/syscall_sysv.go b/syscall_sysv.go
index 1c5f6015905a2c1894b6e6e5445cdde9418e312a..8666d367353de2deb11e62a1d64792f19bb2905c 100644
--- a/syscall_sysv.go
+++ b/syscall_sysv.go
@@ -13,7 +13,7 @@ 	"runtime"
 	"sync"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego/internal/strings"
+	"codeberg.org/puregotk/purego/internal/strings"
 )
 
 var syscall15XABI0 uintptr
diff --git a/syscall_test.go b/syscall_test.go
index 44ebe08a25afaadfd31af05fc7edafd85bd12b93..112c26996046261b7e943b68739dbecedf34b5f5 100644
--- a/syscall_test.go
+++ b/syscall_test.go
@@ -9,8 +9,8 @@ 	"runtime"
 	"testing"
 	"unsafe"
 
-	"github.com/jwijenbergh/purego"
-	"github.com/jwijenbergh/purego/internal/load"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/internal/load"
 )
 
 func TestOS(t *testing.T) {

