From a96d55ceacb591b0c2b1149c21fe4c36ac752ced Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 30 May 2026 00:22:16 -0700
Subject: [PATCH] Go module: Move new files to codeberg

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/struct_arm64.go b/struct_arm64.go
index 5f347c83d0097cefc80dcc3b4e2b25a459be1196..e4b728ab6054f6f7cdf362383f82994a2600eab2 100644
--- a/struct_arm64.go
+++ b/struct_arm64.go
@@ -11,7 +11,7 @@ 	"strconv"
 	stdstrings "strings"
 	"unsafe"
 
-	"github.com/ebitengine/purego/internal/strings"
+	"codeberg.org/puregotk/purego/internal/strings"
 )
 
 func getStruct(outType reflect.Type, syscall syscall15Args) (v reflect.Value) {
diff --git a/syscall_bench_test.go b/syscall_bench_test.go
index ce923b93edb7488d458520671a476559ebce4db6..d642769fb676c04db83ea3ee8ecf559961a9c198 100644
--- a/syscall_bench_test.go
+++ b/syscall_bench_test.go
@@ -9,8 +9,8 @@ 	"os"
 	"path/filepath"
 	"testing"
 
-	"github.com/ebitengine/purego"
-	"github.com/ebitengine/purego/internal/load"
+	"codeberg.org/puregotk/purego"
+	"codeberg.org/puregotk/purego/internal/load"
 )
 
 // BenchmarkCallingMethods compares RegisterFunc, SyscallN, and Callback methods

