From 22e534a58ed60114128ddb46663a0df2c2a8d05f Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 12 May 2026 00:03:56 -0700
Subject: [PATCH] build: Use correct metainfo directory for Flatpak

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/build.py b/build.py
index d6ba406846b7193c7b0af261e78666735c9bfabe..bcb9a5e5ef17442054a70ccbdd89584fb6a21fd0 100755
--- a/build.py
+++ b/build.py
@@ -238,7 +238,8 @@     with open(desktop_path, "w") as f:
         f.write(desktop)
 
     # Create AppStream metadata
-    os.makedirs(f"{appdir}/usr/share/metainfo", exist_ok=True)
+    metainfo_dir = f"{appdir}/share/metainfo" if flatpak else f"{appdir}/usr/share/metainfo"
+    os.makedirs(metainfo_dir, exist_ok=True)
     date_str = datetime.now().strftime("%Y-%m-%d")
     appstream_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
 <component type="desktop-application">
@@ -274,7 +275,7 @@     <release version="{VERSION}" date="{date_str}"/>
   </releases>
 </component>
 """
-    with open(f"{appdir}/usr/share/metainfo/{app_id}.appdata.xml", "w") as f:
+    with open(f"{metainfo_dir}/{app_id}.appdata.xml", "w") as f:
         f.write(appstream_xml)
 
     # Copy the icon file

