From 09a7a199237e1311578aec480c26df093b6ed026 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 21 Sep 2025 12:30:28 -0700
Subject: [PATCH] feat: Add support for getting `n` of the last commits and creating patches from them

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index fcc3102919553f6c4b07ca4b807e014a6d831a14..b8211ba53004abbd868d67f6e8fcfe02961e325e 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -17,6 +17,10 @@ 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
 )
 
+const (
+	commitsKey = "commits"
+)
+
 var (
 	errRepoDirRemoteMissing = errors.New("remote is missing from repo directory")
 )
@@ -24,7 +28,7 @@
 var submitCommand = &cobra.Command{
 	Use:     "submit [repo-dir]",
 	Aliases: []string{"sub", "s"},
-	Short:   "Submit a patch to your ledger based on a repo's last commit",
+	Short:   "Submit a patch to your ledger based on a repo's last -n commits",
 	Args:    cobra.RangeArgs(0, 1),
 	RunE: func(cmd *cobra.Command, args []string) error {
 		if err := viper.BindPFlags(cmd.PersistentFlags()); err != nil {
@@ -51,6 +55,11 @@ 				return err
 			}
 		}
 
+		numCommits := viper.GetUint(commitsKey)
+		if numCommits == 0 {
+			numCommits = 1
+		}
+
 		log = log.With("sourceRepoDirectory", sourceRepoDirectory)
 
 		log.Debug("Opening source repo")
@@ -67,19 +76,19 @@ 		if err != nil {
 			return err
 		}
 
-		var remoteURL string
+		var remoteDir string
 		if len(remotes) > 0 && len(remotes[0].Config().URLs) > 0 {
 			ep, err := transport.NewEndpoint(remotes[0].Config().URLs[0])
 			if err != nil {
 				return err
 			}
 
-			remoteURL = filepath.Join(filepath.Clean(ep.Host), filepath.Clean(strings.TrimPrefix(ep.Path, "/")))
+			remoteDir = filepath.Join(filepath.Clean(ep.Host), filepath.Clean(strings.TrimPrefix(ep.Path, "/")))
 		} else {
 			return errRepoDirRemoteMissing
 		}
 
-		log = log.With("remoteURL", remoteURL)
+		log = log.With("remoteDirectory", remoteDir)
 
 		ref, err := repo.Head()
 		if err != nil {
@@ -91,60 +100,73 @@ 		if err != nil {
 			return err
 		}
 
-		var patch *object.Patch
-		if commit.NumParents() > 0 {
-			parent, err := commit.Parent(0)
-			if err != nil {
-				return err
-			}
+		currentCommit := commit
+
+		for i := uint(0); i < numCommits && currentCommit != nil; i++ {
+			var patch *object.Patch
+			if currentCommit.NumParents() > 0 {
+				parent, err := currentCommit.Parent(0)
+				if err != nil {
+					return err
+				}
 
-			patch, err = parent.Patch(commit)
-			if err != nil {
-				return err
-			}
-		} else {
-			patch, err = commit.Patch(nil)
-			if err != nil {
-				return err
+				patch, err = parent.Patch(currentCommit)
+				if err != nil {
+					return err
+				}
+			} else {
+				patch, err = currentCommit.Patch(nil)
+				if err != nil {
+					return err
+				}
 			}
-		}
 
-		patchText := fmt.Sprintf(`From %v Mon Sep 17 00:00:00 2001
+			patchText := fmt.Sprintf(`From %v Mon Sep 17 00:00:00 2001
 From: %v
 Date: %v
 Subject: [PATCH] %v---
 %v
-`, commit.Hash.String(), commit.Author.String(), commit.Author.When.Format(time.RFC1123Z), commit.Message, patch.String())
+`, currentCommit.Hash.String(), currentCommit.Author.String(), currentCommit.Author.When.Format(time.RFC1123Z), currentCommit.Message, patch.String())
+
+			var (
+				patchFileName = fmt.Sprintf("%v", currentCommit.Author.When.Unix()) + "-" + url.PathEscape(strings.Split(currentCommit.Message, "\n")[0]) + ".patch"
+				patchFilePath = filepath.Join(ledgerRepoDirectory, remoteDir, patchFileName)
+			)
 
-		var (
-			patchFileName = fmt.Sprintf("%v", commit.Author.When.Unix()) + "-" + url.PathEscape(strings.Split(commit.Message, "\n")[0]) + ".patch"
-			patchFilePath = filepath.Join(ledgerRepoDirectory, remoteURL, patchFileName)
-		)
+			commitLog := log.With("patchFilePath", patchFilePath)
+
+			commitLog.Debug("Writing patch to ledger repo")
 
-		log = log.With("patchFilePath", patchFilePath)
+			if err := os.MkdirAll(filepath.Dir(patchFilePath), os.ModePerm); err != nil {
+				return err
+			}
 
-		log.Debug("Writing patch to ledger repo")
+			if err := os.WriteFile(patchFilePath, []byte(patchText), os.ModePerm); err != nil {
+				return err
+			}
 
-		if err := os.MkdirAll(filepath.Dir(patchFilePath), os.ModePerm); err != nil {
-			return err
-		}
+			patchFilePathRel, err := filepath.Rel(ledgerRepoDirectory, patchFilePath)
+			if err != nil {
+				return err
+			}
 
-		if err := os.WriteFile(patchFilePath, []byte(patchText), os.ModePerm); err != nil {
-			return err
-		}
+			fmt.Println("Patch", patchFilePathRel, "submitted successfully")
 
-		patchFilePathRel, err := filepath.Rel(ledgerRepoDirectory, patchFilePath)
-		if err != nil {
-			return err
+			if i+1 < numCommits && currentCommit.NumParents() > 0 {
+				currentCommit, err = currentCommit.Parent(0)
+				if err != nil {
+					return err
+				}
+			}
 		}
-
-		fmt.Println("Patch", patchFilePathRel, "submitted successfully")
 
 		return nil
 	},
 }
 
 func init() {
+	submitCommand.PersistentFlags().UintP(commitsKey, "n", 1, "Number of commits to submit patches for")
+
 	viper.AutomaticEnv()
 
 	indexCommand.AddCommand(submitCommand)

