From f076a5e140e20635601fae53457aaf518f9c011f Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 22 Sep 2025 23:49:29 -0700
Subject: [PATCH] fix: Get patch file correctly for `pledge log` by diffing commits to get subtree

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/log.go b/cmd/pledge/cmd/log.go
index 354499a644419db30fdde1519986076df95bceda..d6c7f06453e558ec9a449aa6c1cf691872569380 100644
--- a/cmd/pledge/cmd/log.go
+++ b/cmd/pledge/cmd/log.go
@@ -12,7 +12,6 @@ 	"time"
 
 	"github.com/go-git/go-git/v6"
 	"github.com/go-git/go-git/v6/plumbing/object"
-	"github.com/go-git/go-git/v6/plumbing/storer"
 	"github.com/spf13/cobra"
 )
 
@@ -50,16 +49,31 @@ 		go func() {
 			defer logInput.Close()
 
 			if err = commitLog.ForEach(func(c *object.Commit) error {
+				var patch *object.Patch
+				if c.NumParents() > 0 {
+					parent, err := c.Parent(0)
+					if err != nil {
+						return err
+					}
+
+					patch, err = parent.Patch(c)
+					if err != nil {
+						return err
+					}
+				} else {
+					patch, err = c.Patch(nil)
+					if err != nil {
+						return err
+					}
+				}
+
 				var repo string
-				if tree, err := c.Tree(); err == nil {
-					tree.Files().ForEach(func(f *object.File) error {
-						if strings.HasSuffix(f.Name, ".patch") {
-							repo = filepath.Dir(f.Name)
+				for _, fp := range patch.FilePatches() {
+					if _, to := fp.Files(); to != nil && strings.HasSuffix(to.Path(), ".patch") {
+						repo = filepath.Dir(to.Path())
 
-							return storer.ErrStop
-						}
-						return nil
-					})
+						break
+					}
 				}
 
 				fmt.Fprintf(logInput, "Repo: %v\nFrom %v Mon Sep 17 00:00:00 2001\nFrom: %v\nDate: %v\nSubject: [PATCH] %v---\n\n", repo, c.Hash.String(), c.Author.String(), c.Author.When.Format(time.RFC1123Z), c.Message)

