From 929272be26061a45fc35be40c6e46f13dd745cbe Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 23 Sep 2025 23:01:14 -0700
Subject: [PATCH] refactor: Use the `slog` context methods consistently

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/check.go b/cmd/pledge/cmd/check.go
index 231234c9ecf6931378b44747aac09ee1f3aa5c3e..3ee123eb9ec4996b9b8bb7d3d3f918c78e3df4f1 100644
--- a/cmd/pledge/cmd/check.go
+++ b/cmd/pledge/cmd/check.go
@@ -35,7 +35,7 @@ 		}
 
 		log := log.With("ledgerRepoDirectory", ledgerRepoDirectory)
 
-		log.Debug("Checking for today's commit in ledger repo")
+		log.DebugContext(ctx, "Checking for today's commit in ledger repo")
 
 		ref, err := ledgerRepo.Head()
 		if err != nil {
@@ -123,7 +123,7 @@ 		pagerCmd.Stdin = logOutput
 		pagerCmd.Stdout = os.Stdout
 		pagerCmd.Stderr = os.Stderr
 
-		log.Debug("Writing commit output to pager", "pager", pager)
+		log.DebugContext(ctx, "Writing commit output to pager", "pager", pager)
 
 		if err := pagerCmd.Run(); err != nil {
 			return err
diff --git a/cmd/pledge/cmd/index.go b/cmd/pledge/cmd/index.go
index 0af70ec6510086508bfb56cc628a20e8d3ebcf48..0533091b96082541be5f5dfae28b52643030e762 100644
--- a/cmd/pledge/cmd/index.go
+++ b/cmd/pledge/cmd/index.go
@@ -39,7 +39,7 @@
 			if viper.IsSet(configKey) {
 				viper.SetConfigFile(viper.GetString(configKey))
 
-				log.Debug("Config key set, reading from file", "path", viper.GetViper().ConfigFileUsed())
+				log.DebugContext(cmd.Context(), "Config key set, reading from file", "path", viper.GetViper().ConfigFileUsed())
 
 				if err := viper.ReadInConfig(); err != nil {
 					return err
@@ -51,7 +51,7 @@
 				viper.SetConfigName(configName)
 				viper.AddConfigPath(configBase)
 
-				log.Debug("Config key not set, reading from default location", "path", filepath.Join(configBase, configName))
+				log.DebugContext(cmd.Context(), "Config key not set, reading from default location", "path", filepath.Join(configBase, configName))
 
 				if err := viper.ReadInConfig(); err != nil && !errors.As(err, &viper.ConfigFileNotFoundError{}) {
 					return err
@@ -68,7 +68,7 @@ 	ledgerRepoDirectory := viper.GetString(ledgerKey)
 
 	log = log.With("ledgerRepoDirectory", ledgerRepoDirectory)
 
-	log.Debug("Checking whether ledger repo directory exists")
+	log.DebugContext(ctx, "Checking whether ledger repo directory exists")
 
 	if _, err := os.Stat(ledgerRepoDirectory); err != nil {
 		if errors.Is(err, os.ErrNotExist) {
@@ -80,7 +80,7 @@
 		return nil, "", err
 	}
 
-	log.Debug("Opening ledger repo")
+	log.DebugContext(ctx, "Opening ledger repo")
 
 	repo, err := git.PlainOpen(ledgerRepoDirectory)
 	if err != nil {
@@ -88,7 +88,7 @@ 		return nil, "", err
 	}
 
 	if viper.GetBool(pullKey) {
-		log.Debug("Pulling ledger repo")
+		log.DebugContext(ctx, "Pulling ledger repo")
 
 		worktree, err := repo.Worktree()
 		if err != nil {
diff --git a/cmd/pledge/cmd/log.go b/cmd/pledge/cmd/log.go
index d6c7f06453e558ec9a449aa6c1cf691872569380..fd23d072d5bb06baaa10b9767c312661a8d73243 100644
--- a/cmd/pledge/cmd/log.go
+++ b/cmd/pledge/cmd/log.go
@@ -31,7 +31,7 @@ 		}
 
 		log := log.With("ledgerRepoDirectory", ledgerRepoDirectory)
 
-		log.Debug("Getting commit log from ledger repo")
+		log.DebugContext(cmd.Context(), "Getting commit log from ledger repo")
 
 		ref, err := ledgerRepo.Head()
 		if err != nil {
@@ -94,7 +94,7 @@ 		pagerCmd.Stdin = logOutput
 		pagerCmd.Stdout = os.Stdout
 		pagerCmd.Stderr = os.Stderr
 
-		log.Debug("Writing log output to pager", "pager", pager)
+		log.DebugContext(ctx, "Writing log output to pager", "pager", pager)
 
 		if err := pagerCmd.Run(); err != nil {
 			return err
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index 2e743292927a914cda741956243d792723fe0b8c..66ea864003bfc0cde7e15d04709eb545878a9001 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -66,14 +66,14 @@ 		}
 
 		log = log.With("sourceRepoDirectory", sourceRepoDirectory)
 
-		log.Debug("Opening source repo")
+		log.DebugContext(ctx, "Opening source repo")
 
 		repo, err := git.PlainOpen(sourceRepoDirectory)
 		if err != nil {
 			return err
 		}
 
-		log.Debug("Getting remote URL")
+		log.DebugContext(ctx, "Getting remote URL")
 
 		remotes, err := repo.Remotes()
 		if err != nil {
@@ -145,7 +145,7 @@ 			)
 
 			commitLog := log.With("patchFilePath", patchFilePath)
 
-			commitLog.Debug("Writing patch to ledger repo")
+			commitLog.DebugContext(ctx, "Writing patch to ledger repo")
 
 			if err := os.MkdirAll(filepath.Dir(patchFilePath), os.ModePerm); err != nil {
 				return err
@@ -160,7 +160,7 @@ 			if err != nil {
 				return err
 			}
 
-			commitLog.Debug("Adding patch file to git")
+			commitLog.DebugContext(ctx, "Adding patch file to git")
 
 			_, err = worktree.Add(patchFilePathRel)
 			if err != nil {
@@ -176,7 +176,7 @@ 				}
 
 				commitLog = commitLog.With("keyID", keyID)
 
-				commitLog.Debug("Reading private key to sign commit with")
+				commitLog.DebugContext(ctx, "Reading private key to sign commit with")
 
 				secretKey, err := pgp.GetPGPSecretKey(ctx, keyID)
 				if err != nil {
@@ -186,7 +186,7 @@
 				signKey = secretKey.GetEntity()
 			}
 
-			commitLog.Debug("Committing patch file with original commit info")
+			commitLog.DebugContext(ctx, "Committing patch file with original commit info")
 
 			if _, err = worktree.Commit(currentCommit.Message, &git.CommitOptions{
 				Author: &object.Signature{
@@ -221,7 +221,7 @@
 			return nil
 		}
 
-		log.Debug("Pushing commits to remote repo")
+		log.DebugContext(ctx, "Pushing commits to remote repo")
 
 		if err := ledgerRepo.Push(&git.PushOptions{}); err != nil {
 			return err

