From 7d58a0a5c715dae5f15adcda5b07bc4b2884c398 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 22 Sep 2025 23:50:15 -0700
Subject: [PATCH] refactor: Rework `pledge check` to only fetch the newest commit and read the patch file from git

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/check.go b/cmd/pledge/cmd/check.go
index 130d6feb332b4b511cb1bf7dd82beb5661fc71d9..231234c9ecf6931378b44747aac09ee1f3aa5c3e 100644
--- a/cmd/pledge/cmd/check.go
+++ b/cmd/pledge/cmd/check.go
@@ -11,9 +11,7 @@ 	"path/filepath"
 	"strings"
 	"time"
 
-	"github.com/go-git/go-git/v6"
 	"github.com/go-git/go-git/v6/plumbing/object"
-	"github.com/go-git/go-git/v6/plumbing/storer"
 	"github.com/spf13/cobra"
 )
 
@@ -44,57 +42,75 @@ 		if err != nil {
 			return err
 		}
 
-		commitLog, err := ledgerRepo.Log(&git.LogOptions{From: ref.Hash()})
+		commit, err := ledgerRepo.CommitObject(ref.Hash())
 		if err != nil {
 			return err
 		}
-		defer commitLog.Close()
 
 		var (
 			now      = time.Now()
 			today    = time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, now.Location())
 			tomorrow = today.AddDate(0, 0, 1)
-
-			todayCommit *object.Commit
 		)
-		if err = commitLog.ForEach(func(c *object.Commit) error {
-			commitDate := c.Author.When
-			if commitDate.After(today) && commitDate.Before(tomorrow) {
-				todayCommit = c
+		commitDate := commit.Author.When
+		if !(commitDate.After(today) && commitDate.Before(tomorrow)) {
+			return errNoCommitFoundForToday
+		}
 
-				return storer.ErrStop
+		var patch *object.Patch
+		if commit.NumParents() > 0 {
+			parent, err := commit.Parent(0)
+			if err != nil {
+				return err
 			}
 
-			if commitDate.Before(today) {
-				return storer.ErrStop
+			patch, err = parent.Patch(commit)
+			if err != nil {
+				return err
+			}
+		} else {
+			patch, err = commit.Patch(nil)
+			if err != nil {
+				return err
 			}
-
-			return nil
-		}); err != nil {
-			return err
 		}
 
-		if todayCommit == nil {
-			return errNoCommitFoundForToday
+		var (
+			repo          string
+			patchFilePath string
+		)
+		for _, fp := range patch.FilePatches() {
+			if _, to := fp.Files(); to != nil && strings.HasSuffix(to.Path(), ".patch") {
+				repo = filepath.Dir(to.Path())
+				patchFilePath = to.Path()
+
+				break
+			}
 		}
 
 		logOutput, logInput := io.Pipe()
 		go func() {
 			defer logInput.Close()
 
-			if tree, err := todayCommit.Tree(); err == nil {
-				tree.Files().ForEach(func(f *object.File) error {
-					if strings.HasSuffix(f.Name, ".patch") {
-						repo := filepath.Dir(f.Name)
-						if content, err := f.Contents(); err == nil {
-							fmt.Fprintf(logInput, "Repo: %v\n%s", repo, content)
-						}
+			fmt.Fprintf(logInput, "Repo: %v\n", repo)
+
+			tree, err := commit.Tree()
+			if err != nil {
+				return
+			}
+
+			file, err := tree.File(patchFilePath)
+			if err != nil {
+				return
+			}
 
-						return storer.ErrStop
-					}
-					return nil
-				})
+			reader, err := file.Reader()
+			if err != nil {
+				return
 			}
+			defer reader.Close()
+
+			io.Copy(logInput, reader)
 		}()
 
 		pager := os.Getenv("PAGER")
@@ -107,7 +123,7 @@ 		pagerCmd.Stdin = logOutput
 		pagerCmd.Stdout = os.Stdout
 		pagerCmd.Stderr = os.Stderr
 
-		log.Debug("Writing log output to pager", "pager", pager)
+		log.Debug("Writing commit output to pager", "pager", pager)
 
 		if err := pagerCmd.Run(); err != nil {
 			return err

