From 9546186258b2b832930d4be944432611c4270daa Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 20 Sep 2025 00:12:43 -0700
Subject: [PATCH] refactor: Simplify patch text creation and add support for multi-line commits

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index 5ff1d19fc27d76ac38d10ee0243fdf441944fe71..9eec18cbe65e9cbd28a56ac42d8a698c972cf786 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -3,6 +3,7 @@
 import (
 	"context"
 	"fmt"
+	"net/url"
 	"os"
 	"strings"
 	"time"
@@ -76,9 +77,7 @@ 		if err != nil {
 			return err
 		}
 
-		subject := strings.Split(commit.Message, "\n")[0]
-
-		patchFileName := fmt.Sprintf("%v", time.Now().Unix()) + "-" + subject + ".patch"
+		patchFileName := fmt.Sprintf("%v", time.Now().Unix()) + "-" + url.PathEscape(strings.Split(commit.Message, "\n")[0]) + ".patch"
 
 		log = log.With("patchFilename", patchFileName)
 
@@ -98,16 +97,15 @@
 		patchText := fmt.Sprintf(`From %v Mon Sep 17 00:00:00 2001
 From: %v
 Date: %v
-Subject: [PATCH] %v
-
-Signed-off-by: %v
----
+Subject: [PATCH] %v---
 %v
-`, commit.Hash.String(), commit.Author.String(), commit.Author.When.Format(time.RFC1123Z), subject, commit.Author.String(), patch.String())
+`, commit.Hash.String(), commit.Author.String(), commit.Author.When.Format(time.RFC1123Z), commit.Message, patch.String())
 
 		log = log.With("patchText", patchText)
 
-		log.Debug("Writing patch")
+		log.Debug("Writing patch to ledger repo")
+
+		fmt.Println(patchText)
 
 		return nil
 	},

