From 623292d6963c8065dd16e73f75cc304aefd2aeaa Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 21 Sep 2025 12:55:30 -0700
Subject: [PATCH] feat: Commit and push changes to ledger as part of submit

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index 8be1a044f74bee30acb1fcfe7ce9344ba937000a..d4d4f8a576d72d097d132b982f364bde1736664a 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -38,7 +38,7 @@
 		ctx, cancel := context.WithCancel(cmd.Context())
 		defer cancel()
 
-		_, ledgerRepoDirectory, err := verifyLedgerRepo(ctx, log)
+		ledgerRepo, ledgerRepoDirectory, err := verifyLedgerRepo(ctx, log)
 		if err != nil {
 			return err
 		}
@@ -102,6 +102,11 @@ 		}
 
 		currentCommit := commit
 
+		worktree, err := ledgerRepo.Worktree()
+		if err != nil {
+			return err
+		}
+
 		for i := uint(0); i < numCommits && currentCommit != nil; i++ {
 			var patch *object.Patch
 			if currentCommit.NumParents() > 0 {
@@ -150,7 +155,26 @@ 			if err != nil {
 				return err
 			}
 
-			fmt.Println("Patch", patchFilePathRel, "submitted successfully")
+			commitLog.Debug("Adding patch file to git")
+
+			_, err = worktree.Add(patchFilePathRel)
+			if err != nil {
+				return err
+			}
+
+			commitLog.Debug("Committing patch file with original commit info")
+
+			if _, err = worktree.Commit(currentCommit.Message, &git.CommitOptions{
+				Author: &object.Signature{
+					Name:  currentCommit.Author.Name,
+					Email: currentCommit.Author.Email,
+					When:  currentCommit.Author.When,
+				},
+			}); err != nil {
+				return err
+			}
+
+			fmt.Println("Patch", patchFilePathRel, "committed successfully")
 
 			if i+1 < numCommits && currentCommit.NumParents() > 0 {
 				currentCommit, err = currentCommit.Parent(0)
@@ -159,6 +183,15 @@ 					return err
 				}
 			}
 		}
+
+		log.Debug("Pushing commits to remote repo")
+
+		err = ledgerRepo.Push(&git.PushOptions{})
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("All patches pushed to ledger repo remote")
 
 		return nil
 	},

