From 54ee1687bf0b4944d45793caef7549e1b449af9c Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 20 Sep 2025 00:46:55 -0700
Subject: [PATCH] feat: Support repos with just one commit

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index 9eec18cbe65e9cbd28a56ac42d8a698c972cf786..7bc515189e73e7eba19d961e15011f716c537b29 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -77,21 +77,22 @@ 		if err != nil {
 			return err
 		}
 
-		patchFileName := fmt.Sprintf("%v", time.Now().Unix()) + "-" + url.PathEscape(strings.Split(commit.Message, "\n")[0]) + ".patch"
-
-		log = log.With("patchFilename", patchFileName)
-
-		var parent *object.Commit
+		var patch *object.Patch
 		if commit.NumParents() > 0 {
-			parent, err = commit.Parent(0)
+			parent, err := commit.Parent(0)
 			if err != nil {
 				return err
 			}
-		}
 
-		patch, err := parent.Patch(commit)
-		if err != nil {
-			return err
+			patch, err = parent.Patch(commit)
+			if err != nil {
+				return err
+			}
+		} else {
+			patch, err = commit.Patch(nil)
+			if err != nil {
+				return err
+			}
 		}
 
 		patchText := fmt.Sprintf(`From %v Mon Sep 17 00:00:00 2001

