From 8240b95797a54590604bf89b2de7509d2a44b423 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 20 Sep 2025 00:48:14 -0700
Subject: [PATCH] feat: Transform the remote URL into a file path properly

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/cmd/pledge/cmd/submit.go b/cmd/pledge/cmd/submit.go
index 7bc515189e73e7eba19d961e15011f716c537b29..fa48d1bf97d57ff4072a74133eaf6b10f173e81f 100644
--- a/cmd/pledge/cmd/submit.go
+++ b/cmd/pledge/cmd/submit.go
@@ -5,13 +5,19 @@ 	"context"
 	"fmt"
 	"net/url"
 	"os"
+	"path/filepath"
 	"strings"
 	"time"
 
 	"github.com/go-git/go-git/v6"
 	"github.com/go-git/go-git/v6/plumbing/object"
+	"github.com/go-git/go-git/v6/plumbing/transport"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+)
+
+var (
+	errRepoDirRemoteMissing = errors.New("remote is missing from repo directory")
 )
 
 var submitCommand = &cobra.Command{
@@ -62,7 +68,14 @@ 		}
 
 		var remoteURL string
 		if len(remotes) > 0 && len(remotes[0].Config().URLs) > 0 {
-			remoteURL = remotes[0].Config().URLs[0]
+			ep, err := transport.NewEndpoint(remotes[0].Config().URLs[0])
+			if err != nil {
+				return err
+			}
+
+			remoteURL = filepath.Join(filepath.Clean(ep.Host), filepath.Clean(strings.TrimPrefix(ep.Path, "/")))
+		} else {
+			return errRepoDirRemoteMissing
 		}
 
 		log = log.With("remoteURL", remoteURL)

