From 5bb8c3359c53cc9ea92f457da221af41926468d8 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Thu, 19 Mar 2026 22:04:32 -0700
Subject: [PATCH] Examples: Bump to GNOME 50 and latest library versions

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/examples/myapp-gnome-gomod/page.codeberg.puregotk.MyAppGnomeGomod.json b/examples/myapp-gnome-gomod/page.codeberg.puregotk.MyAppGnomeGomod.json
index 7317a132469fea9767648bfd2f4a9f6fab640bc0..5bdd227ca61f5a26687a3fdd9a051c37de42760d 100644
--- a/examples/myapp-gnome-gomod/page.codeberg.puregotk.MyAppGnomeGomod.json
+++ b/examples/myapp-gnome-gomod/page.codeberg.puregotk.MyAppGnomeGomod.json
@@ -1,7 +1,7 @@
 {
   "id": "page.codeberg.puregotk.MyAppGnomeGomod",
   "runtime": "org.gnome.Platform",
-  "runtime-version": "49",
+  "runtime-version": "50",
   "sdk": "org.gnome.Sdk",
   "sdk-extensions": ["org.freedesktop.Sdk.Extension.golang"],
   "command": "myapp-gnome-gomod",
diff --git a/examples/myapp-gnome-meson/page.codeberg.puregotk.MyAppGnomeMeson.json b/examples/myapp-gnome-meson/page.codeberg.puregotk.MyAppGnomeMeson.json
index 97ba8e61bd0c3850f5bf3057603fb53ca9abbb84..bd6c75523f6269a29d5188d742169440b85eeb31 100644
--- a/examples/myapp-gnome-meson/page.codeberg.puregotk.MyAppGnomeMeson.json
+++ b/examples/myapp-gnome-meson/page.codeberg.puregotk.MyAppGnomeMeson.json
@@ -1,7 +1,7 @@
 {
   "id": "page.codeberg.puregotk.MyAppGnomeMeson",
   "runtime": "org.gnome.Platform",
-  "runtime-version": "49",
+  "runtime-version": "50",
   "sdk": "org.gnome.Sdk",
   "sdk-extensions": ["org.freedesktop.Sdk.Extension.golang"],
   "command": "myapp-gnome-meson",
diff --git a/examples/myapp-gnome-meson/src/meson.build b/examples/myapp-gnome-meson/src/meson.build
index b1b9c972bf6354ff0dd19c6f83eb0eb609deeee9..08c3112456a406cb72c9242af95f77e32ee2f4d6 100644
--- a/examples/myapp-gnome-meson/src/meson.build
+++ b/examples/myapp-gnome-meson/src/meson.build
@@ -20,7 +20,7 @@ myapp_gnome_meson_source_go_files = files('main.go')
 
 myapp_gnome_meson_deps = [
     dependency('gtk4'),
-    dependency('libadwaita-1', version: '>= 1.4'),
+    dependency('libadwaita-1', version: '>= 1.9'),
     dependency('mylib-gtk-meson-0.1', version: '>= 0.1.0'),
 ]
 
diff --git a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
index b33950d434cb16d92e819d741ec5056056e66031..5d18b76e1000a9b3f20a29e6de0e300e6f2e2e5a 100644
--- a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
+++ b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
@@ -451,6 +451,19 @@ 	gtk.XGtkAccessibleAnnounce(x.GoPointer(), MessageVar, PriorityVar)
 
 }
 
+// Retrieves the accessible identifier for the accessible object.
+//
+// This functionality can be overridden by `GtkAccessible`
+// implementations.
+//
+// It is left to the accessible implementation to define the scope
+// and uniqueness of the identifier.
+func (x *MainApplicationWindow) GetAccessibleId() string {
+
+	cret := gtk.XGtkAccessibleGetAccessibleId(x.GoPointer())
+	return cret
+}
+
 // Retrieves the accessible parent for an accessible object.
 //
 // This function returns `NULL` for top level widgets.
diff --git a/examples/mylib-gtk-meson/page.codeberg.puregotk.MyLibGtkMesonExample.json b/examples/mylib-gtk-meson/page.codeberg.puregotk.MyLibGtkMesonExample.json
index b03e66d24bf59bf68e6940a2ef333f9f2306e121..741ddf9c732e20ba6f75b087143cee7c9c183196 100644
--- a/examples/mylib-gtk-meson/page.codeberg.puregotk.MyLibGtkMesonExample.json
+++ b/examples/mylib-gtk-meson/page.codeberg.puregotk.MyLibGtkMesonExample.json
@@ -1,7 +1,7 @@
 {
   "id": "page.codeberg.puregotk.MyLibGtkMesonExample",
   "runtime": "org.gnome.Platform",
-  "runtime-version": "49",
+  "runtime-version": "50",
   "sdk": "org.gnome.Sdk",
   "sdk-extensions": ["org.freedesktop.Sdk.Extension.golang"],
   "command": "example.js",
diff --git a/examples/mylib-gtk-meson/src/docs.toml.in b/examples/mylib-gtk-meson/src/docs.toml.in
index efc46fa0b76b3f28078a9a2d6e048edc224d731c..88b357a70ad993aab1fc34a74e7ee508bc39335f 100644
--- a/examples/mylib-gtk-meson/src/docs.toml.in
+++ b/examples/mylib-gtk-meson/src/docs.toml.in
@@ -18,7 +18,7 @@
 [dependencies."Adw-1"]
     name = "Adw"
     description = "Building blocks for modern GNOME applications"
-    docs_url = "https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.7/"
+    docs_url = "https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.9/"
 
 [theme]
     show_index_summary = true
diff --git a/examples/mylib-gtk-meson/src/meson.build b/examples/mylib-gtk-meson/src/meson.build
index bec0ab5e90a362014be2a41745c94e3d6101fc8e..766ffae33659c112d35ee9829370dcf2a7c82f55 100644
--- a/examples/mylib-gtk-meson/src/meson.build
+++ b/examples/mylib-gtk-meson/src/meson.build
@@ -43,7 +43,7 @@ )
 
 mylib_gtk_meson_deps = [
     dependency('gtk4'),
-    dependency('libadwaita-1', version: '>= 1.4'),
+    dependency('libadwaita-1', version: '>= 1.9'),
 ]
 
 mylib_gtk_meson_blueprints = custom_target(
@@ -137,7 +137,7 @@     name: 'mylib-gtk-meson',
     filebase: 'mylib-gtk-meson-' + api_version,
     version: meson.project_version(),
     subdirs: 'mylib-gtk-meson',
-    requires: ['gtk4 >= 4.17.5', 'libadwaita-1 >= 1.7.5'],
+    requires: ['gtk4 >= 4.22.1', 'libadwaita-1 >= 1.9.0'],
     install_dir: get_option('libdir') / 'pkgconfig',
 )
 

