From ffb1bc0afd987d3ad007e67ce91862a46803c434 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 13 Apr 2026 22:44:39 -0700
Subject: [PATCH] GIR: Process `throws="1"` correctly even if there are no parameters

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/gir/types/types.go b/internal/gir/types/types.go
index d348c4dc8a1beaf2d0fc8f0508553cb179edd70e..22aee1496933bc603a5be9718fd3bfcd039bac64 100644
--- a/internal/gir/types/types.go
+++ b/internal/gir/types/types.go
@@ -675,16 +675,11 @@ 	return snaked + "Var"
 }
 
 func (p *Parameters) Template(ns string, ifacens string, kinds KindMap, throws bool) funcArgsTemplate {
-	if p == nil {
-		return funcArgsTemplate{}
-	}
-	params := p.Parameters
-	if len(params) == 0 {
-		return funcArgsTemplate{}
-	}
 	args := funcArgsTemplate{}
-	for _, par := range params {
-		args.Add(par, ifacens, ns, kinds)
+	if p != nil {
+		for _, par := range p.Parameters {
+			args.Add(par, ifacens, ns, kinds)
+		}
 	}
 	if throws {
 		args.AddThrows(ns)

