From 2f5010d598c760216f4774ce33235d8e37d93310 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 13 Apr 2026 22:53:27 -0700
Subject: [PATCH] Gen: Re-run script

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
index 5d18b76e1000a9b3f20a29e6de0e300e6f2e2e5a..fe3609bf7bba313e7aa4abfd511ee4b60f0caaaa 100644
--- a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
+++ b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
@@ -37,9 +37,7 @@ var xMainApplicationWindowShowToast func(uintptr, string)
 
 // Shows a toast notification with the given message inside the window.
 func (x *MainApplicationWindow) ShowToast(MessageVar string) {
-
 	xMainApplicationWindowShowToast(x.GoPointer(), MessageVar)
-
 }
 
 func (c *MainApplicationWindow) GoPointer() uintptr {
@@ -80,7 +78,6 @@ 		fa.Ptr = clsPtr
 		cbFn := *cb
 
 		cbFn(fa)
-
 	}
 	cbRefPtr := purego.NewCallback(fcb)
 	glib.SaveCallback(cbPtr, cbRefPtr)
@@ -91,36 +88,28 @@ // Emits the [signal@Gio.ActionGroup::action-added] signal on @action_group.
 //
 // This function should only be called by [type@Gio.ActionGroup] implementations.
 func (x *MainApplicationWindow) ActionAdded(ActionNameVar string) {
-
 	gio.XGActionGroupActionAdded(x.GoPointer(), ActionNameVar)
-
 }
 
 // Emits the [signal@Gio.ActionGroup::action-enabled-changed] signal on @action_group.
 //
 // This function should only be called by [type@Gio.ActionGroup] implementations.
 func (x *MainApplicationWindow) ActionEnabledChanged(ActionNameVar string, EnabledVar bool) {
-
 	gio.XGActionGroupActionEnabledChanged(x.GoPointer(), ActionNameVar, EnabledVar)
-
 }
 
 // Emits the [signal@Gio.ActionGroup::action-removed] signal on @action_group.
 //
 // This function should only be called by [type@Gio.ActionGroup] implementations.
 func (x *MainApplicationWindow) ActionRemoved(ActionNameVar string) {
-
 	gio.XGActionGroupActionRemoved(x.GoPointer(), ActionNameVar)
-
 }
 
 // Emits the [signal@Gio.ActionGroup::action-state-changed] signal on @action_group.
 //
 // This function should only be called by [type@Gio.ActionGroup] implementations.
 func (x *MainApplicationWindow) ActionStateChanged(ActionNameVar string, StateVar *glib.Variant) {
-
 	gio.XGActionGroupActionStateChanged(x.GoPointer(), ActionNameVar, StateVar)
-
 }
 
 // Activate the named action within @action_group.
@@ -157,9 +146,7 @@ //
 // exit (0);
 // ```
 func (x *MainApplicationWindow) ActivateAction(ActionNameVar string, ParameterVar *glib.Variant) {
-
 	gio.XGActionGroupActivateAction(x.GoPointer(), ActionNameVar, ParameterVar)
-
 }
 
 // Request for the state of the named action within @action_group to be
@@ -174,9 +161,7 @@ // See [method@Gio.ActionGroup.get_action_state_hint].
 //
 // If the @value GVariant is floating, it is consumed.
 func (x *MainApplicationWindow) ChangeActionState(ActionNameVar string, ValueVar *glib.Variant) {
-
 	gio.XGActionGroupChangeActionState(x.GoPointer(), ActionNameVar, ValueVar)
-
 }
 
 // Checks if the named action within @action_group is currently enabled.
@@ -184,7 +169,6 @@ //
 // An action must be enabled in order to be activated or in order to
 // have its state changed from outside callers.
 func (x *MainApplicationWindow) GetActionEnabled(ActionNameVar string) bool {
-
 	cret := gio.XGActionGroupGetActionEnabled(x.GoPointer(), ActionNameVar)
 	return cret
 }
@@ -203,7 +187,6 @@ // The parameter type of a particular action will never change but it is
 // possible for an action to be removed and for a new action to be added
 // with the same name but a different parameter type.
 func (x *MainApplicationWindow) GetActionParameterType(ActionNameVar string) *glib.VariantType {
-
 	cret := gio.XGActionGroupGetActionParameterType(x.GoPointer(), ActionNameVar)
 	return cret
 }
@@ -217,7 +200,6 @@ //
 // The return value (if non-`NULL`) should be freed with
 // [method@GLib.Variant.unref] when it is no longer required.
 func (x *MainApplicationWindow) GetActionState(ActionNameVar string) *glib.Variant {
-
 	cret := gio.XGActionGroupGetActionState(x.GoPointer(), ActionNameVar)
 	return cret
 }
@@ -241,7 +223,6 @@ //
 // The return value (if non-`NULL`) should be freed with
 // [method@GLib.Variant.unref] when it is no longer required.
 func (x *MainApplicationWindow) GetActionStateHint(ActionNameVar string) *glib.Variant {
-
 	cret := gio.XGActionGroupGetActionStateHint(x.GoPointer(), ActionNameVar)
 	return cret
 }
@@ -263,14 +244,12 @@ // The state type of a particular action will never change but it is
 // possible for an action to be removed and for a new action to be added
 // with the same name but a different state type.
 func (x *MainApplicationWindow) GetActionStateType(ActionNameVar string) *glib.VariantType {
-
 	cret := gio.XGActionGroupGetActionStateType(x.GoPointer(), ActionNameVar)
 	return cret
 }
 
 // Checks if the named action exists within @action_group.
 func (x *MainApplicationWindow) HasAction(ActionNameVar string) bool {
-
 	cret := gio.XGActionGroupHasAction(x.GoPointer(), ActionNameVar)
 	return cret
 }
@@ -280,7 +259,6 @@ //
 // The caller is responsible for freeing the list with [func@GLib.strfreev] when
 // it is no longer required.
 func (x *MainApplicationWindow) ListActions() []string {
-
 	cret := gio.XGActionGroupListActions(x.GoPointer())
 	return cret
 }
@@ -313,7 +291,6 @@ // fields (as indicated by having a non-`NULL` reference passed in) are
 // filled.  If the action doesn’t exist, `FALSE` is returned and the
 // fields may or may not have been modified.
 func (x *MainApplicationWindow) QueryAction(ActionNameVar string, EnabledVar *bool, ParameterTypeVar **glib.VariantType, StateTypeVar **glib.VariantType, StateHintVar **glib.Variant, StateVar **glib.Variant) bool {
-
 	cret := gio.XGActionGroupQueryAction(x.GoPointer(), ActionNameVar, EnabledVar, ParameterTypeVar, StateTypeVar, StateHintVar, StateVar)
 	return cret
 }
@@ -325,9 +302,7 @@ // as @action then the old action is dropped from the action map.
 //
 // The action map takes its own reference on @action.
 func (x *MainApplicationWindow) AddAction(ActionVar gio.Action) {
-
 	gio.XGActionMapAddAction(x.GoPointer(), ActionVar.GoPointer())
-
 }
 
 // A convenience function for creating multiple [class@Gio.SimpleAction]
@@ -374,9 +349,7 @@ //	}
 //
 // ```
 func (x *MainApplicationWindow) AddActionEntries(EntriesVar []gio.ActionEntry, NEntriesVar int32, UserDataVar uintptr) {
-
 	gio.XGActionMapAddActionEntries(x.GoPointer(), EntriesVar, NEntriesVar, UserDataVar)
-
 }
 
 // Looks up the action with the name @action_name in @action_map.
@@ -400,9 +373,7 @@ // Removes the named action from the action map.
 //
 // If no action of this name is in the map then nothing happens.
 func (x *MainApplicationWindow) RemoveAction(ActionNameVar string) {
-
 	gio.XGActionMapRemoveAction(x.GoPointer(), ActionNameVar)
-
 }
 
 // Remove actions from a [iface@Gio.ActionMap]. This is meant as the reverse of
@@ -431,9 +402,7 @@ //	}
 //
 // ```
 func (x *MainApplicationWindow) RemoveActionEntries(EntriesVar []gio.ActionEntry, NEntriesVar int32) {
-
 	gio.XGActionMapRemoveActionEntries(x.GoPointer(), EntriesVar, NEntriesVar)
-
 }
 
 // Requests the user's screen reader to announce the given message.
@@ -446,9 +415,7 @@ //
 // Also, by using this API, you can ensure that the message
 // does not interrupts the user's current screen reader output.
 func (x *MainApplicationWindow) Announce(MessageVar string, PriorityVar gtk.AccessibleAnnouncementPriority) {
-
 	gtk.XGtkAccessibleAnnounce(x.GoPointer(), MessageVar, PriorityVar)
-
 }
 
 // Retrieves the accessible identifier for the accessible object.
@@ -459,7 +426,6 @@ //
 // It is left to the accessible implementation to define the scope
 // and uniqueness of the identifier.
 func (x *MainApplicationWindow) GetAccessibleId() string {
-
 	cret := gtk.XGtkAccessibleGetAccessibleId(x.GoPointer())
 	return cret
 }
@@ -482,7 +448,6 @@ }
 
 // Retrieves the accessible role of an accessible object.
 func (x *MainApplicationWindow) GetAccessibleRole() gtk.AccessibleRole {
-
 	cret := gtk.XGtkAccessibleGetAccessibleRole(x.GoPointer())
 	return cret
 }
@@ -507,7 +472,6 @@ // This functionality can be overridden by `GtkAccessible`
 // implementations, e.g. to get the bounds from an ignored
 // child widget.
 func (x *MainApplicationWindow) GetBounds(XVar *int32, YVar *int32, WidthVar *int32, HeightVar *int32) bool {
-
 	cret := gtk.XGtkAccessibleGetBounds(x.GoPointer(), XVar, YVar, WidthVar, HeightVar)
 	return cret
 }
@@ -546,30 +510,23 @@ // This functionality can be overridden by `GtkAccessible`
 // implementations, e.g. to get platform state from an ignored
 // child widget, as is the case for `GtkText` wrappers.
 func (x *MainApplicationWindow) GetPlatformState(StateVar gtk.AccessiblePlatformState) bool {
-
 	cret := gtk.XGtkAccessibleGetPlatformState(x.GoPointer(), StateVar)
 	return cret
 }
 
 // Resets the accessible property to its default value.
 func (x *MainApplicationWindow) ResetProperty(PropertyVar gtk.AccessibleProperty) {
-
 	gtk.XGtkAccessibleResetProperty(x.GoPointer(), PropertyVar)
-
 }
 
 // Resets the accessible relation to its default value.
 func (x *MainApplicationWindow) ResetRelation(RelationVar gtk.AccessibleRelation) {
-
 	gtk.XGtkAccessibleResetRelation(x.GoPointer(), RelationVar)
-
 }
 
 // Resets the accessible state to its default value.
 func (x *MainApplicationWindow) ResetState(StateVar gtk.AccessibleState) {
-
 	gtk.XGtkAccessibleResetState(x.GoPointer(), StateVar)
-
 }
 
 // Sets the parent and sibling of an accessible object.
@@ -582,9 +539,7 @@ // interface, you will use this function to ensure that the parent of each
 // child widget is the metadata object, and the parent of each metadata
 // object is the container widget.
 func (x *MainApplicationWindow) SetAccessibleParent(ParentVar gtk.Accessible, NextSiblingVar gtk.Accessible) {
-
 	gtk.XGtkAccessibleSetAccessibleParent(x.GoPointer(), ParentVar.GoPointer(), NextSiblingVar.GoPointer())
-
 }
 
 // Updates the next accessible sibling.
@@ -592,9 +547,7 @@ //
 // That might be useful when a new child of a custom accessible
 // is created, and it needs to be linked to a previous child.
 func (x *MainApplicationWindow) UpdateNextAccessibleSibling(NewSiblingVar gtk.Accessible) {
-
 	gtk.XGtkAccessibleUpdateNextAccessibleSibling(x.GoPointer(), NewSiblingVar.GoPointer())
-
 }
 
 // Informs ATs that the platform state has changed.
@@ -603,9 +556,7 @@ // This function should be used by `GtkAccessible` implementations that
 // have a platform state but are not widgets. Widgets handle platform
 // states automatically.
 func (x *MainApplicationWindow) UpdatePlatformState(StateVar gtk.AccessiblePlatformState) {
-
 	gtk.XGtkAccessibleUpdatePlatformState(x.GoPointer(), StateVar)
-
 }
 
 // Updates a list of accessible properties.
@@ -627,9 +578,7 @@ //	-1);
 //
 // ```
 func (x *MainApplicationWindow) UpdateProperty(FirstPropertyVar gtk.AccessibleProperty, varArgs ...interface{}) {
-
 	gtk.XGtkAccessibleUpdateProperty(x.GoPointer(), FirstPropertyVar, varArgs...)
-
 }
 
 // Updates an array of accessible properties.
@@ -639,9 +588,7 @@ // property change must be communicated to assistive technologies.
 //
 // This function is meant to be used by language bindings.
 func (x *MainApplicationWindow) UpdatePropertyValue(NPropertiesVar int32, PropertiesVar []gtk.AccessibleProperty, ValuesVar []gobject.Value) {
-
 	gtk.XGtkAccessibleUpdatePropertyValue(x.GoPointer(), NPropertiesVar, PropertiesVar, ValuesVar)
-
 }
 
 // Updates a list of accessible relations.
@@ -663,9 +610,7 @@ //	-1);
 //
 // ```
 func (x *MainApplicationWindow) UpdateRelation(FirstRelationVar gtk.AccessibleRelation, varArgs ...interface{}) {
-
 	gtk.XGtkAccessibleUpdateRelation(x.GoPointer(), FirstRelationVar, varArgs...)
-
 }
 
 // Updates an array of accessible relations.
@@ -675,9 +620,7 @@ // relation change must be communicated to assistive technologies.
 //
 // This function is meant to be used by language bindings.
 func (x *MainApplicationWindow) UpdateRelationValue(NRelationsVar int32, RelationsVar []gtk.AccessibleRelation, ValuesVar []gobject.Value) {
-
 	gtk.XGtkAccessibleUpdateRelationValue(x.GoPointer(), NRelationsVar, RelationsVar, ValuesVar)
-
 }
 
 // Updates a list of accessible states.
@@ -700,9 +643,7 @@ //	-1);
 //
 // ```
 func (x *MainApplicationWindow) UpdateState(FirstStateVar gtk.AccessibleState, varArgs ...interface{}) {
-
 	gtk.XGtkAccessibleUpdateState(x.GoPointer(), FirstStateVar, varArgs...)
-
 }
 
 // Updates an array of accessible states.
@@ -712,9 +653,7 @@ // state change must be communicated to assistive technologies.
 //
 // This function is meant to be used by language bindings.
 func (x *MainApplicationWindow) UpdateStateValue(NStatesVar int32, StatesVar []gtk.AccessibleState, ValuesVar []gobject.Value) {
-
 	gtk.XGtkAccessibleUpdateStateValue(x.GoPointer(), NStatesVar, StatesVar, ValuesVar)
-
 }
 
 // Gets the ID of the @buildable object.
@@ -722,7 +661,6 @@ //
 // `GtkBuilder` sets the name based on the ID attribute
 // of the `&lt;object&gt;` tag used to construct the @buildable.
 func (x *MainApplicationWindow) GetBuildableId() string {
-
 	cret := gtk.XGtkBuildableGetBuildableId(x.GoPointer())
 	return cret
 }
@@ -762,27 +700,21 @@ //
 // This is the translation from @self's surface coordinates into
 // @self's widget coordinates.
 func (x *MainApplicationWindow) GetSurfaceTransform(XVar *float64, YVar *float64) {
-
 	gtk.XGtkNativeGetSurfaceTransform(x.GoPointer(), XVar, YVar)
-
 }
 
 // Realizes a `GtkNative`.
 //
 // This should only be used by subclasses.
 func (x *MainApplicationWindow) Realize() {
-
 	gtk.XGtkNativeRealize(x.GoPointer())
-
 }
 
 // Unrealizes a `GtkNative`.
 //
 // This should only be used by subclasses.
 func (x *MainApplicationWindow) Unrealize() {
-
 	gtk.XGtkNativeUnrealize(x.GoPointer())
-
 }
 
 // Returns the display that this `GtkRoot` is on.
@@ -829,9 +761,7 @@ // To set the focus to a particular widget in the root, it is usually
 // more convenient to use [method@Gtk.Widget.grab_focus] instead of
 // this function.
 func (x *MainApplicationWindow) SetFocus(FocusVar *gtk.Widget) {
-
 	gtk.XGtkRootSetFocus(x.GoPointer(), FocusVar.GoPointer())
-
 }
 
 func init() {
@@ -849,5 +779,4 @@
 	core.PuregoSafeRegister(&xMainApplicationWindowGLibType, libs, "mylib_gtk_meson_main_application_window_get_type")
 
 	core.PuregoSafeRegister(&xMainApplicationWindowShowToast, libs, "mylib_gtk_meson_main_application_window_show_toast")
-
 }

