From 35ee3be8d43f848c8cd5fcf8ed8537a04d52dff1 Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 14 Apr 2026 23:32:10 -0700
Subject: [PATCH] Gen: Re-run script

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
index fe3609bf7bba313e7aa4abfd511ee4b60f0caaaa..0a2337e0ef2df35662fddb2aa32e105e8187187b 100644
--- a/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
+++ b/examples/mylib-gtk-meson-go/mylibgtkmeson/window.go
@@ -766,7 +766,7 @@ }
 
 func init() {
 	core.SetPackageName("MYLIBGTKMESON", "mylib-gtk-meson-0.1")
-	core.SetSharedLibraries("MYLIBGTKMESON", []string{"libmylib-gtk-meson-0.1.so"})
+	core.SetSharedLibraries("MYLIBGTKMESON", []string{"libmylib-gtk-meson-0.1.so", "libmylib-gtk-meson-0.1.dylib"})
 	var libs []uintptr
 	for _, libPath := range core.GetPaths("MYLIBGTKMESON") {
 		lib, err := purego.Dlopen(libPath, purego.RTLD_NOW|purego.RTLD_GLOBAL)

