From 6b8e5efc073dd4a0338e1540efa2baeb7021c3ad Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Mon, 13 Apr 2026 22:44:52 -0700
Subject: [PATCH] Gen: Re-run script

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/v4/gdk/gdkcicpparams.go b/v4/gdk/gdkcicpparams.go
index fce7a710a9e6a323870f784ce16f3c0340464877..0f43faf0391b2e0b81c428bd9dd78f09b4d288f3 100644
--- a/v4/gdk/gdkcicpparams.go
+++ b/v4/gdk/gdkcicpparams.go
@@ -96,7 +96,7 @@ 	cls.Ptr = cret
 	return cls
 }
 
-var xCicpParamsBuildColorState func(uintptr) *ColorState
+var xCicpParamsBuildColorState func(uintptr, **glib.Error) *ColorState
 
 // Creates a new `GdkColorState` object for the cicp parameters in @self.
 //
@@ -106,7 +106,7 @@ // with an error message that can be presented to the user.
 func (x *CicpParams) BuildColorState() (*ColorState, error) {
 	var cerr *glib.Error
 
-	cret := xCicpParamsBuildColorState(x.GoPointer())
+	cret := xCicpParamsBuildColorState(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gdk/gdkcontentdeserializer.go b/v4/gdk/gdkcontentdeserializer.go
index 4725552ffa1e4094d553243c131eef04802e2129..1ad2615667b5b504a4e6210605425c31d0887f72 100644
--- a/v4/gdk/gdkcontentdeserializer.go
+++ b/v4/gdk/gdkcontentdeserializer.go
@@ -235,7 +235,7 @@ // to enable subclasses to chain up correctly.
 func (x *ContentDeserializer) LegacyPropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := gio.XGAsyncResultLegacyPropagateError(x.GoPointer())
+	cret := gio.XGAsyncResultLegacyPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gdk/gdkcontentserializer.go b/v4/gdk/gdkcontentserializer.go
index 59ef8f077a4b2a29b56afdae7b00377228992875..7eec0f778a54999e947bdebd00492e6b3ebc1f58 100644
--- a/v4/gdk/gdkcontentserializer.go
+++ b/v4/gdk/gdkcontentserializer.go
@@ -236,7 +236,7 @@ // to enable subclasses to chain up correctly.
 func (x *ContentSerializer) LegacyPropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := gio.XGAsyncResultLegacyPropagateError(x.GoPointer())
+	cret := gio.XGAsyncResultLegacyPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gdk/gdkdisplay.go b/v4/gdk/gdkdisplay.go
index 62b7e2c05d0f2ae5856aebeecc8c0ce16d0abf6c..cbff4ad62bb795f55492930c371778cc1e05b7d6 100644
--- a/v4/gdk/gdkdisplay.go
+++ b/v4/gdk/gdkdisplay.go
@@ -58,7 +58,7 @@ func (x *Display) Close() {
 	xDisplayClose(x.GoPointer())
 }
 
-var xDisplayCreateGlContext func(uintptr) uintptr
+var xDisplayCreateGlContext func(uintptr, **glib.Error) uintptr
 
 // Creates a new `GdkGLContext` for the `GdkDisplay`.
 //
@@ -73,7 +73,7 @@ func (x *Display) CreateGlContext() (*GLContext, error) {
 	var cls *GLContext
 	var cerr *glib.Error
 
-	cret := xDisplayCreateGlContext(x.GoPointer())
+	cret := xDisplayCreateGlContext(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -371,7 +371,7 @@ func (x *Display) NotifyStartupComplete(StartupIdVar string) {
 	xDisplayNotifyStartupComplete(x.GoPointer(), StartupIdVar)
 }
 
-var xDisplayPrepareGl func(uintptr) bool
+var xDisplayPrepareGl func(uintptr, **glib.Error) bool
 
 // Checks that OpenGL is available for @self and ensures that it is
 // properly initialized.
@@ -390,7 +390,7 @@ // might make use of OpenGL.
 func (x *Display) PrepareGl() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xDisplayPrepareGl(x.GoPointer())
+	cret := xDisplayPrepareGl(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gdk/gdkglcontext.go b/v4/gdk/gdkglcontext.go
index 56df9e05cae3218372c4be20917d67bc6e1b12b3..6cf2470c9ff48d837ece157d8e36fcd5146c5935 100644
--- a/v4/gdk/gdkglcontext.go
+++ b/v4/gdk/gdkglcontext.go
@@ -251,7 +251,7 @@ func (x *GLContext) MakeCurrent() {
 	xGLContextMakeCurrent(x.GoPointer())
 }
 
-var xGLContextRealize func(uintptr) bool
+var xGLContextRealize func(uintptr, **glib.Error) bool
 
 // Realizes the given `GdkGLContext`.
 //
@@ -259,7 +259,7 @@ // It is safe to call this function on a realized `GdkGLContext`.
 func (x *GLContext) Realize() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xGLContextRealize(x.GoPointer())
+	cret := xGLContextRealize(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gdk/gdksurface.go b/v4/gdk/gdksurface.go
index 89b6be7ff7d422cabc7f4802dd6e8da908da036b..61dfe7f764cdfe661c0f4a50b08bc180e94091e0 100644
--- a/v4/gdk/gdksurface.go
+++ b/v4/gdk/gdksurface.go
@@ -107,7 +107,7 @@ 	cls.Ptr = cret
 	return cls
 }
 
-var xSurfaceCreateGlContext func(uintptr) uintptr
+var xSurfaceCreateGlContext func(uintptr, **glib.Error) uintptr
 
 // Creates a new `GdkGLContext` for the `GdkSurface`.
 //
@@ -119,7 +119,7 @@ func (x *Surface) CreateGlContext() (*GLContext, error) {
 	var cls *GLContext
 	var cerr *glib.Error
 
-	cret := xSurfaceCreateGlContext(x.GoPointer())
+	cret := xSurfaceCreateGlContext(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -154,14 +154,14 @@ 	cret := xSurfaceCreateSimilarSurface(x.GoPointer(), ContentVar, WidthVar, HeightVar)
 	return cret
 }
 
-var xSurfaceCreateVulkanContext func(uintptr) uintptr
+var xSurfaceCreateVulkanContext func(uintptr, **glib.Error) uintptr
 
 // Sets an error and returns %NULL.
 func (x *Surface) CreateVulkanContext() (*VulkanContext, error) {
 	var cls *VulkanContext
 	var cerr *glib.Error
 
-	cret := xSurfaceCreateVulkanContext(x.GoPointer())
+	cret := xSurfaceCreateVulkanContext(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
diff --git a/v4/gdkpixbuf/gdk-pixbuf-loader.go b/v4/gdkpixbuf/gdk-pixbuf-loader.go
index 4adaff9ac963b15fbad2b3404c3f75b8c90f68e6..8af4f668c7f8544041e6077abfa3a9bd3cf3266f 100644
--- a/v4/gdkpixbuf/gdk-pixbuf-loader.go
+++ b/v4/gdkpixbuf/gdk-pixbuf-loader.go
@@ -266,7 +266,7 @@ 	}
 	return cls, cerr
 }
 
-var xPixbufLoaderClose func(uintptr) bool
+var xPixbufLoaderClose func(uintptr, **glib.Error) bool
 
 // Informs a pixbuf loader that no further writes with
 // gdk_pixbuf_loader_write() will occur, so that it can free its
@@ -286,7 +286,7 @@ // you will need to explicitly release any reference you hold.
 func (x *PixbufLoader) Close() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xPixbufLoaderClose(x.GoPointer())
+	cret := xPixbufLoaderClose(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gasyncresult.go b/v4/gio/gasyncresult.go
index 2df0a8501660b50c1c72d9e87bc7de60678e32bc..d9bff9ccc6da0e38dc2ff747551558c29e765d86 100644
--- a/v4/gio/gasyncresult.go
+++ b/v4/gio/gasyncresult.go
@@ -275,7 +275,7 @@ // to enable subclasses to chain up correctly.
 func (x *AsyncResultBase) LegacyPropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := XGAsyncResultLegacyPropagateError(x.GoPointer())
+	cret := XGAsyncResultLegacyPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
@@ -286,7 +286,7 @@ var (
 	XGAsyncResultGetSourceObject      func(uintptr) uintptr
 	XGAsyncResultGetUserData          func(uintptr) uintptr
 	XGAsyncResultIsTagged             func(uintptr, uintptr) bool
-	XGAsyncResultLegacyPropagateError func(uintptr) bool
+	XGAsyncResultLegacyPropagateError func(uintptr, **glib.Error) bool
 )
 
 func init() {
diff --git a/v4/gio/gcancellable.go b/v4/gio/gcancellable.go
index e27775595c74496169f15e6535cb20e5ef1b5ce0..2d109758cc4d237ebb9c1ac68f35a1702ab18865 100644
--- a/v4/gio/gcancellable.go
+++ b/v4/gio/gcancellable.go
@@ -433,14 +433,14 @@ func (x *Cancellable) Reset() {
 	xCancellableReset(x.GoPointer())
 }
 
-var xCancellableSetErrorIfCancelled func(uintptr) bool
+var xCancellableSetErrorIfCancelled func(uintptr, **glib.Error) bool
 
 // If the @cancellable is cancelled, sets the error to notify
 // that the operation was cancelled.
 func (x *Cancellable) SetErrorIfCancelled() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xCancellableSetErrorIfCancelled(x.GoPointer())
+	cret := xCancellableSetErrorIfCancelled(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gcredentials.go b/v4/gio/gcredentials.go
index 341084451a9ddbf444aa8c80f67826af14df6b3e..a3e07ae752fd28a2b962ae2ef5fddddd95106b9a 100644
--- a/v4/gio/gcredentials.go
+++ b/v4/gio/gcredentials.go
@@ -102,7 +102,7 @@ 	cret := xCredentialsGetNative(x.GoPointer(), NativeTypeVar)
 	return cret
 }
 
-var xCredentialsGetUnixPid func(uintptr) int
+var xCredentialsGetUnixPid func(uintptr, **glib.Error) int
 
 // Tries to get the UNIX process identifier from @credentials. This
 // method is only available on UNIX platforms.
@@ -113,14 +113,14 @@ // about the UNIX process ID.
 func (x *Credentials) GetUnixPid() (int, error) {
 	var cerr *glib.Error
 
-	cret := xCredentialsGetUnixPid(x.GoPointer())
+	cret := xCredentialsGetUnixPid(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
 	return cret, cerr
 }
 
-var xCredentialsGetUnixUser func(uintptr) uint
+var xCredentialsGetUnixUser func(uintptr, **glib.Error) uint
 
 // Tries to get the UNIX user identifier from @credentials. This
 // method is only available on UNIX platforms.
@@ -135,7 +135,7 @@ // particularly in language bindings.
 func (x *Credentials) GetUnixUser() (uint, error) {
 	var cerr *glib.Error
 
-	cret := xCredentialsGetUnixUser(x.GoPointer())
+	cret := xCredentialsGetUnixUser(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gdbusmessage.go b/v4/gio/gdbusmessage.go
index 04590607d51fa271eb5373e45189af35745b7fb8..bf3ff05e4c94ba774c561f8038393de6f4549037 100644
--- a/v4/gio/gdbusmessage.go
+++ b/v4/gio/gdbusmessage.go
@@ -100,7 +100,7 @@ 	cls.Ptr = cret
 	return cls
 }
 
-var xDBusMessageCopy func(uintptr) uintptr
+var xDBusMessageCopy func(uintptr, **glib.Error) uintptr
 
 // Copies @message. The copy is a deep copy and the returned
 // #GDBusMessage is completely identical except that it is guaranteed
@@ -112,7 +112,7 @@ func (x *DBusMessage) Copy() (*DBusMessage, error) {
 	var cls *DBusMessage
 	var cerr *glib.Error
 
-	cret := xDBusMessageCopy(x.GoPointer())
+	cret := xDBusMessageCopy(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -577,7 +577,7 @@ 	}
 	return cret, cerr
 }
 
-var xDBusMessageToGerror func(uintptr) bool
+var xDBusMessageToGerror func(uintptr, **glib.Error) bool
 
 // If @message is not of type %G_DBUS_MESSAGE_TYPE_ERROR does
 // nothing and returns %FALSE.
@@ -589,7 +589,7 @@ // well as the first string item in @message's body.
 func (x *DBusMessage) ToGerror() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xDBusMessageToGerror(x.GoPointer())
+	cret := xDBusMessageToGerror(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/ginputstream.go b/v4/gio/ginputstream.go
index 8181f23e7af960468c7974b4c7822ad8218eb4c1..3113112dc5c2df67475a18805a684fe33d9db738 100644
--- a/v4/gio/ginputstream.go
+++ b/v4/gio/ginputstream.go
@@ -713,7 +713,7 @@ 	}
 	return cret, cerr
 }
 
-var xInputStreamSetPending func(uintptr) bool
+var xInputStreamSetPending func(uintptr, **glib.Error) bool
 
 // Sets @stream to have actions pending. If the pending flag is
 // already set or @stream is closed, it will return %FALSE and set
@@ -721,7 +721,7 @@ // @error.
 func (x *InputStream) SetPending() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xInputStreamSetPending(x.GoPointer())
+	cret := xInputStreamSetPending(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/giostream.go b/v4/gio/giostream.go
index ec4b6b531040da98fa079b8848d545f5e22c7843..868b9069c2152f93856ae2e55fe6818ff5c9165e 100644
--- a/v4/gio/giostream.go
+++ b/v4/gio/giostream.go
@@ -622,7 +622,7 @@ 	cret := xIOStreamIsClosed(x.GoPointer())
 	return cret
 }
 
-var xIOStreamSetPending func(uintptr) bool
+var xIOStreamSetPending func(uintptr, **glib.Error) bool
 
 // Sets @stream to have actions pending. If the pending flag is
 // already set or @stream is closed, it will return %FALSE and set
@@ -630,7 +630,7 @@ // @error.
 func (x *IOStream) SetPending() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xIOStreamSetPending(x.GoPointer())
+	cret := xIOStreamSetPending(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/goutputstream.go b/v4/gio/goutputstream.go
index 57a8d40f36e37c89e53ea81f0f5d9068f085d92a..22fbc2495e23aa141fecc23d7f4efc56d42594cb 100644
--- a/v4/gio/goutputstream.go
+++ b/v4/gio/goutputstream.go
@@ -733,7 +733,7 @@ 	cret := xOutputStreamPrintf(x.GoPointer(), BytesWrittenVar, CancellableVar.GoPointer(), ErrorVar, FormatVar, varArgs...)
 	return cret
 }
 
-var xOutputStreamSetPending func(uintptr) bool
+var xOutputStreamSetPending func(uintptr, **glib.Error) bool
 
 // Sets @stream to have actions pending. If the pending flag is
 // already set or @stream is closed, it will return %FALSE and set
@@ -741,7 +741,7 @@ // @error.
 func (x *OutputStream) SetPending() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xOutputStreamSetPending(x.GoPointer())
+	cret := xOutputStreamSetPending(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gsimpleasyncresult.go b/v4/gio/gsimpleasyncresult.go
index c402818e67b15304a01b13e08e935073517fc1eb..900fa27f013f94d917f70f370dd323948d3ec2a1 100644
--- a/v4/gio/gsimpleasyncresult.go
+++ b/v4/gio/gsimpleasyncresult.go
@@ -368,7 +368,7 @@ 	cret := xSimpleAsyncResultGetSourceTag(x.GoPointer())
 	return cret
 }
 
-var xSimpleAsyncResultPropagateError func(uintptr) bool
+var xSimpleAsyncResultPropagateError func(uintptr, **glib.Error) bool
 
 // Propagates an error from within the simple asynchronous result to
 // a given destination.
@@ -379,7 +379,7 @@ // function will return %TRUE with @dest set appropriately.
 func (x *SimpleAsyncResult) PropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xSimpleAsyncResultPropagateError(x.GoPointer())
+	cret := xSimpleAsyncResultPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
@@ -533,7 +533,7 @@ // to enable subclasses to chain up correctly.
 func (x *SimpleAsyncResult) LegacyPropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := XGAsyncResultLegacyPropagateError(x.GoPointer())
+	cret := XGAsyncResultLegacyPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gsocket.go b/v4/gio/gsocket.go
index 6147847ce4e24de33cff5de483a8d9b4a2ab2713..93aa32f3485879023ca7f5167c3ee52aa3bb8291 100644
--- a/v4/gio/gsocket.go
+++ b/v4/gio/gsocket.go
@@ -496,7 +496,7 @@ 	}
 	return cret, cerr
 }
 
-var xSocketCheckConnectResult func(uintptr) bool
+var xSocketCheckConnectResult func(uintptr, **glib.Error) bool
 
 // Checks and resets the pending connect error for the socket.
 // This is used to check for errors when g_socket_connect() is
@@ -504,14 +504,14 @@ // used in non-blocking mode.
 func (x *Socket) CheckConnectResult() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xSocketCheckConnectResult(x.GoPointer())
+	cret := xSocketCheckConnectResult(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
 	return cret, cerr
 }
 
-var xSocketClose func(uintptr) bool
+var xSocketClose func(uintptr, **glib.Error) bool
 
 // Closes the socket, shutting down any active connection.
 //
@@ -545,7 +545,7 @@ // does.)
 func (x *Socket) Close() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xSocketClose(x.GoPointer())
+	cret := xSocketClose(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
@@ -736,7 +736,7 @@ 	cret := xSocketGetBroadcast(x.GoPointer())
 	return cret
 }
 
-var xSocketGetCredentials func(uintptr) uintptr
+var xSocketGetCredentials func(uintptr, **glib.Error) uintptr
 
 // Returns the credentials of the foreign process connected to this
 // socket, if any (e.g. it is only supported for %G_SOCKET_FAMILY_UNIX
@@ -762,7 +762,7 @@ func (x *Socket) GetCredentials() (*Credentials, error) {
 	var cls *Credentials
 	var cerr *glib.Error
 
-	cret := xSocketGetCredentials(x.GoPointer())
+	cret := xSocketGetCredentials(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -813,7 +813,7 @@ 	cret := xSocketGetListenBacklog(x.GoPointer())
 	return cret
 }
 
-var xSocketGetLocalAddress func(uintptr) uintptr
+var xSocketGetLocalAddress func(uintptr, **glib.Error) uintptr
 
 // Try to get the local address of a bound socket. This is only
 // useful if the socket has been bound to a local address,
@@ -822,7 +822,7 @@ func (x *Socket) GetLocalAddress() (*SocketAddress, error) {
 	var cls *SocketAddress
 	var cerr *glib.Error
 
-	cret := xSocketGetLocalAddress(x.GoPointer())
+	cret := xSocketGetLocalAddress(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -888,7 +888,7 @@ 	cret := xSocketGetProtocol(x.GoPointer())
 	return cret
 }
 
-var xSocketGetRemoteAddress func(uintptr) uintptr
+var xSocketGetRemoteAddress func(uintptr, **glib.Error) uintptr
 
 // Try to get the remote address of a connected socket. This is only
 // useful for connection oriented sockets that have been connected.
@@ -896,7 +896,7 @@ func (x *Socket) GetRemoteAddress() (*SocketAddress, error) {
 	var cls *SocketAddress
 	var cerr *glib.Error
 
-	cret := xSocketGetRemoteAddress(x.GoPointer())
+	cret := xSocketGetRemoteAddress(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -1049,7 +1049,7 @@ 	}
 	return cret, cerr
 }
 
-var xSocketListen func(uintptr) bool
+var xSocketListen func(uintptr, **glib.Error) bool
 
 // Marks the socket as a server socket, i.e. a socket that is used
 // to accept incoming requests using g_socket_accept().
@@ -1062,7 +1062,7 @@ // g_socket_set_listen_backlog().
 func (x *Socket) Listen() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xSocketListen(x.GoPointer())
+	cret := xSocketListen(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gio/gsocketconnection.go b/v4/gio/gsocketconnection.go
index 501d052f728d1281709393621d80af77744a71d4..e1fd7b4e61c756550a977c1538d97a3ba9f51bf4 100644
--- a/v4/gio/gsocketconnection.go
+++ b/v4/gio/gsocketconnection.go
@@ -254,14 +254,14 @@ 	}
 	return cret, cerr
 }
 
-var xSocketConnectionGetLocalAddress func(uintptr) uintptr
+var xSocketConnectionGetLocalAddress func(uintptr, **glib.Error) uintptr
 
 // Try to get the local address of a socket connection.
 func (x *SocketConnection) GetLocalAddress() (*SocketAddress, error) {
 	var cls *SocketAddress
 	var cerr *glib.Error
 
-	cret := xSocketConnectionGetLocalAddress(x.GoPointer())
+	cret := xSocketConnectionGetLocalAddress(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
@@ -274,7 +274,7 @@ 	}
 	return cls, cerr
 }
 
-var xSocketConnectionGetRemoteAddress func(uintptr) uintptr
+var xSocketConnectionGetRemoteAddress func(uintptr, **glib.Error) uintptr
 
 // Try to get the remote address of a socket connection.
 //
@@ -288,7 +288,7 @@ func (x *SocketConnection) GetRemoteAddress() (*SocketAddress, error) {
 	var cls *SocketAddress
 	var cerr *glib.Error
 
-	cret := xSocketConnectionGetRemoteAddress(x.GoPointer())
+	cret := xSocketConnectionGetRemoteAddress(x.GoPointer(), &cerr)
 
 	if cret == 0 {
 		return nil, cerr
diff --git a/v4/gio/gtask.go b/v4/gio/gtask.go
index f7101dc7d3aca22abb5743d2e8a6bb3785f455e1..16a89ea5a9329bee169817a651717563dacd9243 100644
--- a/v4/gio/gtask.go
+++ b/v4/gio/gtask.go
@@ -777,7 +777,7 @@ 	cret := xTaskHadError(x.GoPointer())
 	return cret
 }
 
-var xTaskPropagateBoolean func(uintptr) bool
+var xTaskPropagateBoolean func(uintptr, **glib.Error) bool
 
 // Gets the result of @task as a #gboolean.
 //
@@ -789,14 +789,14 @@ // error) to the caller, you may only call it once.
 func (x *Task) PropagateBoolean() (bool, error) {
 	var cerr *glib.Error
 
-	cret := xTaskPropagateBoolean(x.GoPointer())
+	cret := xTaskPropagateBoolean(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
 	return cret, cerr
 }
 
-var xTaskPropagateInt func(uintptr) int
+var xTaskPropagateInt func(uintptr, **glib.Error) int
 
 // Gets the result of @task as an integer (#gssize).
 //
@@ -808,14 +808,14 @@ // error) to the caller, you may only call it once.
 func (x *Task) PropagateInt() (int, error) {
 	var cerr *glib.Error
 
-	cret := xTaskPropagateInt(x.GoPointer())
+	cret := xTaskPropagateInt(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
 	return cret, cerr
 }
 
-var xTaskPropagatePointer func(uintptr) uintptr
+var xTaskPropagatePointer func(uintptr, **glib.Error) uintptr
 
 // Gets the result of @task as a pointer, and transfers ownership
 // of that value to the caller.
@@ -828,7 +828,7 @@ // error) to the caller, you may only call it once.
 func (x *Task) PropagatePointer() (uintptr, error) {
 	var cerr *glib.Error
 
-	cret := xTaskPropagatePointer(x.GoPointer())
+	cret := xTaskPropagatePointer(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
@@ -1214,7 +1214,7 @@ // to enable subclasses to chain up correctly.
 func (x *Task) LegacyPropagateError() (bool, error) {
 	var cerr *glib.Error
 
-	cret := XGAsyncResultLegacyPropagateError(x.GoPointer())
+	cret := XGAsyncResultLegacyPropagateError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/glib/gerror.go b/v4/glib/gerror.go
index f17c785d8c887f043508fbf3b7fd0b894233b10b..f389a3613bb836ca4526906f2c82211aec7ebc7c 100644
--- a/v4/glib/gerror.go
+++ b/v4/glib/gerror.go
@@ -120,14 +120,14 @@ 	cret := xErrorMatches(x.GoPointer(), DomainVar, CodeVar)
 	return cret
 }
 
-var xClearError func()
+var xClearError func(**Error)
 
 // If @err or `*err` is %NULL, does nothing. Otherwise,
 // calls g_error_free() on `*err` and sets `*err` to %NULL.
 func ClearError() error {
 	var cerr *Error
 
-	xClearError()
+	xClearError(&cerr)
 	if cerr == nil {
 		return nil
 	}
diff --git a/v4/glib/giochannel.go b/v4/glib/giochannel.go
index 5635b4ce456ec376033129244a64d73c5fce0b61..14ee6a70895d687c3e9f4bfed6058ae6c020c8ad 100644
--- a/v4/glib/giochannel.go
+++ b/v4/glib/giochannel.go
@@ -160,13 +160,13 @@ func (x *IOChannel) Close() {
 	xIOChannelClose(x.GoPointer())
 }
 
-var xIOChannelFlush func(uintptr) IOStatus
+var xIOChannelFlush func(uintptr, **Error) IOStatus
 
 // Flushes the write buffer for the GIOChannel.
 func (x *IOChannel) Flush() (IOStatus, error) {
 	var cerr *Error
 
-	cret := xIOChannelFlush(x.GoPointer())
+	cret := xIOChannelFlush(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/glib/gmarkup.go b/v4/glib/gmarkup.go
index 06235d319f0b4fd5a92cde6dd77ddf59ca474f85..193ea12b8f3793f51ba2df1adc5fe9a63e38a853 100644
--- a/v4/glib/gmarkup.go
+++ b/v4/glib/gmarkup.go
@@ -41,7 +41,7 @@ 	cret := xNewMarkupParseContext(ParserVar, FlagsVar, UserDataVar, NewCallback(UserDataDnotifyVar))
 	return cret
 }
 
-var xMarkupParseContextEndParse func(uintptr) bool
+var xMarkupParseContextEndParse func(uintptr, **Error) bool
 
 // Signals to the #GMarkupParseContext that all data has been
 // fed into the parse context with g_markup_parse_context_parse().
@@ -51,7 +51,7 @@ // for example if elements are still open.
 func (x *MarkupParseContext) EndParse() (bool, error) {
 	var cerr *Error
 
-	cret := xMarkupParseContextEndParse(x.GoPointer())
+	cret := xMarkupParseContextEndParse(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/glib/gregex.go b/v4/glib/gregex.go
index 6f93b7c5ef32595c67ee18cbfa1dd42874926762..5f1c031b3e86d91fe0d0c406bf687179ce314e88 100644
--- a/v4/glib/gregex.go
+++ b/v4/glib/gregex.go
@@ -442,7 +442,7 @@ 	cret := xMatchInfoMatches(x.GoPointer())
 	return cret
 }
 
-var xMatchInfoNext func(uintptr) bool
+var xMatchInfoNext func(uintptr, **Error) bool
 
 // Scans for the next match using the same parameters of the previous
 // call to g_regex_match_full() or g_regex_match() that returned
@@ -453,7 +453,7 @@ // cannot free it before calling this function.
 func (x *MatchInfo) Next() (bool, error) {
 	var cerr *Error
 
-	cret := xMatchInfoNext(x.GoPointer())
+	cret := xMatchInfoNext(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gtk/gtkprintjob.go b/v4/gtk/gtkprintjob.go
index ee3ce66a84a43a6399bb813c2e0dc0acf3045115..bff97bca7a4df17825bbb1a7306a6c1104cf546f 100644
--- a/v4/gtk/gtkprintjob.go
+++ b/v4/gtk/gtkprintjob.go
@@ -181,14 +181,14 @@ 	cret := xPrintJobGetStatus(x.GoPointer())
 	return cret
 }
 
-var xPrintJobGetSurface func(uintptr) *cairo.Surface
+var xPrintJobGetSurface func(uintptr, **glib.Error) *cairo.Surface
 
 // Gets a cairo surface onto which the pages of
 // the print job should be rendered.
 func (x *PrintJob) GetSurface() (*cairo.Surface, error) {
 	var cerr *glib.Error
 
-	cret := xPrintJobGetSurface(x.GoPointer())
+	cret := xPrintJobGetSurface(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}
diff --git a/v4/gtk/gtkprintoperation.go b/v4/gtk/gtkprintoperation.go
index ff4bda21fca0a85c7d938743062d5a94717fb8bc..c6442336a140cd7a1da5b94dff928d78d59d52d4 100644
--- a/v4/gtk/gtkprintoperation.go
+++ b/v4/gtk/gtkprintoperation.go
@@ -673,7 +673,7 @@ 	cret := xPrintOperationGetEmbedPageSetup(x.GoPointer())
 	return cret
 }
 
-var xPrintOperationGetError func(uintptr)
+var xPrintOperationGetError func(uintptr, **glib.Error)
 
 // Call this when the result of a print operation is
 // %GTK_PRINT_OPERATION_RESULT_ERROR.
@@ -686,7 +686,7 @@ // The returned `GError` will contain more details on what went wrong.
 func (x *PrintOperation) GetError() error {
 	var cerr *glib.Error
 
-	xPrintOperationGetError(x.GoPointer())
+	xPrintOperationGetError(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return nil
 	}
diff --git a/v4/gtk/gtkrecentmanager.go b/v4/gtk/gtkrecentmanager.go
index 50469624ea5ff7c8e0bcba1704aa055d65a91776..a319a01b991042c4ccfde52933200f54a398f962 100644
--- a/v4/gtk/gtkrecentmanager.go
+++ b/v4/gtk/gtkrecentmanager.go
@@ -676,13 +676,13 @@ 	}
 	return cret, cerr
 }
 
-var xRecentManagerPurgeItems func(uintptr) int32
+var xRecentManagerPurgeItems func(uintptr, **glib.Error) int32
 
 // Purges every item from the recently used resources list.
 func (x *RecentManager) PurgeItems() (int32, error) {
 	var cerr *glib.Error
 
-	cret := xRecentManagerPurgeItems(x.GoPointer())
+	cret := xRecentManagerPurgeItems(x.GoPointer(), &cerr)
 	if cerr == nil {
 		return cret, nil
 	}

