From 31fb96ca3ba5d137dc838e0328157c2baeace9ad Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 14 Apr 2026 23:29:47 -0700
Subject: [PATCH] GIR: Document where Homebrew and MacPorts store libraries

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/internal/core/core.go b/internal/core/core.go
index cb1196ce1d386c3bf1f9118b21f2c25129d60fcc..0081d4f2d59ce78ceb95292755063e89cd77524d 100644
--- a/internal/core/core.go
+++ b/internal/core/core.go
@@ -30,12 +30,16 @@ // paths to where the shared object files should be located
 // this is unique per architecture
 // Debian/Ubuntu has it split into specific arch folder, Fedora is just /usr/lib64
 // Flatpak uses /app/lib for application libraries and runtimes don't vendor `pkg-config` as the fallback
+// Homebrew on macOS uses /opt/homebrew/lib (arm64) or /usr/local/lib/ (amd64)
+// MacPorts on macOS /opt/local/lib/
 // see:
 // https://fedora.pkgs.org/38/fedora-x86_64/gtk4-4.10.1-1.fc38.x86_64.rpm.html
 // https://fedora.pkgs.org/38/fedora-aarch64/gtk4-4.10.1-1.fc38.aarch64.rpm.html
 // https://ubuntu.pkgs.org/23.04/ubuntu-main-amd64/libgtk-4-1_4.10.1+ds-2ubuntu1_amd64.deb.html
 // https://ubuntu.pkgs.org/23.04/ubuntu-main-arm64/libgtk-4-1_4.10.1+ds-2ubuntu1_arm64.deb.html
 // https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html (see --libdir)
+// https://docs.brew.sh/Installation
+// https://trac.macports.org/wiki/FAQ (see "Can I install software in /usr/local while using MacPorts? ¶")
 var paths = map[string][]string{
 	"linux_amd64":  {"/app/lib/", "/usr/lib/x86_64-linux-gnu/", "/usr/lib64/", "/usr/lib/"},
 	"linux_arm64":  {"/app/lib/", "/usr/lib/aarch64-linux-gnu/", "/usr/lib64/", "/usr/lib/"},

