From c9a9c5e42eb97c394fc452cb2cf38bdff0707fbc Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sun, 19 Apr 2026 15:33:48 -0700
Subject: [PATCH] Examples: Use common gitignore for `mylib-gtk-meson` too

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/examples/mylib-gtk-meson/.gitignore b/examples/mylib-gtk-meson/.gitignore
index 070bbf540f481f378f4d824b2d478caad538b2fa..7e69e3ebe204ee28e2d026f2e9f0369a0dd2a296 100644
--- a/examples/mylib-gtk-meson/.gitignore
+++ b/examples/mylib-gtk-meson/.gitignore
@@ -1,7 +1,8 @@
-_build
-repo
-builddir
-.flatpak-builder
 *.gresource
 *.ui
 *.mo
+LINGUAS
+.flatpak-builder
+builddir
+repo
+vendor

