From 2c08ca0d275c13b84d3b0bc981820265512bdaeb Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Sat, 25 Oct 2025 15:20:39 -0700
Subject: [PATCH] Strip trailing slashes from forge URLs when fetching stars

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/src/bz-full-view.c b/src/bz-full-view.c
index de106035a548afe4cbf7de9d63a8028f69002d26..e824bf9f497923ade55c2bea0a2d84a3448e0b62 100644
--- a/src/bz-full-view.c
+++ b/src/bz-full-view.c
@@ -791,6 +791,10 @@   forge_link = bz_entry_get_forge_url (entry);
   if (forge_link == NULL)
     goto done;
 
+  // Remove trailing `/` from forge URLs if it exists
+  if (g_str_has_suffix (forge_link, "/"))
+    forge_link = g_strndup (forge_link, strlen (forge_link) - 1);
+
   if (g_regex_match_simple (
           "https://github.com/.*/.*",
           forge_link,

