From 1a4d2fb86a5aec041c46012c393cb96ecd084d7a Mon Sep 17 00:00:00 2001
From: Felicitas Pojtinger <felicitas@pojtinger.com>
Date: Tue, 17 Mar 2026 23:59:00 -0700
Subject: [PATCH] feat: Add simple DB content list CLI via serial

Signed-off-by: Felicitas Pojtinger <felicitas@pojtinger.com>
---
diff --git a/go.mod b/go.mod
index 0d3587642009d23c1e13e4a073f50220530404ef..38e7e509b825334af0c4cf5e03f56e71d5565b40 100644
--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,11 @@
 go 1.25.0
 
 require (
-	github.com/creack/goselect v0.1.2 // indirect
-	go.bug.st/serial v1.6.4 // indirect
-	golang.org/x/sys v0.19.0 // indirect
+	github.com/sigurn/crc16 v0.0.0-20240131213347-83fcde1e29d1
+	go.bug.st/serial v1.6.4
+)
+
+require (
+	github.com/creack/goselect v0.1.3 // indirect
+	golang.org/x/sys v0.42.0 // indirect
 )
diff --git a/go.sum b/go.sum
index 3223a0270ea508aee4e7646a75b5a42903d66901..a82a7998565d993f2cd17131583352b792a41b01 100644
--- a/go.sum
+++ b/go.sum
@@ -1,6 +1,16 @@
-github.com/creack/goselect v0.1.2 h1:2DNy14+JPjRBgPzAd1thbQp4BSIihxcBf0IXhQXDRa0=
-github.com/creack/goselect v0.1.2/go.mod h1:a/NhLweNvqIYMuxcMOuWY516Cimucms3DglDzQP3hKY=
+github.com/creack/goselect v0.1.3 h1:MaGNMclRo7P2Jl21hBpR1Cn33ITSbKP6E49RtfblLKc=
+github.com/creack/goselect v0.1.3/go.mod h1:a/NhLweNvqIYMuxcMOuWY516Cimucms3DglDzQP3hKY=
+github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
+github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
+github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
+github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
+github.com/sigurn/crc16 v0.0.0-20240131213347-83fcde1e29d1 h1:NVK+OqnavpyFmUiKfUMHrpvbCi2VFoWTrcpI7aDaJ2I=
+github.com/sigurn/crc16 v0.0.0-20240131213347-83fcde1e29d1/go.mod h1:9/etS5gpQq9BJsJMWg1wpLbfuSnkm8dPF6FdW2JXVhA=
+github.com/stretchr/testify v1.8.4 h1:CcVxjf3Q8PM0mHUKJCdn+eZZtm5yQwehR5yeSVQQcUk=
+github.com/stretchr/testify v1.8.4/go.mod h1:sz/lmYIOXD/1dqDmKjjqLyZ2RngseejIcXlSw2iwfAo=
 go.bug.st/serial v1.6.4 h1:7FmqNPgVp3pu2Jz5PoPtbZ9jJO5gnEnZIvnI1lzve8A=
 go.bug.st/serial v1.6.4/go.mod h1:nofMJxTeNVny/m6+KaafC6vJGj3miwQZ6vW4BZUGJPI=
-golang.org/x/sys v0.19.0 h1:q5f1RH2jigJ1MoAWp2KTp3gm5zAGFUTarQZ5U386+4o=
-golang.org/x/sys v0.19.0/go.mod h1:/VUhepiaJMQUp4+oa/7Zr1D23ma6VTLIYjOOTFZPUcA=
+golang.org/x/sys v0.42.0 h1:omrd2nAlyT5ESRdCLYdm3+fMfNFE/+Rf4bDIQImRJeo=
+golang.org/x/sys v0.42.0/go.mod h1:4GL1E5IUh+htKOUEOaiffhrAeqysfVGipDYzABqnCmw=
+gopkg.in/yaml.v3 v3.0.1 h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=
+gopkg.in/yaml.v3 v3.0.1/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
diff --git a/main.go b/main.go
new file mode 100644
index 0000000000000000000000000000000000000000..7a2646f7e263ff6d586d870bd7b0301665a1d103
--- /dev/null
+++ b/main.go
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"go.bug.st/serial"
+
+	"codeberg.org/pojntfx/runway/pkg/cmp"
+	"codeberg.org/pojntfx/runway/pkg/dlp"
+)
+
+func main() {
+	device := flag.String("device", "/dev/ttyUSB0", "Serial device")
+
+	flag.Parse()
+
+	port, err := serial.Open(
+		*device,
+		&serial.Mode{
+			BaudRate: cmp.InitialBaud,
+			DataBits: 8,
+			Parity:   serial.NoParity,
+			StopBits: serial.OneStopBit,
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer port.Close()
+
+	port.SetReadTimeout(30 * time.Second)
+
+	slog.Info("Waiting for HotSync button being pressed ...")
+
+	baud := cmp.Handshake(port)
+
+	slog.Info("Connected", "baud", baud)
+
+	xid := byte(1)
+	startIndex := uint16(0)
+	for {
+		entries, lastIndex, more := dlp.ReadDBList(port, &xid, startIndex)
+		for _, e := range entries {
+			fmt.Println(e.Name)
+		}
+
+		if !more {
+			break
+		}
+
+		startIndex = lastIndex + 1
+	}
+
+	dlp.EndOfSync(port, &xid)
+}

